<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>gnomad_qc.v4.annotations.compute_coverage &mdash; gnomad_qc master documentation</title>
      <link rel="stylesheet" href="../../../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../../../_static/theme_overrides.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../../_static/jquery.js"></script>
        <script src="../../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../../../../" id="documentation_options" src="../../../../_static/documentation_options.js"></script>
        <script src="../../../../_static/doctools.js"></script>
        <script src="../../../../_static/sphinx_highlight.js"></script>
        <script src="https://unpkg.com/mermaid@10.2.0/dist/mermaid.min.js"></script>
        <script>mermaid.initialize({startOnLoad:true});</script>
    <script src="../../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../../index.html" class="icon icon-home">
            gnomad_qc
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../api_reference/index.html">API Reference</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/broadinstitute/gnomad_qc/tags">Change Log</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../../index.html">gnomad_qc</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">gnomad_qc.v4.annotations.compute_coverage</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for gnomad_qc.v4.annotations.compute_coverage</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;Script to compute coverage statistics on gnomAD v4 exomes.&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Optional</span>

<span class="kn">import</span> <span class="nn">hail</span> <span class="k">as</span> <span class="nn">hl</span>
<span class="kn">from</span> <span class="nn">gnomad.resources.grch38.reference_data</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">telomeres_and_centromeres</span><span class="p">,</span>
    <span class="n">vep_context</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">gnomad.utils.annotations</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">build_freq_stratification_list</span><span class="p">,</span>
    <span class="n">generate_freq_group_membership_array</span><span class="p">,</span>
    <span class="n">qual_hist_expr</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">gnomad.utils.sparse_mt</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">compute_coverage_stats</span><span class="p">,</span>
    <span class="n">compute_stats_per_ref_site</span><span class="p">,</span>
    <span class="n">get_allele_number_agg_func</span><span class="p">,</span>
<span class="p">)</span>

<span class="kn">from</span> <span class="nn">gnomad_qc.resource_utils</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">PipelineResourceCollection</span><span class="p">,</span>
    <span class="n">PipelineStepResourceCollection</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v4.resources.annotations</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">get_all_sites_an_and_qual_hists</span><span class="p">,</span>
    <span class="n">get_downsampling</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v4.resources.basics</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">calling_intervals</span><span class="p">,</span>
    <span class="n">get_gnomad_v4_genomes_vds</span><span class="p">,</span>
    <span class="n">get_gnomad_v4_vds</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v4.resources.meta</span> <span class="kn">import</span> <span class="n">meta</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v4.resources.release</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">release_all_sites_an</span><span class="p">,</span>
    <span class="n">release_all_sites_an_tsv_path</span><span class="p">,</span>
    <span class="n">release_coverage</span><span class="p">,</span>
    <span class="n">release_coverage_tsv_path</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="nb">format</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%(levelname)s</span><span class="s2"> (</span><span class="si">%(name)s</span><span class="s2"> </span><span class="si">%(lineno)s</span><span class="s2">): </span><span class="si">%(message)s</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s2">&quot;coverage&quot;</span><span class="p">)</span>
<span class="n">logger</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>


<div class="viewcode-block" id="get_exomes_group_membership_ht"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/compute_coverage.html#gnomad_qc.v4.annotations.compute_coverage.get_exomes_group_membership_ht">[docs]</a><span class="k">def</span> <span class="nf">get_exomes_group_membership_ht</span><span class="p">(</span>
    <span class="n">meta_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">ds_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">non_ukb_ds_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get exomes group membership HT for all sites allele number stratification.</span>

<span class="sd">    :param meta_ht: Metadata HT.</span>
<span class="sd">    :param ds_ht: Full frequency downsampling HT.</span>
<span class="sd">    :param non_ukb_ds_ht: Non-UKB frequency downsampling HT.</span>
<span class="sd">    :return: Group membership HT.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Filter to release samples.</span>
    <span class="n">meta_ht</span> <span class="o">=</span> <span class="n">meta_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">meta_ht</span><span class="o">.</span><span class="n">release</span><span class="p">)</span>

    <span class="c1"># Filter to non-UKB samples.</span>
    <span class="n">non_ukb_meta_ht</span> <span class="o">=</span> <span class="n">meta_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="o">~</span><span class="n">meta_ht</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">ukb_sample</span><span class="p">)</span>

    <span class="c1"># Create group membership HT.</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">generate_freq_group_membership_array</span><span class="p">(</span>
        <span class="n">meta_ht</span><span class="p">,</span>
        <span class="n">build_freq_stratification_list</span><span class="p">(</span>
            <span class="n">sex_expr</span><span class="o">=</span><span class="n">meta_ht</span><span class="o">.</span><span class="n">sex_imputation</span><span class="o">.</span><span class="n">sex_karyotype</span><span class="p">,</span>
            <span class="n">gen_anc_expr</span><span class="o">=</span><span class="n">meta_ht</span><span class="o">.</span><span class="n">population_inference</span><span class="o">.</span><span class="n">pop</span><span class="p">,</span>
            <span class="n">downsampling_expr</span><span class="o">=</span><span class="n">ds_ht</span><span class="p">[</span><span class="n">meta_ht</span><span class="o">.</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">downsampling</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="n">downsamplings</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">ds_ht</span><span class="o">.</span><span class="n">downsamplings</span><span class="p">),</span>
        <span class="n">ds_gen_anc_counts</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">ds_ht</span><span class="o">.</span><span class="n">ds_pop_counts</span><span class="p">),</span>
    <span class="p">)</span>

    <span class="c1"># Add non-UKB group membership to HT.</span>
    <span class="n">non_ukb_ht</span> <span class="o">=</span> <span class="n">generate_freq_group_membership_array</span><span class="p">(</span>
        <span class="n">non_ukb_meta_ht</span><span class="p">,</span>
        <span class="n">build_freq_stratification_list</span><span class="p">(</span>
            <span class="n">sex_expr</span><span class="o">=</span><span class="n">non_ukb_meta_ht</span><span class="o">.</span><span class="n">sex_imputation</span><span class="o">.</span><span class="n">sex_karyotype</span><span class="p">,</span>
            <span class="n">gen_anc_expr</span><span class="o">=</span><span class="n">non_ukb_meta_ht</span><span class="o">.</span><span class="n">population_inference</span><span class="o">.</span><span class="n">pop</span><span class="p">,</span>
            <span class="n">downsampling_expr</span><span class="o">=</span><span class="n">non_ukb_ds_ht</span><span class="p">[</span><span class="n">non_ukb_meta_ht</span><span class="o">.</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">downsampling</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="n">downsamplings</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">non_ukb_ds_ht</span><span class="o">.</span><span class="n">downsamplings</span><span class="p">),</span>
        <span class="n">ds_gen_anc_counts</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">non_ukb_ds_ht</span><span class="o">.</span><span class="n">ds_pop_counts</span><span class="p">),</span>
    <span class="p">)</span>
    <span class="n">n_non_ukb</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">len</span><span class="p">(</span><span class="n">non_ukb_ht</span><span class="o">.</span><span class="n">freq_meta</span><span class="p">))</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="n">group_membership</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">group_membership</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">coalesce</span><span class="p">(</span>
                <span class="n">non_ukb_ht</span><span class="p">[</span><span class="n">ht</span><span class="o">.</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">group_membership</span><span class="p">,</span>
                <span class="n">hl</span><span class="o">.</span><span class="n">range</span><span class="p">(</span><span class="n">n_non_ukb</span><span class="p">)</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="kc">False</span><span class="p">),</span>
            <span class="p">)</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="c1"># Add non-UKB group metadata and sample counts to HT.</span>
    <span class="n">non_ukb_globals</span> <span class="o">=</span> <span class="n">non_ukb_ht</span><span class="o">.</span><span class="n">index_globals</span><span class="p">()</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span>
        <span class="n">freq_meta</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">freq_meta</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span>
            <span class="n">non_ukb_globals</span><span class="o">.</span><span class="n">freq_meta</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
                <span class="k">lambda</span> <span class="n">d</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">dict</span><span class="p">(</span><span class="n">d</span><span class="o">.</span><span class="n">items</span><span class="p">()</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="s2">&quot;subset&quot;</span><span class="p">,</span> <span class="s2">&quot;non_ukb&quot;</span><span class="p">)))</span>
            <span class="p">)</span>
        <span class="p">)</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">d</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">dict</span><span class="p">(</span>
                <span class="n">d</span><span class="o">.</span><span class="n">items</span><span class="p">()</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;pop&quot;</span><span class="p">,</span> <span class="p">(</span><span class="s2">&quot;gen_anc&quot;</span><span class="p">,</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span> <span class="n">x</span><span class="p">))</span>
            <span class="p">)</span>
        <span class="p">),</span>
        <span class="n">freq_meta_sample_count</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">freq_meta_sample_count</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span>
            <span class="n">non_ukb_globals</span><span class="o">.</span><span class="n">freq_meta_sample_count</span>
        <span class="p">),</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">ht</span></div>


<div class="viewcode-block" id="get_genomes_group_membership_ht"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/compute_coverage.html#gnomad_qc.v4.annotations.compute_coverage.get_genomes_group_membership_ht">[docs]</a><span class="k">def</span> <span class="nf">get_genomes_group_membership_ht</span><span class="p">(</span>
    <span class="n">meta_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get genomes group membership HT for all sites allele number stratification.</span>

<span class="sd">    :param meta_ht: Metadata HT.</span>
<span class="sd">    :return: Group membership HT.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Filter to release samples.</span>
    <span class="n">meta_ht</span> <span class="o">=</span> <span class="n">meta_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">meta_ht</span><span class="o">.</span><span class="n">release</span><span class="p">)</span>

    <span class="c1"># Create group membership HT.</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">generate_freq_group_membership_array</span><span class="p">(</span>
        <span class="n">meta_ht</span><span class="p">,</span>
        <span class="n">build_freq_stratification_list</span><span class="p">(</span>
            <span class="n">sex_expr</span><span class="o">=</span><span class="n">meta_ht</span><span class="o">.</span><span class="n">sex_imputation</span><span class="o">.</span><span class="n">sex_karyotype</span><span class="p">,</span>
            <span class="n">gen_anc_expr</span><span class="o">=</span><span class="n">meta_ht</span><span class="o">.</span><span class="n">population_inference</span><span class="o">.</span><span class="n">pop</span><span class="p">,</span>
        <span class="p">),</span>
    <span class="p">)</span>

    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span>
        <span class="n">freq_meta</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">freq_meta</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">d</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">dict</span><span class="p">(</span>
                <span class="n">d</span><span class="o">.</span><span class="n">items</span><span class="p">()</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;pop&quot;</span><span class="p">,</span> <span class="p">(</span><span class="s2">&quot;gen_anc&quot;</span><span class="p">,</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span> <span class="n">x</span><span class="p">))</span>
            <span class="p">)</span>
        <span class="p">),</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">ht</span></div>


<div class="viewcode-block" id="adjust_interval_padding"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/compute_coverage.html#gnomad_qc.v4.annotations.compute_coverage.adjust_interval_padding">[docs]</a><span class="k">def</span> <span class="nf">adjust_interval_padding</span><span class="p">(</span><span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">padding</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Adjust interval padding in HT.</span>

<span class="sd">    .. warning::</span>

<span class="sd">        This function can lead to overlapping intervals, so it is not recommended for</span>
<span class="sd">        most applications. For example, it can be used to filter a variant list to all</span>
<span class="sd">        variants within the returned interval list, but would not work for getting an</span>
<span class="sd">        aggregate statistic for each interval if the desired output is independent</span>
<span class="sd">        statistics.</span>

<span class="sd">    :param ht: HT to adjust.</span>
<span class="sd">    :param padding: Padding to use.</span>
<span class="sd">    :return: HT with adjusted interval padding.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">ht</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span>
        <span class="n">interval</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">locus_interval</span><span class="p">(</span>
            <span class="n">ht</span><span class="o">.</span><span class="n">interval</span><span class="o">.</span><span class="n">start</span><span class="o">.</span><span class="n">contig</span><span class="p">,</span>
            <span class="n">ht</span><span class="o">.</span><span class="n">interval</span><span class="o">.</span><span class="n">start</span><span class="o">.</span><span class="n">position</span> <span class="o">-</span> <span class="n">padding</span><span class="p">,</span>
            <span class="n">ht</span><span class="o">.</span><span class="n">interval</span><span class="o">.</span><span class="n">end</span><span class="o">.</span><span class="n">position</span> <span class="o">+</span> <span class="n">padding</span><span class="p">,</span>
            <span class="c1"># Include the end of the intervals to capture all variants.</span>
            <span class="n">reference_genome</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">interval</span><span class="o">.</span><span class="n">start</span><span class="o">.</span><span class="n">dtype</span><span class="o">.</span><span class="n">reference_genome</span><span class="p">,</span>
            <span class="n">includes_end</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="compute_an_and_qual_hists_per_ref_site"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/compute_coverage.html#gnomad_qc.v4.annotations.compute_coverage.compute_an_and_qual_hists_per_ref_site">[docs]</a><span class="k">def</span> <span class="nf">compute_an_and_qual_hists_per_ref_site</span><span class="p">(</span>
    <span class="n">vds</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">VariantDataset</span><span class="p">,</span>
    <span class="n">ref_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">interval_ht</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">group_membership_ht</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Compute allele number and quality histograms per reference site.</span>

<span class="sd">    :param vds: Input VDS.</span>
<span class="sd">    :param ref_ht: Reference HT.</span>
<span class="sd">    :param interval_ht: Interval HT.</span>
<span class="sd">    :param group_membership_ht: Group membership HT.</span>
<span class="sd">    :return: HT with allele number and quality histograms per reference site.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">_get_hists</span><span class="p">(</span><span class="n">qual_expr</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">Expression</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">qual_hist_expr</span><span class="p">(</span>
            <span class="n">gq_expr</span><span class="o">=</span><span class="n">qual_expr</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
            <span class="n">dp_expr</span><span class="o">=</span><span class="n">qual_expr</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
            <span class="n">adj_expr</span><span class="o">=</span><span class="n">qual_expr</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">,</span>
            <span class="n">split_adj_and_raw</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="n">entry_agg_funcs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;AN&quot;</span><span class="p">:</span> <span class="n">get_allele_number_agg_func</span><span class="p">(</span><span class="s2">&quot;LGT&quot;</span><span class="p">),</span>
        <span class="s2">&quot;qual_hists&quot;</span><span class="p">:</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">t</span><span class="p">:</span> <span class="p">[</span><span class="n">t</span><span class="o">.</span><span class="n">GQ</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DP</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">adj</span><span class="p">],</span> <span class="n">_get_hists</span><span class="p">),</span>
    <span class="p">}</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Computing allele number and histograms per reference site...&quot;</span><span class="p">)</span>
    <span class="c1"># Below we use just the raw group for qual hist computations because qual hists</span>
    <span class="c1"># has its own built-in adj filtering when adj is passed as an argument and will</span>
    <span class="c1"># produce both adj and raw histograms.</span>

    <span class="n">vmt</span> <span class="o">=</span> <span class="n">vds</span><span class="o">.</span><span class="n">variant_data</span>
    <span class="n">vmt</span> <span class="o">=</span> <span class="n">vmt</span><span class="o">.</span><span class="n">annotate_cols</span><span class="p">(</span><span class="n">sex_karyotype</span><span class="o">=</span><span class="n">vmt</span><span class="o">.</span><span class="n">meta</span><span class="o">.</span><span class="n">sex_imputation</span><span class="o">.</span><span class="n">sex_karyotype</span><span class="p">)</span>
    <span class="n">vds</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">VariantDataset</span><span class="p">(</span><span class="n">vds</span><span class="o">.</span><span class="n">reference_data</span><span class="p">,</span> <span class="n">vmt</span><span class="p">)</span>

    <span class="n">ht</span> <span class="o">=</span> <span class="n">compute_stats_per_ref_site</span><span class="p">(</span>
        <span class="n">vds</span><span class="p">,</span>
        <span class="n">ref_ht</span><span class="p">,</span>
        <span class="n">entry_agg_funcs</span><span class="p">,</span>
        <span class="n">interval_ht</span><span class="o">=</span><span class="n">interval_ht</span><span class="p">,</span>
        <span class="n">group_membership_ht</span><span class="o">=</span><span class="n">group_membership_ht</span><span class="p">,</span>
        <span class="n">entry_keep_fields</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;GQ&quot;</span><span class="p">,</span> <span class="s2">&quot;DP&quot;</span><span class="p">],</span>
        <span class="n">entry_agg_group_membership</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;qual_hists&quot;</span><span class="p">:</span> <span class="p">[{</span><span class="s2">&quot;group&quot;</span><span class="p">:</span> <span class="s2">&quot;raw&quot;</span><span class="p">}]},</span>
        <span class="n">sex_karyotype_field</span><span class="o">=</span><span class="s2">&quot;sex_karyotype&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">qual_hists</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">qual_hists</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

    <span class="k">return</span> <span class="n">ht</span></div>


<div class="viewcode-block" id="add_exomes_interval_annotations"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/compute_coverage.html#gnomad_qc.v4.annotations.compute_coverage.add_exomes_interval_annotations">[docs]</a><span class="k">def</span> <span class="nf">add_exomes_interval_annotations</span><span class="p">(</span><span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">padding</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">150</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Add interval annotations to exomes HT.</span>

<span class="sd">    :param ht: Input HT.</span>
<span class="sd">    :param padding: Number of base pair padding to use for the calling intervals.</span>
<span class="sd">    :return: HT with exome interval annotations.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Add capture and calling intervals as region flags.</span>
    <span class="n">intervals</span> <span class="o">=</span> <span class="p">[(</span><span class="n">c</span><span class="p">,</span> <span class="n">calling_intervals</span><span class="p">(</span><span class="n">c</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">ht</span><span class="p">())</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;broad&quot;</span><span class="p">,</span> <span class="s2">&quot;ukb&quot;</span><span class="p">]]</span>
    <span class="n">intervals</span> <span class="o">=</span> <span class="p">[(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">c</span><span class="si">}</span><span class="s2">_capture&quot;</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">intervals</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span>
        <span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">c</span><span class="si">}</span><span class="s2">_calling&quot;</span><span class="p">,</span> <span class="n">adjust_interval_padding</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">padding</span><span class="p">))</span> <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">intervals</span>
    <span class="p">]</span>

    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="o">**</span><span class="p">{</span><span class="sa">f</span><span class="s2">&quot;outside_</span><span class="si">{</span><span class="n">c</span><span class="si">}</span><span class="s2">_region&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">is_missing</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="n">ht</span><span class="o">.</span><span class="n">locus</span><span class="p">])</span> <span class="k">for</span> <span class="n">c</span><span class="p">,</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">intervals</span><span class="p">}</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">ht</span></div>


<div class="viewcode-block" id="get_coverage_resources"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/compute_coverage.html#gnomad_qc.v4.annotations.compute_coverage.get_coverage_resources">[docs]</a><span class="k">def</span> <span class="nf">get_coverage_resources</span><span class="p">(</span>
    <span class="n">test</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
    <span class="n">overwrite</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
    <span class="n">data_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">calling_interval_name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">calling_interval_padding</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">PipelineResourceCollection</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get PipelineResourceCollection for all resources needed in the coverage pipeline.</span>

<span class="sd">    :param test: Whether to gather all resources for testing.</span>
<span class="sd">    :param overwrite: Whether to overwrite resources if they exist.</span>
<span class="sd">    :param data_type: One of &#39;exomes&#39; or &#39;genomes&#39;.</span>
<span class="sd">    :param calling_interval_name: Name of calling intervals to use.</span>
<span class="sd">    :param calling_interval_padding: Padding to use for calling intervals.</span>
<span class="sd">    :return: PipelineResourceCollection containing resources for all steps of the</span>
<span class="sd">        coverage pipeline.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Create resource collection for each step of the coverage pipeline.</span>
    <span class="k">if</span> <span class="n">calling_interval_name</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">calling_interval_padding</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">coverage_input_resources</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;interval list&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;interval_ht&quot;</span><span class="p">:</span> <span class="n">calling_intervals</span><span class="p">(</span>
                    <span class="n">interval_name</span><span class="o">=</span><span class="n">calling_interval_name</span><span class="p">,</span>
                    <span class="n">calling_interval_padding</span><span class="o">=</span><span class="n">calling_interval_padding</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">},</span>
        <span class="p">}</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">coverage_input_resources</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="c1"># Initialize coverage pipeline resource collection.</span>
    <span class="n">coverage_pipeline</span> <span class="o">=</span> <span class="n">PipelineResourceCollection</span><span class="p">(</span>
        <span class="n">pipeline_name</span><span class="o">=</span><span class="s2">&quot;coverage&quot;</span><span class="p">,</span>
        <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">,</span>
        <span class="n">pipeline_resources</span><span class="o">=</span><span class="n">coverage_input_resources</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">compute_coverage_ht</span> <span class="o">=</span> <span class="n">PipelineStepResourceCollection</span><span class="p">(</span>
        <span class="s2">&quot;--compute-coverage-ht&quot;</span><span class="p">,</span>
        <span class="n">input_resources</span><span class="o">=</span><span class="n">coverage_input_resources</span><span class="p">,</span>
        <span class="n">output_resources</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;coverage_ht&quot;</span><span class="p">:</span> <span class="n">release_coverage</span><span class="p">(</span>
                <span class="n">data_type</span><span class="o">=</span><span class="n">data_type</span><span class="p">,</span> <span class="n">public</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">test</span><span class="o">=</span><span class="n">test</span>
            <span class="p">)</span>
        <span class="p">},</span>
    <span class="p">)</span>
    <span class="n">downsampling_resources</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;downsampling Tables&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;ds_ht&quot;</span><span class="p">:</span> <span class="n">get_downsampling</span><span class="p">(),</span>
            <span class="s2">&quot;non_ukb_ds_ht&quot;</span><span class="p">:</span> <span class="n">get_downsampling</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="s2">&quot;non_ukb&quot;</span><span class="p">),</span>
        <span class="p">}</span>
    <span class="p">}</span>
    <span class="n">compute_allele_number_ht</span> <span class="o">=</span> <span class="n">PipelineStepResourceCollection</span><span class="p">(</span>
        <span class="s2">&quot;--compute-all-sites-an-and-qual-hist-ht&quot;</span><span class="p">,</span>
        <span class="n">input_resources</span><span class="o">=</span><span class="p">{</span>
            <span class="o">**</span><span class="n">coverage_input_resources</span><span class="p">,</span>
            <span class="o">**</span><span class="p">{</span>
                <span class="s2">&quot;metadata Table&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;meta_ht&quot;</span><span class="p">:</span> <span class="n">meta</span><span class="p">(</span><span class="n">data_type</span><span class="o">=</span><span class="n">data_type</span><span class="p">)},</span>
                <span class="o">**</span><span class="n">downsampling_resources</span><span class="p">,</span>
            <span class="p">},</span>
        <span class="p">},</span>
        <span class="n">output_resources</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;an_qual_hist_ht&quot;</span><span class="p">:</span> <span class="n">get_all_sites_an_and_qual_hists</span><span class="p">(</span>
                <span class="n">data_type</span><span class="o">=</span><span class="n">data_type</span><span class="p">,</span> <span class="n">test</span><span class="o">=</span><span class="n">test</span>
            <span class="p">)</span>
        <span class="p">},</span>
    <span class="p">)</span>
    <span class="n">export_coverage_files</span> <span class="o">=</span> <span class="n">PipelineStepResourceCollection</span><span class="p">(</span>
        <span class="s2">&quot;--export-coverage-release-files&quot;</span><span class="p">,</span>
        <span class="n">output_resources</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;coverage_tsv&quot;</span><span class="p">:</span> <span class="n">release_coverage_tsv_path</span><span class="p">(</span><span class="n">data_type</span><span class="p">,</span> <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">),</span>
            <span class="s2">&quot;release_ht&quot;</span><span class="p">:</span> <span class="n">release_coverage</span><span class="p">(</span>
                <span class="n">data_type</span><span class="o">=</span><span class="n">data_type</span><span class="p">,</span> <span class="n">public</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">,</span> <span class="n">stratify</span><span class="o">=</span><span class="kc">False</span>
            <span class="p">),</span>
        <span class="p">},</span>
        <span class="n">pipeline_input_steps</span><span class="o">=</span><span class="p">[</span><span class="n">compute_coverage_ht</span><span class="p">],</span>
    <span class="p">)</span>
    <span class="n">export_an_files</span> <span class="o">=</span> <span class="n">PipelineStepResourceCollection</span><span class="p">(</span>
        <span class="s2">&quot;--export-all-sites-an-release-files&quot;</span><span class="p">,</span>
        <span class="n">output_resources</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;allele_number_ht&quot;</span><span class="p">:</span> <span class="n">release_all_sites_an</span><span class="p">(</span>
                <span class="n">data_type</span><span class="o">=</span><span class="n">data_type</span><span class="p">,</span>
                <span class="n">public</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">,</span>
            <span class="p">),</span>
            <span class="s2">&quot;allele_number_tsv&quot;</span><span class="p">:</span> <span class="n">release_all_sites_an_tsv_path</span><span class="p">(</span><span class="n">data_type</span><span class="p">,</span> <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">),</span>
        <span class="p">},</span>
        <span class="n">pipeline_input_steps</span><span class="o">=</span><span class="p">[</span><span class="n">compute_allele_number_ht</span><span class="p">],</span>
    <span class="p">)</span>

    <span class="c1"># Add all steps to the coverage pipeline resource collection.</span>
    <span class="n">coverage_pipeline</span><span class="o">.</span><span class="n">add_steps</span><span class="p">(</span>
        <span class="p">{</span>
            <span class="s2">&quot;compute_coverage_ht&quot;</span><span class="p">:</span> <span class="n">compute_coverage_ht</span><span class="p">,</span>
            <span class="s2">&quot;compute_allele_number_ht&quot;</span><span class="p">:</span> <span class="n">compute_allele_number_ht</span><span class="p">,</span>
            <span class="s2">&quot;export_coverage_files&quot;</span><span class="p">:</span> <span class="n">export_coverage_files</span><span class="p">,</span>
            <span class="s2">&quot;export_all_sites_an_release_files&quot;</span><span class="p">:</span> <span class="n">export_an_files</span><span class="p">,</span>
        <span class="p">}</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">coverage_pipeline</span></div>


<div class="viewcode-block" id="main"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/compute_coverage.html#gnomad_qc.v4.annotations.compute_coverage.main">[docs]</a><span class="k">def</span> <span class="nf">main</span><span class="p">(</span><span class="n">args</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compute coverage statistics, including mean, median_approx, and coverage over certain DPs.&quot;&quot;&quot;</span>
    <span class="n">hl</span><span class="o">.</span><span class="n">init</span><span class="p">(</span>
        <span class="n">log</span><span class="o">=</span><span class="s2">&quot;/coverage.log&quot;</span><span class="p">,</span>
        <span class="n">default_reference</span><span class="o">=</span><span class="s2">&quot;GRCh38&quot;</span><span class="p">,</span>
        <span class="n">tmp_dir</span><span class="o">=</span><span class="s2">&quot;gs://gnomad-tmp-30day&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># SSA Logs are easier to troubleshoot with.</span>
    <span class="n">hl</span><span class="o">.</span><span class="n">_set_flags</span><span class="p">(</span><span class="n">use_ssa_logs</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">)</span>

    <span class="n">test_2_partitions</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">test_2_partitions</span>
    <span class="n">test_chr22_chrx_chry</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">test_chr22_chrx_chry</span>
    <span class="n">test</span> <span class="o">=</span> <span class="n">test_2_partitions</span> <span class="ow">or</span> <span class="n">test_chr22_chrx_chry</span>
    <span class="n">overwrite</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">overwrite</span>
    <span class="n">data_type</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">data_type</span>
    <span class="n">n_partitions</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">n_partitions</span>

    <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;exomes&quot;</span><span class="p">:</span>
        <span class="n">calling_interval_padding</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">calling_interval_padding</span>
        <span class="n">calling_interval_name</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">calling_interval_name</span>
        <span class="n">adjust_padding</span> <span class="o">=</span> <span class="n">calling_interval_padding</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">50</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">calling_interval_padding</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">calling_interval_name</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">adjust_padding</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="n">coverage_resources</span> <span class="o">=</span> <span class="n">get_coverage_resources</span><span class="p">(</span>
        <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">,</span>
        <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">,</span>
        <span class="n">data_type</span><span class="o">=</span><span class="n">data_type</span><span class="p">,</span>
        <span class="n">calling_interval_name</span><span class="o">=</span><span class="n">calling_interval_name</span><span class="p">,</span>
        <span class="c1"># We only have interval lists with 0 or 50bp padding, so if a different padding</span>
        <span class="c1"># is requested, use 0bp and adjust it based on the desired padding.</span>
        <span class="n">calling_interval_padding</span><span class="o">=</span><span class="mi">0</span> <span class="k">if</span> <span class="n">adjust_padding</span> <span class="k">else</span> <span class="n">calling_interval_padding</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">compute_coverage_ht</span> <span class="ow">or</span> <span class="n">args</span><span class="o">.</span><span class="n">compute_all_sites_an_and_qual_hist_ht</span><span class="p">:</span>
            <span class="c1"># Read in context Table.</span>
            <span class="n">ref_ht</span> <span class="o">=</span> <span class="n">vep_context</span><span class="o">.</span><span class="n">versions</span><span class="p">[</span><span class="s2">&quot;105&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">test_chr22_chrx_chry</span><span class="p">:</span>
                <span class="n">chrom</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;chr22&quot;</span><span class="p">,</span> <span class="s2">&quot;chrX&quot;</span><span class="p">,</span> <span class="s2">&quot;chrY&quot;</span><span class="p">]</span>
                <span class="n">ref_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">filter_intervals</span><span class="p">(</span>
                    <span class="n">ref_ht</span><span class="p">,</span> <span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">parse_locus_interval</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">chrom</span><span class="p">]</span>
                <span class="p">)</span>
            <span class="k">elif</span> <span class="n">test_2_partitions</span><span class="p">:</span>
                <span class="n">ref_ht</span> <span class="o">=</span> <span class="n">ref_ht</span><span class="o">.</span><span class="n">_filter_partitions</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">5</span><span class="p">))</span>

            <span class="c1"># Retain only &#39;locus&#39; annotation from context Table.</span>
            <span class="n">ref_ht</span> <span class="o">=</span> <span class="n">ref_ht</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">select</span><span class="p">()</span><span class="o">.</span><span class="n">distinct</span><span class="p">()</span>

            <span class="c1"># Read in VDS.</span>
            <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;exomes&quot;</span><span class="p">:</span>
                <span class="n">vds</span> <span class="o">=</span> <span class="n">get_gnomad_v4_vds</span><span class="p">(</span>
                    <span class="n">release_only</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                    <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">,</span>
                    <span class="n">filter_partitions</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span> <span class="k">if</span> <span class="n">test_2_partitions</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
                    <span class="n">annotate_meta</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                    <span class="n">chrom</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chr22&quot;</span><span class="p">,</span> <span class="s2">&quot;chrX&quot;</span><span class="p">,</span> <span class="s2">&quot;chrY&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">test_chr22_chrx_chry</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="n">interval_ht</span> <span class="o">=</span> <span class="n">coverage_resources</span><span class="o">.</span><span class="n">interval_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>
                <span class="k">if</span> <span class="n">adjust_padding</span><span class="p">:</span>
                    <span class="n">interval_ht</span> <span class="o">=</span> <span class="n">adjust_interval_padding</span><span class="p">(</span>
                        <span class="n">interval_ht</span><span class="p">,</span> <span class="n">calling_interval_padding</span>
                    <span class="p">)</span>

            <span class="k">else</span><span class="p">:</span>
                <span class="n">vds</span> <span class="o">=</span> <span class="n">get_gnomad_v4_genomes_vds</span><span class="p">(</span>
                    <span class="n">release_only</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                    <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">,</span>
                    <span class="n">filter_partitions</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span> <span class="k">if</span> <span class="n">test_2_partitions</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
                    <span class="n">annotate_meta</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                    <span class="n">chrom</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chr22&quot;</span><span class="p">,</span> <span class="s2">&quot;chrX&quot;</span><span class="p">,</span> <span class="s2">&quot;chrY&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">test_chr22_chrx_chry</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="n">interval_ht</span> <span class="o">=</span> <span class="kc">None</span>

                <span class="c1"># For genomes, remove the centromeres and telomeres from the reference</span>
                <span class="c1"># HT if they exist.</span>
                <span class="n">ref_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">filter_intervals</span><span class="p">(</span>
                    <span class="n">ref_ht</span><span class="p">,</span>
                    <span class="n">telomeres_and_centromeres</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span><span class="o">.</span><span class="n">interval</span><span class="o">.</span><span class="n">collect</span><span class="p">(),</span>
                    <span class="n">keep</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="n">ref_ht</span> <span class="o">=</span> <span class="n">ref_ht</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">new_temp_file</span><span class="p">(</span><span class="s2">&quot;ref&quot;</span><span class="p">,</span> <span class="s2">&quot;ht&quot;</span><span class="p">))</span>

        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">compute_coverage_ht</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running compute coverage...&quot;</span><span class="p">)</span>
            <span class="n">res</span> <span class="o">=</span> <span class="n">coverage_resources</span><span class="o">.</span><span class="n">compute_coverage_ht</span>
            <span class="n">res</span><span class="o">.</span><span class="n">check_resource_existence</span><span class="p">()</span>

            <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">stratify_by_ukb_and_platform</span><span class="p">:</span>
                <span class="n">meta_expr</span> <span class="o">=</span> <span class="n">vds</span><span class="o">.</span><span class="n">variant_data</span><span class="o">.</span><span class="n">meta</span>
                <span class="n">ukb_expr</span> <span class="o">=</span> <span class="n">meta_expr</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">ukb_sample</span>
                <span class="n">strata</span> <span class="o">=</span> <span class="p">[</span>
                    <span class="p">{</span><span class="s2">&quot;ukb_strata&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span><span class="n">ukb_expr</span><span class="p">,</span> <span class="s2">&quot;ukb&quot;</span><span class="p">,</span> <span class="s2">&quot;non_ukb&quot;</span><span class="p">)},</span>
                    <span class="p">{</span>
                        <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">or_missing</span><span class="p">(</span><span class="o">~</span><span class="n">ukb_expr</span><span class="p">,</span> <span class="s2">&quot;non_ukb&quot;</span><span class="p">),</span>
                        <span class="s2">&quot;platform&quot;</span><span class="p">:</span> <span class="n">meta_expr</span><span class="o">.</span><span class="n">platform_inference</span><span class="o">.</span><span class="n">qc_platform</span><span class="p">,</span>
                    <span class="p">},</span>
                <span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">strata</span> <span class="o">=</span> <span class="kc">None</span>

            <span class="c1"># Compute coverage stats.</span>
            <span class="n">coverage_ht</span> <span class="o">=</span> <span class="n">compute_coverage_stats</span><span class="p">(</span>
                <span class="n">vds</span><span class="p">,</span>
                <span class="n">ref_ht</span><span class="p">,</span>
                <span class="n">interval_ht</span><span class="o">=</span><span class="n">interval_ht</span><span class="p">,</span>
                <span class="n">strata_expr</span><span class="o">=</span><span class="n">strata</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="c1"># Checkpoint Table.</span>
            <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">stratify_by_ukb_and_platform</span><span class="p">:</span>
                <span class="n">coverage_ht</span> <span class="o">=</span> <span class="n">coverage_ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span>
                    <span class="n">coverage_stats_meta</span><span class="o">=</span><span class="n">coverage_ht</span><span class="o">.</span><span class="n">coverage_stats_meta</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
                        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">dict</span><span class="p">(</span>
                            <span class="n">x</span><span class="o">.</span><span class="n">items</span><span class="p">()</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
                                <span class="k">lambda</span> <span class="n">m</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span>
                                    <span class="n">m</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;ukb_strata&quot;</span><span class="p">,</span> <span class="p">(</span><span class="s2">&quot;subset&quot;</span><span class="p">,</span> <span class="n">m</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span> <span class="n">m</span>
                                <span class="p">)</span>
                            <span class="p">)</span>
                        <span class="p">)</span>
                    <span class="p">)</span>
                <span class="p">)</span>
            <span class="n">coverage_ht</span> <span class="o">=</span> <span class="n">coverage_ht</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span>
                <span class="n">hl</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">new_temp_file</span><span class="p">(</span><span class="s2">&quot;coverage&quot;</span><span class="p">,</span> <span class="s2">&quot;ht&quot;</span><span class="p">)</span>
            <span class="p">)</span>

            <span class="c1"># Naive coalesce and write out final Table.</span>
            <span class="n">coverage_ht</span> <span class="o">=</span> <span class="n">coverage_ht</span><span class="o">.</span><span class="n">naive_coalesce</span><span class="p">(</span><span class="n">n_partitions</span><span class="p">)</span>
            <span class="n">coverage_ht</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">coverage_ht</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">compute_all_sites_an_and_qual_hist_ht</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="s2">&quot;Running compute all sites allele number and quality histograms HT...&quot;</span>
            <span class="p">)</span>
            <span class="n">res</span> <span class="o">=</span> <span class="n">coverage_resources</span><span class="o">.</span><span class="n">compute_allele_number_ht</span>
            <span class="n">res</span><span class="o">.</span><span class="n">check_resource_existence</span><span class="p">()</span>

            <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;exomes&quot;</span><span class="p">:</span>
                <span class="n">group_membership_ht</span> <span class="o">=</span> <span class="n">get_exomes_group_membership_ht</span><span class="p">(</span>
                    <span class="n">res</span><span class="o">.</span><span class="n">meta_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">(),</span>
                    <span class="n">res</span><span class="o">.</span><span class="n">ds_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">(),</span>
                    <span class="n">res</span><span class="o">.</span><span class="n">non_ukb_ds_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">(),</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">group_membership_ht</span> <span class="o">=</span> <span class="n">get_genomes_group_membership_ht</span><span class="p">(</span>
                    <span class="n">res</span><span class="o">.</span><span class="n">meta_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">(),</span>
                <span class="p">)</span>
            <span class="n">group_membership_ht</span> <span class="o">=</span> <span class="n">group_membership_ht</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span>
                <span class="n">hl</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">new_temp_file</span><span class="p">(</span><span class="s2">&quot;group_membership&quot;</span><span class="p">,</span> <span class="s2">&quot;ht&quot;</span><span class="p">)</span>
            <span class="p">)</span>

            <span class="n">an_ht</span> <span class="o">=</span> <span class="n">compute_an_and_qual_hists_per_ref_site</span><span class="p">(</span>
                <span class="n">vds</span><span class="p">,</span>
                <span class="n">ref_ht</span><span class="p">,</span>
                <span class="n">interval_ht</span><span class="o">=</span><span class="n">interval_ht</span><span class="p">,</span>
                <span class="n">group_membership_ht</span><span class="o">=</span><span class="n">group_membership_ht</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">an_ht</span> <span class="o">=</span> <span class="n">an_ht</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">new_temp_file</span><span class="p">(</span><span class="s2">&quot;an&quot;</span><span class="p">,</span> <span class="s2">&quot;ht&quot;</span><span class="p">))</span>
            <span class="c1"># Naive coalesce and write out the intermediate HT.</span>
            <span class="n">an_ht</span> <span class="o">=</span> <span class="n">an_ht</span><span class="o">.</span><span class="n">naive_coalesce</span><span class="p">(</span><span class="n">n_partitions</span><span class="p">)</span>
            <span class="n">an_ht</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">an_qual_hist_ht</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">export_coverage_release_files</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Exporting coverage tsv...&quot;</span><span class="p">)</span>
            <span class="n">res</span> <span class="o">=</span> <span class="n">coverage_resources</span><span class="o">.</span><span class="n">export_coverage_files</span>
            <span class="n">res</span><span class="o">.</span><span class="n">check_resource_existence</span><span class="p">()</span>
            <span class="n">ht</span> <span class="o">=</span> <span class="n">res</span><span class="o">.</span><span class="n">coverage_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>
            <span class="k">if</span> <span class="s2">&quot;coverage_stats&quot;</span> <span class="ow">in</span> <span class="n">ht</span><span class="o">.</span><span class="n">row</span><span class="p">:</span>
                <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
                    <span class="o">**</span><span class="p">{</span><span class="n">k</span><span class="p">:</span> <span class="n">ht</span><span class="o">.</span><span class="n">coverage_stats</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="n">k</span><span class="p">]</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">ht</span><span class="o">.</span><span class="n">coverage_stats</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span>
                <span class="p">)</span>
            <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;coverage_stats_meta&quot;</span><span class="p">,</span> <span class="s2">&quot;coverage_stats_meta_sample_count&quot;</span><span class="p">)</span>
            <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">release_ht</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">)</span>
            <span class="n">ht</span><span class="o">.</span><span class="n">export</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">coverage_tsv</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">export_all_sites_an_release_files</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Exporting all sites AN HT...&quot;</span><span class="p">)</span>
            <span class="n">res</span> <span class="o">=</span> <span class="n">coverage_resources</span><span class="o">.</span><span class="n">export_all_sites_an_release_files</span>
            <span class="n">res</span><span class="o">.</span><span class="n">check_resource_existence</span><span class="p">()</span>

            <span class="n">ht</span> <span class="o">=</span> <span class="n">res</span><span class="o">.</span><span class="n">an_qual_hist_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>
            <span class="c1"># Select only the AN and write out final Table.</span>
            <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;AN&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;exomes&quot;</span><span class="p">:</span>
                <span class="n">ht</span> <span class="o">=</span> <span class="n">add_exomes_interval_annotations</span><span class="p">(</span><span class="n">ht</span><span class="p">,</span> <span class="n">calling_interval_padding</span><span class="p">)</span>

            <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">allele_number_ht</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">)</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Exporting all sites AN tsv...&quot;</span><span class="p">)</span>
            <span class="c1"># Only export the adj AN for all release samples.</span>
            <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">AN</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">AN</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">ht</span><span class="o">.</span><span class="n">export</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">allele_number_tsv</span><span class="p">)</span>

    <span class="k">finally</span><span class="p">:</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">copy_log</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;gs://gnomad-tmp-4day/coverage/compute_coverage.log&quot;</span><span class="p">)</span></div>


<span class="k">def</span> <span class="nf">get_script_argument_parser</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get script argument parser.&quot;&quot;&quot;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">()</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--overwrite&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Overwrite existing hail Tables.&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--test-2-partitions&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span>
            <span class="s2">&quot;Whether to run a test using only the first 2 partitions of the VDS test&quot;</span>
            <span class="s2">&quot; dataset.&quot;</span>
        <span class="p">),</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--test-chr22-chrx-chry&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span>
            <span class="s2">&quot;Whether to run a test using only the chr22, chrX, and chrY chromosomes of&quot;</span>
            <span class="s2">&quot; the VDS test dataset.&quot;</span>
        <span class="p">),</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--n-partitions&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Number of partitions to use for the output Table.&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mi">5000</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--data-type&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span>
            <span class="s2">&quot;Data type to compute coverage or AN and quality hist all sites Table &quot;</span>
            <span class="s2">&quot;for. One of &#39;exomes&#39; or &#39;genomes&#39;.&quot;</span>
        <span class="p">),</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;exomes&quot;</span><span class="p">,</span> <span class="s2">&quot;genomes&quot;</span><span class="p">],</span>
        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;exomes&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--compute-coverage-ht&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Compute the coverage HT.&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--compute-all-sites-an-and-qual-hist-ht&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Compute the all sites allele number and quality histogram HT.&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--stratify-by-ukb-and-platform&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span>
            <span class="s2">&quot;Whether to compute coverage stratified by UKB/non-UKB and platform. Only &quot;</span>
            <span class="s2">&quot;applicable if --data-type is exomes.&quot;</span>
        <span class="p">),</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--calling-interval-name&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span>
            <span class="s2">&quot;Name of calling intervals to use for interval coverage. One of: &#39;ukb&#39;,&quot;</span>
            <span class="s2">&quot; &#39;broad&#39;, &#39;intersection&#39;, or &#39;union&#39;. Only applicable if --data-type is &quot;</span>
            <span class="s2">&quot;exomes.&quot;</span>
        <span class="p">),</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;ukb&quot;</span><span class="p">,</span> <span class="s2">&quot;broad&quot;</span><span class="p">,</span> <span class="s2">&quot;intersection&quot;</span><span class="p">,</span> <span class="s2">&quot;union&quot;</span><span class="p">],</span>
        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;union&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--calling-interval-padding&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span>
            <span class="s2">&quot;Number of base pair padding to use on the calling intervals. One of 0, &quot;</span>
            <span class="s2">&quot; 50, or 150 bp.  Only applicable if --data-type is exomes.&quot;</span>
        <span class="p">),</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
        <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span> <span class="mi">150</span><span class="p">],</span>
        <span class="n">default</span><span class="o">=</span><span class="mi">150</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--export-coverage-release-files&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Exports coverage release HT and TSV file.&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--export-all-sites-an-release-files&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Export all sites AN release HT and TSV file.&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">parser</span>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">get_script_argument_parser</span><span class="p">()</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">stratify_by_ukb_and_platform</span> <span class="ow">and</span> <span class="n">args</span><span class="o">.</span><span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;genomes&quot;</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;Stratifying by UKB/non-UKB and platform is only applicable for exomes.&quot;</span>
        <span class="p">)</span>

    <span class="n">main</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p></p>
  </div>

   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>