<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>gnomad_qc.v4.annotations.fix_freq_an &mdash; gnomad_qc master documentation</title>
      <link rel="stylesheet" href="../../../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../../../_static/theme_overrides.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../../_static/jquery.js"></script>
        <script src="../../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../../../../" id="documentation_options" src="../../../../_static/documentation_options.js"></script>
        <script src="../../../../_static/doctools.js"></script>
        <script src="../../../../_static/sphinx_highlight.js"></script>
        <script src="https://unpkg.com/mermaid@10.2.0/dist/mermaid.min.js"></script>
        <script>mermaid.initialize({startOnLoad:true});</script>
    <script src="../../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../../index.html" class="icon icon-home">
            gnomad_qc
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../api_reference/index.html">API Reference</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/broadinstitute/gnomad_qc/tags">Change Log</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../../index.html">gnomad_qc</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">gnomad_qc.v4.annotations.fix_freq_an</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for gnomad_qc.v4.annotations.fix_freq_an</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Updates the v4.0 freq HT with the correct AN for the v4.1 release.</span>

<span class="sd">There is an error in the v4.0 freq HT where the AN can be incorrect for variants that</span>
<span class="sd">are only present in one of the UKB or non-UKB subsets. This is because the frequency</span>
<span class="sd">data was computed on each subset separately and then combined. In order to filter the</span>
<span class="sd">dataset to only samples that are in the UKB or non-UKB subsets, the</span>
<span class="sd">`hail.vds.filter_samples` was used. Unfortunately, the behavior of this function was</span>
<span class="sd">not the expected behavior due to the following line of code:</span>

<span class="sd">    .. code-block::</span>

<span class="sd">        variant_data = variant_data.filter_rows(hl.agg.count() &gt; 0)</span>

<span class="sd">This line of code filters any row that has no genotypes in the sample subset. When the</span>
<span class="sd">VariantDataset is densified prior to the frequency calculations this results in no</span>
<span class="sd">reference data being filled in for that row, unless a row is part of multiallelic site that is not</span>
<span class="sd">exclusive to the sample subset, and therefore a missing AN for that row. When combining</span>
<span class="sd">the frequency data for the UKB and non-UKB subsets, the AN for these rows will</span>
<span class="sd">only have the AN for the subset that has non-ref genotypes for that row.</span>

<span class="sd">This script performs the following steps:</span>

<span class="sd">    - Generates allele number and GQ/DP histograms for all sites in the exome calling</span>
<span class="sd">      interval.</span>

<span class="sd">        - For raw genotypes, the AN is just the sum of the ploidy (more specifics on</span>
<span class="sd">          this below) for all genotypes.</span>
<span class="sd">        - For adj genotypes, the AN is the sum of the ploidy for all genotypes that pass</span>
<span class="sd">          the GQ, DP, and allele balance adj filters. Since this is for a site, rather</span>
<span class="sd">          than a specific variant, this number will not always be the same as the AN for</span>
<span class="sd">          a specific variant at that site. This AN can be smaller if it is a</span>
<span class="sd">          multi-allelic site where there are non-ref genotypes for another variant that</span>
<span class="sd">          fails the adj filters.</span>
<span class="sd">        - For raw genotypes on non-PAR regions of the X or Y chromosome, the AN is the</span>
<span class="sd">          sum of the ploidy after adjusting for sex karyotype, BUT NOT taking into</span>
<span class="sd">          account the genotype when adjusting the ploidy. For instance, a het genotype</span>
<span class="sd">          for an XY sample will still have a ploidy of 1, even though the genotype is a</span>
<span class="sd">          het and is therefore set to missing for the frequency calculations.</span>

<span class="sd">    - Generates allele number and GQ/DP histograms for the frequency correction.</span>

<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">logging</span>

<span class="kn">import</span> <span class="nn">hail</span> <span class="k">as</span> <span class="nn">hl</span>
<span class="kn">from</span> <span class="nn">gnomad.resources.grch38.reference_data</span> <span class="kn">import</span> <span class="n">vep_context</span>
<span class="kn">from</span> <span class="nn">gnomad.sample_qc.sex</span> <span class="kn">import</span> <span class="n">adjusted_sex_ploidy_expr</span>
<span class="kn">from</span> <span class="nn">gnomad.utils.annotations</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">agg_by_strata</span><span class="p">,</span>
    <span class="n">annotate_and_index_source_mt_for_sex_ploidy</span><span class="p">,</span>
    <span class="n">get_gq_dp_adj_expr</span><span class="p">,</span>
    <span class="n">get_het_ab_adj_expr</span><span class="p">,</span>
    <span class="n">get_is_haploid_expr</span><span class="p">,</span>
    <span class="n">merge_freq_arrays</span><span class="p">,</span>
    <span class="n">merge_histograms</span><span class="p">,</span>
    <span class="n">qual_hist_expr</span><span class="p">,</span>
    <span class="n">set_xx_y_metrics_to_na_expr</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">gnomad.utils.filtering</span> <span class="kn">import</span> <span class="n">filter_arrays_by_meta</span>
<span class="kn">from</span> <span class="nn">gnomad.utils.sparse_mt</span> <span class="kn">import</span> <span class="n">compute_stats_per_ref_site</span>

<span class="kn">from</span> <span class="nn">gnomad_qc.v4.annotations.compute_coverage</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">adjust_interval_padding</span><span class="p">,</span>
    <span class="n">get_exomes_group_membership_ht</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v4.annotations.generate_freq</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">compute_inbreeding_coeff</span><span class="p">,</span>
    <span class="n">correct_for_high_ab_hets</span><span class="p">,</span>
    <span class="n">create_final_freq_ht</span><span class="p">,</span>
    <span class="n">generate_faf_grpmax</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v4.resources.annotations</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">get_all_sites_an_and_qual_hists</span><span class="p">,</span>
    <span class="n">get_downsampling</span><span class="p">,</span>
    <span class="n">get_freq</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v4.resources.basics</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">calling_intervals</span><span class="p">,</span>
    <span class="n">get_gnomad_v4_vds</span><span class="p">,</span>
    <span class="n">get_logging_path</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v4.resources.meta</span> <span class="kn">import</span> <span class="n">meta</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v4.resources.release</span> <span class="kn">import</span> <span class="n">release_all_sites_an</span>

<span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span>
    <span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">,</span>
    <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%(asctime)s</span><span class="s2">: </span><span class="si">%(message)s</span><span class="s2">&quot;</span><span class="p">,</span>
    <span class="n">datefmt</span><span class="o">=</span><span class="s2">&quot;%m/</span><span class="si">%d</span><span class="s2">/%Y %I:%M:%S %p&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s2">&quot;gnomAD_4.1_exomes_AN_fix&quot;</span><span class="p">)</span>
<span class="n">logger</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>


<div class="viewcode-block" id="prep_vds_for_all_sites_stats"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/fix_freq_an.html#gnomad_qc.v4.annotations.fix_freq_an.prep_vds_for_all_sites_stats">[docs]</a><span class="k">def</span> <span class="nf">prep_vds_for_all_sites_stats</span><span class="p">(</span><span class="n">vds</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">VariantDataset</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">VariantDataset</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Prepare VDS for all sites stats.</span>

<span class="sd">    Adds the following annotations to the VDS:</span>

<span class="sd">        - &#39;ploidy&#39;: The ploidy of the genotype adjusted for sex karyotype.</span>
<span class="sd">        - &#39;adj&#39;: The adj filter for GQ and DP.</span>

<span class="sd">    The following steps are performed on the reference data MT:</span>

<span class="sd">        - Segments the reference data MT at the PAR boundaries on chrX and chrY.</span>

<span class="sd">            - This is done to ensure that entries are annotated with the correct</span>
<span class="sd">              &#39;ploidy&#39; field when adjusting for sex karyotype.</span>
<span class="sd">            - If the END field of an entry spans a PAR boundary, the entry will be</span>
<span class="sd">              split into two separate entries so that a locus &#39;in_non_par&#39; annotation</span>
<span class="sd">              can be used to correctly annotate if the full reference block defined by</span>
<span class="sd">              an entry is non-PAR or not.</span>

<span class="sd">        - Annotates each entry with the ploidy of the reference block adjusting for</span>
<span class="sd">          sex karyotype.</span>

<span class="sd">            - chrY non-PAR XX entries are set to missing ploidy.</span>
<span class="sd">            - chrX and chrY non-PAR XY entries are set to ploidy 1.</span>
<span class="sd">            - All other entries are set to ploidy 2.</span>

<span class="sd">        - Annotates each entry with the adj filter for GQ and DP taking into account</span>
<span class="sd">          sex karyotype for DP.</span>

<span class="sd">    The following steps are performed on the variant data MT:</span>

<span class="sd">        - Annotates each entry with the ploidy of the genotype adjusting for sex</span>
<span class="sd">          karyotype.</span>

<span class="sd">            - chrY non-PAR XX entries are set to missing ploidy.</span>
<span class="sd">            - chrX and chrY non-PAR XY entries are set to ploidy 1 if LGT is defined.</span>

<span class="sd">                - This deviates from our standard ploidy adjustment where if the</span>
<span class="sd">                  genotype is a het, the ploidy is set to missing. This is because we</span>
<span class="sd">                  are adjusting ploidy before splitting multi-allelic sites and this</span>
<span class="sd">                  missing annotation will be propagated to all split entries including</span>
<span class="sd">                  the reference genotypes causing discrepancies in adjusting ploidy</span>
<span class="sd">                  before splitting and after splitting. Since we are using this VDS</span>
<span class="sd">                  downstream to compute a reference AN, we want to keep this as a</span>
<span class="sd">                  ploidy of 1 even if the genotype is a het.</span>
<span class="sd">            - All other entries are set to the ploidy of the genotype.</span>

<span class="sd">        - Annotates each entry with the adj filter for GQ and DP taking into account</span>
<span class="sd">          sex karyotype for DP.</span>

<span class="sd">          - The sex karyotype and locus are used instead of the genotype to determine</span>
<span class="sd">            if the genotype is haploid or not. This is for the same reason as described</span>
<span class="sd">            above for the ploidy adjustment of non-PAR XY het genotypes.</span>

<span class="sd">    :param vds: Hail VDS to annotate adj onto variant data.</span>
<span class="sd">    :return: Hail VDS with adj annotation.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">rmt</span> <span class="o">=</span> <span class="n">vds</span><span class="o">.</span><span class="n">reference_data</span>
    <span class="n">vmt</span> <span class="o">=</span> <span class="n">vds</span><span class="o">.</span><span class="n">variant_data</span>
    <span class="n">vmt</span> <span class="o">=</span> <span class="n">vmt</span><span class="o">.</span><span class="n">annotate_cols</span><span class="p">(</span><span class="n">sex_karyotype</span><span class="o">=</span><span class="n">vmt</span><span class="o">.</span><span class="n">meta</span><span class="o">.</span><span class="n">sex_imputation</span><span class="o">.</span><span class="n">sex_karyotype</span><span class="p">)</span>

    <span class="c1"># In order to correctly annotate the ploidy on the reference data MT, there needs to</span>
    <span class="c1"># be an END field where the start and end position are both in PAR or both in</span>
    <span class="c1"># non-PAR regions. We can make sure this is the case by segmenting the reference</span>
    <span class="c1"># data MT at the PAR boundaries.</span>
    <span class="c1"># This can be done by building an array of intervals that cover the entire reference</span>
    <span class="c1"># genome, then segment the intervals at the PAR boundaries and pass that interval</span>
    <span class="c1"># list to hl.vds.segment_reference_blocks. This will segment the reference data MT</span>
    <span class="c1"># so any sample with an END field that spans a PAR boundary will be split into two</span>
    <span class="c1"># separate entries.</span>
    <span class="n">rg</span> <span class="o">=</span> <span class="n">rmt</span><span class="o">.</span><span class="n">locus</span><span class="o">.</span><span class="n">dtype</span><span class="o">.</span><span class="n">reference_genome</span>
    <span class="n">rg_intervals</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">parse_locus_interval</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;[</span><span class="si">{</span><span class="n">contig</span><span class="si">}</span><span class="s2">:1-</span><span class="si">{</span><span class="n">rg</span><span class="o">.</span><span class="n">lengths</span><span class="p">[</span><span class="n">contig</span><span class="p">]</span><span class="si">}</span><span class="s2">]&quot;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">contig</span> <span class="ow">in</span> <span class="n">rg</span><span class="o">.</span><span class="n">contigs</span><span class="p">[:</span><span class="mi">24</span><span class="p">]</span>
        <span class="p">]</span>
    <span class="p">)</span>
    <span class="n">rg_intervals</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="o">.</span><span class="n">parallelize</span><span class="p">(</span>
        <span class="n">rg_intervals</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span><span class="n">interval</span><span class="o">=</span><span class="n">x</span><span class="p">)),</span>
        <span class="n">schema</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">tstruct</span><span class="p">(</span><span class="n">interval</span><span class="o">=</span><span class="n">rg_intervals</span><span class="o">.</span><span class="n">dtype</span><span class="o">.</span><span class="n">element_type</span><span class="p">),</span>
        <span class="n">key</span><span class="o">=</span><span class="s2">&quot;interval&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">par_boundaries</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">par_interval</span> <span class="ow">in</span> <span class="n">rg</span><span class="o">.</span><span class="n">par</span><span class="p">:</span>
        <span class="n">par_boundaries</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">par_interval</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>
        <span class="n">par_boundaries</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">par_interval</span><span class="o">.</span><span class="n">end</span><span class="p">)</span>

    <span class="n">rg_intervals</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">segment_intervals</span><span class="p">(</span><span class="n">rg_intervals</span><span class="p">,</span> <span class="n">par_boundaries</span><span class="p">)</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">new_temp_file</span><span class="p">(</span><span class="s2">&quot;segment_par_boundaries&quot;</span><span class="p">,</span> <span class="s2">&quot;ht&quot;</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="n">rmt</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">segment_reference_blocks</span><span class="p">(</span><span class="n">rmt</span><span class="p">,</span> <span class="n">rg_intervals</span><span class="p">)</span>

    <span class="c1"># Annotating the cols and rows with annotations used in the filter_entries to</span>
    <span class="c1"># prevent the need to recompute these annotations for every entry in the</span>
    <span class="c1"># filter_entries.</span>
    <span class="n">rmt</span> <span class="o">=</span> <span class="n">rmt</span><span class="o">.</span><span class="n">annotate_cols</span><span class="p">(</span><span class="n">sex_karyotype</span><span class="o">=</span><span class="n">vmt</span><span class="o">.</span><span class="n">cols</span><span class="p">()[</span><span class="n">rmt</span><span class="o">.</span><span class="n">col_key</span><span class="p">]</span><span class="o">.</span><span class="n">sex_karyotype</span><span class="p">)</span>
    <span class="n">rmt</span> <span class="o">=</span> <span class="n">rmt</span><span class="o">.</span><span class="n">annotate_rows</span><span class="p">(</span><span class="n">in_non_par</span><span class="o">=~</span><span class="n">rmt</span><span class="o">.</span><span class="n">locus</span><span class="o">.</span><span class="n">in_autosome_or_par</span><span class="p">())</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Annotating reference data MT with GQ and DP adj as well as ploidy...&quot;</span><span class="p">)</span>
    <span class="n">rmt</span> <span class="o">=</span> <span class="n">rmt</span><span class="o">.</span><span class="n">annotate_entries</span><span class="p">(</span>
        <span class="n">adj</span><span class="o">=</span><span class="n">get_gq_dp_adj_expr</span><span class="p">(</span>
            <span class="n">rmt</span><span class="o">.</span><span class="n">GQ</span><span class="p">,</span> <span class="n">rmt</span><span class="o">.</span><span class="n">DP</span><span class="p">,</span> <span class="n">locus_expr</span><span class="o">=</span><span class="n">rmt</span><span class="o">.</span><span class="n">locus</span><span class="p">,</span> <span class="n">karyotype_expr</span><span class="o">=</span><span class="n">rmt</span><span class="o">.</span><span class="n">sex_karyotype</span>
        <span class="p">),</span>
        <span class="n">ploidy</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span>
            <span class="n">rmt</span><span class="o">.</span><span class="n">in_non_par</span>
            <span class="o">&amp;</span> <span class="n">get_is_haploid_expr</span><span class="p">(</span>
                <span class="n">locus_expr</span><span class="o">=</span><span class="n">rmt</span><span class="o">.</span><span class="n">locus</span><span class="p">,</span> <span class="n">karyotype_expr</span><span class="o">=</span><span class="n">rmt</span><span class="o">.</span><span class="n">sex_karyotype</span>
            <span class="p">),</span>
            <span class="mi">1</span><span class="p">,</span>
            <span class="mi">2</span><span class="p">,</span>
        <span class="p">),</span>
    <span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Adjusting sex ploidy for variant data MT and annotating with DP/GQ adj...&quot;</span>
    <span class="p">)</span>
    <span class="c1"># An optimization that annotates the locus&#39;s source matrix table with the</span>
    <span class="c1"># fields in the case statements below, so they are not re-computed for every entry.</span>
    <span class="n">c_idx</span><span class="p">,</span> <span class="n">r_idx</span> <span class="o">=</span> <span class="n">annotate_and_index_source_mt_for_sex_ploidy</span><span class="p">(</span>
        <span class="n">locus_expr</span><span class="o">=</span><span class="n">vmt</span><span class="o">.</span><span class="n">locus</span><span class="p">,</span> <span class="n">karyotype_expr</span><span class="o">=</span><span class="n">vmt</span><span class="o">.</span><span class="n">sex_karyotype</span>
    <span class="p">)</span>
    <span class="n">ploidy_expr</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">case</span><span class="p">(</span><span class="n">missing_false</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="o">~</span><span class="n">r_idx</span><span class="o">.</span><span class="n">in_non_par</span><span class="p">,</span> <span class="n">vmt</span><span class="o">.</span><span class="n">LGT</span><span class="o">.</span><span class="n">ploidy</span><span class="p">)</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">c_idx</span><span class="o">.</span><span class="n">xx</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">r_idx</span><span class="o">.</span><span class="n">y_par</span> <span class="o">|</span> <span class="n">r_idx</span><span class="o">.</span><span class="n">y_nonpar</span><span class="p">),</span> <span class="n">hl</span><span class="o">.</span><span class="n">missing</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">tint32</span><span class="p">))</span>
        <span class="c1"># For XY non-PAR het genotypes, the reference ploidy is 1, so we set it to one</span>
        <span class="c1"># even if the genotype is a het. This is to keep the reference ploidy consistent</span>
        <span class="c1"># with the adjusted ploidy after splitting multi-allelic sites.</span>
        <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">c_idx</span><span class="o">.</span><span class="n">xy</span> <span class="o">&amp;</span> <span class="n">r_idx</span><span class="o">.</span><span class="n">in_non_par</span> <span class="o">&amp;</span> <span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">vmt</span><span class="o">.</span><span class="n">LGT</span><span class="o">.</span><span class="n">ploidy</span><span class="p">),</span> <span class="mi">1</span><span class="p">)</span>
        <span class="o">.</span><span class="n">default</span><span class="p">(</span><span class="n">vmt</span><span class="o">.</span><span class="n">LGT</span><span class="o">.</span><span class="n">ploidy</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="c1"># Use the locus and karyotype to annotate the GQ and DP adj correctly for sex</span>
    <span class="c1"># ploidy.</span>
    <span class="n">vmt</span> <span class="o">=</span> <span class="n">vmt</span><span class="o">.</span><span class="n">annotate_entries</span><span class="p">(</span>
        <span class="n">adj</span><span class="o">=</span><span class="n">get_gq_dp_adj_expr</span><span class="p">(</span>
            <span class="n">vmt</span><span class="o">.</span><span class="n">GQ</span><span class="p">,</span> <span class="n">vmt</span><span class="o">.</span><span class="n">DP</span><span class="p">,</span> <span class="n">locus_expr</span><span class="o">=</span><span class="n">vmt</span><span class="o">.</span><span class="n">locus</span><span class="p">,</span> <span class="n">karyotype_expr</span><span class="o">=</span><span class="n">vmt</span><span class="o">.</span><span class="n">sex_karyotype</span>
        <span class="p">),</span>
        <span class="n">ploidy</span><span class="o">=</span><span class="n">ploidy_expr</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">VariantDataset</span><span class="p">(</span><span class="n">rmt</span><span class="p">,</span> <span class="n">vmt</span><span class="p">)</span></div>


<div class="viewcode-block" id="compute_an_and_hists_het_fail_adj_ab"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/fix_freq_an.html#gnomad_qc.v4.annotations.fix_freq_an.compute_an_and_hists_het_fail_adj_ab">[docs]</a><span class="k">def</span> <span class="nf">compute_an_and_hists_het_fail_adj_ab</span><span class="p">(</span>
    <span class="n">vds</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">VariantDataset</span><span class="p">,</span>
    <span class="n">group_membership_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">freq_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Compute allele number and histograms for het fail adj ab and non-PAR XY het.</span>

<span class="sd">    This module computes the allele number and quality histograms for only the</span>
<span class="sd">    genotypes that are het and fail the adj allele balance or are non-PAR XY het</span>
<span class="sd">    genotypes.</span>

<span class="sd">    :param vds: Input VariantDataset.</span>
<span class="sd">    :param group_membership_ht: Table of samples group memberships.</span>
<span class="sd">    :param freq_ht: Table of frequency data.</span>
<span class="sd">    :return: Table of allele number and histograms for het fail adj ab.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">vmt</span> <span class="o">=</span> <span class="n">vds</span><span class="o">.</span><span class="n">variant_data</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Filtering variant data MT entries to those passing GQ and DP adj thresholds, &quot;</span>
        <span class="s2">&quot;but failing the het allele balance adj threshold...&quot;</span>
    <span class="p">)</span>
    <span class="c1"># Use the locus and karyotype to determine the GQ and DP adj correctly for sex</span>
    <span class="c1"># ploidy on unsplit data. The adj allele balance filter is not dependent on sex</span>
    <span class="c1"># ploidy, so we can annotate that before adjusting the ploidy of the genotype.</span>
    <span class="n">vmt</span> <span class="o">=</span> <span class="n">vmt</span><span class="o">.</span><span class="n">select_cols</span><span class="p">(</span>
        <span class="n">sex_karyotype</span><span class="o">=</span><span class="n">vmt</span><span class="o">.</span><span class="n">meta</span><span class="o">.</span><span class="n">sex_imputation</span><span class="o">.</span><span class="n">sex_karyotype</span><span class="p">,</span>
        <span class="n">xy</span><span class="o">=</span><span class="n">vmt</span><span class="o">.</span><span class="n">meta</span><span class="o">.</span><span class="n">sex_imputation</span><span class="o">.</span><span class="n">sex_karyotype</span> <span class="o">==</span> <span class="s2">&quot;XY&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">vmt</span> <span class="o">=</span> <span class="n">vmt</span><span class="o">.</span><span class="n">annotate_rows</span><span class="p">(</span><span class="n">in_non_par</span><span class="o">=~</span><span class="n">vmt</span><span class="o">.</span><span class="n">locus</span><span class="o">.</span><span class="n">in_autosome_or_par</span><span class="p">())</span>
    <span class="n">vmt</span> <span class="o">=</span> <span class="n">vmt</span><span class="o">.</span><span class="n">annotate_entries</span><span class="p">(</span>
        <span class="n">adj</span><span class="o">=</span><span class="n">get_gq_dp_adj_expr</span><span class="p">(</span>
            <span class="n">vmt</span><span class="o">.</span><span class="n">GQ</span><span class="p">,</span> <span class="n">vmt</span><span class="o">.</span><span class="n">DP</span><span class="p">,</span> <span class="n">locus_expr</span><span class="o">=</span><span class="n">vmt</span><span class="o">.</span><span class="n">locus</span><span class="p">,</span> <span class="n">karyotype_expr</span><span class="o">=</span><span class="n">vmt</span><span class="o">.</span><span class="n">sex_karyotype</span>
        <span class="p">),</span>
        <span class="n">fail_het_ab_adj</span><span class="o">=~</span><span class="n">get_het_ab_adj_expr</span><span class="p">(</span><span class="n">vmt</span><span class="o">.</span><span class="n">LGT</span><span class="p">,</span> <span class="n">vmt</span><span class="o">.</span><span class="n">DP</span><span class="p">,</span> <span class="n">vmt</span><span class="o">.</span><span class="n">LAD</span><span class="p">),</span>
        <span class="n">is_nonpar_xy_het</span><span class="o">=</span><span class="n">vmt</span><span class="o">.</span><span class="n">in_non_par</span> <span class="o">&amp;</span> <span class="n">vmt</span><span class="o">.</span><span class="n">xy</span> <span class="o">&amp;</span> <span class="n">vmt</span><span class="o">.</span><span class="n">LGT</span><span class="o">.</span><span class="n">is_het</span><span class="p">(),</span>
    <span class="p">)</span>

    <span class="c1"># Keep only the entries where the genotype passes the DP and GQ adj filters,</span>
    <span class="c1"># but fails the adj allele balance filter. This optimization avoids splitting</span>
    <span class="c1"># multi-allelic entries that do not contribute to the aggregate counts.</span>
    <span class="n">vmt</span> <span class="o">=</span> <span class="n">vmt</span><span class="o">.</span><span class="n">filter_entries</span><span class="p">((</span><span class="n">vmt</span><span class="o">.</span><span class="n">fail_het_ab_adj</span> <span class="o">&amp;</span> <span class="n">vmt</span><span class="o">.</span><span class="n">adj</span><span class="p">)</span> <span class="o">|</span> <span class="n">vmt</span><span class="o">.</span><span class="n">is_nonpar_xy_het</span><span class="p">)</span>

    <span class="c1"># Only need to keep rows where the locus is in the freq HT and there is at least</span>
    <span class="c1"># one non-ref genotype. This saves on computation by not keeping rows that will be</span>
    <span class="c1"># filtered out later or don&#39;t have any genotypes that contribute to the aggregate</span>
    <span class="c1"># counts.</span>
    <span class="n">locus_freq_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">(</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">ir</span><span class="o">.</span><span class="n">TableKeyBy</span><span class="p">(</span>
            <span class="n">freq_ht</span><span class="o">.</span><span class="n">_tir</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;locus&quot;</span><span class="p">],</span> <span class="n">is_sorted</span><span class="o">=</span><span class="kc">True</span>
        <span class="p">)</span>  <span class="c1"># Prevents hail from running sort on HT which is already sorted.</span>
    <span class="p">)</span>
    <span class="n">vmt</span> <span class="o">=</span> <span class="n">vmt</span><span class="o">.</span><span class="n">filter_rows</span><span class="p">(</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">locus_freq_ht</span><span class="p">[</span><span class="n">vmt</span><span class="o">.</span><span class="n">locus</span><span class="p">])</span> <span class="o">&amp;</span> <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">vmt</span><span class="o">.</span><span class="n">LGT</span><span class="o">.</span><span class="n">is_non_ref</span><span class="p">())</span>
    <span class="p">)</span>
    <span class="n">vmt</span> <span class="o">=</span> <span class="n">vmt</span><span class="o">.</span><span class="n">select_entries</span><span class="p">(</span><span class="s2">&quot;LA&quot;</span><span class="p">,</span> <span class="s2">&quot;LGT&quot;</span><span class="p">,</span> <span class="s2">&quot;DP&quot;</span><span class="p">,</span> <span class="s2">&quot;GQ&quot;</span><span class="p">,</span> <span class="s2">&quot;LAD&quot;</span><span class="p">,</span> <span class="s2">&quot;adj&quot;</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Splitting multi-allelic sites...&quot;</span><span class="p">)</span>
    <span class="n">vmt</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">experimental</span><span class="o">.</span><span class="n">sparse_split_multi</span><span class="p">(</span><span class="n">vmt</span><span class="p">)</span>

    <span class="c1"># Now that the multi-allelic sites are split, we can filter genotypes that are non</span>
    <span class="c1"># ref (we don&#39;t want to count the ref), and correctly handle the allele balance</span>
    <span class="c1"># filtering of het non-ref genotypes.</span>
    <span class="n">vmt</span> <span class="o">=</span> <span class="n">vmt</span><span class="o">.</span><span class="n">filter_rows</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">freq_ht</span><span class="p">[</span><span class="n">vmt</span><span class="o">.</span><span class="n">row_key</span><span class="p">]))</span>

    <span class="c1"># Redo the &#39;is_nonpar_xy_het&#39; and &#39;fail_het_ab_adj&#39; annotations after splitting</span>
    <span class="c1"># multi-allelic sites to correctly handle the allele specific adj allele balance</span>
    <span class="c1"># and non-PAR XY het genotypes.</span>
    <span class="n">gt_expr</span> <span class="o">=</span> <span class="n">adjusted_sex_ploidy_expr</span><span class="p">(</span><span class="n">vmt</span><span class="o">.</span><span class="n">locus</span><span class="p">,</span> <span class="n">vmt</span><span class="o">.</span><span class="n">GT</span><span class="p">,</span> <span class="n">vmt</span><span class="o">.</span><span class="n">sex_karyotype</span><span class="p">)</span>
    <span class="n">vmt</span> <span class="o">=</span> <span class="n">vmt</span><span class="o">.</span><span class="n">annotate_entries</span><span class="p">(</span>
        <span class="n">is_nonpar_xy_het</span><span class="o">=</span><span class="n">vmt</span><span class="o">.</span><span class="n">in_non_par</span> <span class="o">&amp;</span> <span class="n">vmt</span><span class="o">.</span><span class="n">xy</span> <span class="o">&amp;</span> <span class="n">vmt</span><span class="o">.</span><span class="n">GT</span><span class="o">.</span><span class="n">is_het</span><span class="p">(),</span>
        <span class="n">GT</span><span class="o">=</span><span class="n">gt_expr</span><span class="p">,</span>
        <span class="n">fail_het_ab_adj</span><span class="o">=~</span><span class="n">get_het_ab_adj_expr</span><span class="p">(</span><span class="n">gt_expr</span><span class="p">,</span> <span class="n">vmt</span><span class="o">.</span><span class="n">DP</span><span class="p">,</span> <span class="n">vmt</span><span class="o">.</span><span class="n">AD</span><span class="p">),</span>
    <span class="p">)</span>
    <span class="n">vmt</span> <span class="o">=</span> <span class="n">vmt</span><span class="o">.</span><span class="n">filter_entries</span><span class="p">(</span>
        <span class="p">(</span><span class="n">vmt</span><span class="o">.</span><span class="n">GT</span><span class="o">.</span><span class="n">is_non_ref</span><span class="p">()</span> <span class="o">&amp;</span> <span class="n">vmt</span><span class="o">.</span><span class="n">fail_het_ab_adj</span><span class="p">)</span> <span class="o">|</span> <span class="n">vmt</span><span class="o">.</span><span class="n">is_nonpar_xy_het</span>
    <span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Annotating variant data MT with DP and GQ qual hists for het fail adj allele &quot;</span>
        <span class="s2">&quot;balance...&quot;</span>
    <span class="p">)</span>
    <span class="n">adj_hist_include_expr</span> <span class="o">=</span> <span class="n">vmt</span><span class="o">.</span><span class="n">adj</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">vmt</span><span class="o">.</span><span class="n">fail_het_ab_adj</span> <span class="o">|</span> <span class="n">vmt</span><span class="o">.</span><span class="n">is_nonpar_xy_het</span><span class="p">)</span>
    <span class="n">vmt</span> <span class="o">=</span> <span class="n">vmt</span><span class="o">.</span><span class="n">annotate_rows</span><span class="p">(</span>
        <span class="n">qual_hists_het_fail_adj_ab_or_is_nonpar_xy_het</span><span class="o">=</span><span class="n">qual_hist_expr</span><span class="p">(</span>
            <span class="n">gq_expr</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">or_missing</span><span class="p">(</span><span class="n">adj_hist_include_expr</span><span class="p">,</span> <span class="n">vmt</span><span class="o">.</span><span class="n">GQ</span><span class="p">),</span>
            <span class="n">dp_expr</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">or_missing</span><span class="p">(</span><span class="n">adj_hist_include_expr</span><span class="p">,</span> <span class="n">vmt</span><span class="o">.</span><span class="n">DP</span><span class="p">),</span>
        <span class="p">),</span>
        <span class="c1"># NOTE: This is not needed for the adjustment of the raw qual hists. They are</span>
        <span class="c1"># not impacted by the non-PAR XY het genotypes being set to None because the</span>
        <span class="c1"># GQ and DP are maintained. Adj qual hists are impacted by the non-PAR XY het</span>
        <span class="c1"># genotypes being set to None because the adj annotation is dependent on the</span>
        <span class="c1"># GT, and therefore the GQ and DP are not included in the histogram</span>
        <span class="c1"># calculations.</span>
        <span class="n">raw_qual_hists_is_nonpar_xy_het</span><span class="o">=</span><span class="n">qual_hist_expr</span><span class="p">(</span>
            <span class="n">gq_expr</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">or_missing</span><span class="p">(</span><span class="n">vmt</span><span class="o">.</span><span class="n">is_nonpar_xy_het</span><span class="p">,</span> <span class="n">vmt</span><span class="o">.</span><span class="n">GQ</span><span class="p">),</span>
            <span class="n">dp_expr</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">or_missing</span><span class="p">(</span><span class="n">vmt</span><span class="o">.</span><span class="n">is_nonpar_xy_het</span><span class="p">,</span> <span class="n">vmt</span><span class="o">.</span><span class="n">DP</span><span class="p">),</span>
        <span class="p">),</span>
    <span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Computing allele number for het fail adj allele balance...&quot;</span><span class="p">)</span>
    <span class="n">an_nonpar_xy_het_meta</span> <span class="o">=</span> <span class="p">[{</span><span class="s2">&quot;group&quot;</span><span class="p">:</span> <span class="s2">&quot;raw&quot;</span><span class="p">},</span> <span class="p">{</span><span class="s2">&quot;group&quot;</span><span class="p">:</span> <span class="s2">&quot;raw&quot;</span><span class="p">,</span> <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="s2">&quot;non_ukb&quot;</span><span class="p">}]</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">agg_by_strata</span><span class="p">(</span>
        <span class="n">vmt</span><span class="o">.</span><span class="n">select_entries</span><span class="p">(</span>
            <span class="s2">&quot;adj&quot;</span><span class="p">,</span>
            <span class="c1"># Adjust the ploidy for non-PAR XY het genotypes to 1.</span>
            <span class="n">ploidy</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span><span class="n">vmt</span><span class="o">.</span><span class="n">is_nonpar_xy_het</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">vmt</span><span class="o">.</span><span class="n">GT</span><span class="o">.</span><span class="n">ploidy</span><span class="p">),</span>
            <span class="n">is_nonpar_xy_het</span><span class="o">=</span><span class="n">vmt</span><span class="o">.</span><span class="n">is_nonpar_xy_het</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="p">{</span>
            <span class="s2">&quot;AN_het_fail_adj_ab_or_nonpar_xy_het&quot;</span><span class="p">:</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">t</span><span class="p">:</span> <span class="n">t</span><span class="o">.</span><span class="n">ploidy</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">sum</span><span class="p">),</span>
            <span class="s2">&quot;AN_nonpar_xy_het&quot;</span><span class="p">:</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">t</span><span class="p">:</span> <span class="n">t</span><span class="o">.</span><span class="n">is_nonpar_xy_het</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">),</span>
        <span class="p">},</span>
        <span class="n">select_fields</span><span class="o">=</span><span class="p">[</span>
            <span class="s2">&quot;qual_hists_het_fail_adj_ab_or_is_nonpar_xy_het&quot;</span><span class="p">,</span>
            <span class="s2">&quot;raw_qual_hists_is_nonpar_xy_het&quot;</span><span class="p">,</span>
        <span class="p">],</span>
        <span class="n">group_membership_ht</span><span class="o">=</span><span class="n">group_membership_ht</span><span class="p">,</span>
        <span class="n">entry_agg_group_membership</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;AN_nonpar_xy_het&quot;</span><span class="p">:</span> <span class="n">an_nonpar_xy_het_meta</span><span class="p">},</span>
    <span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">naive_coalesce</span><span class="p">(</span><span class="mi">1000</span><span class="p">)</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">new_temp_file</span><span class="p">(</span><span class="s2">&quot;an_qual_hist_adjust&quot;</span><span class="p">,</span> <span class="s2">&quot;ht&quot;</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="c1"># Adjust the raw groups using the count of &#39;non_par_xy_hets&#39; rather than the sum of</span>
    <span class="c1"># the ploidies of the failed AB adj hets.</span>
    <span class="n">freq_meta</span> <span class="o">=</span> <span class="n">group_membership_ht</span><span class="o">.</span><span class="n">index_globals</span><span class="p">()</span><span class="o">.</span><span class="n">freq_meta</span>
    <span class="n">an_nonpar_xy_het_meta</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">an_nonpar_xy_het_meta</span><span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
        <span class="n">AN_adjust</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">m</span><span class="p">,</span> <span class="n">an</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span>
                <span class="n">an_nonpar_xy_het_meta</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">m</span><span class="p">),</span>
                <span class="n">ht</span><span class="o">.</span><span class="n">AN_nonpar_xy_het</span><span class="p">[</span><span class="n">an_nonpar_xy_het_meta</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">m</span><span class="p">)],</span>
                <span class="n">an</span><span class="p">,</span>
            <span class="p">),</span>
            <span class="n">freq_meta</span><span class="p">,</span>
            <span class="n">ht</span><span class="o">.</span><span class="n">AN_het_fail_adj_ab_or_nonpar_xy_het</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="n">qual_hists_adjust</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
            <span class="n">qual_hists</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">qual_hists_het_fail_adj_ab_or_is_nonpar_xy_het</span><span class="p">,</span>
            <span class="n">raw_qual_hists</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">raw_qual_hists_is_nonpar_xy_het</span><span class="p">,</span>
        <span class="p">),</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">ht</span></div>


<div class="viewcode-block" id="compute_allele_number_per_ref_site"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/fix_freq_an.html#gnomad_qc.v4.annotations.fix_freq_an.compute_allele_number_per_ref_site">[docs]</a><span class="k">def</span> <span class="nf">compute_allele_number_per_ref_site</span><span class="p">(</span>
    <span class="n">vds</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">VariantDataset</span><span class="p">,</span>
    <span class="n">reference_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">interval_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">group_membership_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Compute allele number per reference site and histograms for frequency correction.</span>

<span class="sd">    :param vds: Input VariantDataset.</span>
<span class="sd">    :param reference_ht: Table of reference sites.</span>
<span class="sd">    :param interval_ht: Table of intervals.</span>
<span class="sd">    :param group_membership_ht: Table of samples group memberships.</span>
<span class="sd">    :return: Table of AN and GQ/DP hists per reference site.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">_get_hists</span><span class="p">(</span><span class="n">qual_expr</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">Expression</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">qual_hist_expr</span><span class="p">(</span>
            <span class="n">gq_expr</span><span class="o">=</span><span class="n">qual_expr</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
            <span class="n">dp_expr</span><span class="o">=</span><span class="n">qual_expr</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
            <span class="n">adj_expr</span><span class="o">=</span><span class="n">qual_expr</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">,</span>
            <span class="n">split_adj_and_raw</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="n">entry_agg_funcs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;AN&quot;</span><span class="p">:</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">t</span><span class="p">:</span> <span class="n">t</span><span class="o">.</span><span class="n">ploidy</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">sum</span><span class="p">),</span>
        <span class="s2">&quot;qual_hists&quot;</span><span class="p">:</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">t</span><span class="p">:</span> <span class="p">[</span><span class="n">t</span><span class="o">.</span><span class="n">GQ</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DP</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">adj</span><span class="p">],</span> <span class="n">_get_hists</span><span class="p">),</span>
    <span class="p">}</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Computing allele number and histograms per reference site...&quot;</span><span class="p">)</span>
    <span class="c1"># Below we use just the raw group for qual hist computations because qual hists</span>
    <span class="c1"># has its own built-in adj filtering when adj is passed as an argument and will</span>
    <span class="c1"># produce both adj and raw histograms.</span>

    <span class="n">ht</span> <span class="o">=</span> <span class="n">compute_stats_per_ref_site</span><span class="p">(</span>
        <span class="n">vds</span><span class="p">,</span>
        <span class="n">reference_ht</span><span class="p">,</span>
        <span class="n">entry_agg_funcs</span><span class="p">,</span>
        <span class="n">interval_ht</span><span class="o">=</span><span class="n">interval_ht</span><span class="p">,</span>
        <span class="n">group_membership_ht</span><span class="o">=</span><span class="n">group_membership_ht</span><span class="p">,</span>
        <span class="n">entry_keep_fields</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;GQ&quot;</span><span class="p">,</span> <span class="s2">&quot;DP&quot;</span><span class="p">,</span> <span class="s2">&quot;ploidy&quot;</span><span class="p">],</span>
        <span class="n">entry_agg_group_membership</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;qual_hists&quot;</span><span class="p">:</span> <span class="p">[{</span><span class="s2">&quot;group&quot;</span><span class="p">:</span> <span class="s2">&quot;raw&quot;</span><span class="p">}]},</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">ht</span></div>


<div class="viewcode-block" id="get_all_sites_nonref_adjustment"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/fix_freq_an.html#gnomad_qc.v4.annotations.fix_freq_an.get_all_sites_nonref_adjustment">[docs]</a><span class="k">def</span> <span class="nf">get_all_sites_nonref_adjustment</span><span class="p">(</span>
    <span class="n">vds</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">VariantDataset</span><span class="p">,</span>
    <span class="n">group_membership_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Compute allele number and histograms for het non ref adjustments.</span>

<span class="sd">    The adjustment numbers computed in `compute_an_and_hists_het_fail_adj_ab` are</span>
<span class="sd">    correct for per allele AN and histogram adjustments for the frequency data, but</span>
<span class="sd">    they are not correct for the all sites AN and histogram adjustments. This is</span>
<span class="sd">    because when computing the sum of all per allele adjustment values, the</span>
<span class="sd">    heterozygous non-ref genotypes can be counted 2 times if both alleles are adjusted.</span>

<span class="sd">    This function will compute the number of het non-ref genotypes that are being</span>
<span class="sd">    counted twice so the all sites AN and histograms can be adjusted correctly.</span>

<span class="sd">    :param vds: Input VariantDataset.</span>
<span class="sd">    :param group_membership_ht: Table of samples group memberships.</span>
<span class="sd">    :return: Table of allele number and histograms for het fail adj ab.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">vmt</span> <span class="o">=</span> <span class="n">vds</span><span class="o">.</span><span class="n">variant_data</span>
    <span class="n">vmt</span> <span class="o">=</span> <span class="n">vmt</span><span class="o">.</span><span class="n">select_cols</span><span class="p">(</span>
        <span class="n">sex_karyotype</span><span class="o">=</span><span class="n">vmt</span><span class="o">.</span><span class="n">meta</span><span class="o">.</span><span class="n">sex_imputation</span><span class="o">.</span><span class="n">sex_karyotype</span><span class="p">,</span>
        <span class="n">xy</span><span class="o">=</span><span class="n">vmt</span><span class="o">.</span><span class="n">meta</span><span class="o">.</span><span class="n">sex_imputation</span><span class="o">.</span><span class="n">sex_karyotype</span> <span class="o">==</span> <span class="s2">&quot;XY&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">vmt</span> <span class="o">=</span> <span class="n">vmt</span><span class="o">.</span><span class="n">annotate_rows</span><span class="p">(</span><span class="n">in_non_par</span><span class="o">=~</span><span class="n">vmt</span><span class="o">.</span><span class="n">locus</span><span class="o">.</span><span class="n">in_autosome_or_par</span><span class="p">())</span>

    <span class="c1"># Filter to only het non-ref genotypes that are either failing the adj allele</span>
    <span class="c1"># balance for both alleles or are non-PAR XY het genotypes.</span>
    <span class="n">vmt</span> <span class="o">=</span> <span class="n">vmt</span><span class="o">.</span><span class="n">filter_entries</span><span class="p">(</span>
        <span class="n">vmt</span><span class="o">.</span><span class="n">LGT</span><span class="o">.</span><span class="n">is_het_non_ref</span><span class="p">()</span>
        <span class="o">&amp;</span> <span class="p">(</span>
            <span class="p">(</span><span class="n">vmt</span><span class="o">.</span><span class="n">in_non_par</span> <span class="o">&amp;</span> <span class="n">vmt</span><span class="o">.</span><span class="n">xy</span><span class="p">)</span>
            <span class="o">|</span> <span class="p">(</span>
                <span class="p">(</span><span class="n">vmt</span><span class="o">.</span><span class="n">LAD</span><span class="p">[</span><span class="n">vmt</span><span class="o">.</span><span class="n">LGT</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">/</span> <span class="n">vmt</span><span class="o">.</span><span class="n">DP</span> <span class="o">&lt;</span> <span class="mf">0.2</span><span class="p">)</span>
                <span class="o">&amp;</span> <span class="p">(</span><span class="n">vmt</span><span class="o">.</span><span class="n">LAD</span><span class="p">[</span><span class="n">vmt</span><span class="o">.</span><span class="n">LGT</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span> <span class="o">/</span> <span class="n">vmt</span><span class="o">.</span><span class="n">DP</span> <span class="o">&lt;</span> <span class="mf">0.2</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="n">vmt</span> <span class="o">=</span> <span class="n">vmt</span><span class="o">.</span><span class="n">filter_rows</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">vmt</span><span class="o">.</span><span class="n">LGT</span><span class="p">)))</span>

    <span class="c1"># Add the DQ/DP adj filter to the entries.</span>
    <span class="n">vmt</span> <span class="o">=</span> <span class="n">vmt</span><span class="o">.</span><span class="n">annotate_entries</span><span class="p">(</span>
        <span class="n">adj</span><span class="o">=</span><span class="n">get_gq_dp_adj_expr</span><span class="p">(</span>
            <span class="n">vmt</span><span class="o">.</span><span class="n">GQ</span><span class="p">,</span> <span class="n">vmt</span><span class="o">.</span><span class="n">DP</span><span class="p">,</span> <span class="n">locus_expr</span><span class="o">=</span><span class="n">vmt</span><span class="o">.</span><span class="n">locus</span><span class="p">,</span> <span class="n">karyotype_expr</span><span class="o">=</span><span class="n">vmt</span><span class="o">.</span><span class="n">sex_karyotype</span>
        <span class="p">),</span>
    <span class="p">)</span>

    <span class="c1"># Calculate adj histograms for the het non-ref genotypes.</span>
    <span class="n">vmt</span> <span class="o">=</span> <span class="n">vmt</span><span class="o">.</span><span class="n">annotate_rows</span><span class="p">(</span>
        <span class="n">qual_hists_het_nonref</span><span class="o">=</span><span class="n">qual_hist_expr</span><span class="p">(</span>
            <span class="n">gq_expr</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">or_missing</span><span class="p">(</span><span class="n">vmt</span><span class="o">.</span><span class="n">adj</span><span class="p">,</span> <span class="n">vmt</span><span class="o">.</span><span class="n">GQ</span><span class="p">),</span>
            <span class="n">dp_expr</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">or_missing</span><span class="p">(</span><span class="n">vmt</span><span class="o">.</span><span class="n">adj</span><span class="p">,</span> <span class="n">vmt</span><span class="o">.</span><span class="n">DP</span><span class="p">),</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="n">vmt</span> <span class="o">=</span> <span class="n">vmt</span><span class="o">.</span><span class="n">select_entries</span><span class="p">(</span>
        <span class="s2">&quot;adj&quot;</span><span class="p">,</span>
        <span class="n">is_nonpar_xy_het</span><span class="o">=</span><span class="n">vmt</span><span class="o">.</span><span class="n">in_non_par</span> <span class="o">&amp;</span> <span class="n">vmt</span><span class="o">.</span><span class="n">xy</span> <span class="o">&amp;</span> <span class="n">vmt</span><span class="o">.</span><span class="n">LGT</span><span class="o">.</span><span class="n">is_het_non_ref</span><span class="p">(),</span>
        <span class="n">ploidy</span><span class="o">=</span><span class="n">adjusted_sex_ploidy_expr</span><span class="p">(</span><span class="n">vmt</span><span class="o">.</span><span class="n">locus</span><span class="p">,</span> <span class="n">vmt</span><span class="o">.</span><span class="n">LGT</span><span class="p">,</span> <span class="n">vmt</span><span class="o">.</span><span class="n">sex_karyotype</span><span class="p">)</span><span class="o">.</span><span class="n">ploidy</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># Compute the number of het non-ref genotypes that are being counted twice.</span>
    <span class="c1"># XY het genotypes in non-PAR regions are set to ploidy 1, so we can use the</span>
    <span class="c1"># ploidy to determine AN.</span>
    <span class="c1"># The raw AN adjustment is only the count of XY het genotypes in non-PAR regions</span>
    <span class="c1"># since it doesn&#39;t include adj genotypes.</span>
    <span class="n">an_nonref_raw_meta</span> <span class="o">=</span> <span class="p">[{</span><span class="s2">&quot;group&quot;</span><span class="p">:</span> <span class="s2">&quot;raw&quot;</span><span class="p">},</span> <span class="p">{</span><span class="s2">&quot;group&quot;</span><span class="p">:</span> <span class="s2">&quot;raw&quot;</span><span class="p">,</span> <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="s2">&quot;non_ukb&quot;</span><span class="p">}]</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">agg_by_strata</span><span class="p">(</span>
        <span class="n">vmt</span><span class="o">.</span><span class="n">annotate_entries</span><span class="p">(</span><span class="n">ploidy</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span><span class="n">vmt</span><span class="o">.</span><span class="n">is_nonpar_xy_het</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">vmt</span><span class="o">.</span><span class="n">ploidy</span><span class="p">)),</span>
        <span class="p">{</span>
            <span class="s2">&quot;AN_het_nonref_adj&quot;</span><span class="p">:</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">t</span><span class="p">:</span> <span class="n">t</span><span class="o">.</span><span class="n">ploidy</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">sum</span><span class="p">),</span>
            <span class="s2">&quot;AN_het_nonref_raw&quot;</span><span class="p">:</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">t</span><span class="p">:</span> <span class="n">t</span><span class="o">.</span><span class="n">is_nonpar_xy_het</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">),</span>
        <span class="p">},</span>
        <span class="n">select_fields</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;qual_hists_het_nonref&quot;</span><span class="p">],</span>
        <span class="n">group_membership_ht</span><span class="o">=</span><span class="n">group_membership_ht</span><span class="p">,</span>
        <span class="n">entry_agg_group_membership</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;AN_het_nonref_raw&quot;</span><span class="p">:</span> <span class="n">an_nonref_raw_meta</span><span class="p">},</span>
    <span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;alleles&quot;</span><span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">new_temp_file</span><span class="p">(</span><span class="s2">&quot;an_qual_hist_het_nonref_adjust&quot;</span><span class="p">,</span> <span class="s2">&quot;ht&quot;</span><span class="p">))</span>

    <span class="c1"># Modify the raw groups to use &#39;AN_het_nonref_raw&#39; instead of &#39;AN_het_nonref_adj&#39;.</span>
    <span class="n">freq_meta</span> <span class="o">=</span> <span class="n">group_membership_ht</span><span class="o">.</span><span class="n">index_globals</span><span class="p">()</span><span class="o">.</span><span class="n">freq_meta</span>
    <span class="n">an_nonref_raw_meta</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">an_nonref_raw_meta</span><span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
        <span class="s2">&quot;qual_hists_het_nonref&quot;</span><span class="p">,</span>
        <span class="n">AN_het_nonref</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">m</span><span class="p">,</span> <span class="n">an</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span>
                <span class="n">an_nonref_raw_meta</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">m</span><span class="p">),</span>
                <span class="n">ht</span><span class="o">.</span><span class="n">AN_het_nonref_raw</span><span class="p">[</span><span class="n">an_nonref_raw_meta</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">m</span><span class="p">)],</span>
                <span class="n">an</span><span class="p">,</span>
            <span class="p">),</span>
            <span class="n">freq_meta</span><span class="p">,</span>
            <span class="n">ht</span><span class="o">.</span><span class="n">AN_het_nonref_adj</span><span class="p">,</span>
        <span class="p">),</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">ht</span></div>


<div class="viewcode-block" id="get_sub_hist_expr"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/fix_freq_an.html#gnomad_qc.v4.annotations.fix_freq_an.get_sub_hist_expr">[docs]</a><span class="k">def</span> <span class="nf">get_sub_hist_expr</span><span class="p">(</span>
    <span class="n">hist_expr</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">StructExpression</span><span class="p">,</span>
    <span class="n">sub_hist_expr</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">StructExpression</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">StructExpression</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Subtract sub histograms from histograms.</span>

<span class="sd">    :param hist_expr: Histograms to subtract from.</span>
<span class="sd">    :param sub_hist_expr: Histograms to subtract.</span>
<span class="sd">    :return: Subtracted histograms.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sub_hist_expr</span> <span class="o">=</span> <span class="n">sub_hist_expr</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="o">**</span><span class="p">{</span>
            <span class="n">k</span><span class="p">:</span> <span class="n">v</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
                <span class="n">bin_freq</span><span class="o">=</span><span class="n">v</span><span class="o">.</span><span class="n">bin_freq</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="o">-</span><span class="n">x</span><span class="p">),</span>
                <span class="n">n_smaller</span><span class="o">=-</span><span class="n">v</span><span class="o">.</span><span class="n">n_smaller</span><span class="p">,</span>
                <span class="n">n_larger</span><span class="o">=-</span><span class="n">v</span><span class="o">.</span><span class="n">n_larger</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">sub_hist_expr</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
        <span class="p">}</span>
    <span class="p">)</span>
    <span class="n">hist_expr</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">sub_hist_expr</span><span class="p">),</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
            <span class="o">**</span><span class="p">{</span><span class="n">k</span><span class="p">:</span> <span class="n">merge_histograms</span><span class="p">([</span><span class="n">v</span><span class="p">,</span> <span class="n">sub_hist_expr</span><span class="p">[</span><span class="n">k</span><span class="p">]])</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">hist_expr</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
        <span class="p">),</span>
        <span class="n">hist_expr</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">hist_expr</span></div>


<div class="viewcode-block" id="adjust_per_site_an_and_hists_for_frequency"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/fix_freq_an.html#gnomad_qc.v4.annotations.fix_freq_an.adjust_per_site_an_and_hists_for_frequency">[docs]</a><span class="k">def</span> <span class="nf">adjust_per_site_an_and_hists_for_frequency</span><span class="p">(</span>
    <span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">freq_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">het_fail_adj_ab_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Adjust allele number and histograms for frequency correction.</span>

<span class="sd">    The all sites values on `ht` are only the reference AN and quality histograms</span>
<span class="sd">    at the site level (keyed by locus) and don&#39;t take into account the differences in</span>
<span class="sd">    the alleles. This function will adjust the all sites AN and quality histograms</span>
<span class="sd">    by the `het_fail_adj_ab_ht` (computed by `compute_an_and_hists_het_fail_adj_ab`)</span>
<span class="sd">    so they can be used for the variant frequency correction.</span>

<span class="sd">    The reasons a variant&#39;s AN and quality histograms can be different from the all</span>
<span class="sd">    sites values are:</span>

<span class="sd">        - The reference AN adj filter doesn&#39;t take into account the allele balance adj</span>
<span class="sd">          filter, so the variant AN can be smaller than the reference AN if there are</span>
<span class="sd">          samples with het genotypes that fail the adj allele balance filter.</span>
<span class="sd">        - The reference AN doesn&#39;t remove the non-PAR het genotypes in XY samples, so</span>
<span class="sd">          the variant AN can be smaller than the reference AN if there are XY samples</span>
<span class="sd">          with a het genotype in a non-PAR region on chrX or chrY.</span>

<span class="sd">    :param ht: Table of AN and GQ/DP hists per reference site.</span>
<span class="sd">    :param freq_ht: Table of frequency data.</span>
<span class="sd">    :param het_fail_adj_ab_ht: Table of variant data histograms for het fail adj ab.</span>
<span class="sd">    :return: Table of adjusted AN and GQ/DP hists for frequency correction.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Adjusting allele number and histograms for het fail adj allele balance...&quot;</span>
    <span class="p">)</span>
    <span class="n">global_annotations</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">index_globals</span><span class="p">()</span>
    <span class="n">freq_correction</span> <span class="o">=</span> <span class="n">ht</span><span class="p">[</span><span class="n">freq_ht</span><span class="o">.</span><span class="n">locus</span><span class="p">]</span>
    <span class="n">het_fail_adj_ab</span> <span class="o">=</span> <span class="n">het_fail_adj_ab_ht</span><span class="p">[</span><span class="n">freq_ht</span><span class="o">.</span><span class="n">key</span><span class="p">]</span>
    <span class="n">qual_hists</span> <span class="o">=</span> <span class="n">freq_correction</span><span class="o">.</span><span class="n">qual_hists</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">qual_hists</span>
    <span class="n">het_fail_adj_ab_qual_hists</span> <span class="o">=</span> <span class="n">het_fail_adj_ab</span><span class="o">.</span><span class="n">qual_hists_adjust</span><span class="o">.</span><span class="n">qual_hists</span>

    <span class="c1"># Convert the het fail adj allele balance histograms to negative values so that</span>
    <span class="c1"># they can be subtracted from the partial adj (GQ/DP pass) histograms using</span>
    <span class="c1"># merge_histograms to create complete adj (GQ, DP, and AB pass) histograms.</span>
    <span class="n">freq_correction_ht</span> <span class="o">=</span> <span class="n">freq_ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
        <span class="n">AN</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">coalesce</span><span class="p">(</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
                <span class="k">lambda</span> <span class="n">an</span><span class="p">,</span> <span class="n">an_fail_adj</span><span class="p">,</span> <span class="n">m</span><span class="p">:</span> <span class="n">an</span> <span class="o">-</span> <span class="n">hl</span><span class="o">.</span><span class="n">coalesce</span><span class="p">(</span><span class="n">an_fail_adj</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>
                <span class="n">freq_correction</span><span class="o">.</span><span class="n">AN</span><span class="p">,</span>
                <span class="n">het_fail_adj_ab</span><span class="o">.</span><span class="n">AN_adjust</span><span class="p">,</span>
                <span class="n">global_annotations</span><span class="o">.</span><span class="n">strata_meta</span><span class="p">,</span>
            <span class="p">),</span>
            <span class="n">freq_correction</span><span class="o">.</span><span class="n">AN</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="n">qual_hists</span><span class="o">=</span><span class="n">freq_correction</span><span class="o">.</span><span class="n">qual_hists</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
            <span class="n">qual_hists</span><span class="o">=</span><span class="n">get_sub_hist_expr</span><span class="p">(</span><span class="n">qual_hists</span><span class="p">,</span> <span class="n">het_fail_adj_ab_qual_hists</span><span class="p">)</span>
        <span class="p">),</span>
    <span class="p">)</span>
    <span class="n">freq_correction_ht</span> <span class="o">=</span> <span class="n">freq_correction_ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span><span class="o">**</span><span class="n">global_annotations</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">freq_correction_ht</span></div>


<div class="viewcode-block" id="adjust_all_sites_an_and_hists"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/fix_freq_an.html#gnomad_qc.v4.annotations.fix_freq_an.adjust_all_sites_an_and_hists">[docs]</a><span class="k">def</span> <span class="nf">adjust_all_sites_an_and_hists</span><span class="p">(</span>
    <span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">het_fail_adj_ab_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">het_nonref_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Adjust all sites allele number and histograms Table.</span>

<span class="sd">    :param ht: Table of AN and GQ/DP hists per reference site.</span>
<span class="sd">    :param het_fail_adj_ab_ht: Table of variant data histograms for het fail adj ab.</span>
<span class="sd">    :param het_nonref_ht: Table of AN and GQ/DP hists for only het non-ref calls.</span>
<span class="sd">    :return: Adjusted all sites AN and GQ/DP hists Table.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Get the sum of het GTs that fail the adj allele balance filter for all alleles at</span>
    <span class="c1"># each locus.</span>
    <span class="n">het_fail_adj_ab_ht</span> <span class="o">=</span> <span class="n">het_fail_adj_ab_ht</span><span class="o">.</span><span class="n">group_by</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
        <span class="n">AN_adjust</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">array_sum</span><span class="p">(</span><span class="n">het_fail_adj_ab_ht</span><span class="o">.</span><span class="n">AN_adjust</span><span class="p">),</span>
        <span class="n">qual_hists_adjust</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
            <span class="o">**</span><span class="p">{</span>
                <span class="n">k</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                    <span class="n">bin_edges</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">take</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">bin_edges</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">],</span>
                    <span class="n">bin_freq</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">array_sum</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">bin_freq</span><span class="p">),</span>
                    <span class="n">n_smaller</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">n_smaller</span><span class="p">),</span>
                    <span class="n">n_larger</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">v</span><span class="o">.</span><span class="n">n_larger</span><span class="p">),</span>
                <span class="p">)</span>
                <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">het_fail_adj_ab_ht</span><span class="o">.</span><span class="n">qual_hists_adjust</span><span class="o">.</span><span class="n">qual_hists</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
            <span class="p">}</span>
        <span class="p">),</span>
    <span class="p">)</span>

    <span class="c1"># Annotate the all sites AN and qual hists Table with the number of hets that fail</span>
    <span class="c1"># the adj allele balance filter and het non-refs at the locus and add the AN</span>
    <span class="c1"># corrected for the allele balance filter to the all sites AN and qual hists Table.</span>
    <span class="n">het_fail_adj_ab</span> <span class="o">=</span> <span class="n">het_fail_adj_ab_ht</span><span class="p">[</span><span class="n">ht</span><span class="o">.</span><span class="n">locus</span><span class="p">]</span>
    <span class="n">het_nonref</span> <span class="o">=</span> <span class="n">het_nonref_ht</span><span class="p">[</span><span class="n">ht</span><span class="o">.</span><span class="n">locus</span><span class="p">]</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
        <span class="n">AN_pre_adjust</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">AN</span><span class="p">,</span>
        <span class="n">qual_hists_pre_adjust</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">qual_hists</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
        <span class="n">AN_adjust</span><span class="o">=</span><span class="n">het_fail_adj_ab</span><span class="o">.</span><span class="n">AN_adjust</span><span class="p">,</span>
        <span class="n">AN_het_nonref</span><span class="o">=</span><span class="n">het_nonref</span><span class="o">.</span><span class="n">AN_het_nonref</span><span class="p">,</span>
        <span class="n">qual_hists_adjust</span><span class="o">=</span><span class="n">het_fail_adj_ab</span><span class="o">.</span><span class="n">qual_hists_adjust</span><span class="p">,</span>
        <span class="n">qual_hists_het_nonref</span><span class="o">=</span><span class="n">het_nonref</span><span class="o">.</span><span class="n">qual_hists_het_nonref</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="c1"># Remove duplicate counts of the het non-ref sites from the adjustment histograms.</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="n">_tmp_hist_adjust</span><span class="o">=</span><span class="n">get_sub_hist_expr</span><span class="p">(</span>
            <span class="n">ht</span><span class="o">.</span><span class="n">qual_hists_adjust</span><span class="p">,</span> <span class="n">ht</span><span class="o">.</span><span class="n">qual_hists_het_nonref</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="c1"># When present, remove duplicate counts of the het non-ref sites from the</span>
    <span class="c1"># adjustment AN and then adjust the AN based on result. Adjust qual histograms with</span>
    <span class="c1"># the previously het non-ref adjusted histograms.</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="n">AN</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">AN_adjust</span><span class="p">),</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">enumerate</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">AN_pre_adjust</span><span class="p">)</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
                <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span>
                    <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                    <span class="o">-</span> <span class="p">(</span>
                        <span class="n">ht</span><span class="o">.</span><span class="n">AN_adjust</span><span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
                        <span class="o">-</span> <span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span>
                            <span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">AN_het_nonref</span><span class="p">),</span> <span class="n">ht</span><span class="o">.</span><span class="n">AN_het_nonref</span><span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]],</span> <span class="mi">0</span>
                        <span class="p">)</span>
                    <span class="p">)</span>
                <span class="p">)</span>
            <span class="p">),</span>
            <span class="n">ht</span><span class="o">.</span><span class="n">AN_pre_adjust</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="n">qual_hists</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">qual_hists_pre_adjust</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
            <span class="n">qual_hists</span><span class="o">=</span><span class="n">get_sub_hist_expr</span><span class="p">(</span>
                <span class="n">ht</span><span class="o">.</span><span class="n">qual_hists_pre_adjust</span><span class="o">.</span><span class="n">qual_hists</span><span class="p">,</span> <span class="n">ht</span><span class="o">.</span><span class="n">_tmp_hist_adjust</span>
            <span class="p">)</span>
        <span class="p">),</span>
    <span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;_tmp_hist_adjust&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">ht</span></div>


<div class="viewcode-block" id="update_freq_an_and_hists"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/fix_freq_an.html#gnomad_qc.v4.annotations.fix_freq_an.update_freq_an_and_hists">[docs]</a><span class="k">def</span> <span class="nf">update_freq_an_and_hists</span><span class="p">(</span>
    <span class="n">freq_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">freq_correction_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Update frequency HT AN field and histograms with the correct AN from the AN HT.</span>

<span class="sd">    This module updates the freq_ht AN field which is an array of ints with the correct</span>
<span class="sd">    AN from the AN HT integer array annotation. It uses the freq_correction_ht</span>
<span class="sd">    dictionary and the freq_ht dictionary to match the indexing of array annotations so</span>
<span class="sd">    each group is correctly updated. It then recomputes AF, AC/AN.</span>

<span class="sd">    :param freq_ht: Frequency HT to update.</span>
<span class="sd">    :param freq_correction_ht: HT to update AN and histograms from.</span>
<span class="sd">    :return: Updated frequency HT.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">freq_correction_meta</span> <span class="o">=</span> <span class="n">freq_correction_ht</span><span class="o">.</span><span class="n">index_globals</span><span class="p">()</span><span class="o">.</span><span class="n">strata_meta</span>

    <span class="c1"># Set all freq_ht&#39;s ANs to 0 so can use the merge_freq function to update</span>
    <span class="c1"># the ANs (expects int32s).</span>
    <span class="n">freq_ht</span> <span class="o">=</span> <span class="n">freq_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="n">freq</span><span class="o">=</span><span class="n">freq_ht</span><span class="o">.</span><span class="n">freq</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
                <span class="n">AC</span><span class="o">=</span><span class="n">x</span><span class="o">.</span><span class="n">AC</span><span class="p">,</span> <span class="n">AN</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">int32</span><span class="p">(</span><span class="mi">0</span><span class="p">),</span> <span class="n">homozygote_count</span><span class="o">=</span><span class="n">x</span><span class="o">.</span><span class="n">homozygote_count</span><span class="p">,</span> <span class="n">AF</span><span class="o">=</span><span class="n">x</span><span class="o">.</span><span class="n">AF</span>
            <span class="p">)</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="c1"># Create freq array annotation of structs with AC and homozygote_count set to 0 to</span>
    <span class="c1"># use merge_freq function (expects int32s).</span>
    <span class="n">freq_correction_ht</span> <span class="o">=</span> <span class="n">freq_correction_ht</span><span class="o">.</span><span class="n">transmute</span><span class="p">(</span>
        <span class="n">freq</span><span class="o">=</span><span class="n">freq_correction_ht</span><span class="o">.</span><span class="n">AN</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                <span class="n">AC</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">AN</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">int32</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="n">homozygote_count</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">AF</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">missing</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">tfloat64</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="n">freq_ht</span> <span class="o">=</span> <span class="n">freq_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="n">freq_correction_freq</span><span class="o">=</span><span class="n">freq_correction_ht</span><span class="p">[</span><span class="n">freq_ht</span><span class="o">.</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">freq</span>
    <span class="p">)</span>
    <span class="n">freq_ht</span> <span class="o">=</span> <span class="n">freq_ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span><span class="n">freq_correction_strata_meta</span><span class="o">=</span><span class="n">freq_correction_meta</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Merging freq arrays from v4 and freq correction HT...&quot;</span><span class="p">)</span>
    <span class="n">freq</span><span class="p">,</span> <span class="n">freq_meta</span> <span class="o">=</span> <span class="n">merge_freq_arrays</span><span class="p">(</span>
        <span class="p">[</span><span class="n">freq_ht</span><span class="o">.</span><span class="n">freq</span><span class="p">,</span> <span class="n">freq_ht</span><span class="o">.</span><span class="n">freq_correction_freq</span><span class="p">],</span>
        <span class="p">[</span><span class="n">freq_ht</span><span class="o">.</span><span class="n">freq_meta</span><span class="p">,</span> <span class="n">freq_ht</span><span class="o">.</span><span class="n">freq_correction_strata_meta</span><span class="p">],</span>
    <span class="p">)</span>
    <span class="n">freq_ht</span> <span class="o">=</span> <span class="n">freq_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">freq</span><span class="o">=</span><span class="n">freq</span><span class="p">)</span>
    <span class="n">freq_ht</span> <span class="o">=</span> <span class="n">freq_ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span><span class="n">freq_meta</span><span class="o">=</span><span class="n">freq_meta</span><span class="p">)</span>
    <span class="n">freq_ht</span> <span class="o">=</span> <span class="n">freq_ht</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;freq_correction_freq&quot;</span><span class="p">,</span> <span class="s2">&quot;freq_correction_strata_meta&quot;</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Updating freq_ht with corrected GQ and DP histograms...&quot;</span><span class="p">)</span>
    <span class="n">corrected_index</span> <span class="o">=</span> <span class="n">freq_correction_ht</span><span class="p">[</span><span class="n">freq_ht</span><span class="o">.</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">qual_hists</span>
    <span class="n">freq_ht</span> <span class="o">=</span> <span class="n">freq_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="n">qual_hists</span><span class="o">=</span><span class="n">freq_ht</span><span class="o">.</span><span class="n">qual_hists</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
            <span class="n">gq_hist_all</span><span class="o">=</span><span class="n">corrected_index</span><span class="o">.</span><span class="n">qual_hists</span><span class="o">.</span><span class="n">gq_hist_all</span><span class="p">,</span>
            <span class="n">dp_hist_all</span><span class="o">=</span><span class="n">corrected_index</span><span class="o">.</span><span class="n">qual_hists</span><span class="o">.</span><span class="n">dp_hist_all</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="n">raw_qual_hists</span><span class="o">=</span><span class="n">freq_ht</span><span class="o">.</span><span class="n">raw_qual_hists</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
            <span class="n">gq_hist_all</span><span class="o">=</span><span class="n">corrected_index</span><span class="o">.</span><span class="n">raw_qual_hists</span><span class="o">.</span><span class="n">gq_hist_all</span><span class="p">,</span>
            <span class="n">dp_hist_all</span><span class="o">=</span><span class="n">corrected_index</span><span class="o">.</span><span class="n">raw_qual_hists</span><span class="o">.</span><span class="n">dp_hist_all</span><span class="p">,</span>
        <span class="p">),</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">freq_ht</span></div>


<div class="viewcode-block" id="drop_gatk_groupings"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/fix_freq_an.html#gnomad_qc.v4.annotations.fix_freq_an.drop_gatk_groupings">[docs]</a><span class="k">def</span> <span class="nf">drop_gatk_groupings</span><span class="p">(</span><span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Drop GATK groupings from the frequency HT.</span>

<span class="sd">    :param ht: Frequency HT to drop GATK groupings from.</span>
<span class="sd">    :return: Frequency HT with GATK groupings dropped.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">freq_meta</span><span class="p">,</span> <span class="n">array_exprs</span> <span class="o">=</span> <span class="n">filter_arrays_by_meta</span><span class="p">(</span>
        <span class="n">ht</span><span class="o">.</span><span class="n">freq_meta</span><span class="p">,</span>
        <span class="p">{</span>
            <span class="s2">&quot;freq&quot;</span><span class="p">:</span> <span class="n">ht</span><span class="o">.</span><span class="n">freq</span><span class="p">,</span>
            <span class="s2">&quot;high_ab_hets_by_group&quot;</span><span class="p">:</span> <span class="n">ht</span><span class="o">.</span><span class="n">high_ab_hets_by_group</span><span class="p">,</span>
            <span class="s2">&quot;freq_meta_sample_count&quot;</span><span class="p">:</span> <span class="n">ht</span><span class="o">.</span><span class="n">index_globals</span><span class="p">()</span><span class="o">.</span><span class="n">freq_meta_sample_count</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="n">items_to_filter</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;gatk_version&quot;</span><span class="p">],</span>
        <span class="n">keep</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="n">freq</span><span class="o">=</span><span class="n">array_exprs</span><span class="p">[</span><span class="s2">&quot;freq&quot;</span><span class="p">],</span>
        <span class="n">high_ab_hets_by_group</span><span class="o">=</span><span class="n">array_exprs</span><span class="p">[</span><span class="s2">&quot;high_ab_hets_by_group&quot;</span><span class="p">],</span>
    <span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span>
        <span class="n">freq_meta</span><span class="o">=</span><span class="n">freq_meta</span><span class="p">,</span>
        <span class="n">freq_meta_sample_count</span><span class="o">=</span><span class="n">array_exprs</span><span class="p">[</span><span class="s2">&quot;freq_meta_sample_count&quot;</span><span class="p">],</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">ht</span></div>


<div class="viewcode-block" id="main"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/fix_freq_an.html#gnomad_qc.v4.annotations.fix_freq_an.main">[docs]</a><span class="k">def</span> <span class="nf">main</span><span class="p">(</span><span class="n">args</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Script to generate all sites AN and v4.0 exomes frequency fix.&quot;&quot;&quot;</span>
    <span class="n">overwrite</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">overwrite</span>
    <span class="n">use_test_dataset</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">use_test_dataset</span>
    <span class="n">test_chr22_chrx_chry</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">test_chr22_chrx_chry</span>
    <span class="n">test_n_partitions</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">test_n_partitions</span>
    <span class="n">test</span> <span class="o">=</span> <span class="n">use_test_dataset</span> <span class="ow">or</span> <span class="n">test_n_partitions</span> <span class="ow">or</span> <span class="n">test_chr22_chrx_chry</span>
    <span class="n">af_threshold</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">af_threshold</span>

    <span class="n">hl</span><span class="o">.</span><span class="n">init</span><span class="p">(</span>
        <span class="n">log</span><span class="o">=</span><span class="s2">&quot;/frequency_an_update.log&quot;</span><span class="p">,</span>
        <span class="n">default_reference</span><span class="o">=</span><span class="s2">&quot;GRCh38&quot;</span><span class="p">,</span>
        <span class="n">tmp_dir</span><span class="o">=</span><span class="s2">&quot;gs://gnomad-tmp-30day&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">hl</span><span class="o">.</span><span class="n">_set_flags</span><span class="p">(</span><span class="n">use_ssa_logs</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">)</span>

    <span class="n">freq_ht</span> <span class="o">=</span> <span class="n">get_freq</span><span class="p">(</span><span class="s2">&quot;4.0&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">if</span> <span class="p">(</span>
            <span class="n">args</span><span class="o">.</span><span class="n">regenerate_ref_an_hists</span>
            <span class="ow">or</span> <span class="n">args</span><span class="o">.</span><span class="n">calculate_var_an_hists_adjustment</span>
            <span class="ow">or</span> <span class="n">args</span><span class="o">.</span><span class="n">calculate_all_sites_het_nonref_adjustment</span>
        <span class="p">):</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Regenerating reference AN and GQ/DP histograms...&quot;</span><span class="p">)</span>

            <span class="n">ref_ht</span> <span class="o">=</span> <span class="n">vep_context</span><span class="o">.</span><span class="n">versions</span><span class="p">[</span><span class="s2">&quot;105&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>
            <span class="n">ref_ht</span> <span class="o">=</span> <span class="n">ref_ht</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">select</span><span class="p">()</span><span class="o">.</span><span class="n">distinct</span><span class="p">()</span>

            <span class="k">if</span> <span class="n">test_chr22_chrx_chry</span><span class="p">:</span>
                <span class="n">chrom</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;chr22&quot;</span><span class="p">,</span> <span class="s2">&quot;chrX&quot;</span><span class="p">,</span> <span class="s2">&quot;chrY&quot;</span><span class="p">]</span>
                <span class="n">ref_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">filter_intervals</span><span class="p">(</span>
                    <span class="n">ref_ht</span><span class="p">,</span> <span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">parse_locus_interval</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">chrom</span><span class="p">]</span>
                <span class="p">)</span>

            <span class="n">vds</span> <span class="o">=</span> <span class="n">get_gnomad_v4_vds</span><span class="p">(</span>
                <span class="n">release_only</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">test</span><span class="o">=</span><span class="n">use_test_dataset</span> <span class="ow">or</span> <span class="n">test_chr22_chrx_chry</span><span class="p">,</span>
                <span class="n">filter_partitions</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span> <span class="k">if</span> <span class="n">test_n_partitions</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
                <span class="n">annotate_meta</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">chrom</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chr22&quot;</span><span class="p">,</span> <span class="s2">&quot;chrX&quot;</span><span class="p">,</span> <span class="s2">&quot;chrY&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">test_chr22_chrx_chry</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="n">group_membership_ht</span> <span class="o">=</span> <span class="n">get_exomes_group_membership_ht</span><span class="p">(</span>
                <span class="n">meta</span><span class="p">()</span><span class="o">.</span><span class="n">ht</span><span class="p">(),</span>
                <span class="n">get_downsampling</span><span class="p">()</span><span class="o">.</span><span class="n">ht</span><span class="p">(),</span>
                <span class="n">get_downsampling</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="s2">&quot;non_ukb&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">ht</span><span class="p">(),</span>
            <span class="p">)</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">new_temp_file</span><span class="p">(</span><span class="s2">&quot;group_membership&quot;</span><span class="p">,</span> <span class="s2">&quot;ht&quot;</span><span class="p">))</span>

            <span class="n">interval_ht</span> <span class="o">=</span> <span class="n">adjust_interval_padding</span><span class="p">(</span>
                <span class="n">calling_intervals</span><span class="p">(</span>
                    <span class="n">interval_name</span><span class="o">=</span><span class="s2">&quot;union&quot;</span><span class="p">,</span> <span class="n">calling_interval_padding</span><span class="o">=</span><span class="mi">0</span>
                <span class="p">)</span><span class="o">.</span><span class="n">ht</span><span class="p">(),</span>
                <span class="mi">150</span><span class="p">,</span>
            <span class="p">)</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">new_temp_file</span><span class="p">(</span><span class="s2">&quot;interval&quot;</span><span class="p">,</span> <span class="s2">&quot;ht&quot;</span><span class="p">))</span>

            <span class="c1"># Filter out interval HT to only include intervals that have at least one</span>
            <span class="c1"># locus in the VDS variant data. This saves on computation by reducing the</span>
            <span class="c1"># number of ref loci we compute AN and hists on during the test</span>
            <span class="k">if</span> <span class="n">test</span><span class="p">:</span>
                <span class="n">tmp_interval_ht</span> <span class="o">=</span> <span class="n">interval_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">in_interval</span><span class="o">=</span><span class="n">interval_ht</span><span class="o">.</span><span class="n">interval</span><span class="p">)</span>
                <span class="n">tmp_ht</span> <span class="o">=</span> <span class="n">vds</span><span class="o">.</span><span class="n">variant_data</span><span class="o">.</span><span class="n">rows</span><span class="p">()</span>
                <span class="n">tmp_ht</span> <span class="o">=</span> <span class="n">tmp_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
                    <span class="n">in_interval</span><span class="o">=</span><span class="n">tmp_interval_ht</span><span class="p">[</span><span class="n">tmp_ht</span><span class="o">.</span><span class="n">locus</span><span class="p">]</span><span class="o">.</span><span class="n">in_interval</span>
                <span class="p">)</span>
                <span class="n">test_intervals</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">literal</span><span class="p">(</span>
                    <span class="n">tmp_ht</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">collect_as_set</span><span class="p">(</span><span class="n">tmp_ht</span><span class="o">.</span><span class="n">in_interval</span><span class="p">))</span>
                <span class="p">)</span>
                <span class="n">interval_ht</span> <span class="o">=</span> <span class="n">interval_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                    <span class="n">test_intervals</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">interval_ht</span><span class="o">.</span><span class="n">interval</span><span class="p">)</span>
                <span class="p">)</span>

            <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">regenerate_ref_an_hists</span><span class="p">:</span>
                <span class="n">compute_allele_number_per_ref_site</span><span class="p">(</span>
                    <span class="n">prep_vds_for_all_sites_stats</span><span class="p">(</span><span class="n">vds</span><span class="p">),</span>
                    <span class="n">ref_ht</span><span class="p">,</span>
                    <span class="n">interval_ht</span><span class="p">,</span>
                    <span class="n">group_membership_ht</span><span class="p">,</span>
                <span class="p">)</span><span class="o">.</span><span class="n">naive_coalesce</span><span class="p">(</span><span class="mi">10000</span><span class="p">)</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;gs://gnomad</span><span class="si">{</span><span class="s1">&#39;-tmp-4day&#39;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">test</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;&#39;</span><span class="si">}</span><span class="s2">/v4.1/temp/frequency_fix/all_sites_an_before_adjustment.ht&quot;</span><span class="p">,</span>
                    <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">,</span>
                <span class="p">)</span>

            <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">calculate_var_an_hists_adjustment</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                    <span class="s2">&quot;Calculating adjustments for all sites AN and GQ/DP histograms...&quot;</span>
                <span class="p">)</span>
                <span class="n">compute_an_and_hists_het_fail_adj_ab</span><span class="p">(</span>
                    <span class="n">vds</span><span class="p">,</span>
                    <span class="n">group_membership_ht</span><span class="p">,</span>
                    <span class="n">freq_ht</span><span class="p">,</span>
                <span class="p">)</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;gs://gnomad</span><span class="si">{</span><span class="s1">&#39;-tmp-4day&#39;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">test</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;&#39;</span><span class="si">}</span><span class="s2">/v4.1/temp/frequency_fix/het_fail_adj_ab_ht.ht&quot;</span><span class="p">,</span>
                    <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">,</span>
                <span class="p">)</span>

            <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">calculate_all_sites_het_nonref_adjustment</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Calculating all sites het non ref adjustment...&quot;</span><span class="p">)</span>
                <span class="n">get_all_sites_nonref_adjustment</span><span class="p">(</span>
                    <span class="n">vds</span><span class="p">,</span>
                    <span class="n">group_membership_ht</span><span class="p">,</span>
                <span class="p">)</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;gs://gnomad</span><span class="si">{</span><span class="s1">&#39;-tmp-4day&#39;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">test</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;&#39;</span><span class="si">}</span><span class="s2">/v4.1/temp/frequency_fix/het_nonref_adjust.ht&quot;</span><span class="p">,</span>
                    <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">,</span>
                <span class="p">)</span>

        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">adjust_all_sites_an_and_hists</span><span class="p">:</span>
            <span class="n">an_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;gs://gnomad</span><span class="si">{</span><span class="s1">&#39;-tmp-4day&#39;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">test</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;&#39;</span><span class="si">}</span><span class="s2">/v4.1/temp/frequency_fix/all_sites_an_before_adjustment.ht&quot;</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">het_fail_adj_ab_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;gs://gnomad</span><span class="si">{</span><span class="s1">&#39;-tmp-4day&#39;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">test</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;&#39;</span><span class="si">}</span><span class="s2">/v4.1/temp/frequency_fix/het_fail_adj_ab_ht.ht&quot;</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">het_nonref_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;gs://gnomad</span><span class="si">{</span><span class="s1">&#39;-tmp-4day&#39;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">test</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;&#39;</span><span class="si">}</span><span class="s2">/v4.1/temp/frequency_fix/het_nonref_adjust.ht&quot;</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">an_ht</span> <span class="o">=</span> <span class="n">adjust_all_sites_an_and_hists</span><span class="p">(</span>
                <span class="n">an_ht</span><span class="p">,</span> <span class="n">het_fail_adj_ab_ht</span><span class="p">,</span> <span class="n">het_nonref_ht</span>
            <span class="p">)</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;gs://gnomad</span><span class="si">{</span><span class="s1">&#39;-tmp-4day&#39;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">test</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;&#39;</span><span class="si">}</span><span class="s2">/v4.1/temp/frequency_fix/all_sites_an_after_adjustment.ht&quot;</span><span class="p">,</span>
                <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">an_ht</span> <span class="o">=</span> <span class="n">an_ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;AN&quot;</span><span class="p">,</span> <span class="s2">&quot;qual_hists&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span>
                <span class="n">get_all_sites_an_and_qual_hists</span><span class="p">(</span><span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">)</span><span class="o">.</span><span class="n">path</span><span class="p">,</span>
                <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">an_ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;AN&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="n">release_all_sites_an</span><span class="p">(</span><span class="n">public</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">)</span><span class="o">.</span><span class="n">path</span><span class="p">,</span>
                <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">adjust_freq_an_hists</span><span class="p">:</span>
            <span class="n">an_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;gs://gnomad</span><span class="si">{</span><span class="s1">&#39;-tmp-4day&#39;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">test</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;&#39;</span><span class="si">}</span><span class="s2">/v4.1/temp/frequency_fix/all_sites_an_before_adjustment.ht&quot;</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">het_fail_adj_ab_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;gs://gnomad</span><span class="si">{</span><span class="s1">&#39;-tmp-4day&#39;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">test</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;&#39;</span><span class="si">}</span><span class="s2">/v4.1/temp/frequency_fix/het_fail_adj_ab_ht.ht&quot;</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="n">freq_correction_ht</span> <span class="o">=</span> <span class="n">adjust_per_site_an_and_hists_for_frequency</span><span class="p">(</span>
                <span class="n">an_ht</span><span class="p">,</span> <span class="n">freq_ht</span><span class="p">,</span> <span class="n">het_fail_adj_ab_ht</span>
            <span class="p">)</span>
            <span class="n">freq_correction_ht</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;gs://gnomad</span><span class="si">{</span><span class="s1">&#39;-tmp-4day&#39;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">test</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;&#39;</span><span class="si">}</span><span class="s2">/v4.1/temp/frequency_fix/freq_correction.ht&quot;</span><span class="p">,</span>
                <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">update_freq_an_and_hists</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Updating AN in freq HT...&quot;</span><span class="p">)</span>

            <span class="n">freq_correction_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;gs://gnomad</span><span class="si">{</span><span class="s1">&#39;-tmp-4day&#39;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">test</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;&#39;</span><span class="si">}</span><span class="s2">/v4.1/temp/frequency_fix/freq_correction.ht&quot;</span>
            <span class="p">)</span>
            <span class="c1"># The new correction HT has used &quot;gen_anc&quot; as the pop key from the</span>
            <span class="c1"># start so replacing it with &quot;pop&quot; here to match the freq HT. Updating the</span>
            <span class="c1"># freq_ht to &quot;gen_anc&quot; is the appropriate fix but it requires far more code</span>
            <span class="c1"># updates to the imported functions of generate_freq.</span>
            <span class="n">freq_correction_ht</span> <span class="o">=</span> <span class="n">freq_correction_ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span>
                <span class="n">strata_meta</span><span class="o">=</span><span class="n">freq_correction_ht</span><span class="o">.</span><span class="n">strata_meta</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
                    <span class="k">lambda</span> <span class="n">d</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">dict</span><span class="p">(</span>
                        <span class="n">d</span><span class="o">.</span><span class="n">items</span><span class="p">()</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
                            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;gen_anc&quot;</span><span class="p">,</span> <span class="p">(</span><span class="s2">&quot;pop&quot;</span><span class="p">,</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span> <span class="n">x</span><span class="p">)</span>
                        <span class="p">)</span>
                    <span class="p">)</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="n">freq_ht</span> <span class="o">=</span> <span class="n">get_freq</span><span class="p">(</span>
                <span class="n">version</span><span class="o">=</span><span class="s2">&quot;4.0&quot;</span><span class="p">,</span>
                <span class="n">hom_alt_adjusted</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                <span class="n">finalized</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="p">)</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>

            <span class="k">if</span> <span class="n">test</span><span class="p">:</span>
                <span class="n">freq_ht</span> <span class="o">=</span> <span class="n">freq_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">freq_correction_ht</span><span class="p">[</span><span class="n">freq_ht</span><span class="o">.</span><span class="n">key</span><span class="p">]))</span>

            <span class="n">freq_ht</span> <span class="o">=</span> <span class="n">drop_gatk_groupings</span><span class="p">(</span><span class="n">freq_ht</span><span class="p">)</span>
            <span class="n">ht</span> <span class="o">=</span> <span class="n">update_freq_an_and_hists</span><span class="p">(</span><span class="n">freq_ht</span><span class="p">,</span> <span class="n">freq_correction_ht</span><span class="p">)</span>

            <span class="n">ht</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="n">get_freq</span><span class="p">(</span>
                    <span class="n">version</span><span class="o">=</span><span class="s2">&quot;4.1&quot;</span><span class="p">,</span>
                    <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">,</span>
                    <span class="n">hom_alt_adjusted</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                    <span class="n">finalized</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                <span class="p">)</span><span class="o">.</span><span class="n">path</span><span class="p">,</span>
                <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">update_af_dependent_anns</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Updating AF dependent annotations...&quot;</span><span class="p">)</span>

            <span class="n">ht</span> <span class="o">=</span> <span class="n">get_freq</span><span class="p">(</span>
                <span class="n">version</span><span class="o">=</span><span class="s2">&quot;4.1&quot;</span><span class="p">,</span>
                <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">,</span>
                <span class="n">hom_alt_adjusted</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                <span class="n">finalized</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="p">)</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Setting XX samples call stats to missing on chrY...&quot;</span><span class="p">)</span>
            <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">freq</span><span class="o">=</span><span class="n">set_xx_y_metrics_to_na_expr</span><span class="p">(</span><span class="n">ht</span><span class="p">))</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Correcting call stats, qual AB hists, and age hists...&quot;</span><span class="p">)</span>
            <span class="n">ht</span> <span class="o">=</span> <span class="n">correct_for_high_ab_hets</span><span class="p">(</span><span class="n">ht</span><span class="p">,</span> <span class="n">af_threshold</span><span class="o">=</span><span class="n">af_threshold</span><span class="p">)</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Computing FAF &amp; grpmax...&quot;</span><span class="p">)</span>
            <span class="n">ht</span> <span class="o">=</span> <span class="n">generate_faf_grpmax</span><span class="p">(</span><span class="n">ht</span><span class="p">)</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Calculating InbreedingCoeff...&quot;</span><span class="p">)</span>
            <span class="n">ht</span> <span class="o">=</span> <span class="n">compute_inbreeding_coeff</span><span class="p">(</span><span class="n">ht</span><span class="p">)</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;High AB het corrected frequency HT schema...&quot;</span><span class="p">)</span>
            <span class="n">ht</span><span class="o">.</span><span class="n">describe</span><span class="p">()</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Writing corrected frequency Table...&quot;</span><span class="p">)</span>
            <span class="n">ht</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="n">get_freq</span><span class="p">(</span>
                    <span class="n">version</span><span class="o">=</span><span class="s2">&quot;4.1&quot;</span><span class="p">,</span>
                    <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">,</span>
                    <span class="n">hom_alt_adjusted</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                    <span class="n">finalized</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                <span class="p">)</span><span class="o">.</span><span class="n">path</span><span class="p">,</span>
                <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">finalize_freq_ht</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Writing final frequency Table...&quot;</span><span class="p">)</span>
            <span class="n">freq_ht</span> <span class="o">=</span> <span class="n">get_freq</span><span class="p">(</span>
                <span class="n">version</span><span class="o">=</span><span class="s2">&quot;4.1&quot;</span><span class="p">,</span>
                <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">,</span>
                <span class="n">hom_alt_adjusted</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">finalized</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="p">)</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>
            <span class="n">freq_ht</span> <span class="o">=</span> <span class="n">create_final_freq_ht</span><span class="p">(</span><span class="n">freq_ht</span><span class="p">)</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Final frequency HT schema...&quot;</span><span class="p">)</span>
            <span class="n">freq_ht</span><span class="o">.</span><span class="n">describe</span><span class="p">()</span>
            <span class="n">freq_ht</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="n">get_freq</span><span class="p">(</span><span class="n">version</span><span class="o">=</span><span class="s2">&quot;4.1&quot;</span><span class="p">,</span> <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">,</span> <span class="n">finalized</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">path</span><span class="p">,</span>
                <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">,</span>
            <span class="p">)</span>

    <span class="k">finally</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Copying log to logging bucket...&quot;</span><span class="p">)</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">copy_log</span><span class="p">(</span><span class="n">get_logging_path</span><span class="p">(</span><span class="s2">&quot;frequency_fix&quot;</span><span class="p">))</span></div>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">()</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--use-test-dataset&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Runs a test on the gnomad test dataset.&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--test-n-partitions&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span>
            <span class="s2">&quot;Use only N partitions of the VDS as input for testing purposes. Defaults&quot;</span>
            <span class="s2">&quot;to 2 if passed without a value.&quot;</span>
        <span class="p">),</span>
        <span class="n">nargs</span><span class="o">=</span><span class="s2">&quot;?&quot;</span><span class="p">,</span>
        <span class="n">const</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--test-chr22-chrx-chry&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span>
            <span class="s2">&quot;Whether to run a test using only the chr22, chrX, and chrY chromosomes of&quot;</span>
            <span class="s2">&quot; the VDS test dataset.&quot;</span>
        <span class="p">),</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--overwrite&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Overwrites existing files.&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--regenerate-ref-an-hists&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Calculate reference allele number and GQ/DP histograms for all sites.&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--calculate-var-an-hists-adjustment&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span>
            <span class="s2">&quot;Calculate adjustments that need to be made to the all sites AN and GQ/DP &quot;</span>
            <span class="s2">&quot;histograms in order to be used for the multi-allelic split frequency &quot;</span>
            <span class="s2">&quot;Table.&quot;</span>
        <span class="p">),</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--calculate-all-sites-het-nonref-adjustment&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span>
            <span class="s2">&quot;Calculate the number of het non-ref genotypes that are being counted &quot;</span>
            <span class="s2">&quot;twice so the all sites AN and histograms can be adjusted correctly.&quot;</span>
        <span class="p">),</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--adjust-all-sites-an-and-hists&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Adjust the all sites AN and GQ/DP histograms for the frequency Table.&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--adjust-freq-an-hists&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Adjust the all sites AN and GQ/DP histograms for the frequency Table.&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--update-freq-an-and-hists&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span>
            <span class="s2">&quot;Update frequency HT AN field and GQ and DP hists with the correct AN and&quot;</span>
            <span class="s2">&quot; hists from the AN HT.&quot;</span>
        <span class="p">),</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--update-af-dependent-anns&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span>
            <span class="s2">&quot;Update AF dependent annotations in frequency HT with the new AF based on&quot;</span>
            <span class="s2">&quot; updated AN.&quot;</span>
        <span class="p">),</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--af-threshold&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span>
            <span class="s2">&quot;Threshold at which to adjust site group frequencies at sites for&quot;</span>
            <span class="s2">&quot; homozygous alternate depletion present in GATK versions released prior to&quot;</span>
            <span class="s2">&quot; 4.1.4.1.&quot;</span>
        <span class="p">),</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--finalize-freq-ht&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span>
            <span class="s2">&quot;Finalize frequency Table by dropping unnecessary fields and renaming&quot;</span>
            <span class="s2">&quot; remaining fields.&quot;</span>
        <span class="p">),</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>

    <span class="n">main</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p></p>
  </div>

   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>