<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>gnomad_qc.v4.annotations.generate_freq_genomes &mdash; gnomad_qc master documentation</title>
      <link rel="stylesheet" href="../../../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../../../_static/theme_overrides.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../../_static/jquery.js"></script>
        <script src="../../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../../../../" id="documentation_options" src="../../../../_static/documentation_options.js"></script>
        <script src="../../../../_static/doctools.js"></script>
        <script src="../../../../_static/sphinx_highlight.js"></script>
        <script src="https://unpkg.com/mermaid@10.2.0/dist/mermaid.min.js"></script>
        <script>mermaid.initialize({startOnLoad:true});</script>
    <script src="../../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../../index.html" class="icon icon-home">
            gnomad_qc
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../api_reference/index.html">API Reference</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/broadinstitute/gnomad_qc/tags">Change Log</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../../index.html">gnomad_qc</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">gnomad_qc.v4.annotations.generate_freq_genomes</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for gnomad_qc.v4.annotations.generate_freq_genomes</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Script to create frequencies HT for v4.0 genomes.</span>

<span class="sd">This script is written specifically for the v4.0 genomes release to handle the</span>
<span class="sd">addition and subtraction of samples from the HGDP + 1KG subset. The HGDP + 1KG</span>
<span class="sd">subset has updated sample QC annotations. The addition of new samples from the</span>
<span class="sd">densified MT will include new variants, which will require the recomputation of AN</span>
<span class="sd">for the v3.1 releases samples, and merging it to call stats of the v3.1.2 release</span>
<span class="sd">sites HT and of updated HGDP + 1KG subset. In addition, the code will also get the</span>
<span class="sd">other call stats related annotations, including: filtering allele frequencies (&#39;faf&#39;),</span>
<span class="sd">&#39;grpmax&#39;, &#39;gen_anc_faf_max&#39; and &#39;InbreedingCoeff&#39; as done in v4.0 exomes.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Any</span><span class="p">,</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">List</span><span class="p">,</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">,</span> <span class="n">Union</span>

<span class="kn">import</span> <span class="nn">hail</span> <span class="k">as</span> <span class="nn">hl</span>
<span class="kn">from</span> <span class="nn">gnomad.resources.grch38.gnomad</span> <span class="kn">import</span> <span class="n">GEN_ANC_GROUPS_TO_REMOVE_FOR_GRPMAX</span><span class="p">,</span> <span class="n">SUBSETS</span>
<span class="kn">from</span> <span class="nn">gnomad.sample_qc.sex</span> <span class="kn">import</span> <span class="n">adjust_sex_ploidy</span>
<span class="kn">from</span> <span class="nn">gnomad.utils.annotations</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">annotate_adj</span><span class="p">,</span>
    <span class="n">annotate_freq</span><span class="p">,</span>
    <span class="n">bi_allelic_site_inbreeding_expr</span><span class="p">,</span>
    <span class="n">faf_expr</span><span class="p">,</span>
    <span class="n">gen_anc_faf_max_expr</span><span class="p">,</span>
    <span class="n">generate_freq_group_membership_array</span><span class="p">,</span>
    <span class="n">grpmax_expr</span><span class="p">,</span>
    <span class="n">merge_freq_arrays</span><span class="p">,</span>
    <span class="n">missing_callstats_expr</span><span class="p">,</span>
    <span class="n">set_xx_y_metrics_to_na_expr</span><span class="p">,</span>
    <span class="n">update_structured_annotations</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">gnomad.utils.filtering</span> <span class="kn">import</span> <span class="n">filter_arrays_by_meta</span>
<span class="kn">from</span> <span class="nn">gnomad.utils.release</span> <span class="kn">import</span> <span class="n">make_freq_index_dict_from_meta</span>
<span class="kn">from</span> <span class="nn">gnomad.utils.slack</span> <span class="kn">import</span> <span class="n">slack_notifications</span>
<span class="kn">from</span> <span class="nn">hail.utils</span> <span class="kn">import</span> <span class="n">new_temp_file</span>

<span class="kn">from</span> <span class="nn">gnomad_qc.resource_utils</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">PipelineResourceCollection</span><span class="p">,</span>
    <span class="n">PipelineStepResourceCollection</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.slack_creds</span> <span class="kn">import</span> <span class="n">slack_token</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v3.resources.annotations</span> <span class="kn">import</span> <span class="n">get_freq</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v3.resources.basics</span> <span class="kn">import</span> <span class="n">get_gnomad_v3_vds</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v3.resources.basics</span> <span class="kn">import</span> <span class="n">meta</span> <span class="k">as</span> <span class="n">v3_meta</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v3.resources.release</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">hgdp_tgp_subset</span><span class="p">,</span>
    <span class="n">hgdp_tgp_subset_annotations</span><span class="p">,</span>
    <span class="n">release_sites</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v3.resources.sample_qc</span> <span class="kn">import</span> <span class="n">relatedness</span> <span class="k">as</span> <span class="n">v3_relatedness</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v3.utils</span> <span class="kn">import</span> <span class="n">hom_alt_depletion_fix</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v4.resources.annotations</span> <span class="kn">import</span> <span class="n">get_freq</span> <span class="k">as</span> <span class="n">v4_get_freq</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v4.resources.annotations</span> <span class="kn">import</span> <span class="n">hgdp_tgp_updated_callstats</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v4.resources.basics</span> <span class="kn">import</span> <span class="n">meta</span> <span class="k">as</span> <span class="n">v4_meta</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v4.resources.sample_qc</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">hgdp_recomputed_freemix</span><span class="p">,</span>
    <span class="n">hgdp_tgp_duplicated_to_exomes</span><span class="p">,</span>
    <span class="n">hgdp_tgp_meta_updated</span><span class="p">,</span>
    <span class="n">hgdp_tgp_pop_outliers</span><span class="p">,</span>
    <span class="n">hgdp_tgp_populations_updated</span><span class="p">,</span>
    <span class="n">hgdp_tgp_related_samples_to_drop</span><span class="p">,</span>
    <span class="n">hgdp_tgp_related_to_nonsubset</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v4.resources.sample_qc</span> <span class="kn">import</span> <span class="n">relatedness</span> <span class="k">as</span> <span class="n">v4_relatedness</span>

<span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="nb">format</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%(levelname)s</span><span class="s2"> (</span><span class="si">%(name)s</span><span class="s2"> </span><span class="si">%(lineno)s</span><span class="s2">): </span><span class="si">%(message)s</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span>
    <span class="s2">&quot;Create v4.0 frequencies HT with updated HGDP/TGP metadata and new annotations&quot;</span>
<span class="p">)</span>
<span class="n">logger</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>

<span class="n">SUBSETS</span> <span class="o">=</span> <span class="n">SUBSETS</span><span class="p">[</span><span class="s2">&quot;v3&quot;</span><span class="p">]</span>
<span class="sd">&quot;&quot;&quot;List of subsets in the v3.1/v4.0 genomes release.&quot;&quot;&quot;</span>
<span class="n">POP_MAP</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;bantusafrica&quot;</span><span class="p">:</span> <span class="s2">&quot;bantusouthafrica&quot;</span><span class="p">,</span>
    <span class="s2">&quot;biakaPygmy&quot;</span><span class="p">:</span> <span class="s2">&quot;biaka&quot;</span><span class="p">,</span>
    <span class="s2">&quot;italian&quot;</span><span class="p">:</span> <span class="s2">&quot;bergamoitalian&quot;</span><span class="p">,</span>
    <span class="s2">&quot;mbutiPygmy&quot;</span><span class="p">:</span> <span class="s2">&quot;mbuti&quot;</span><span class="p">,</span>
    <span class="s2">&quot;melanesian&quot;</span><span class="p">:</span> <span class="s2">&quot;bougainville&quot;</span><span class="p">,</span>
    <span class="s2">&quot;mongola&quot;</span><span class="p">:</span> <span class="s2">&quot;mongolian&quot;</span><span class="p">,</span>
    <span class="s2">&quot;miaozu&quot;</span><span class="p">:</span> <span class="s2">&quot;miao&quot;</span><span class="p">,</span>
    <span class="s2">&quot;yizu&quot;</span><span class="p">:</span> <span class="s2">&quot;yi&quot;</span><span class="p">,</span>
    <span class="s2">&quot;oth&quot;</span><span class="p">:</span> <span class="s2">&quot;remaining&quot;</span><span class="p">,</span>
<span class="p">}</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Map of HGDP populations that need to be renamed in the v4.0 genomes release. Also includes</span>
<span class="sd">the renaming of &#39;oth&#39; to &#39;remaining&#39;.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="n">JOIN_FREQS</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;release&quot;</span><span class="p">,</span> <span class="s2">&quot;pop_diff&quot;</span><span class="p">,</span> <span class="s2">&quot;added&quot;</span><span class="p">,</span> <span class="s2">&quot;subtracted&quot;</span><span class="p">]</span>
<span class="sd">&quot;&quot;&quot;Frequency Tables to join for creation of the v4.0 genomes release sites HT.&quot;&quot;&quot;</span>
<span class="n">FREQ_GLOBALS</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;freq_meta&quot;</span><span class="p">,</span> <span class="s2">&quot;freq_meta_sample_count&quot;</span><span class="p">)</span>
<span class="sd">&quot;&quot;&quot;Global annotations on the frequency Table.&quot;&quot;&quot;</span>


<div class="viewcode-block" id="filter_to_test"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/generate_freq_genomes.html#gnomad_qc.v4.annotations.generate_freq_genomes.filter_to_test">[docs]</a><span class="k">def</span> <span class="nf">filter_to_test</span><span class="p">(</span>
    <span class="n">t</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">VariantDataset</span><span class="p">],</span>
    <span class="n">gene_on_chrx</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Union</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">VariantDataset</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Filter to a region in PCSK9 or TRPC5 in Table or MatrixTable for testing purposes.</span>

<span class="sd">    :param t: Table or MatrixTable to filter.</span>
<span class="sd">    :param gene_on_chrx: Whether to filter to TRPC5 (in the non-PAR region), instead of</span>
<span class="sd">        PCSK9, for testing chrX.</span>
<span class="sd">    :return: Table or MatrixTable filtered to a region in PCSK9 or TRPC5.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">gene_on_chrx</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Filtering to TRPC5 on chrX in MT for testing purposes...&quot;</span><span class="p">)</span>
        <span class="n">test_locus</span> <span class="o">=</span> <span class="s2">&quot;chrX:111776000-111786000&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Filtering to 10kb in PCSK9 in MT for testing purposes...&quot;</span><span class="p">)</span>
        <span class="n">test_locus</span> <span class="o">=</span> <span class="s2">&quot;chr1:55039447-55064852&quot;</span>

    <span class="n">test_interval</span> <span class="o">=</span> <span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">parse_locus_interval</span><span class="p">(</span><span class="n">test_locus</span><span class="p">,</span> <span class="n">reference_genome</span><span class="o">=</span><span class="s2">&quot;GRCh38&quot;</span><span class="p">)]</span>

    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">VariantDataset</span><span class="p">):</span>
        <span class="n">t</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">filter_intervals</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">test_interval</span><span class="p">,</span> <span class="n">split_reference_blocks</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">t</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">filter_intervals</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">test_interval</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">t</span></div>


<div class="viewcode-block" id="get_hgdp_tgp_related_to_nonsubset"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/generate_freq_genomes.html#gnomad_qc.v4.annotations.generate_freq_genomes.get_hgdp_tgp_related_to_nonsubset">[docs]</a><span class="k">def</span> <span class="nf">get_hgdp_tgp_related_to_nonsubset</span><span class="p">(</span>
    <span class="n">v3_meta_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">rel_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get the samples in the HGDP + 1KG subset that were related to samples outside the subset in the v3.1 release and were not included in the v3.1 release.</span>

<span class="sd">    :param v3_meta_ht: Table with the v3.1 release metadata.</span>
<span class="sd">    :param rel_ht: Table with the v3.1 release relatedness, here we use the results</span>
<span class="sd">        from pc_relate of the full dataset.</span>
<span class="sd">    :return: Table with the samples in the HGDP + 1KG subset that are related to</span>
<span class="sd">        samples outside the subset in v3 release.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Filter to samples that are either in the HGDP + 1KG subset or in the v3.1 release.</span>
    <span class="n">v3_meta_ht</span> <span class="o">=</span> <span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="n">hgdp_tgp</span><span class="o">=</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">subsets</span><span class="o">.</span><span class="n">tgp</span> <span class="o">|</span> <span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">subsets</span><span class="o">.</span><span class="n">hgdp</span>
    <span class="p">)</span>
    <span class="n">v3_meta_ht</span> <span class="o">=</span> <span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">release</span> <span class="o">|</span> <span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">hgdp_tgp</span><span class="p">)</span>

    <span class="n">rel_ht</span> <span class="o">=</span> <span class="n">rel_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="o">**</span><span class="p">{</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s2">_meta&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                <span class="n">v3_release</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">coalesce</span><span class="p">(</span><span class="n">v3_meta_ht</span><span class="p">[</span><span class="n">rel_ht</span><span class="p">[</span><span class="n">x</span><span class="p">]</span><span class="o">.</span><span class="n">s</span><span class="p">]</span><span class="o">.</span><span class="n">release</span><span class="p">,</span> <span class="kc">False</span><span class="p">),</span>
                <span class="n">hgdp_tgp</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">coalesce</span><span class="p">(</span><span class="n">v3_meta_ht</span><span class="p">[</span><span class="n">rel_ht</span><span class="p">[</span><span class="n">x</span><span class="p">]</span><span class="o">.</span><span class="n">s</span><span class="p">]</span><span class="o">.</span><span class="n">hgdp_tgp</span><span class="p">,</span> <span class="kc">False</span><span class="p">),</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;i&quot;</span><span class="p">,</span> <span class="s2">&quot;j&quot;</span><span class="p">]</span>
        <span class="p">}</span>
    <span class="p">)</span>

    <span class="n">rel_ht</span> <span class="o">=</span> <span class="n">rel_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
        <span class="c1"># Filter to pairs with 2nd degree or closer relatedness.</span>
        <span class="p">(</span><span class="n">rel_ht</span><span class="o">.</span><span class="n">relationship</span> <span class="o">!=</span> <span class="s2">&quot;unrelated&quot;</span><span class="p">)</span>
        <span class="c1"># Filter to pairs where at least one of the samples was in the v3.1 release.</span>
        <span class="o">&amp;</span> <span class="p">(</span><span class="n">rel_ht</span><span class="o">.</span><span class="n">i_meta</span><span class="o">.</span><span class="n">v3_release</span> <span class="o">|</span> <span class="n">rel_ht</span><span class="o">.</span><span class="n">j_meta</span><span class="o">.</span><span class="n">v3_release</span><span class="p">)</span>
        <span class="c1"># Filter to pairs where one and only one of the samples is in the HGDP +</span>
        <span class="c1"># 1KG subset.</span>
        <span class="o">&amp;</span> <span class="p">((</span><span class="n">hl</span><span class="o">.</span><span class="n">int</span><span class="p">(</span><span class="n">rel_ht</span><span class="o">.</span><span class="n">i_meta</span><span class="o">.</span><span class="n">hgdp_tgp</span><span class="p">)</span> <span class="o">+</span> <span class="n">hl</span><span class="o">.</span><span class="n">int</span><span class="p">(</span><span class="n">rel_ht</span><span class="o">.</span><span class="n">j_meta</span><span class="o">.</span><span class="n">hgdp_tgp</span><span class="p">))</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span>
        <span class="c1"># Exclude pairs where the HGDP/1KG sample in the pair is also a v3.1 release</span>
        <span class="c1"># sample.</span>
        <span class="o">&amp;</span> <span class="o">~</span><span class="p">(</span><span class="n">rel_ht</span><span class="o">.</span><span class="n">i_meta</span><span class="o">.</span><span class="n">hgdp_tgp</span> <span class="o">&amp;</span> <span class="n">rel_ht</span><span class="o">.</span><span class="n">i_meta</span><span class="o">.</span><span class="n">v3_release</span><span class="p">)</span>
        <span class="o">&amp;</span> <span class="o">~</span><span class="p">(</span><span class="n">rel_ht</span><span class="o">.</span><span class="n">j_meta</span><span class="o">.</span><span class="n">hgdp_tgp</span> <span class="o">&amp;</span> <span class="n">rel_ht</span><span class="o">.</span><span class="n">j_meta</span><span class="o">.</span><span class="n">v3_release</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="c1"># Filter the relatedness HT twice to get the HGDP/1KG sample in each pair.</span>
    <span class="n">ht1</span> <span class="o">=</span> <span class="n">rel_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">rel_ht</span><span class="o">.</span><span class="n">i_meta</span><span class="o">.</span><span class="n">hgdp_tgp</span><span class="p">)</span><span class="o">.</span><span class="n">key_by</span><span class="p">()</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;i&quot;</span><span class="p">)</span>
    <span class="n">ht1</span> <span class="o">=</span> <span class="n">ht1</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="n">s</span><span class="o">=</span><span class="n">ht1</span><span class="o">.</span><span class="n">i</span><span class="o">.</span><span class="n">s</span><span class="p">)</span>
    <span class="n">ht2</span> <span class="o">=</span> <span class="n">rel_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">rel_ht</span><span class="o">.</span><span class="n">j_meta</span><span class="o">.</span><span class="n">hgdp_tgp</span><span class="p">)</span><span class="o">.</span><span class="n">key_by</span><span class="p">()</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;j&quot;</span><span class="p">)</span>
    <span class="n">ht2</span> <span class="o">=</span> <span class="n">ht2</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="n">s</span><span class="o">=</span><span class="n">ht2</span><span class="o">.</span><span class="n">j</span><span class="o">.</span><span class="n">s</span><span class="p">)</span>

    <span class="c1"># Merge the two HTs and remove the v3.1 prefix from the sample IDs.</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht1</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">ht2</span><span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="n">s</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">s</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;v3.1::&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="s2">&quot;s&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">distinct</span><span class="p">()</span>

    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">naive_coalesce</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span><span class="n">new_temp_file</span><span class="p">(</span><span class="s2">&quot;related_to_nonsubset_ht&quot;</span><span class="p">))</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;</span><span class="si">%d</span><span class="s2"> HGDP/1KG samples are related to samples outside the subset&quot;</span><span class="p">,</span> <span class="n">ht</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">ht</span><span class="o">.</span><span class="n">select_globals</span><span class="p">()</span></div>


<div class="viewcode-block" id="get_hgdp_tgp_v4_exome_duplicates"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/generate_freq_genomes.html#gnomad_qc.v4.annotations.generate_freq_genomes.get_hgdp_tgp_v4_exome_duplicates">[docs]</a><span class="k">def</span> <span class="nf">get_hgdp_tgp_v4_exome_duplicates</span><span class="p">(</span>
    <span class="n">v3_meta_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">v4_meta_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">rel_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get the samples in the HGDP + 1KG subset that are duplicates of an exome in the v4.0 release.</span>

<span class="sd">    The duplicated samples are defined as samples that were HGDP + 1KG subset samples</span>
<span class="sd">    in the v3.1 release and are also in the v4.0 exomes release. The duplicated samples</span>
<span class="sd">    have to be removed because we will have combined frequencies rom v4.0 exomes and</span>
<span class="sd">    genomes.</span>

<span class="sd">    :param v3_meta_ht: Table with the v3.1 release metadata.</span>
<span class="sd">    :param v4_meta_ht: Table with the v4.0 exomes release metadata.</span>
<span class="sd">    :param rel_ht: Table with the v3.1 and v4.0 joint relatedness, it&#39;s based on</span>
<span class="sd">        cuKING relatedness results.</span>
<span class="sd">    :return: Table with the samples in the HGDP + 1KG subset that are duplicates of an</span>
<span class="sd">        exome in the v4.0 exomes release.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Get HGDP + 1KG subset samples in v3.1 meta.</span>
    <span class="n">v3_meta_ht</span> <span class="o">=</span> <span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">subsets</span><span class="o">.</span><span class="n">hgdp</span> <span class="o">|</span> <span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">subsets</span><span class="o">.</span><span class="n">tgp</span><span class="p">)</span>

    <span class="c1"># Get release samples in v4.0 exomes.</span>
    <span class="n">v4_meta_ht</span> <span class="o">=</span> <span class="n">v4_meta_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">v4_meta_ht</span><span class="o">.</span><span class="n">release</span><span class="p">)</span>

    <span class="c1"># Get samples that are in the v3.1 genomes and are also in the v4.0 exomes.</span>
    <span class="n">rel_ht</span> <span class="o">=</span> <span class="n">rel_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">rel_ht</span><span class="o">.</span><span class="n">gnomad_v3_duplicate</span><span class="p">)</span>

    <span class="c1"># Check if the duplicates are included in the v4.0 exomes release and belong to the</span>
    <span class="c1"># HGDP + 1KG subset.</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">rel_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="o">**</span><span class="p">{</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s2">_meta&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                <span class="n">in_v4_exomes_release</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">v4_meta_ht</span><span class="p">[</span><span class="n">rel_ht</span><span class="p">[</span><span class="n">x</span><span class="p">]</span><span class="o">.</span><span class="n">s</span><span class="p">]),</span>
                <span class="n">in_hgdp_tgp_subset</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">v3_meta_ht</span><span class="p">[</span><span class="n">rel_ht</span><span class="p">[</span><span class="n">x</span><span class="p">]</span><span class="o">.</span><span class="n">s</span><span class="p">]),</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;i&quot;</span><span class="p">,</span> <span class="s2">&quot;j&quot;</span><span class="p">]</span>
        <span class="p">}</span>
    <span class="p">)</span>

    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
        <span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">i_meta</span><span class="o">.</span><span class="n">in_v4_exomes_release</span> <span class="o">&amp;</span> <span class="n">ht</span><span class="o">.</span><span class="n">j_meta</span><span class="o">.</span><span class="n">in_hgdp_tgp_subset</span><span class="p">)</span>
        <span class="o">|</span> <span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">j_meta</span><span class="o">.</span><span class="n">in_v4_exomes_release</span> <span class="o">&amp;</span> <span class="n">ht</span><span class="o">.</span><span class="n">i_meta</span><span class="o">.</span><span class="n">in_hgdp_tgp_subset</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="c1"># Filter the relatedness HT twice to get the HGDP/1KG sample in each pair.</span>
    <span class="n">ht1</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">i_meta</span><span class="o">.</span><span class="n">in_hgdp_tgp_subset</span><span class="p">)</span><span class="o">.</span><span class="n">key_by</span><span class="p">()</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;i&quot;</span><span class="p">)</span>
    <span class="n">ht1</span> <span class="o">=</span> <span class="n">ht1</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="n">s</span><span class="o">=</span><span class="n">ht1</span><span class="o">.</span><span class="n">i</span><span class="o">.</span><span class="n">s</span><span class="p">)</span>
    <span class="c1"># This table is likely empty because of the ordering when computing relatedness,</span>
    <span class="c1"># but included for completeness.</span>
    <span class="n">ht2</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">j_meta</span><span class="o">.</span><span class="n">in_hgdp_tgp_subset</span><span class="p">)</span><span class="o">.</span><span class="n">key_by</span><span class="p">()</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;j&quot;</span><span class="p">)</span>
    <span class="n">ht2</span> <span class="o">=</span> <span class="n">ht2</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="n">s</span><span class="o">=</span><span class="n">ht2</span><span class="o">.</span><span class="n">j</span><span class="o">.</span><span class="n">s</span><span class="p">)</span>

    <span class="c1"># Merge the two HTs and remove the v3.1 prefix from the sample IDs.</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht1</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">ht2</span><span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="n">s</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">s</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;v3.1::&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="s2">&quot;s&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">distinct</span><span class="p">()</span>

    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">naive_coalesce</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span><span class="n">new_temp_file</span><span class="p">(</span><span class="s2">&quot;duplicate_in_v4_exomes&quot;</span><span class="p">))</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;</span><span class="si">%d</span><span class="s2"> HGDP/1KG samples are duplicated in the v4.0 exomes release&quot;</span><span class="p">,</span> <span class="n">ht</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">ht</span><span class="o">.</span><span class="n">select_globals</span><span class="p">()</span></div>


<div class="viewcode-block" id="add_updated_sample_qc_annotations"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/generate_freq_genomes.html#gnomad_qc.v4.annotations.generate_freq_genomes.add_updated_sample_qc_annotations">[docs]</a><span class="k">def</span> <span class="nf">add_updated_sample_qc_annotations</span><span class="p">(</span><span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Add updated sample QC annotations to the HGDP + 1KG subset.</span>

<span class="sd">    .. note::</span>

<span class="sd">        The following annotations need to be updated for the v4.0 genomes release based</span>
<span class="sd">        on the latest sample QC results of the subset:</span>

<span class="sd">            - `hgdp_tgp_meta.subcontinental_pca.outlier`: to apply the updated pop</span>
<span class="sd">              outlier filter implemented by Alicia Martin&#39;s group.</span>
<span class="sd">            - `gnomad_sample_filters.hard_filtered`: to apply the recomputed freemix</span>
<span class="sd">              filter for HGDP samples.</span>
<span class="sd">            - `gnomad_sample_filters.related_to_nonsubset`: to further filter out</span>
<span class="sd">              samples that are related to samples outside the subset but were not</span>
<span class="sd">              included in the v3 release.</span>
<span class="sd">            - `gnomad_sample_filters.v4_exome_duplicate`: to further filter out the</span>
<span class="sd">              samples in the HGDP + 1KG subset that are duplicates of an exome in the</span>
<span class="sd">              v4.0 release.</span>
<span class="sd">            - `relatedness_inference.related`: to apply the updated relatedness</span>
<span class="sd">              inference implemented by Alicia Martin&#39;s group.</span>

<span class="sd">    :param ht: Table with the HGDP + 1KG subset metadata from the v3.1.2 release.</span>
<span class="sd">    :return: Table with updated sample QC annotations.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Load all updated sample QC Tables.</span>
    <span class="n">contamination_ht</span> <span class="o">=</span> <span class="n">hgdp_recomputed_freemix</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>
    <span class="n">pop_outliers_ht</span> <span class="o">=</span> <span class="n">hgdp_tgp_pop_outliers</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>
    <span class="n">populations_ht</span> <span class="o">=</span> <span class="n">hgdp_tgp_populations_updated</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>
    <span class="n">relatedness_ht</span> <span class="o">=</span> <span class="n">hgdp_tgp_related_samples_to_drop</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>
    <span class="n">related_to_nonsubset_ht</span> <span class="o">=</span> <span class="n">hgdp_tgp_related_to_nonsubset</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>
    <span class="n">duplicated_to_exomes_ht</span> <span class="o">=</span> <span class="n">hgdp_tgp_duplicated_to_exomes</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>

    <span class="c1"># TODO: get the global &amp; subcontinental PCs for the updated HGDP + 1KG subset:</span>
    <span class="c1">#  hgdp_tgp_meta.global_pca_scores,</span>
    <span class="c1">#  hgdp_tgp_meta.subcontinental_pca.pca_scores,</span>
    <span class="c1">#  hgdp_tgp_meta.subcontinental_pca.pca_scores_outliers_removed,</span>
    <span class="c1">#  could update the meta HT after MVP release</span>

    <span class="c1"># Update annotations impacted by the recomputed freemix.</span>
    <span class="n">freemix</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">coalesce</span><span class="p">(</span>
        <span class="n">contamination_ht</span><span class="p">[</span><span class="n">ht</span><span class="o">.</span><span class="n">s</span><span class="p">]</span><span class="o">.</span><span class="n">recomputed_contam_rate</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span> <span class="n">ht</span><span class="o">.</span><span class="n">bam_metrics</span><span class="o">.</span><span class="n">freemix</span>
    <span class="p">)</span>
    <span class="n">hard_filters</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">gnomad_sample_filters</span><span class="o">.</span><span class="n">hard_filters</span> <span class="o">|</span> <span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span>
        <span class="n">freemix</span> <span class="o">&gt;</span> <span class="mf">5.0</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">set</span><span class="p">({</span><span class="s2">&quot;contamination&quot;</span><span class="p">}),</span> <span class="n">hl</span><span class="o">.</span><span class="n">empty_set</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">tstr</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="n">hard_filtered</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">len</span><span class="p">(</span><span class="n">hard_filters</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span>
    <span class="c1"># Update annotations impacted by the updated relatedness inference.</span>
    <span class="n">related_subset</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">relatedness_ht</span><span class="p">[</span><span class="n">ht</span><span class="o">.</span><span class="n">key</span><span class="p">])</span>
    <span class="c1"># Update the relatedness to nonsubset samples.</span>
    <span class="n">related_nonsubset</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">related_to_nonsubset_ht</span><span class="p">[</span><span class="n">ht</span><span class="o">.</span><span class="n">key</span><span class="p">])</span>
    <span class="c1"># Get the duplicated samples in the HGDP + 1KG subset.</span>
    <span class="n">duplicated_to_exomes</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">duplicated_to_exomes_ht</span><span class="p">[</span><span class="n">ht</span><span class="o">.</span><span class="n">key</span><span class="p">])</span>
    <span class="c1"># Update annotations impacted by the updated HGDP + 1KG metadata.</span>
    <span class="n">outlier</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">pop_outliers_ht</span><span class="p">[</span><span class="n">ht</span><span class="o">.</span><span class="n">key</span><span class="p">])</span>
    <span class="n">populations</span> <span class="o">=</span> <span class="n">populations_ht</span><span class="p">[</span><span class="n">ht</span><span class="o">.</span><span class="n">key</span><span class="p">]</span>
    <span class="c1"># Update gnomAD inferred population &#39;oth&#39; to be &#39;remaining&#39;.</span>
    <span class="n">gnomad_pop</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">gnomad_population_inference</span><span class="o">.</span><span class="n">pop</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;oth&quot;</span><span class="p">,</span> <span class="s2">&quot;remaining&quot;</span><span class="p">)</span>

    <span class="n">gnomad_release</span> <span class="o">=</span> <span class="p">(</span>
        <span class="o">~</span><span class="n">hard_filtered</span>
        <span class="o">&amp;</span> <span class="o">~</span><span class="n">outlier</span>
        <span class="o">&amp;</span> <span class="o">~</span><span class="n">related_subset</span>
        <span class="o">&amp;</span> <span class="o">~</span><span class="n">related_nonsubset</span>
        <span class="o">&amp;</span> <span class="o">~</span><span class="n">duplicated_to_exomes</span>
    <span class="p">)</span>

    <span class="n">sample_annotations_to_update</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;bam_metrics&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;freemix&quot;</span><span class="p">:</span> <span class="n">freemix</span><span class="p">},</span>
        <span class="s2">&quot;gnomad_population_inference&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;pop&quot;</span><span class="p">:</span> <span class="n">gnomad_pop</span><span class="p">},</span>
        <span class="s2">&quot;gnomad_sample_filters&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;hard_filters&quot;</span><span class="p">:</span> <span class="n">hard_filters</span><span class="p">,</span>
            <span class="s2">&quot;hard_filtered&quot;</span><span class="p">:</span> <span class="n">hard_filtered</span><span class="p">,</span>
            <span class="s2">&quot;related_to_nonsubset&quot;</span><span class="p">:</span> <span class="n">related_nonsubset</span><span class="p">,</span>
            <span class="s2">&quot;v4_exome_duplicate&quot;</span><span class="p">:</span> <span class="n">duplicated_to_exomes</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;gnomad_high_quality&quot;</span><span class="p">:</span> <span class="n">ht</span><span class="o">.</span><span class="n">gnomad_high_quality</span> <span class="o">&amp;</span> <span class="o">~</span><span class="n">hard_filtered</span><span class="p">,</span>
        <span class="s2">&quot;gnomad_release&quot;</span><span class="p">:</span> <span class="n">gnomad_release</span><span class="p">,</span>
        <span class="s2">&quot;relatedness_inference&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;related&quot;</span><span class="p">:</span> <span class="n">related_subset</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;hgdp_tgp_meta&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;subcontinental_pca&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;outlier&quot;</span><span class="p">:</span> <span class="n">outlier</span><span class="p">},</span>
            <span class="s2">&quot;population&quot;</span><span class="p">:</span> <span class="n">populations</span><span class="o">.</span><span class="n">population</span><span class="p">,</span>
            <span class="s2">&quot;latitude&quot;</span><span class="p">:</span> <span class="n">populations</span><span class="o">.</span><span class="n">latitude</span><span class="p">,</span>
            <span class="s2">&quot;longitude&quot;</span><span class="p">:</span> <span class="n">populations</span><span class="o">.</span><span class="n">longitude</span><span class="p">,</span>
            <span class="s2">&quot;gnomad_labeled_subpop&quot;</span><span class="p">:</span> <span class="n">populations</span><span class="o">.</span><span class="n">population</span><span class="o">.</span><span class="n">lower</span><span class="p">(),</span>
        <span class="p">},</span>
        <span class="s2">&quot;high_quality&quot;</span><span class="p">:</span> <span class="o">~</span><span class="n">hard_filtered</span> <span class="o">&amp;</span> <span class="o">~</span><span class="n">outlier</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">updated_ht</span> <span class="o">=</span> <span class="n">update_structured_annotations</span><span class="p">(</span>
        <span class="n">ht</span><span class="p">,</span>
        <span class="n">sample_annotations_to_update</span><span class="p">,</span>
        <span class="n">annotation_update_label</span><span class="o">=</span><span class="s2">&quot;sample_annotations_updated&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">updated_ht</span> <span class="o">=</span> <span class="n">updated_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">()</span>

    <span class="n">updated_ht</span> <span class="o">=</span> <span class="n">updated_ht</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span><span class="n">new_temp_file</span><span class="p">(</span><span class="s2">&quot;hgdp_tgp_meta_update&quot;</span><span class="p">,</span> <span class="s2">&quot;ht&quot;</span><span class="p">))</span>
    <span class="n">updated_counts</span> <span class="o">=</span> <span class="n">updated_ht</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
            <span class="n">n_hard_filtered</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">(</span>
                <span class="n">updated_ht</span><span class="o">.</span><span class="n">gnomad_sample_filters</span><span class="o">.</span><span class="n">hard_filtered</span>
            <span class="p">),</span>
            <span class="n">n_related_subset</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">(</span>
                <span class="n">updated_ht</span><span class="o">.</span><span class="n">relatedness_inference</span><span class="o">.</span><span class="n">related</span>
            <span class="p">),</span>
            <span class="n">n_related_nonsubset</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">(</span>
                <span class="o">~</span><span class="n">updated_ht</span><span class="o">.</span><span class="n">relatedness_inference</span><span class="o">.</span><span class="n">related</span>
                <span class="o">&amp;</span> <span class="n">updated_ht</span><span class="o">.</span><span class="n">gnomad_sample_filters</span><span class="o">.</span><span class="n">related_to_nonsubset</span>
            <span class="p">),</span>
            <span class="n">n_duplicate_to_exomes</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">(</span>
                <span class="o">~</span><span class="n">updated_ht</span><span class="o">.</span><span class="n">relatedness_inference</span><span class="o">.</span><span class="n">related</span>
                <span class="o">&amp;</span> <span class="o">~</span><span class="n">updated_ht</span><span class="o">.</span><span class="n">gnomad_sample_filters</span><span class="o">.</span><span class="n">related_to_nonsubset</span>
                <span class="o">&amp;</span> <span class="n">updated_ht</span><span class="o">.</span><span class="n">gnomad_sample_filters</span><span class="o">.</span><span class="n">v4_exome_duplicate</span>
            <span class="p">),</span>
            <span class="n">n_outlier</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">(</span>
                <span class="n">updated_ht</span><span class="o">.</span><span class="n">hgdp_tgp_meta</span><span class="o">.</span><span class="n">subcontinental_pca</span><span class="o">.</span><span class="n">outlier</span>
            <span class="p">),</span>
            <span class="n">n_release</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">(</span><span class="n">updated_ht</span><span class="o">.</span><span class="n">gnomad_release</span><span class="p">),</span>
            <span class="n">n_diff</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span>
                <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">counter</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="n">updated_ht</span><span class="o">.</span><span class="n">sample_annotations_updated</span>
            <span class="p">),</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;%d sample(s) hard filtered (%d more samples compared to v3.1.2 subset release);</span>
<span class="sd">        %d samples are pop outliers (%d samples different compared to v3.1.2 subset release);</span>
<span class="sd">        %d samples have different population labels compared to v3.1.2 subset release;</span>
<span class="sd">        %d samples related within the subset (%d samples different compared to v3.1.2 subset release);</span>
<span class="sd">        %d samples further filtered out due to their relatedness to samples outside the subset;</span>
<span class="sd">        %d samples filtered out because they are duplicated in the v4.0 exomes release;</span>
<span class="sd">        %d samples will be in the v4.0 release, compared to 3280 in the v3.1.2 release.&quot;&quot;&quot;</span><span class="p">,</span>
        <span class="n">updated_counts</span><span class="p">[</span><span class="s2">&quot;n_hard_filtered&quot;</span><span class="p">],</span>
        <span class="n">updated_counts</span><span class="p">[</span><span class="s2">&quot;n_diff&quot;</span><span class="p">][</span><span class="s2">&quot;gnomad_sample_filters.hard_filtered&quot;</span><span class="p">],</span>
        <span class="n">updated_counts</span><span class="p">[</span><span class="s2">&quot;n_outlier&quot;</span><span class="p">],</span>
        <span class="n">updated_counts</span><span class="p">[</span><span class="s2">&quot;n_diff&quot;</span><span class="p">][</span><span class="s2">&quot;hgdp_tgp_meta.subcontinental_pca.outlier&quot;</span><span class="p">],</span>
        <span class="n">updated_counts</span><span class="p">[</span><span class="s2">&quot;n_diff&quot;</span><span class="p">][</span><span class="s2">&quot;hgdp_tgp_meta.population&quot;</span><span class="p">],</span>
        <span class="n">updated_counts</span><span class="p">[</span><span class="s2">&quot;n_related_subset&quot;</span><span class="p">],</span>
        <span class="n">updated_counts</span><span class="p">[</span><span class="s2">&quot;n_diff&quot;</span><span class="p">][</span><span class="s2">&quot;relatedness_inference.related&quot;</span><span class="p">],</span>
        <span class="n">updated_counts</span><span class="p">[</span><span class="s2">&quot;n_related_nonsubset&quot;</span><span class="p">],</span>
        <span class="n">updated_counts</span><span class="p">[</span><span class="s2">&quot;n_duplicate_to_exomes&quot;</span><span class="p">],</span>
        <span class="n">updated_counts</span><span class="p">[</span><span class="s2">&quot;n_release&quot;</span><span class="p">],</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">updated_ht</span></div>


<div class="viewcode-block" id="get_updated_release_samples"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/generate_freq_genomes.html#gnomad_qc.v4.annotations.generate_freq_genomes.get_updated_release_samples">[docs]</a><span class="k">def</span> <span class="nf">get_updated_release_samples</span><span class="p">(</span><span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get the samples in the HGDP + 1KG subset that will be added, subtracted, or have different pop labels in the v4.0 release compared to the v3.1 release.</span>

<span class="sd">    Three sets of samples will be obtained:</span>
<span class="sd">        - samples that will have different pop labels in the v4.0 genomes release:</span>
<span class="sd">          samples in the to-be-split &#39;Han&#39; and &#39;Papuan&#39; populations AND their</span>
<span class="sd">          &#39;gnomad_release&#39; status hasn&#39;t changed.</span>
<span class="sd">        - samples that will be added to the v4.0 genomes release: samples where</span>
<span class="sd">          &#39;gnomad_release&#39; status has changed and &#39;gnomad_release&#39; is now True.</span>
<span class="sd">        - samples that will be removed from the v4.0 genomes release: samples where</span>
<span class="sd">          &#39;gnomad_release&#39; status has changed and &#39;gnomad_release&#39; is now False.</span>

<span class="sd">    :param ht: Table with the HGDP + 1KG subset metadata with updated sample</span>
<span class="sd">        annotations.</span>
<span class="sd">    :return: Tuple of Tables with samples that have different pop labels in the v4</span>
<span class="sd">        release, to be added, and to be subtracted.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Filter to all samples with a change in the &#39;gnomad_release&#39; status.</span>
    <span class="n">release_diff_ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
        <span class="n">ht</span><span class="o">.</span><span class="n">sample_annotations_updated</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;gnomad_release&quot;</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="n">release_add_ht</span> <span class="o">=</span> <span class="n">release_diff_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">release_diff_ht</span><span class="o">.</span><span class="n">gnomad_release</span><span class="p">)</span>
    <span class="n">release_subtract_ht</span> <span class="o">=</span> <span class="n">release_diff_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="o">~</span><span class="n">release_diff_ht</span><span class="o">.</span><span class="n">gnomad_release</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        %d samples will be added after the new filters</span>
<span class="sd">        %d sample(s) will be removed after the new filters&quot;&quot;&quot;</span><span class="p">,</span>
        <span class="n">release_add_ht</span><span class="o">.</span><span class="n">count</span><span class="p">(),</span>
        <span class="n">release_subtract_ht</span><span class="o">.</span><span class="n">count</span><span class="p">(),</span>
    <span class="p">)</span>

    <span class="c1"># Filter to all samples with a pop name split and no change in release status.</span>
    <span class="n">split_pops</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">literal</span><span class="p">([</span><span class="s2">&quot;PapuanHighlands&quot;</span><span class="p">,</span> <span class="s2">&quot;PapuanSepik&quot;</span><span class="p">,</span> <span class="s2">&quot;Han&quot;</span><span class="p">,</span> <span class="s2">&quot;NorthernHan&quot;</span><span class="p">])</span>
    <span class="n">pop_diff_ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
        <span class="n">split_pops</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">hgdp_tgp_meta</span><span class="o">.</span><span class="n">population</span><span class="p">)</span>
        <span class="o">&amp;</span> <span class="n">ht</span><span class="o">.</span><span class="n">gnomad_release</span>
        <span class="o">&amp;</span> <span class="o">~</span><span class="n">ht</span><span class="o">.</span><span class="n">sample_annotations_updated</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;gnomad_release&quot;</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%d</span><span class="s2"> samples in the split Papuan &amp; Han set&quot;</span><span class="p">,</span> <span class="n">pop_diff_ht</span><span class="o">.</span><span class="n">count</span><span class="p">())</span>

    <span class="k">return</span> <span class="n">pop_diff_ht</span><span class="p">,</span> <span class="n">release_add_ht</span><span class="p">,</span> <span class="n">release_subtract_ht</span></div>


<div class="viewcode-block" id="calculate_callstats_for_selected_samples"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/generate_freq_genomes.html#gnomad_qc.v4.annotations.generate_freq_genomes.calculate_callstats_for_selected_samples">[docs]</a><span class="k">def</span> <span class="nf">calculate_callstats_for_selected_samples</span><span class="p">(</span>
    <span class="n">mt</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">,</span>
    <span class="n">samples_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">subsets</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate the call stats for samples in `samples_ht`.</span>

<span class="sd">    :param mt: MatrixTable with the HGDP + 1KG subset.</span>
<span class="sd">    :param samples_ht: Table with selected samples and their metadata.</span>
<span class="sd">    :param subsets: Optional List of subsets to be included in the frequency metadata</span>
<span class="sd">        &#39;freq_meta&#39; annotation returned by `annotate_freq`. e.g. [&#39;hgdp&#39;], [&#39;tgp&#39;],</span>
<span class="sd">        [&#39;hgdp&#39;, &#39;tgp&#39;].</span>
<span class="sd">    :return: Table with the call stats for the selected samples.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Filtering MT to sample subset and annotating with metadata...&quot;</span><span class="p">)</span>
    <span class="n">mt</span> <span class="o">=</span> <span class="n">mt</span><span class="o">.</span><span class="n">filter_cols</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">samples_ht</span><span class="p">[</span><span class="n">mt</span><span class="o">.</span><span class="n">col_key</span><span class="p">]))</span>
    <span class="n">mt</span> <span class="o">=</span> <span class="n">mt</span><span class="o">.</span><span class="n">annotate_cols</span><span class="p">(</span><span class="o">**</span><span class="n">samples_ht</span><span class="p">[</span><span class="n">mt</span><span class="o">.</span><span class="n">col_key</span><span class="p">])</span>

    <span class="n">ht</span> <span class="o">=</span> <span class="n">annotate_freq</span><span class="p">(</span>
        <span class="n">mt</span><span class="p">,</span>
        <span class="n">sex_expr</span><span class="o">=</span><span class="n">mt</span><span class="o">.</span><span class="n">gnomad_sex_imputation</span><span class="o">.</span><span class="n">sex_karyotype</span><span class="p">,</span>
        <span class="n">gen_anc_expr</span><span class="o">=</span><span class="p">(</span>
            <span class="n">mt</span><span class="o">.</span><span class="n">hgdp_tgp_meta</span><span class="o">.</span><span class="n">population</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">subsets</span>
            <span class="k">else</span> <span class="n">mt</span><span class="o">.</span><span class="n">gnomad_population_inference</span><span class="o">.</span><span class="n">pop</span>
        <span class="p">),</span>
        <span class="n">annotate_mt</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">subsets</span><span class="p">:</span>
        <span class="n">freq_meta</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">{</span><span class="o">**</span><span class="n">x</span><span class="p">,</span> <span class="o">**</span><span class="p">{</span><span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="s2">&quot;|&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">subsets</span><span class="p">)}}</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">freq_meta</span><span class="p">)</span>
        <span class="p">]</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span><span class="n">freq_meta</span><span class="o">=</span><span class="n">freq_meta</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">ht</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span><span class="n">new_temp_file</span><span class="p">(</span><span class="s2">&quot;hgdp_tgp_subset_freq&quot;</span><span class="p">,</span> <span class="s2">&quot;ht&quot;</span><span class="p">))</span></div>


<div class="viewcode-block" id="concatenate_annotations"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/generate_freq_genomes.html#gnomad_qc.v4.annotations.generate_freq_genomes.concatenate_annotations">[docs]</a><span class="k">def</span> <span class="nf">concatenate_annotations</span><span class="p">(</span>
    <span class="n">hts</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">],</span>
    <span class="n">global_field_names</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="n">FREQ_GLOBALS</span><span class="p">,</span>
    <span class="n">row_field_names</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;freq&quot;</span><span class="p">,),</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Concatenate annotations on multiple Tables into a single Table.</span>

<span class="sd">    :param hts: List of Tables with annotations to be concatenated.</span>
<span class="sd">    :param global_field_names: Global field names to concatenate.</span>
<span class="sd">    :param row_field_names: Row field names to concatenate.</span>
<span class="sd">    :return: Table with concatenated annotations.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">global_field_names</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">global_field_names</span><span class="p">)</span>
    <span class="n">row_field_names</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">row_field_names</span><span class="p">)</span>
    <span class="n">hts</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">ht</span><span class="o">.</span><span class="n">select_globals</span><span class="p">(</span><span class="o">*</span><span class="n">global_field_names</span><span class="p">)</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="o">*</span><span class="n">row_field_names</span><span class="p">)</span> <span class="k">for</span> <span class="n">ht</span> <span class="ow">in</span> <span class="n">hts</span>
    <span class="p">]</span>
    <span class="n">concat_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="o">.</span><span class="n">multi_way_zip_join</span><span class="p">(</span>
        <span class="n">hts</span><span class="p">,</span>
        <span class="n">data_field_name</span><span class="o">=</span><span class="s2">&quot;ann_array&quot;</span><span class="p">,</span>
        <span class="n">global_field_name</span><span class="o">=</span><span class="s2">&quot;global_array&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">concat_ht</span> <span class="o">=</span> <span class="n">concat_ht</span><span class="o">.</span><span class="n">transmute</span><span class="p">(</span>
        <span class="o">**</span><span class="p">{</span><span class="n">f</span><span class="p">:</span> <span class="n">concat_ht</span><span class="o">.</span><span class="n">ann_array</span><span class="o">.</span><span class="n">flatmap</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="n">f</span><span class="p">])</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">row_field_names</span><span class="p">}</span>
    <span class="p">)</span>
    <span class="n">concat_ht</span> <span class="o">=</span> <span class="n">concat_ht</span><span class="o">.</span><span class="n">transmute_globals</span><span class="p">(</span>
        <span class="o">**</span><span class="p">{</span>
            <span class="n">g</span><span class="p">:</span> <span class="n">concat_ht</span><span class="o">.</span><span class="n">global_array</span><span class="o">.</span><span class="n">flatmap</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="n">g</span><span class="p">])</span>
            <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">global_field_names</span>
        <span class="p">},</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">concat_ht</span></div>


<div class="viewcode-block" id="get_hgdp_tgp_callstats_for_selected_samples"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/generate_freq_genomes.html#gnomad_qc.v4.annotations.generate_freq_genomes.get_hgdp_tgp_callstats_for_selected_samples">[docs]</a><span class="k">def</span> <span class="nf">get_hgdp_tgp_callstats_for_selected_samples</span><span class="p">(</span>
    <span class="n">mt</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">,</span> <span class="n">samples_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">compute_freq_all</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate call stats for samples in `samples_ht` grouped by all stratifications in the v3.1 release call stats.</span>

<span class="sd">    This function calculates call stats for:</span>
<span class="sd">        - all samples provided in the `samples_ht` (if `compute_freq_all` is True).</span>
<span class="sd">        - samples in only the HGDP subset.</span>
<span class="sd">        - samples in only the 1KG (tgp) subset.</span>

<span class="sd">    The call stats in these three Tables are then concatenate to create a single Table</span>
<span class="sd">    with all call stats.</span>

<span class="sd">    :param mt: MatrixTable with the HGDP + 1KG subset.</span>
<span class="sd">    :param samples_ht: Table with the samples to filter to before computing call stats.</span>
<span class="sd">    :param compute_freq_all: Whether to compute the call stats for all samples in</span>
<span class="sd">        `samples_ht` instead of only the subset call stats. If False, only the HGDP and</span>
<span class="sd">        1KG subset call stats are computed. Default is True.</span>
<span class="sd">    :return: Table with the call stats for the selected samples.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">projects</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">subset_freq_hts</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Calculating call stats for selected samples...&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">compute_freq_all</span><span class="p">:</span>
        <span class="n">projects</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="s2">&quot;all&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">))</span>

    <span class="n">projects</span><span class="o">.</span><span class="n">extend</span><span class="p">([(</span><span class="s2">&quot;HGDP&quot;</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;hgdp&quot;</span><span class="p">]),</span> <span class="p">(</span><span class="s2">&quot;1000 Genomes&quot;</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;tgp&quot;</span><span class="p">])])</span>
    <span class="n">project_expr</span> <span class="o">=</span> <span class="n">samples_ht</span><span class="o">.</span><span class="n">hgdp_tgp_meta</span><span class="o">.</span><span class="n">project</span>
    <span class="n">project_n</span> <span class="o">=</span> <span class="n">samples_ht</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">counter</span><span class="p">(</span><span class="n">project_expr</span><span class="p">))</span>

    <span class="k">for</span> <span class="n">project</span><span class="p">,</span> <span class="n">subset</span> <span class="ow">in</span> <span class="n">projects</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">project</span> <span class="o">==</span> <span class="s2">&quot;all&quot;</span><span class="p">:</span>
            <span class="n">subset_ht</span> <span class="o">=</span> <span class="n">samples_ht</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">subset_ht</span> <span class="o">=</span> <span class="n">samples_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">project_expr</span> <span class="o">==</span> <span class="n">project</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">project</span> <span class="o">==</span> <span class="s2">&quot;all&quot;</span> <span class="ow">or</span> <span class="n">project_n</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">project</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">freq_ht</span> <span class="o">=</span> <span class="n">calculate_callstats_for_selected_samples</span><span class="p">(</span><span class="n">mt</span><span class="p">,</span> <span class="n">subset_ht</span><span class="p">,</span> <span class="n">subset</span><span class="p">)</span>
            <span class="n">freq_ht</span> <span class="o">=</span> <span class="n">freq_ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;freq&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">select_globals</span><span class="p">(</span><span class="o">*</span><span class="n">FREQ_GLOBALS</span><span class="p">)</span>
            <span class="n">subset_freq_hts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">freq_ht</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Concatenating call stats Tables for selected samples...&quot;</span><span class="p">)</span>
    <span class="n">freq_ht</span> <span class="o">=</span> <span class="n">concatenate_annotations</span><span class="p">(</span><span class="n">subset_freq_hts</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">freq_ht</span></div>


<div class="viewcode-block" id="filter_freq_arrays"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/generate_freq_genomes.html#gnomad_qc.v4.annotations.generate_freq_genomes.filter_freq_arrays">[docs]</a><span class="k">def</span> <span class="nf">filter_freq_arrays</span><span class="p">(</span>
    <span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">items_to_filter</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]],</span>
    <span class="n">keep</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">combine_operator</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;and&quot;</span><span class="p">,</span>
    <span class="n">annotations</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;freq&quot;</span><span class="p">,),</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Use `filter_arrays_by_meta` to filter annotations and globals by &#39;freq_meta&#39; and annotate them back on `ht`.</span>

<span class="sd">    Filter `annotations`, &#39;freq_meta&#39; and &#39;freq_meta_sample_count&#39; fields to only</span>
<span class="sd">    `items_to_filter` by using the &#39;freq_meta&#39; array field values.</span>

<span class="sd">    :param ht: Input Table.</span>
<span class="sd">    :param items_to_filter: Items to filter by.</span>
<span class="sd">    :param keep: Whether to keep or remove items. Default is True.</span>
<span class="sd">    :param combine_operator: Operator (&quot;and&quot; or &quot;or&quot;) to use when combining items in</span>
<span class="sd">        &#39;items_to_filter&#39;. Default is &quot;and&quot;.</span>
<span class="sd">    :param annotations: Annotations in &#39;ht&#39; to filter by `items_to_filter`.</span>
<span class="sd">    :return: Table with filtered &#39;annotations&#39; and &#39;freq_meta&#39; array fields.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">freq_meta</span><span class="p">,</span> <span class="n">array_exprs</span> <span class="o">=</span> <span class="n">filter_arrays_by_meta</span><span class="p">(</span>
        <span class="n">ht</span><span class="o">.</span><span class="n">freq_meta</span><span class="p">,</span>
        <span class="p">{</span>
            <span class="o">**</span><span class="p">{</span><span class="n">a</span><span class="p">:</span> <span class="n">ht</span><span class="p">[</span><span class="n">a</span><span class="p">]</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">annotations</span><span class="p">},</span>
            <span class="s2">&quot;freq_meta_sample_count&quot;</span><span class="p">:</span> <span class="n">ht</span><span class="o">.</span><span class="n">index_globals</span><span class="p">()</span><span class="o">.</span><span class="n">freq_meta_sample_count</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="n">items_to_filter</span><span class="o">=</span><span class="n">items_to_filter</span><span class="p">,</span>
        <span class="n">keep</span><span class="o">=</span><span class="n">keep</span><span class="p">,</span>
        <span class="n">combine_operator</span><span class="o">=</span><span class="n">combine_operator</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="o">**</span><span class="p">{</span><span class="n">a</span><span class="p">:</span> <span class="n">array_exprs</span><span class="p">[</span><span class="n">a</span><span class="p">]</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">annotations</span><span class="p">})</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span>
        <span class="n">freq_meta</span><span class="o">=</span><span class="n">freq_meta</span><span class="p">,</span>
        <span class="n">freq_meta_sample_count</span><span class="o">=</span><span class="n">array_exprs</span><span class="p">[</span><span class="s2">&quot;freq_meta_sample_count&quot;</span><span class="p">],</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">ht</span></div>


<div class="viewcode-block" id="annotate_v3_subsets_sample_count"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/generate_freq_genomes.html#gnomad_qc.v4.annotations.generate_freq_genomes.annotate_v3_subsets_sample_count">[docs]</a><span class="k">def</span> <span class="nf">annotate_v3_subsets_sample_count</span><span class="p">(</span><span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get the freq sample counts from the v3.1 subsets.</span>

<span class="sd">    The freq sample counts on the v3.1.2 release sites HT only contains the counts for</span>
<span class="sd">    the non subset groupings. This function adds the freq sample counts from the v3.1</span>
<span class="sd">    subset frequency groups to the v3.1.2 sample count array to give a sample count</span>
<span class="sd">    array matching the v3.1.2 &#39;freq_meta&#39; global annotation.</span>

<span class="sd">    :param ht: Table of the 3.1.2 release sites HT, which only contains the freq</span>
<span class="sd">       sample counts for non subset groupings.</span>
<span class="sd">    :return: Table with the freq sample counts from the v3.1 subset groups added.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sc</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">index_globals</span><span class="p">()</span><span class="o">.</span><span class="n">freq_sample_count</span>

    <span class="k">for</span> <span class="n">subset</span> <span class="ow">in</span> <span class="n">SUBSETS</span><span class="p">:</span>
        <span class="n">freq_subset_ht</span> <span class="o">=</span> <span class="n">get_freq</span><span class="p">(</span>
            <span class="n">version</span><span class="o">=</span><span class="s2">&quot;3.1&quot;</span><span class="p">,</span> <span class="n">subset</span><span class="o">=</span><span class="n">subset</span><span class="p">,</span> <span class="n">het_nonref_patch</span><span class="o">=</span><span class="kc">True</span>
        <span class="p">)</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>
        <span class="n">sc_subset</span> <span class="o">=</span> <span class="n">freq_subset_ht</span><span class="o">.</span><span class="n">index_globals</span><span class="p">()</span><span class="o">.</span><span class="n">freq_sample_count</span>
        <span class="n">sc</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">sc_subset</span><span class="p">)</span>

    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span><span class="n">freq_meta_sample_count</span><span class="o">=</span><span class="n">sc</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">ht</span></div>


<div class="viewcode-block" id="update_pop_labels"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/generate_freq_genomes.html#gnomad_qc.v4.annotations.generate_freq_genomes.update_pop_labels">[docs]</a><span class="k">def</span> <span class="nf">update_pop_labels</span><span class="p">(</span><span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">pop_map</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Any</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Update the population labels in the &#39;freq_meta&#39; annotation on `ht`.</span>

<span class="sd">    :param ht: Table with population labels to update.</span>
<span class="sd">    :param pop_map: Dictionary mapping old to new population labels.</span>
<span class="sd">    :return: Table with updated population labels in &#39;freq_meta&#39; annotation.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">pop_map</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">literal</span><span class="p">(</span><span class="n">pop_map</span><span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span>
        <span class="n">freq_meta</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">freq_meta</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">d</span><span class="p">:</span> <span class="n">d</span><span class="o">.</span><span class="n">map_values</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">or_else</span><span class="p">(</span><span class="n">pop_map</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="n">x</span><span class="p">))</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">ht</span></div>


<div class="viewcode-block" id="join_release_ht_with_subsets"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/generate_freq_genomes.html#gnomad_qc.v4.annotations.generate_freq_genomes.join_release_ht_with_subsets">[docs]</a><span class="k">def</span> <span class="nf">join_release_ht_with_subsets</span><span class="p">(</span>
    <span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">freq_hts</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Join the release HT with the call stats for added and subtracted samples.</span>

<span class="sd">    :param ht: gnomAD v3.1.2 release sites Table.</span>
<span class="sd">    :param freq_hts: Dictionary with the call stats Tables for added and subtracted</span>
<span class="sd">        samples.</span>
<span class="sd">    :return: Table with `ht` and all Tables in `freq_hts` joined.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Adding freq sample counts from v3 subsets...&quot;</span><span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">annotate_v3_subsets_sample_count</span><span class="p">(</span><span class="n">ht</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Removing &#39;Han&#39; and &#39;Papuan&#39; pops from freq and freq_meta...&quot;</span><span class="p">)</span>
    <span class="n">pops_to_remove</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;pop&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;han&quot;</span><span class="p">,</span> <span class="s2">&quot;papuan&quot;</span><span class="p">]}</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">filter_freq_arrays</span><span class="p">(</span><span class="n">ht</span><span class="p">,</span> <span class="n">pops_to_remove</span><span class="p">,</span> <span class="n">keep</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">combine_operator</span><span class="o">=</span><span class="s2">&quot;or&quot;</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Updating HGDP pop labels and &#39;oth&#39; -&gt; &#39;remaining&#39;...&quot;</span><span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">update_pop_labels</span><span class="p">(</span><span class="n">ht</span><span class="p">,</span> <span class="n">POP_MAP</span><span class="p">)</span>

    <span class="n">freq_hts</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;release&quot;</span><span class="p">:</span> <span class="n">ht</span><span class="p">,</span> <span class="o">**</span><span class="n">freq_hts</span><span class="p">}</span>
    <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">freq_ht</span> <span class="ow">in</span> <span class="n">freq_hts</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;There are </span><span class="si">%i</span><span class="s2"> variants in the </span><span class="si">%s</span><span class="s2"> HT...&quot;</span><span class="p">,</span> <span class="n">freq_ht</span><span class="o">.</span><span class="n">count</span><span class="p">(),</span> <span class="n">name</span><span class="p">)</span>
        <span class="n">freq_hts</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">freq_ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;freq&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">select_globals</span><span class="p">(</span><span class="o">*</span><span class="n">FREQ_GLOBALS</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Joining release HT with AFs for added and subtracted samples...&quot;</span><span class="p">)</span>
    <span class="n">freq_hts</span> <span class="o">=</span> <span class="p">[</span><span class="n">freq_hts</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">JOIN_FREQS</span><span class="p">]</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="o">.</span><span class="n">multi_way_zip_join</span><span class="p">(</span><span class="n">freq_hts</span><span class="p">,</span> <span class="s2">&quot;ann_array&quot;</span><span class="p">,</span> <span class="s2">&quot;global_array&quot;</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Filtering joint HT to rows where at least one group has AC &gt; 0...&quot;</span><span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">ann_array</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">freq</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="k">lambda</span> <span class="n">y</span><span class="p">:</span> <span class="n">y</span><span class="o">.</span><span class="n">AC</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">))))</span>

    <span class="k">return</span> <span class="n">ht</span></div>


<div class="viewcode-block" id="get_group_membership_ht_for_an"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/generate_freq_genomes.html#gnomad_qc.v4.annotations.generate_freq_genomes.get_group_membership_ht_for_an">[docs]</a><span class="k">def</span> <span class="nf">get_group_membership_ht_for_an</span><span class="p">(</span>
    <span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">only_pop_diff</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate a Table with a &#39;group_membership&#39; array for each sample indicating whether the sample belongs to specific stratification groups.</span>

<span class="sd">    `ht` must have the following annotations:</span>
<span class="sd">        - &#39;meta.population_inference.pop&#39;: population label.</span>
<span class="sd">        - &#39;meta.sex_imputation.sex_karyotype`: sex label.</span>
<span class="sd">        - &#39;meta.project_meta.project_subpop&#39;: subpopulation label.</span>
<span class="sd">        - &#39;meta.subsets&#39;: dictionary with subset labels as keys and boolean values.</span>

<span class="sd">    :param ht: Table with the sample metadata.</span>
<span class="sd">    :param only_pop_diff: Whether to only include the population stratification groups</span>
<span class="sd">        for samples in pop_diff. Default is False.</span>
<span class="sd">    :return: Table with the group membership for each sample to be used for computing</span>
<span class="sd">        allele number (AN) per group.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">only_pop_diff</span><span class="p">:</span>
        <span class="n">pop_expr</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">hgdp_tgp_meta</span><span class="o">.</span><span class="n">population</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
        <span class="n">subpop_expr</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">hgdp_tgp_meta</span><span class="o">.</span><span class="n">population</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
        <span class="n">sex_expr</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">gnomad_sex_imputation</span><span class="o">.</span><span class="n">sex_karyotype</span>
        <span class="n">subsets</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;hgdp&quot;</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">pop_expr</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">meta</span><span class="o">.</span><span class="n">population_inference</span><span class="o">.</span><span class="n">pop</span>
        <span class="n">sex_expr</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">meta</span><span class="o">.</span><span class="n">sex_imputation</span><span class="o">.</span><span class="n">sex_karyotype</span>
        <span class="n">subpop_expr</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">meta</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">project_subpop</span>
        <span class="n">subsets</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;all&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">SUBSETS</span>

    <span class="n">hts</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">subset</span> <span class="ow">in</span> <span class="n">subsets</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">subset</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;tgp&quot;</span><span class="p">,</span> <span class="s2">&quot;hgdp&quot;</span><span class="p">]:</span>
            <span class="n">subset_pop_expr</span> <span class="o">=</span> <span class="n">subpop_expr</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">subset_pop_expr</span> <span class="o">=</span> <span class="n">pop_expr</span>

        <span class="c1"># Build strata expressions to get group membership for pop, sex, pop-sex.</span>
        <span class="n">strata_expr</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">{</span><span class="s2">&quot;pop&quot;</span><span class="p">:</span> <span class="n">subset_pop_expr</span><span class="p">},</span>
            <span class="p">{</span><span class="s2">&quot;sex&quot;</span><span class="p">:</span> <span class="n">sex_expr</span><span class="p">},</span>
            <span class="p">{</span><span class="s2">&quot;pop&quot;</span><span class="p">:</span> <span class="n">subset_pop_expr</span><span class="p">,</span> <span class="s2">&quot;sex&quot;</span><span class="p">:</span> <span class="n">sex_expr</span><span class="p">},</span>
        <span class="p">]</span>
        <span class="c1"># For subsets, add strata expressions to get group membership for subset,</span>
        <span class="c1"># subset-pop, subset-sex, and subset-pop-sex.</span>
        <span class="k">if</span> <span class="n">subset</span> <span class="o">!=</span> <span class="s2">&quot;all&quot;</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">only_pop_diff</span><span class="p">:</span>
            <span class="n">subset_expr</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">or_missing</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">meta</span><span class="o">.</span><span class="n">subsets</span><span class="p">[</span><span class="n">subset</span><span class="p">],</span> <span class="n">subset</span><span class="p">)</span>
            <span class="n">strata_expr</span> <span class="o">=</span> <span class="p">[{</span><span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="n">subset_expr</span><span class="p">}]</span> <span class="o">+</span> <span class="p">[</span>
                <span class="p">{</span><span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="n">subset_expr</span><span class="p">,</span> <span class="o">**</span><span class="n">x</span><span class="p">}</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">strata_expr</span>
            <span class="p">]</span>

        <span class="n">subset_ht</span> <span class="o">=</span> <span class="n">generate_freq_group_membership_array</span><span class="p">(</span>
            <span class="n">ht</span><span class="p">,</span> <span class="n">strata_expr</span><span class="p">,</span> <span class="n">remove_zero_sample_groups</span><span class="o">=</span><span class="kc">True</span>
        <span class="p">)</span>
        <span class="n">subset_globals</span> <span class="o">=</span> <span class="n">subset_ht</span><span class="o">.</span><span class="n">index_globals</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">subset</span> <span class="o">!=</span> <span class="s2">&quot;all&quot;</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">only_pop_diff</span><span class="p">:</span>
            <span class="c1"># The group_membership, freq_meta, and freq_meta_sample_count arrays for</span>
            <span class="c1"># subsets is ordered [adj, raw, subset adj, ...] and we want it to be</span>
            <span class="c1"># [subset adj, subset raw, ...] where subset adj and subset raw have the</span>
            <span class="c1"># same sample grouping.</span>
            <span class="n">subset_adj_idx</span> <span class="o">=</span> <span class="mi">2</span>
            <span class="n">subset_ht</span> <span class="o">=</span> <span class="n">subset_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
                <span class="n">group_membership</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
                    <span class="p">[</span><span class="n">subset_ht</span><span class="o">.</span><span class="n">group_membership</span><span class="p">[</span><span class="n">subset_adj_idx</span><span class="p">]]</span>
                <span class="p">)</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">subset_ht</span><span class="o">.</span><span class="n">group_membership</span><span class="p">[</span><span class="n">subset_adj_idx</span><span class="p">:])</span>
            <span class="p">)</span>
            <span class="n">freq_meta_sample_count</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
                <span class="p">[</span><span class="n">subset_globals</span><span class="o">.</span><span class="n">freq_meta_sample_count</span><span class="p">[</span><span class="n">subset_adj_idx</span><span class="p">]]</span>
            <span class="p">)</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">subset_globals</span><span class="o">.</span><span class="n">freq_meta_sample_count</span><span class="p">[</span><span class="n">subset_adj_idx</span><span class="p">:])</span>

            <span class="c1"># Modify the freq_meta annotation to match the new group_membership and</span>
            <span class="c1"># freq_meta_sample_count arrays.</span>
            <span class="n">subset_ht</span> <span class="o">=</span> <span class="n">subset_ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span>
                <span class="n">freq_meta</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">array</span><span class="p">(</span>
                    <span class="p">[</span>
                        <span class="p">{</span><span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="n">subset</span><span class="p">,</span> <span class="s2">&quot;group&quot;</span><span class="p">:</span> <span class="s2">&quot;adj&quot;</span><span class="p">},</span>
                        <span class="p">{</span><span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="n">subset</span><span class="p">,</span> <span class="s2">&quot;group&quot;</span><span class="p">:</span> <span class="s2">&quot;raw&quot;</span><span class="p">},</span>
                    <span class="p">]</span>
                <span class="p">)</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">subset_globals</span><span class="o">.</span><span class="n">freq_meta</span><span class="p">[</span><span class="n">subset_adj_idx</span> <span class="o">+</span> <span class="mi">1</span> <span class="p">:]),</span>
                <span class="n">freq_meta_sample_count</span><span class="o">=</span><span class="n">freq_meta_sample_count</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="k">if</span> <span class="n">only_pop_diff</span><span class="p">:</span>
            <span class="c1"># We only want:</span>
            <span class="c1"># {&#39;group&#39;: &#39;adj&#39;, &#39;pop&#39;: &#39;han&#39;, &#39;subset&#39;: &#39;hgdp&#39;},</span>
            <span class="c1"># {&#39;group&#39;: &#39;adj&#39;, &#39;pop&#39;: &#39;northernhan&#39;, &#39;subset&#39;: &#39;hgdp&#39;},</span>
            <span class="c1"># {&#39;group&#39;: &#39;adj&#39;, &#39;pop&#39;: &#39;han&#39;, &#39;sex&#39;: &#39;XX&#39;, &#39;subset&#39;: &#39;hgdp&#39;},</span>
            <span class="c1"># {&#39;group&#39;: &#39;adj&#39;, &#39;pop&#39;: &#39;han&#39;, &#39;sex&#39;: &#39;XY&#39;, &#39;subset&#39;: &#39;hgdp&#39;},</span>
            <span class="c1"># {&#39;group&#39;: &#39;adj&#39;, &#39;pop&#39;: &#39;northernhan&#39;, &#39;sex&#39;: &#39;XX&#39;, &#39;subset&#39;: &#39;hgdp&#39;},</span>
            <span class="c1"># {&#39;group&#39;: &#39;adj&#39;, &#39;pop&#39;: &#39;northernhan&#39;, &#39;sex&#39;: &#39;XY&#39;, &#39;subset&#39;: &#39;hgdp&#39;}</span>
            <span class="n">subset_ht</span> <span class="o">=</span> <span class="n">subset_ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span>
                <span class="n">freq_meta</span><span class="o">=</span><span class="p">[</span>
                    <span class="p">{</span><span class="o">**</span><span class="n">x</span><span class="p">,</span> <span class="o">**</span><span class="p">{</span><span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="s2">&quot;hgdp&quot;</span><span class="p">}}</span>
                    <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">subset_ht</span><span class="o">.</span><span class="n">freq_meta</span><span class="p">)[</span><span class="mi">2</span><span class="p">:]</span>
                <span class="p">],</span>
                <span class="n">freq_meta_sample_count</span><span class="o">=</span><span class="n">subset_ht</span><span class="o">.</span><span class="n">freq_meta_sample_count</span><span class="p">[</span><span class="mi">2</span><span class="p">:],</span>
            <span class="p">)</span>
            <span class="n">subset_ht</span> <span class="o">=</span> <span class="n">subset_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
                <span class="n">group_membership</span><span class="o">=</span><span class="n">subset_ht</span><span class="o">.</span><span class="n">group_membership</span><span class="p">[</span><span class="mi">2</span><span class="p">:]</span>
            <span class="p">)</span>
            <span class="n">subset_ht</span> <span class="o">=</span> <span class="n">filter_freq_arrays</span><span class="p">(</span>
                <span class="n">subset_ht</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;pop&quot;</span><span class="p">],</span> <span class="n">annotations</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;group_membership&quot;</span><span class="p">]</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Remove &#39;Han&#39; and &#39;Papuan&#39; pops from group_membership, freq_meta, and</span>
            <span class="c1"># freq_meta_sample_count.</span>
            <span class="n">subset_ht</span> <span class="o">=</span> <span class="n">filter_freq_arrays</span><span class="p">(</span>
                <span class="n">subset_ht</span><span class="p">,</span>
                <span class="p">{</span><span class="s2">&quot;pop&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;han&quot;</span><span class="p">,</span> <span class="s2">&quot;papuan&quot;</span><span class="p">]},</span>
                <span class="n">keep</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                <span class="n">combine_operator</span><span class="o">=</span><span class="s2">&quot;or&quot;</span><span class="p">,</span>
                <span class="n">annotations</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;group_membership&quot;</span><span class="p">],</span>
            <span class="p">)</span>

        <span class="c1"># Keep track of which groups should aggregate raw genotypes.</span>
        <span class="n">subset_ht</span> <span class="o">=</span> <span class="n">subset_ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span>
            <span class="n">raw_group</span><span class="o">=</span><span class="n">subset_ht</span><span class="o">.</span><span class="n">freq_meta</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;group&quot;</span><span class="p">,</span> <span class="s2">&quot;NA&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="s2">&quot;raw&quot;</span><span class="p">),</span>
        <span class="p">)</span>
        <span class="n">hts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subset_ht</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">concatenate_annotations</span><span class="p">(</span>
        <span class="n">hts</span><span class="p">,</span>
        <span class="n">row_field_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;group_membership&quot;</span><span class="p">],</span>
        <span class="n">global_field_names</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">FREQ_GLOBALS</span><span class="p">)</span> <span class="o">+</span> <span class="p">[</span><span class="s2">&quot;raw_group&quot;</span><span class="p">],</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="compute_an_by_group_membership"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/generate_freq_genomes.html#gnomad_qc.v4.annotations.generate_freq_genomes.compute_an_by_group_membership">[docs]</a><span class="k">def</span> <span class="nf">compute_an_by_group_membership</span><span class="p">(</span>
    <span class="n">vds</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">VariantDataset</span><span class="p">,</span>
    <span class="n">group_membership_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">variant_filter_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Compute the allele number for new variants in the v4.0 release by call stats group membership.</span>

<span class="sd">    :param vds: VariantDataset with all v3.1 samples (including non-release).</span>
<span class="sd">    :param group_membership_ht: Table with the group membership for each sample. This</span>
<span class="sd">        is generated by `get_group_membership_ht_for_an`.</span>
<span class="sd">    :param variant_filter_ht: Table with all variants that need AN to be computed.</span>
<span class="sd">    :return: Table with the allele number for new variants in the v4.0 release.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">n_samples</span> <span class="o">=</span> <span class="n">group_membership_ht</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
    <span class="n">n_groups</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">group_membership_ht</span><span class="o">.</span><span class="n">group_membership</span><span class="o">.</span><span class="n">take</span><span class="p">(</span><span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>

    <span class="n">vmt</span> <span class="o">=</span> <span class="n">vds</span><span class="o">.</span><span class="n">variant_data</span>
    <span class="n">rmt</span> <span class="o">=</span> <span class="n">vds</span><span class="o">.</span><span class="n">reference_data</span><span class="o">.</span><span class="n">select_cols</span><span class="p">()</span><span class="o">.</span><span class="n">select_rows</span><span class="p">()</span>

    <span class="c1"># Confirm that all variants in variant_filter_ht are present in the</span>
    <span class="c1"># vds.variant_dataset and throw an error if this is not the case since all samples</span>
    <span class="c1"># added to v4.0 genomes should be in this vds.</span>
    <span class="n">vht</span> <span class="o">=</span> <span class="n">vmt</span><span class="o">.</span><span class="n">rows</span><span class="p">()</span>
    <span class="n">vht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">split_multi</span><span class="p">(</span><span class="n">vht</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">variant_filter_ht</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_missing</span><span class="p">(</span><span class="n">vht</span><span class="p">[</span><span class="n">variant_filter_ht</span><span class="o">.</span><span class="n">key</span><span class="p">]))</span>
    <span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;Not all variants in the variant_filter_ht are found in the &quot;</span>
            <span class="s2">&quot;vds.variant_dataset!&quot;</span>
        <span class="p">)</span>

    <span class="c1"># Filter the VDS variant data and reference data to only keep samples that were in</span>
    <span class="c1"># the v3.1 release. We do it this way instead of using hl.vds.filter_samples because</span>
    <span class="c1"># we want to keep all variants that were in the v3.1 VDS not only those found in the</span>
    <span class="c1"># release but `hl.vds.filter_samples` includes vmt =</span>
    <span class="c1"># vmt.filter_rows(hl.agg.count() &gt; 0) by default.</span>
    <span class="n">release_s</span> <span class="o">=</span> <span class="n">vmt</span><span class="o">.</span><span class="n">aggregate_cols</span><span class="p">(</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">group_membership_ht</span><span class="p">[</span><span class="n">vmt</span><span class="o">.</span><span class="n">s</span><span class="p">]),</span> <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">collect_as_set</span><span class="p">(</span><span class="n">vmt</span><span class="o">.</span><span class="n">s</span><span class="p">)</span>
        <span class="p">),</span>
        <span class="n">_localize</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span><span class="o">.</span><span class="n">_persist</span><span class="p">()</span>
    <span class="n">vmt</span> <span class="o">=</span> <span class="n">vmt</span><span class="o">.</span><span class="n">filter_cols</span><span class="p">(</span><span class="n">release_s</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">vmt</span><span class="o">.</span><span class="n">s</span><span class="p">))</span>
    <span class="n">rmt</span> <span class="o">=</span> <span class="n">rmt</span><span class="o">.</span><span class="n">filter_cols</span><span class="p">(</span><span class="n">release_s</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">rmt</span><span class="o">.</span><span class="n">s</span><span class="p">))</span>
    <span class="n">rmt</span> <span class="o">=</span> <span class="n">rmt</span><span class="o">.</span><span class="n">filter_rows</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count</span><span class="p">()</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>

    <span class="c1"># Only keep necessary entries and col annotations for AN calculation from the VDS</span>
    <span class="c1"># so that only the needed entries are densified. AD, DP, GT, and GQ are needed for</span>
    <span class="c1"># adding the adj annotation to compute AN for all adj stratification groups.</span>
    <span class="c1"># sex_karyotype is needed by adjust_sex_ploidy to adjust relevant genotypes on chrX</span>
    <span class="c1"># and chrY non-PAR before computing AN.</span>
    <span class="n">vmt</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">vmt</span><span class="o">.</span><span class="n">select_cols</span><span class="p">(</span>
            <span class="n">sex_karyotype</span><span class="o">=</span><span class="n">vmt</span><span class="o">.</span><span class="n">meta</span><span class="o">.</span><span class="n">sex_imputation</span><span class="o">.</span><span class="n">sex_karyotype</span><span class="p">,</span>
            <span class="n">group_membership</span><span class="o">=</span><span class="n">group_membership_ht</span><span class="p">[</span><span class="n">vmt</span><span class="o">.</span><span class="n">col_key</span><span class="p">]</span><span class="o">.</span><span class="n">group_membership</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">select_entries</span><span class="p">(</span><span class="s2">&quot;LA&quot;</span><span class="p">,</span> <span class="s2">&quot;LAD&quot;</span><span class="p">,</span> <span class="s2">&quot;DP&quot;</span><span class="p">,</span> <span class="s2">&quot;LGT&quot;</span><span class="p">,</span> <span class="s2">&quot;GQ&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">select_rows</span><span class="p">()</span>
    <span class="p">)</span>
    <span class="n">vmt</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">experimental</span><span class="o">.</span><span class="n">sparse_split_multi</span><span class="p">(</span><span class="n">vmt</span><span class="p">,</span> <span class="n">filter_changed_loci</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">vmt</span> <span class="o">=</span> <span class="n">vmt</span><span class="o">.</span><span class="n">semi_join_rows</span><span class="p">(</span><span class="n">variant_filter_ht</span><span class="p">)</span>
    <span class="n">vds</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">VariantDataset</span><span class="p">(</span><span class="n">rmt</span><span class="p">,</span> <span class="n">vmt</span><span class="p">)</span>

    <span class="c1"># NOTE: We need to write and read the VDS here to avoid a bug in split_multi_hts</span>
    <span class="c1"># that causes code 137 memory errors in Hail 0.2.122, this should be fixed in</span>
    <span class="c1"># subsequent versions of Hail.</span>
    <span class="n">tmp_vds_path</span> <span class="o">=</span> <span class="n">new_temp_file</span><span class="p">(</span><span class="s2">&quot;variants_for_an_split&quot;</span><span class="p">,</span> <span class="s2">&quot;vds&quot;</span><span class="p">)</span>

    <span class="n">vds</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">tmp_vds_path</span><span class="p">)</span>
    <span class="n">vds</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">read_vds</span><span class="p">(</span><span class="n">tmp_vds_path</span><span class="p">)</span>

    <span class="c1"># NOTE: The correct thing to do here is the following commented out code, but in</span>
    <span class="c1"># order to be consistent with v3.1 we will add the adj annotation before adjusting</span>
    <span class="c1"># for sex ploidy.</span>
    <span class="c1"># Densify the VDS, adjust GT sex ploidy, and annotate entries with adj. Adj</span>
    <span class="c1"># annotation must happen after sex ploidy adjustment because haploid GTs have</span>
    <span class="c1"># different adj filtering criteria.</span>
    <span class="n">mt</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">to_dense_mt</span><span class="p">(</span><span class="n">vds</span><span class="p">)</span>
    <span class="c1"># mt = adjust_sex_ploidy(mt, mt.sex_karyotype, male_str=&quot;XY&quot;, female_str=&quot;XX&quot;)</span>
    <span class="n">mt</span> <span class="o">=</span> <span class="n">annotate_adj</span><span class="p">(</span><span class="n">mt</span><span class="p">)</span>
    <span class="n">mt</span> <span class="o">=</span> <span class="n">adjust_sex_ploidy</span><span class="p">(</span><span class="n">mt</span><span class="p">,</span> <span class="n">mt</span><span class="o">.</span><span class="n">sex_karyotype</span><span class="p">,</span> <span class="n">xy_str</span><span class="o">=</span><span class="s2">&quot;XY&quot;</span><span class="p">,</span> <span class="n">xx_str</span><span class="o">=</span><span class="s2">&quot;XX&quot;</span><span class="p">)</span>
    <span class="n">mt</span> <span class="o">=</span> <span class="n">mt</span><span class="o">.</span><span class="n">select_entries</span><span class="p">(</span><span class="s2">&quot;GT&quot;</span><span class="p">,</span> <span class="s2">&quot;adj&quot;</span><span class="p">)</span>

    <span class="c1"># Convert MT to HT with a row annotation that is an array of all samples entries</span>
    <span class="c1"># for that variant.</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">mt</span><span class="o">.</span><span class="n">localize_entries</span><span class="p">(</span><span class="s2">&quot;entries&quot;</span><span class="p">,</span> <span class="s2">&quot;cols&quot;</span><span class="p">)</span>

    <span class="c1"># For each stratification group in group_membership, determine the indices of the</span>
    <span class="c1"># samples that belong to that group.</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span>
        <span class="n">indices_by_group</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">range</span><span class="p">(</span><span class="n">n_groups</span><span class="p">)</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">g_i</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">range</span><span class="p">(</span><span class="n">n_samples</span><span class="p">)</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                <span class="k">lambda</span> <span class="n">s_i</span><span class="p">:</span> <span class="n">ht</span><span class="o">.</span><span class="n">cols</span><span class="p">[</span><span class="n">s_i</span><span class="p">]</span><span class="o">.</span><span class="n">group_membership</span><span class="p">[</span><span class="n">g_i</span><span class="p">]</span>
            <span class="p">)</span>
        <span class="p">),</span>
        <span class="n">raw_group</span><span class="o">=</span><span class="n">group_membership_ht</span><span class="o">.</span><span class="n">index_globals</span><span class="p">()</span><span class="o">.</span><span class="n">raw_group</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># Keep only the row annotations needed for the AN calculation.</span>
    <span class="c1"># Use ploidy to determine the number of alleles for each sample at each variant and</span>
    <span class="c1"># then sum the ploidy of all samples for each group indicated by indices_by_group</span>
    <span class="c1"># to get the AN for each stratification group.</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
        <span class="n">adj_array</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">entries</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">e</span><span class="p">:</span> <span class="n">e</span><span class="o">.</span><span class="n">adj</span><span class="p">),</span>
        <span class="n">ploidy_array</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">entries</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">e</span><span class="p">:</span> <span class="n">e</span><span class="o">.</span><span class="n">GT</span><span class="o">.</span><span class="n">ploidy</span><span class="p">),</span>
    <span class="p">)</span>
    <span class="n">agg_expr</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">s_indices</span><span class="p">,</span> <span class="n">raw</span><span class="p">:</span> <span class="n">s_indices</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">i</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span>
                <span class="n">raw</span><span class="p">,</span>
                <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">ploidy_array</span><span class="p">[</span><span class="n">i</span><span class="p">]),</span>
                <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">adj_array</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">ploidy_array</span><span class="p">[</span><span class="n">i</span><span class="p">])),</span>
            <span class="p">)</span>
        <span class="p">),</span>
        <span class="n">ht</span><span class="o">.</span><span class="n">indices_by_group</span><span class="p">,</span>
        <span class="n">ht</span><span class="o">.</span><span class="n">raw_group</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># Turn the array of ANs for each group into a call stats struct with AC, AF, AN,</span>
    <span class="c1"># and homozygote_count to be easily combined with the call stats of the updated</span>
    <span class="c1"># samples.</span>
    <span class="n">agg_expr</span> <span class="o">=</span> <span class="n">agg_expr</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
            <span class="n">AC</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">AF</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">missing</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">tfloat64</span><span class="p">),</span> <span class="n">AN</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">int32</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="n">homozygote_count</span><span class="o">=</span><span class="mi">0</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="n">freq</span><span class="o">=</span><span class="n">agg_expr</span><span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;cols&quot;</span><span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">select_globals</span><span class="p">(</span><span class="o">**</span><span class="n">group_membership_ht</span><span class="o">.</span><span class="n">index_globals</span><span class="p">())</span>

    <span class="k">return</span> <span class="n">ht</span></div>


<div class="viewcode-block" id="generate_v4_genomes_callstats"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/generate_freq_genomes.html#gnomad_qc.v4.annotations.generate_freq_genomes.generate_v4_genomes_callstats">[docs]</a><span class="k">def</span> <span class="nf">generate_v4_genomes_callstats</span><span class="p">(</span>
    <span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">an_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">pop_diff_an_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate the call stats for the v4.0 genomes release.</span>

<span class="sd">    Merge call stats from the v3.1 release, the v3.1 AN of new v4.0 variants,</span>
<span class="sd">    samples with updated population labels, added samples, and removed samples.</span>

<span class="sd">    Also, compute filtering allele frequencies (&#39;faf&#39;), &#39;grpmax&#39;, &#39;gen_anc_faf_max&#39; and</span>
<span class="sd">    &#39;InbreedingCoeff&#39; on the merged call stats.</span>

<span class="sd">    :param ht: Table returned by `join_release_ht_with_subsets`.</span>
<span class="sd">    :param an_ht: Table with the allele number for new variants in the v4.0 release.</span>
<span class="sd">    :param pop_diff_an_ht: Table with the allele number for samples with updated pop</span>
<span class="sd">        labels and variants in the v4.0 release but not present in the callstats of this subset.</span>
<span class="sd">    :return: Table with the updated call stats for the v4.0 genomes release.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Updating AN Table HGDP pop labels and &#39;oth&#39; -&gt; &#39;remaining&#39;...&quot;</span><span class="p">)</span>
    <span class="n">an_ht</span> <span class="o">=</span> <span class="n">update_pop_labels</span><span class="p">(</span><span class="n">an_ht</span><span class="p">,</span> <span class="n">POP_MAP</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Merging call stats from new variants of v3.1 release samples, from pop_diff &quot;</span>
        <span class="s2">&quot;samples and added samples...&quot;</span>
    <span class="p">)</span>
    <span class="n">global_array</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">index_globals</span><span class="p">()</span><span class="o">.</span><span class="n">global_array</span>
    <span class="n">farrays</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">ht</span><span class="o">.</span><span class="n">ann_array</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">freq</span><span class="p">,</span>
        <span class="n">an_ht</span><span class="p">[</span><span class="n">ht</span><span class="o">.</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">freq</span><span class="p">,</span>
        <span class="n">pop_diff_an_ht</span><span class="p">[</span><span class="n">ht</span><span class="o">.</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">freq</span><span class="p">,</span>
    <span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">ht</span><span class="o">.</span><span class="n">ann_array</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">freq</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">)]</span>
    <span class="n">fmeta</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">global_array</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">freq_meta</span><span class="p">,</span>
        <span class="n">an_ht</span><span class="o">.</span><span class="n">index_globals</span><span class="p">()</span><span class="o">.</span><span class="n">freq_meta</span><span class="p">,</span>
        <span class="n">pop_diff_an_ht</span><span class="o">.</span><span class="n">index_globals</span><span class="p">()</span><span class="o">.</span><span class="n">freq_meta</span><span class="p">,</span>
    <span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">global_array</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">freq_meta</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">)]</span>
    <span class="n">count_arrays</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">global_array</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">freq_meta_sample_count</span><span class="p">,</span>
        <span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">len</span><span class="p">(</span><span class="n">an_ht</span><span class="o">.</span><span class="n">freq_meta</span><span class="p">)),</span>
        <span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">len</span><span class="p">(</span><span class="n">pop_diff_an_ht</span><span class="o">.</span><span class="n">freq_meta</span><span class="p">)),</span>
    <span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">global_array</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">freq_meta_sample_count</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">)]</span>

    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="n">farrays</span><span class="o">=</span><span class="n">farrays</span><span class="p">,</span> <span class="n">sub_array</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">ann_array</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">freq</span><span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span><span class="n">new_temp_file</span><span class="p">(</span><span class="s2">&quot;farrays&quot;</span><span class="p">,</span> <span class="s2">&quot;ht&quot;</span><span class="p">))</span>

    <span class="c1"># Merge the call stats from the v3.1 release, v3.1 AN of new v4.0 variants,</span>
    <span class="c1"># pop_diff, and added samples.</span>
    <span class="n">freq_expr</span><span class="p">,</span> <span class="n">freq_meta</span><span class="p">,</span> <span class="n">sample_counts</span> <span class="o">=</span> <span class="n">merge_freq_arrays</span><span class="p">(</span>
        <span class="p">[</span><span class="n">ht</span><span class="o">.</span><span class="n">farrays</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">fmeta</span><span class="p">))],</span>
        <span class="n">fmeta</span><span class="p">,</span>
        <span class="n">count_arrays</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;counts&quot;</span><span class="p">:</span> <span class="n">count_arrays</span><span class="p">},</span>
    <span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">freq</span><span class="o">=</span><span class="n">freq_expr</span><span class="p">)</span>

    <span class="c1"># Needs checkpoint here to avoid hanging.</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span><span class="n">new_temp_file</span><span class="p">(</span><span class="s2">&quot;added&quot;</span><span class="p">,</span> <span class="s2">&quot;ht&quot;</span><span class="p">))</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Merging call stats from subtracted samples...&quot;</span><span class="p">)</span>
    <span class="c1"># Normally `set_negatives_to_zero` should be False, but we set it to True</span>
    <span class="c1"># because there was a bug in the v3.1 frequency code where we used the v3.0</span>
    <span class="c1"># freq_ht for the hotfix of depletion of homozygous frequency calculation,</span>
    <span class="c1"># once a variant is new in v3.1 (i.e. not present in v3.0 freq_ht) and is from</span>
    <span class="c1"># samples with heterozygous non-reference genotypes with high allele balances,</span>
    <span class="c1"># their GT was set to missing in the v3.1 freq_ht, hence we don&#39;t have</span>
    <span class="c1"># non-zero callstats for these variants in v3.1 release while non-zero callstats</span>
    <span class="c1"># are present the corrected &quot;subtract&quot; freq_ht.</span>
    <span class="n">freq_expr</span><span class="p">,</span> <span class="n">freq_meta</span><span class="p">,</span> <span class="n">sample_counts</span> <span class="o">=</span> <span class="n">merge_freq_arrays</span><span class="p">(</span>
        <span class="p">[</span><span class="n">ht</span><span class="o">.</span><span class="n">freq</span><span class="p">,</span> <span class="n">ht</span><span class="o">.</span><span class="n">sub_array</span><span class="p">],</span>
        <span class="p">[</span><span class="n">freq_meta</span><span class="p">,</span> <span class="n">global_array</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">freq_meta</span><span class="p">],</span>
        <span class="n">operation</span><span class="o">=</span><span class="s2">&quot;diff&quot;</span><span class="p">,</span>
        <span class="n">set_negatives_to_zero</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">count_arrays</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;counts&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">sample_counts</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">],</span> <span class="n">global_array</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">freq_meta_sample_count</span><span class="p">]</span>
        <span class="p">},</span>
    <span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="n">freq</span><span class="o">=</span><span class="n">freq_expr</span><span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">select_globals</span><span class="p">(</span>
        <span class="n">freq_meta</span><span class="o">=</span><span class="n">freq_meta</span><span class="p">,</span>
        <span class="n">freq_index_dict</span><span class="o">=</span><span class="n">make_freq_index_dict_from_meta</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">literal</span><span class="p">(</span><span class="n">freq_meta</span><span class="p">)),</span>
        <span class="n">freq_meta_sample_count</span><span class="o">=</span><span class="n">sample_counts</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">],</span>
    <span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span><span class="n">new_temp_file</span><span class="p">(</span><span class="s2">&quot;merged&quot;</span><span class="p">,</span> <span class="s2">&quot;ht&quot;</span><span class="p">))</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Set Y-variant frequency call stats for female-specific metrics to missing &quot;</span>
        <span class="s2">&quot;structs and the allele frequency to missing for any call stats entries with &quot;</span>
        <span class="s2">&quot;AN == 0...&quot;</span>
    <span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="n">freq</span><span class="o">=</span><span class="n">set_xx_y_metrics_to_na_expr</span><span class="p">(</span><span class="n">ht</span><span class="p">)</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">AF</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">or_missing</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">AN</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="n">x</span><span class="o">.</span><span class="n">AF</span><span class="p">))</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="c1"># Change the &#39;pop&#39; keys in the freq_meta array to &#39;gen_anc&#39;.</span>
    <span class="n">freq_meta</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">literal</span><span class="p">(</span>
        <span class="p">[{(</span><span class="s2">&quot;gen_anc&quot;</span> <span class="k">if</span> <span class="n">k</span> <span class="o">==</span> <span class="s2">&quot;pop&quot;</span> <span class="k">else</span> <span class="n">k</span><span class="p">):</span> <span class="n">m</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">m</span><span class="p">}</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">freq_meta</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span><span class="n">freq_meta</span><span class="o">=</span><span class="n">freq_meta</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">ht</span></div>


<div class="viewcode-block" id="finalize_v4_genomes_callstats"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/generate_freq_genomes.html#gnomad_qc.v4.annotations.generate_freq_genomes.finalize_v4_genomes_callstats">[docs]</a><span class="k">def</span> <span class="nf">finalize_v4_genomes_callstats</span><span class="p">(</span>
    <span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">v3_sites_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">v3_meta_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">updated_meta_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Finalize the call stats for the v4.0 genomes release.</span>

<span class="sd">    The following is done to create the final v4.0 genomes call stats:</span>
<span class="sd">        - Compute filtering allele frequencies (&#39;faf&#39;), &#39;grpmax&#39;, &#39;gen_anc_faf_max&#39; and</span>
<span class="sd">          &#39;inbreeding_coeff&#39; on the v4.0 genomes call stats.</span>
<span class="sd">        - Drop downsamplings from the call stats.</span>
<span class="sd">        - Get the quality histograms from the v3.1 release.</span>
<span class="sd">        - Get the age distribution for the v4.0 genomes release.</span>

<span class="sd">    :param ht: Table with the updated call stats for the v4.0 genomes release.</span>
<span class="sd">    :param v3_sites_ht: Table for the v3.1.2 release sites.</span>
<span class="sd">    :param v3_meta_ht: Table for the v3.1 release metadata.</span>
<span class="sd">    :param updated_meta_ht: Table for the updated metadata.</span>
<span class="sd">    :return: Table with the finalized call stats for the v4.0 genomes release.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Drop downsampling call stats for v4.0 genomes release...&quot;</span><span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">filter_freq_arrays</span><span class="p">(</span><span class="n">ht</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;downsampling&quot;</span><span class="p">],</span> <span class="n">keep</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="c1"># Change the &#39;gen_anc&#39; keys in the freq_meta array to &#39;pop&#39; to compute faf.</span>
    <span class="n">freq_meta</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">literal</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="p">{(</span><span class="s2">&quot;pop&quot;</span> <span class="k">if</span> <span class="n">k</span> <span class="o">==</span> <span class="s2">&quot;gen_anc&quot;</span> <span class="k">else</span> <span class="n">k</span><span class="p">):</span> <span class="n">m</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">m</span><span class="p">}</span>
            <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">freq_meta</span><span class="p">)</span>
        <span class="p">]</span>
    <span class="p">)</span>

    <span class="c1"># Compute filtering allele frequency (faf), grpmax, and gen_anc_faf_max.</span>
    <span class="n">faf</span><span class="p">,</span> <span class="n">faf_meta</span> <span class="o">=</span> <span class="n">faf_expr</span><span class="p">(</span>
        <span class="n">ht</span><span class="o">.</span><span class="n">freq</span><span class="p">,</span> <span class="n">freq_meta</span><span class="p">,</span> <span class="n">ht</span><span class="o">.</span><span class="n">locus</span><span class="p">,</span> <span class="n">GEN_ANC_GROUPS_TO_REMOVE_FOR_GRPMAX</span><span class="p">[</span><span class="s2">&quot;v3&quot;</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="n">grpmax</span> <span class="o">=</span> <span class="n">grpmax_expr</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">freq</span><span class="p">,</span> <span class="n">freq_meta</span><span class="p">,</span> <span class="n">GEN_ANC_GROUPS_TO_REMOVE_FOR_GRPMAX</span><span class="p">[</span><span class="s2">&quot;v3&quot;</span><span class="p">])</span>
    <span class="n">grpmax</span> <span class="o">=</span> <span class="n">grpmax</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="n">gen_anc</span><span class="o">=</span><span class="n">grpmax</span><span class="o">.</span><span class="n">pop</span><span class="p">,</span>
        <span class="n">faf95</span><span class="o">=</span><span class="n">faf</span><span class="p">[</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">literal</span><span class="p">(</span><span class="n">faf_meta</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="k">lambda</span> <span class="n">y</span><span class="p">:</span> <span class="n">y</span><span class="o">.</span><span class="n">values</span><span class="p">()</span> <span class="o">==</span> <span class="p">[</span><span class="s2">&quot;adj&quot;</span><span class="p">,</span> <span class="n">grpmax</span><span class="o">.</span><span class="n">pop</span><span class="p">])</span>
        <span class="p">]</span><span class="o">.</span><span class="n">faf95</span><span class="p">,</span>
    <span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;pop&quot;</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Annotating &#39;faf&#39;, &#39;grpmax&#39;, &#39;gen_anc_faf_max&#39; and &#39;InbreedingCoeff&#39;...&quot;</span>
    <span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="n">faf</span><span class="o">=</span><span class="n">faf</span><span class="p">,</span>
        <span class="n">grpmax</span><span class="o">=</span><span class="n">grpmax</span><span class="p">,</span>
        <span class="n">gen_anc_faf_max</span><span class="o">=</span><span class="n">gen_anc_faf_max_expr</span><span class="p">(</span><span class="n">faf</span><span class="p">,</span> <span class="n">faf_meta</span><span class="p">),</span>
        <span class="n">InbreedingCoeff</span><span class="o">=</span><span class="n">bi_allelic_site_inbreeding_expr</span><span class="p">(</span><span class="n">callstats_expr</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">freq</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span>
    <span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Annotating globals &#39;freq_meta&#39;, &#39;freq_meta_sample_count&#39;, &#39;faf_meta&#39;, &quot;</span>
        <span class="s2">&quot;&#39;freq_index_dict&#39; and &#39;faf_index_dict&#39;...&quot;</span>
    <span class="p">)</span>
    <span class="n">faf_index_dict</span> <span class="o">=</span> <span class="n">make_freq_index_dict_from_meta</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">literal</span><span class="p">(</span><span class="n">faf_meta</span><span class="p">))</span>

    <span class="c1"># Change the &#39;pop&#39; keys back to &#39;gen_anc&#39; in the freq_meta and faf_meta arrays.</span>
    <span class="n">freq_meta</span><span class="p">,</span> <span class="n">faf_meta</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">literal</span><span class="p">([{(</span><span class="s2">&quot;gen_anc&quot;</span> <span class="k">if</span> <span class="n">k</span> <span class="o">==</span> <span class="s2">&quot;pop&quot;</span> <span class="k">else</span> <span class="n">k</span><span class="p">):</span> <span class="n">m</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">m</span><span class="p">}</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">meta</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">meta</span> <span class="ow">in</span> <span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">freq_meta</span><span class="p">),</span> <span class="n">faf_meta</span><span class="p">]</span>
    <span class="p">]</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span>
        <span class="n">freq_meta</span><span class="o">=</span><span class="n">freq_meta</span><span class="p">,</span>
        <span class="n">freq_index_dict</span><span class="o">=</span><span class="n">make_freq_index_dict_from_meta</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">literal</span><span class="p">(</span><span class="n">freq_meta</span><span class="p">)),</span>
        <span class="n">faf_meta</span><span class="o">=</span><span class="n">faf_meta</span><span class="p">,</span>
        <span class="n">faf_index_dict</span><span class="o">=</span><span class="n">faf_index_dict</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Getting quality histograms from the v3.1.2 release...&quot;</span><span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">get_histograms</span><span class="p">(</span><span class="n">ht</span><span class="p">,</span> <span class="n">v3_sites_ht</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Getting age distribution for v4.0 genomes release...&quot;</span><span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span>
        <span class="n">age_distribution</span><span class="o">=</span><span class="n">get_age_distribution</span><span class="p">(</span><span class="n">v3_meta_ht</span><span class="p">,</span> <span class="n">updated_meta_ht</span><span class="p">),</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">ht</span></div>


<div class="viewcode-block" id="get_histograms"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/generate_freq_genomes.html#gnomad_qc.v4.annotations.generate_freq_genomes.get_histograms">[docs]</a><span class="k">def</span> <span class="nf">get_histograms</span><span class="p">(</span><span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">v3_sites_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get histograms for the v4.0 genomes release from the v3.1 release.</span>

<span class="sd">    .. note:</span>
<span class="sd">       We didn&#39;t compute the variant histograms for the v4.0 genomes release because</span>
<span class="sd">       we didn&#39;t densify the VDS for all the variants. The histograms for the new</span>
<span class="sd">       variants will be missing, and the histograms for the variants included</span>
<span class="sd">       in v3.1 release will be the same as the v3.1 release.</span>

<span class="sd">    :param ht: Table with the call stats for the v4.0 genomes release.</span>
<span class="sd">    :param v3_sites_ht: Table for the v3.1 release sites.</span>
<span class="sd">    :return: Table with the histograms added for the v4.0 genomes release.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Getting histograms for the v4.0 genomes release...&quot;</span><span class="p">)</span>
    <span class="n">v3_sites</span> <span class="o">=</span> <span class="n">v3_sites_ht</span><span class="p">[</span><span class="n">ht</span><span class="o">.</span><span class="n">key</span><span class="p">]</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="n">histograms</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
            <span class="n">qual_hists</span><span class="o">=</span><span class="n">v3_sites</span><span class="o">.</span><span class="n">qual_hists</span><span class="p">,</span>
            <span class="n">raw_qual_hists</span><span class="o">=</span><span class="n">v3_sites</span><span class="o">.</span><span class="n">raw_qual_hists</span><span class="p">,</span>
            <span class="n">age_hists</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                <span class="n">age_hist_het</span><span class="o">=</span><span class="n">v3_sites</span><span class="o">.</span><span class="n">age_hist_het</span><span class="p">,</span>
                <span class="n">age_hist_hom</span><span class="o">=</span><span class="n">v3_sites</span><span class="o">.</span><span class="n">age_hist_hom</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">ht</span></div>


<div class="viewcode-block" id="set_downsampling_freq_missing"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/generate_freq_genomes.html#gnomad_qc.v4.annotations.generate_freq_genomes.set_downsampling_freq_missing">[docs]</a><span class="k">def</span> <span class="nf">set_downsampling_freq_missing</span><span class="p">(</span><span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">new_variants_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Set the downsampling call stats for new variants in `ht` to missing.</span>

<span class="sd">    :param ht: Table with the call stats for all variants in the v4.0 genomes release.</span>
<span class="sd">    :param new_variants_ht: Table with new variants in the v4.0 genomes release.</span>
<span class="sd">    :return: Table with the downsampling call stats for new variants set to missing.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">downsampling_indices</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">range</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">len</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">freq_meta</span><span class="p">))</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">i</span><span class="p">:</span> <span class="n">ht</span><span class="o">.</span><span class="n">freq_meta</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;downsampling&quot;</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="n">freq</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">new_variants_ht</span><span class="p">[</span><span class="n">ht</span><span class="o">.</span><span class="n">key</span><span class="p">]),</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">enumerate</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">freq</span><span class="p">)</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
                <span class="k">lambda</span> <span class="n">i</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span>
                    <span class="n">downsampling_indices</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span>
                    <span class="n">missing_callstats_expr</span><span class="p">(),</span>
                    <span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                <span class="p">)</span>
            <span class="p">),</span>
            <span class="n">ht</span><span class="o">.</span><span class="n">freq</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">ht</span></div>


<div class="viewcode-block" id="get_age_distribution"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/generate_freq_genomes.html#gnomad_qc.v4.annotations.generate_freq_genomes.get_age_distribution">[docs]</a><span class="k">def</span> <span class="nf">get_age_distribution</span><span class="p">(</span>
    <span class="n">v3_meta_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">subset_updated_meta_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">StructExpression</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get the updated &#39;age_distribution&#39; annotation for v4.0 genomes.</span>

<span class="sd">    :param v3_meta_ht: Table with the v3.1.2 release sample metadata.</span>
<span class="sd">    :param subset_updated_meta_ht: Table with the updated sample metadata for the</span>
<span class="sd">       HGDP + 1KG subset.</span>
<span class="sd">    :return: Expression with the age distribution for samples in `subset_updated_meta`.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Getting age distribution for v4.0 genomes release...&quot;</span><span class="p">)</span>
    <span class="n">v3_release_count</span> <span class="o">=</span> <span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">release</span><span class="p">)</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
    <span class="n">v4_meta_ht</span> <span class="o">=</span> <span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
        <span class="n">age</span><span class="o">=</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">age</span><span class="p">,</span>
        <span class="n">release</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span>
            <span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">subsets</span><span class="o">.</span><span class="n">tgp</span> <span class="o">|</span> <span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">subsets</span><span class="o">.</span><span class="n">hgdp</span><span class="p">,</span>
            <span class="n">subset_updated_meta_ht</span><span class="p">[</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">s</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;v3.1::&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)]</span><span class="o">.</span><span class="n">gnomad_release</span><span class="p">,</span>
            <span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">release</span><span class="p">,</span>
        <span class="p">),</span>
    <span class="p">)</span>
    <span class="n">v4_meta_ht</span> <span class="o">=</span> <span class="n">v4_meta_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">v4_meta_ht</span><span class="o">.</span><span class="n">release</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;There will be </span><span class="si">%s</span><span class="s2"> samples in the v4.0 genome release, compared to </span><span class="si">%s</span><span class="s2"> in v3.1 &quot;</span>
        <span class="s2">&quot;genome release.&quot;</span><span class="p">,</span>
        <span class="n">v4_meta_ht</span><span class="o">.</span><span class="n">count</span><span class="p">(),</span>
        <span class="n">v3_release_count</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">age_expr</span> <span class="o">=</span> <span class="n">v4_meta_ht</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">v4_meta_ht</span><span class="o">.</span><span class="n">age</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">80</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">age_expr</span></div>


<div class="viewcode-block" id="get_pop_diff_v3_vds_group_membership"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/generate_freq_genomes.html#gnomad_qc.v4.annotations.generate_freq_genomes.get_pop_diff_v3_vds_group_membership">[docs]</a><span class="k">def</span> <span class="nf">get_pop_diff_v3_vds_group_membership</span><span class="p">(</span>
    <span class="n">v3_vds</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">VariantDataset</span><span class="p">,</span>
    <span class="n">meta_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get the group membership for samples in pop_diff for the v3.1 VDS.</span>

<span class="sd">    :param v3_vds: VDS with the v3.1 release samples.</span>
<span class="sd">    :param meta_ht: Table with the updated HGDP + 1KG sample metadata.</span>
<span class="sd">    :return: Table with the group membership for samples in pop_diff.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">v3_sites_meta_ht</span> <span class="o">=</span> <span class="n">v3_vds</span><span class="o">.</span><span class="n">variant_data</span><span class="o">.</span><span class="n">cols</span><span class="p">()</span>
    <span class="n">v3_sites_meta_ht</span> <span class="o">=</span> <span class="n">v3_sites_meta_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">v3_sites_meta_ht</span><span class="o">.</span><span class="n">meta</span><span class="o">.</span><span class="n">release</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Annotating call stats group membership for pop diff release samples...&quot;</span>
    <span class="p">)</span>
    <span class="c1"># Need to make sure we are using the names that are in the vds for the release</span>
    <span class="c1"># samples.</span>
    <span class="c1"># Note: the samples in the pop_diff HT are samples in the to-be-split &#39;Han&#39; and</span>
    <span class="c1"># &#39;Papuan&#39; populations AND their &#39;gnomad_release&#39; status hasn&#39;t changed.</span>
    <span class="n">v3_sites_meta_rename_ht</span> <span class="o">=</span> <span class="n">v3_sites_meta_ht</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span>
        <span class="n">s_no_prefix</span><span class="o">=</span><span class="n">v3_sites_meta_ht</span><span class="o">.</span><span class="n">s</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;v3.1::&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="n">pop_diff_sample_ht</span> <span class="o">=</span> <span class="n">get_updated_release_samples</span><span class="p">(</span><span class="n">meta_ht</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">pop_diff_sample_ht</span> <span class="o">=</span> <span class="n">pop_diff_sample_ht</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span>
        <span class="n">s</span><span class="o">=</span><span class="n">v3_sites_meta_rename_ht</span><span class="p">[</span><span class="n">pop_diff_sample_ht</span><span class="o">.</span><span class="n">s</span><span class="p">]</span><span class="o">.</span><span class="n">s</span>
    <span class="p">)</span>
    <span class="n">pop_diff_group_membership_ht</span> <span class="o">=</span> <span class="n">get_group_membership_ht_for_an</span><span class="p">(</span>
        <span class="n">pop_diff_sample_ht</span><span class="p">,</span> <span class="n">only_pop_diff</span><span class="o">=</span><span class="kc">True</span>
    <span class="p">)</span>
    <span class="n">pop_diff_group_membership_ht</span> <span class="o">=</span> <span class="n">pop_diff_group_membership_ht</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span>
        <span class="n">new_temp_file</span><span class="p">(</span><span class="s2">&quot;group_membership_pop_diff&quot;</span><span class="p">,</span> <span class="s2">&quot;ht&quot;</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">pop_diff_group_membership_ht</span></div>


<div class="viewcode-block" id="patch_v4_genomes_callstats"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/generate_freq_genomes.html#gnomad_qc.v4.annotations.generate_freq_genomes.patch_v4_genomes_callstats">[docs]</a><span class="k">def</span> <span class="nf">patch_v4_genomes_callstats</span><span class="p">(</span><span class="n">freq_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Patch the call stats for some inconsistent variant calls found during validity checks.</span>

<span class="sd">    14 variants were found to have inconsistent calls between v3.1 and v4.0 genomes</span>
<span class="sd">    release, we determined the reason for each of these by manual inspection and</span>
<span class="sd">    this function applies the needed patches to these variants.</span>

<span class="sd">    .. note::</span>
<span class="sd">        This is a temporary fix until we can recompute the call stats for all the</span>
<span class="sd">        samples.</span>

<span class="sd">    :param freq_ht: Table with the call stats for the v4.0 genomes release.</span>
<span class="sd">    :return: Table with the patched call stats for the v4.0 genomes release.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># For most of the variants, there is a single Han sample that was 0/1 in v3.1, but</span>
    <span class="c1"># from the dense MT in v4.0 frequencies it was given a missing genotype. These</span>
    <span class="c1"># variants are not in v3, which is why they get a missing genotype in the v4.0</span>
    <span class="c1"># freq. We dont understand why they were 0/1 in v3.1, we would expect them to</span>
    <span class="c1"># have been missing. It should be 0/1 if missing v3 frequency had been handled</span>
    <span class="c1"># correctly in the fix for high AB hets.</span>
    <span class="n">freq_groups1</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">set</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="p">{</span><span class="s2">&quot;group&quot;</span><span class="p">:</span> <span class="s2">&quot;adj&quot;</span><span class="p">,</span> <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="s2">&quot;hgdp&quot;</span><span class="p">,</span> <span class="s2">&quot;gen_anc&quot;</span><span class="p">:</span> <span class="s2">&quot;han&quot;</span><span class="p">},</span>
            <span class="p">{</span><span class="s2">&quot;group&quot;</span><span class="p">:</span> <span class="s2">&quot;adj&quot;</span><span class="p">,</span> <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="s2">&quot;hgdp&quot;</span><span class="p">,</span> <span class="s2">&quot;gen_anc&quot;</span><span class="p">:</span> <span class="s2">&quot;han&quot;</span><span class="p">,</span> <span class="s2">&quot;sex&quot;</span><span class="p">:</span> <span class="s2">&quot;XX&quot;</span><span class="p">},</span>
        <span class="p">]</span>
    <span class="p">)</span>
    <span class="n">patch1_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="o">.</span><span class="n">parallelize</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="p">{</span><span class="s2">&quot;locus&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">parse_locus</span><span class="p">(</span><span class="s2">&quot;chr1:111544780&quot;</span><span class="p">),</span> <span class="s2">&quot;alleles&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;T&quot;</span><span class="p">]},</span>
            <span class="p">{</span><span class="s2">&quot;locus&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">parse_locus</span><span class="p">(</span><span class="s2">&quot;chr11:119744028&quot;</span><span class="p">),</span> <span class="s2">&quot;alleles&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;G&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">]},</span>
            <span class="p">{</span><span class="s2">&quot;locus&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">parse_locus</span><span class="p">(</span><span class="s2">&quot;chr12:82828211&quot;</span><span class="p">),</span> <span class="s2">&quot;alleles&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;CTTTTTTT&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">]},</span>
            <span class="p">{</span>
                <span class="s2">&quot;locus&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">parse_locus</span><span class="p">(</span><span class="s2">&quot;chr16:87738638&quot;</span><span class="p">),</span>
                <span class="s2">&quot;alleles&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;TCATCCATCCACACACCAGCATCTCATC&quot;</span><span class="p">,</span> <span class="s2">&quot;T&quot;</span><span class="p">],</span>
            <span class="p">},</span>
            <span class="p">{</span><span class="s2">&quot;locus&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">parse_locus</span><span class="p">(</span><span class="s2">&quot;chr22:22399516&quot;</span><span class="p">),</span> <span class="s2">&quot;alleles&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;G&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">]},</span>
        <span class="p">]</span>
    <span class="p">)</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">,</span> <span class="s2">&quot;alleles&quot;</span><span class="p">)</span>
    <span class="n">patch1_expr</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">m</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span>
            <span class="n">freq_groups1</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">m</span><span class="p">),</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                <span class="n">AC</span><span class="o">=</span><span class="n">x</span><span class="o">.</span><span class="n">AC</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                <span class="n">AF</span><span class="o">=</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">AC</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">AN</span> <span class="o">+</span> <span class="mi">2</span><span class="p">),</span>
                <span class="n">AN</span><span class="o">=</span><span class="n">x</span><span class="o">.</span><span class="n">AN</span> <span class="o">+</span> <span class="mi">2</span><span class="p">,</span>
                <span class="n">homozygote_count</span><span class="o">=</span><span class="n">x</span><span class="o">.</span><span class="n">homozygote_count</span><span class="p">,</span>
            <span class="p">),</span>
            <span class="n">x</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="n">freq_ht</span><span class="o">.</span><span class="n">freq</span><span class="p">,</span>
        <span class="n">freq_ht</span><span class="o">.</span><span class="n">freq_meta</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">freq_groups2</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">set</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="p">{</span><span class="s2">&quot;group&quot;</span><span class="p">:</span> <span class="s2">&quot;adj&quot;</span><span class="p">,</span> <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="s2">&quot;hgdp&quot;</span><span class="p">,</span> <span class="s2">&quot;gen_anc&quot;</span><span class="p">:</span> <span class="s2">&quot;han&quot;</span><span class="p">},</span>
            <span class="p">{</span><span class="s2">&quot;group&quot;</span><span class="p">:</span> <span class="s2">&quot;adj&quot;</span><span class="p">,</span> <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="s2">&quot;hgdp&quot;</span><span class="p">,</span> <span class="s2">&quot;gen_anc&quot;</span><span class="p">:</span> <span class="s2">&quot;han&quot;</span><span class="p">,</span> <span class="s2">&quot;sex&quot;</span><span class="p">:</span> <span class="s2">&quot;XY&quot;</span><span class="p">},</span>
        <span class="p">]</span>
    <span class="p">)</span>
    <span class="n">patch2_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="o">.</span><span class="n">parallelize</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="p">{</span>
                <span class="s2">&quot;locus&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">parse_locus</span><span class="p">(</span><span class="s2">&quot;chr12:10380059&quot;</span><span class="p">),</span>
                <span class="s2">&quot;alleles&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;GTTTTTTTTTTTTTTT&quot;</span><span class="p">,</span> <span class="s2">&quot;G&quot;</span><span class="p">],</span>
            <span class="p">},</span>
            <span class="p">{</span>
                <span class="s2">&quot;locus&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">parse_locus</span><span class="p">(</span><span class="s2">&quot;chr16:73308279&quot;</span><span class="p">),</span>
                <span class="s2">&quot;alleles&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;T&quot;</span><span class="p">,</span> <span class="s2">&quot;G&quot;</span><span class="p">],</span>
            <span class="p">},</span>
        <span class="p">]</span>
    <span class="p">)</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">,</span> <span class="s2">&quot;alleles&quot;</span><span class="p">)</span>
    <span class="n">patch2_expr</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">m</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span>
            <span class="n">freq_groups2</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">m</span><span class="p">),</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                <span class="n">AC</span><span class="o">=</span><span class="n">x</span><span class="o">.</span><span class="n">AC</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                <span class="n">AF</span><span class="o">=</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">AC</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">AN</span> <span class="o">+</span> <span class="mi">2</span><span class="p">),</span>
                <span class="n">AN</span><span class="o">=</span><span class="n">x</span><span class="o">.</span><span class="n">AN</span> <span class="o">+</span> <span class="mi">2</span><span class="p">,</span>
                <span class="n">homozygote_count</span><span class="o">=</span><span class="n">x</span><span class="o">.</span><span class="n">homozygote_count</span><span class="p">,</span>
            <span class="p">),</span>
            <span class="n">x</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="n">freq_ht</span><span class="o">.</span><span class="n">freq</span><span class="p">,</span>
        <span class="n">freq_ht</span><span class="o">.</span><span class="n">freq_meta</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">freq_groups3</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">set</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="p">{</span><span class="s2">&quot;group&quot;</span><span class="p">:</span> <span class="s2">&quot;adj&quot;</span><span class="p">,</span> <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="s2">&quot;hgdp&quot;</span><span class="p">,</span> <span class="s2">&quot;gen_anc&quot;</span><span class="p">:</span> <span class="s2">&quot;northernhan&quot;</span><span class="p">},</span>
            <span class="p">{</span><span class="s2">&quot;group&quot;</span><span class="p">:</span> <span class="s2">&quot;adj&quot;</span><span class="p">,</span> <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="s2">&quot;hgdp&quot;</span><span class="p">,</span> <span class="s2">&quot;gen_anc&quot;</span><span class="p">:</span> <span class="s2">&quot;northernhan&quot;</span><span class="p">,</span> <span class="s2">&quot;sex&quot;</span><span class="p">:</span> <span class="s2">&quot;XY&quot;</span><span class="p">},</span>
        <span class="p">]</span>
    <span class="p">)</span>
    <span class="n">patch3_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="o">.</span><span class="n">parallelize</span><span class="p">(</span>
        <span class="p">[{</span><span class="s2">&quot;locus&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">parse_locus</span><span class="p">(</span><span class="s2">&quot;chr7:14689208&quot;</span><span class="p">),</span> <span class="s2">&quot;alleles&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;T&quot;</span><span class="p">,</span> <span class="s2">&quot;TTTTTTTTA&quot;</span><span class="p">]}]</span>
    <span class="p">)</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">,</span> <span class="s2">&quot;alleles&quot;</span><span class="p">)</span>
    <span class="n">patch3_expr</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">m</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span>
            <span class="n">freq_groups3</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">m</span><span class="p">),</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                <span class="n">AC</span><span class="o">=</span><span class="n">x</span><span class="o">.</span><span class="n">AC</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                <span class="n">AF</span><span class="o">=</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">AC</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">AN</span> <span class="o">+</span> <span class="mi">2</span><span class="p">),</span>
                <span class="n">AN</span><span class="o">=</span><span class="n">x</span><span class="o">.</span><span class="n">AN</span> <span class="o">+</span> <span class="mi">2</span><span class="p">,</span>
                <span class="n">homozygote_count</span><span class="o">=</span><span class="n">x</span><span class="o">.</span><span class="n">homozygote_count</span><span class="p">,</span>
            <span class="p">),</span>
            <span class="n">x</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="n">freq_ht</span><span class="o">.</span><span class="n">freq</span><span class="p">,</span>
        <span class="n">freq_ht</span><span class="o">.</span><span class="n">freq_meta</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">freq_groups4</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">set</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="p">{</span><span class="s2">&quot;group&quot;</span><span class="p">:</span> <span class="s2">&quot;adj&quot;</span><span class="p">,</span> <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="s2">&quot;hgdp&quot;</span><span class="p">,</span> <span class="s2">&quot;gen_anc&quot;</span><span class="p">:</span> <span class="s2">&quot;northernhan&quot;</span><span class="p">},</span>
            <span class="p">{</span><span class="s2">&quot;group&quot;</span><span class="p">:</span> <span class="s2">&quot;adj&quot;</span><span class="p">,</span> <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="s2">&quot;hgdp&quot;</span><span class="p">,</span> <span class="s2">&quot;gen_anc&quot;</span><span class="p">:</span> <span class="s2">&quot;northernhan&quot;</span><span class="p">,</span> <span class="s2">&quot;sex&quot;</span><span class="p">:</span> <span class="s2">&quot;XX&quot;</span><span class="p">},</span>
        <span class="p">]</span>
    <span class="p">)</span>
    <span class="n">patch4_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="o">.</span><span class="n">parallelize</span><span class="p">(</span>
        <span class="p">[{</span><span class="s2">&quot;locus&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">parse_locus</span><span class="p">(</span><span class="s2">&quot;chr9:43045892&quot;</span><span class="p">),</span> <span class="s2">&quot;alleles&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">]}]</span>
    <span class="p">)</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">,</span> <span class="s2">&quot;alleles&quot;</span><span class="p">)</span>
    <span class="n">patch4_expr</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">m</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span>
            <span class="n">freq_groups4</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">m</span><span class="p">),</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                <span class="n">AC</span><span class="o">=</span><span class="n">x</span><span class="o">.</span><span class="n">AC</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                <span class="n">AF</span><span class="o">=</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">AC</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">AN</span> <span class="o">+</span> <span class="mi">2</span><span class="p">),</span>
                <span class="n">AN</span><span class="o">=</span><span class="n">x</span><span class="o">.</span><span class="n">AN</span> <span class="o">+</span> <span class="mi">2</span><span class="p">,</span>
                <span class="n">homozygote_count</span><span class="o">=</span><span class="n">x</span><span class="o">.</span><span class="n">homozygote_count</span><span class="p">,</span>
            <span class="p">),</span>
            <span class="n">x</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="n">freq_ht</span><span class="o">.</span><span class="n">freq</span><span class="p">,</span>
        <span class="n">freq_ht</span><span class="o">.</span><span class="n">freq_meta</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># chr16:89834177 [&quot;G&quot;,&quot;GGCC&quot;], [&quot;GCC&quot;,&quot;G&quot;], and</span>
    <span class="c1"># [&quot;GCCTGGATAAGCATAGCCCGTGTGAATCTGTGAACCTGCCTGTGCTCACGGTTGGCCGTCGTAGAAGCA&quot;, &quot;G&quot;].</span>
    <span class="c1"># Code added to the HGDP + 1KG subset to remove alleles not in the subset caused</span>
    <span class="c1"># the duplication of a single site that changed position during the minrep. This</span>
    <span class="c1"># results in 2 of the Han samples being given NA instead of 0/0 as their GT. To fix</span>
    <span class="c1"># this we are adding 4 (2 samples * 2 alleles) to AN.</span>
    <span class="n">freq_groups5</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">set</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="p">{</span><span class="s2">&quot;group&quot;</span><span class="p">:</span> <span class="s2">&quot;adj&quot;</span><span class="p">,</span> <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="s2">&quot;hgdp&quot;</span><span class="p">,</span> <span class="s2">&quot;gen_anc&quot;</span><span class="p">:</span> <span class="s2">&quot;northernhan&quot;</span><span class="p">},</span>
            <span class="p">{</span><span class="s2">&quot;group&quot;</span><span class="p">:</span> <span class="s2">&quot;adj&quot;</span><span class="p">,</span> <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="s2">&quot;hgdp&quot;</span><span class="p">,</span> <span class="s2">&quot;gen_anc&quot;</span><span class="p">:</span> <span class="s2">&quot;northernhan&quot;</span><span class="p">,</span> <span class="s2">&quot;sex&quot;</span><span class="p">:</span> <span class="s2">&quot;XY&quot;</span><span class="p">},</span>
        <span class="p">]</span>
    <span class="p">)</span>
    <span class="n">patch5_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="o">.</span><span class="n">parallelize</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="p">{</span><span class="s2">&quot;locus&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">parse_locus</span><span class="p">(</span><span class="s2">&quot;chr16:89834177&quot;</span><span class="p">),</span> <span class="s2">&quot;alleles&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;G&quot;</span><span class="p">,</span> <span class="s2">&quot;GGCC&quot;</span><span class="p">]},</span>
            <span class="p">{</span><span class="s2">&quot;locus&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">parse_locus</span><span class="p">(</span><span class="s2">&quot;chr16:89834177&quot;</span><span class="p">),</span> <span class="s2">&quot;alleles&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;GCC&quot;</span><span class="p">,</span> <span class="s2">&quot;G&quot;</span><span class="p">]},</span>
            <span class="p">{</span>
                <span class="s2">&quot;locus&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">parse_locus</span><span class="p">(</span><span class="s2">&quot;chr16:89834177&quot;</span><span class="p">),</span>
                <span class="s2">&quot;alleles&quot;</span><span class="p">:</span> <span class="p">[</span>
                    <span class="s2">&quot;GCCTGGATAAGCATAGCCCGTGTGAATCTGTGAACCTGCCTGTGCTCACGGTTGGCCGTCGTAGAAGCA&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;G&quot;</span><span class="p">,</span>
                <span class="p">],</span>
            <span class="p">},</span>
        <span class="p">]</span>
    <span class="p">)</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">,</span> <span class="s2">&quot;alleles&quot;</span><span class="p">)</span>
    <span class="n">patch5_expr</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">m</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span>
            <span class="n">freq_groups5</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">m</span><span class="p">),</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                <span class="n">AC</span><span class="o">=</span><span class="n">x</span><span class="o">.</span><span class="n">AC</span><span class="p">,</span>
                <span class="n">AF</span><span class="o">=</span><span class="n">x</span><span class="o">.</span><span class="n">AC</span> <span class="o">/</span> <span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">AN</span> <span class="o">+</span> <span class="mi">4</span><span class="p">),</span>
                <span class="n">AN</span><span class="o">=</span><span class="n">x</span><span class="o">.</span><span class="n">AN</span> <span class="o">+</span> <span class="mi">4</span><span class="p">,</span>
                <span class="n">homozygote_count</span><span class="o">=</span><span class="n">x</span><span class="o">.</span><span class="n">homozygote_count</span><span class="p">,</span>
            <span class="p">),</span>
            <span class="n">x</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="n">freq_ht</span><span class="o">.</span><span class="n">freq</span><span class="p">,</span>
        <span class="n">freq_ht</span><span class="o">.</span><span class="n">freq_meta</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># Frequency groups that need a patch for variant chr9-93596925-ATTTTTTTTTTTTTT-A.</span>
    <span class="c1"># There is a single Han sample causing a discrepancy in the AC and AN count between</span>
    <span class="c1"># v3.1.2 and v4.0. This sample&#39;s GT was set to missing in the v3.1.2 release HT,</span>
    <span class="c1"># but in the computation for v4.0 the sample&#39;s GT was set to 0/1. This is because in</span>
    <span class="c1"># the original code for v3.1 (before the correction to account for het_non_ref</span>
    <span class="c1"># samples), the hl.if_else statement in the correction for high AB hets (now in</span>
    <span class="c1"># gnomad_qc.v3.utils.hom_alt_depletion_fix) didn&#39;t include the _het_non_ref check.</span>
    <span class="c1"># Without that check, the rest of the hl.if_else statement would evaluate to</span>
    <span class="c1"># missing because of the missing v3.0 frequency. However, when the _het_non_ref</span>
    <span class="c1"># check is added, the presence of a False (~_het_non_ref evaluates to False for a</span>
    <span class="c1"># het_non_ref) in the hl.if_else statement causes it to be given a 0/1 instead.</span>
    <span class="c1"># To correct for this we set Han sample GT to missing to match v3.1.2 value.</span>
    <span class="n">freq_groups6</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">set</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="p">{</span><span class="s2">&quot;group&quot;</span><span class="p">:</span> <span class="s2">&quot;adj&quot;</span><span class="p">,</span> <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="s2">&quot;hgdp&quot;</span><span class="p">,</span> <span class="s2">&quot;gen_anc&quot;</span><span class="p">:</span> <span class="s2">&quot;han&quot;</span><span class="p">},</span>
            <span class="p">{</span><span class="s2">&quot;group&quot;</span><span class="p">:</span> <span class="s2">&quot;adj&quot;</span><span class="p">,</span> <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="s2">&quot;hgdp&quot;</span><span class="p">,</span> <span class="s2">&quot;gen_anc&quot;</span><span class="p">:</span> <span class="s2">&quot;han&quot;</span><span class="p">,</span> <span class="s2">&quot;sex&quot;</span><span class="p">:</span> <span class="s2">&quot;XX&quot;</span><span class="p">},</span>
        <span class="p">]</span>
    <span class="p">)</span>
    <span class="n">patch6_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="o">.</span><span class="n">parallelize</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="p">{</span>
                <span class="s2">&quot;locus&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">parse_locus</span><span class="p">(</span><span class="s2">&quot;chr9:93596925&quot;</span><span class="p">),</span>
                <span class="s2">&quot;alleles&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;ATTTTTTTTTTTTTT&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">],</span>
            <span class="p">}</span>
        <span class="p">]</span>
    <span class="p">)</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">,</span> <span class="s2">&quot;alleles&quot;</span><span class="p">)</span>
    <span class="n">patch6_expr</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">m</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span>
            <span class="n">freq_groups6</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">m</span><span class="p">),</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                <span class="n">AC</span><span class="o">=</span><span class="n">x</span><span class="o">.</span><span class="n">AC</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span>
                <span class="n">AF</span><span class="o">=</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">AC</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">AN</span> <span class="o">-</span> <span class="mi">2</span><span class="p">),</span>
                <span class="n">AN</span><span class="o">=</span><span class="n">x</span><span class="o">.</span><span class="n">AN</span> <span class="o">-</span> <span class="mi">2</span><span class="p">,</span>
                <span class="n">homozygote_count</span><span class="o">=</span><span class="n">x</span><span class="o">.</span><span class="n">homozygote_count</span><span class="p">,</span>
            <span class="p">),</span>
            <span class="n">x</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="n">freq_ht</span><span class="o">.</span><span class="n">freq</span><span class="p">,</span>
        <span class="n">freq_ht</span><span class="o">.</span><span class="n">freq_meta</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># Frequency groups that need a patch for variant chr13-20656122-T-TAAAAAAAAGAA.</span>
    <span class="c1"># This variant has an AC raw of 1 and AC adj of 2 because one of the TGP samples</span>
    <span class="c1"># being removed has a raw GT of 0/1 but adj GT of None because it fails</span>
    <span class="c1"># adj. In v3.1 it was None for both and therefore shouldn&#39;t be subtracted</span>
    <span class="c1"># from raw either.</span>
    <span class="n">freq_groups7</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">set</span><span class="p">([{</span><span class="s2">&quot;group&quot;</span><span class="p">:</span> <span class="s2">&quot;raw&quot;</span><span class="p">,</span> <span class="s2">&quot;subset&quot;</span><span class="p">:</span> <span class="s2">&quot;tgp&quot;</span><span class="p">},</span> <span class="p">{</span><span class="s2">&quot;group&quot;</span><span class="p">:</span> <span class="s2">&quot;raw&quot;</span><span class="p">}])</span>
    <span class="n">patch7_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="o">.</span><span class="n">parallelize</span><span class="p">(</span>
        <span class="p">[{</span><span class="s2">&quot;locus&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">parse_locus</span><span class="p">(</span><span class="s2">&quot;chr13:20656122&quot;</span><span class="p">),</span> <span class="s2">&quot;alleles&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;T&quot;</span><span class="p">,</span> <span class="s2">&quot;TAAAAAAAAGAA&quot;</span><span class="p">]}]</span>
    <span class="p">)</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">,</span> <span class="s2">&quot;alleles&quot;</span><span class="p">)</span>
    <span class="n">patch7_expr</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">m</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span>
            <span class="n">freq_groups7</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">m</span><span class="p">),</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                <span class="n">AC</span><span class="o">=</span><span class="n">x</span><span class="o">.</span><span class="n">AC</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                <span class="n">AF</span><span class="o">=</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">AC</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">AN</span> <span class="o">+</span> <span class="mi">2</span><span class="p">),</span>
                <span class="n">AN</span><span class="o">=</span><span class="n">x</span><span class="o">.</span><span class="n">AN</span> <span class="o">+</span> <span class="mi">2</span><span class="p">,</span>
                <span class="n">homozygote_count</span><span class="o">=</span><span class="n">x</span><span class="o">.</span><span class="n">homozygote_count</span><span class="p">,</span>
            <span class="p">),</span>
            <span class="n">x</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="n">freq_ht</span><span class="o">.</span><span class="n">freq</span><span class="p">,</span>
        <span class="n">freq_ht</span><span class="o">.</span><span class="n">freq_meta</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># Patch the call stats for the variants that need it.</span>
    <span class="n">freq_ht</span> <span class="o">=</span> <span class="n">freq_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="n">freq</span><span class="o">=</span><span class="p">(</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">case</span><span class="p">(</span><span class="n">missing_false</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">patch1_ht</span><span class="p">[</span><span class="n">freq_ht</span><span class="o">.</span><span class="n">key</span><span class="p">]),</span> <span class="n">patch1_expr</span><span class="p">)</span>
            <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">patch2_ht</span><span class="p">[</span><span class="n">freq_ht</span><span class="o">.</span><span class="n">key</span><span class="p">]),</span> <span class="n">patch2_expr</span><span class="p">)</span>
            <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">patch3_ht</span><span class="p">[</span><span class="n">freq_ht</span><span class="o">.</span><span class="n">key</span><span class="p">]),</span> <span class="n">patch3_expr</span><span class="p">)</span>
            <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">patch4_ht</span><span class="p">[</span><span class="n">freq_ht</span><span class="o">.</span><span class="n">key</span><span class="p">]),</span> <span class="n">patch4_expr</span><span class="p">)</span>
            <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">patch5_ht</span><span class="p">[</span><span class="n">freq_ht</span><span class="o">.</span><span class="n">key</span><span class="p">]),</span> <span class="n">patch5_expr</span><span class="p">)</span>
            <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">patch6_ht</span><span class="p">[</span><span class="n">freq_ht</span><span class="o">.</span><span class="n">key</span><span class="p">]),</span> <span class="n">patch6_expr</span><span class="p">)</span>
            <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">patch7_ht</span><span class="p">[</span><span class="n">freq_ht</span><span class="o">.</span><span class="n">key</span><span class="p">]),</span> <span class="n">patch7_expr</span><span class="p">)</span>
            <span class="o">.</span><span class="n">default</span><span class="p">(</span><span class="n">freq_ht</span><span class="o">.</span><span class="n">freq</span><span class="p">)</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="n">freq_ht</span> <span class="o">=</span> <span class="n">freq_ht</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span><span class="n">new_temp_file</span><span class="p">(</span><span class="s2">&quot;patched&quot;</span><span class="p">,</span> <span class="s2">&quot;ht&quot;</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">freq_ht</span></div>


<div class="viewcode-block" id="get_v4_genomes_release_resources"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/generate_freq_genomes.html#gnomad_qc.v4.annotations.generate_freq_genomes.get_v4_genomes_release_resources">[docs]</a><span class="k">def</span> <span class="nf">get_v4_genomes_release_resources</span><span class="p">(</span>
    <span class="n">test</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">overwrite</span><span class="p">:</span> <span class="nb">bool</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">PipelineResourceCollection</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get PipelineResourceCollection for all resources needed to create the gnomAD v4.0 genomes release.</span>

<span class="sd">    :param test: Whether to gather all resources for testing.</span>
<span class="sd">    :param overwrite: Whether to overwrite resources if they exist.</span>
<span class="sd">    :return: PipelineResourceCollection containing resources for all steps of the</span>
<span class="sd">        gnomAD v4.0 genomes release pipeline.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Initialize gnomAD v4.0 genomes release pipeline resource collection.</span>
    <span class="n">hgdp_tgp_res</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;meta_ht&quot;</span><span class="p">:</span> <span class="n">hgdp_tgp_subset_annotations</span><span class="p">(</span><span class="n">sample</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">versions</span><span class="p">[</span><span class="s2">&quot;3.1.2&quot;</span><span class="p">],</span>
        <span class="s2">&quot;dense_mt&quot;</span><span class="p">:</span> <span class="n">hgdp_tgp_subset</span><span class="p">(</span><span class="n">dense</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">public</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">versions</span><span class="p">[</span><span class="s2">&quot;3.1.2&quot;</span><span class="p">],</span>
    <span class="p">}</span>
    <span class="n">v3_sites_ht</span> <span class="o">=</span> <span class="n">release_sites</span><span class="p">(</span><span class="n">public</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">versions</span><span class="p">[</span><span class="s2">&quot;3.1.2&quot;</span><span class="p">]</span>
    <span class="n">v4_genome_release_pipeline</span> <span class="o">=</span> <span class="n">PipelineResourceCollection</span><span class="p">(</span>
        <span class="n">pipeline_name</span><span class="o">=</span><span class="s2">&quot;gnomad_v4_genomes_release&quot;</span><span class="p">,</span>
        <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">,</span>
        <span class="n">pipeline_resources</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;Released HGDP + 1KG resources&quot;</span><span class="p">:</span> <span class="n">hgdp_tgp_res</span><span class="p">,</span>
            <span class="s2">&quot;gnomAD v3.1.2 release sites HT&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;v3_sites_ht&quot;</span><span class="p">:</span> <span class="n">v3_sites_ht</span><span class="p">},</span>
        <span class="p">},</span>
    <span class="p">)</span>

    <span class="c1"># Create resource collection for each step of the v4.0 genomes release pipeline.</span>
    <span class="n">get_related_to_nonsubset</span> <span class="o">=</span> <span class="n">PipelineStepResourceCollection</span><span class="p">(</span>
        <span class="s2">&quot;--get-related-to-nonsubset&quot;</span><span class="p">,</span>
        <span class="n">input_resources</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;v3.1 metadata&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;v3_meta_ht&quot;</span><span class="p">:</span> <span class="n">v3_meta</span><span class="p">},</span>
            <span class="s2">&quot;v3.1 relatedness&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;v3_relatedness_ht&quot;</span><span class="p">:</span> <span class="n">v3_relatedness</span><span class="p">},</span>
        <span class="p">},</span>
        <span class="n">output_resources</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;related_to_nonsubset_ht&quot;</span><span class="p">:</span> <span class="n">hgdp_tgp_related_to_nonsubset</span><span class="p">},</span>
    <span class="p">)</span>
    <span class="n">hgdp_tgp_v4_exome_duplicates</span> <span class="o">=</span> <span class="n">PipelineStepResourceCollection</span><span class="p">(</span>
        <span class="s2">&quot;--get-duplicated-to-exomes&quot;</span><span class="p">,</span>
        <span class="n">input_resources</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;v3.1 metadata&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;v3_meta_ht&quot;</span><span class="p">:</span> <span class="n">v3_meta</span><span class="p">},</span>
            <span class="s2">&quot;v4.0 metadata&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;v4_meta_ht&quot;</span><span class="p">:</span> <span class="n">v4_meta</span><span class="p">},</span>
            <span class="s2">&quot;v3.1 and v4.0 relatedness&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;v4_relatedness_ht&quot;</span><span class="p">:</span> <span class="n">v4_relatedness</span><span class="p">()},</span>
        <span class="p">},</span>
        <span class="n">output_resources</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;hgdp_tgp_v4_exome_duplicate_ht&quot;</span><span class="p">:</span> <span class="n">hgdp_tgp_duplicated_to_exomes</span>
        <span class="p">},</span>
    <span class="p">)</span>
    <span class="n">update_annotations</span> <span class="o">=</span> <span class="n">PipelineStepResourceCollection</span><span class="p">(</span>
        <span class="s2">&quot;--update-annotations&quot;</span><span class="p">,</span>
        <span class="n">input_resources</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;Released HGDP + 1KG sample metadata&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;meta_ht&quot;</span><span class="p">:</span> <span class="n">hgdp_tgp_res</span><span class="p">[</span><span class="s2">&quot;meta_ht&quot;</span><span class="p">]}</span>
        <span class="p">},</span>
        <span class="n">output_resources</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;updated_meta_ht&quot;</span><span class="p">:</span> <span class="n">hgdp_tgp_meta_updated</span><span class="p">},</span>
    <span class="p">)</span>
    <span class="n">get_callstats_for_updated_samples</span> <span class="o">=</span> <span class="n">PipelineStepResourceCollection</span><span class="p">(</span>
        <span class="s2">&quot;--get-callstats-for-updated-samples&quot;</span><span class="p">,</span>
        <span class="n">pipeline_input_steps</span><span class="o">=</span><span class="p">[</span><span class="n">update_annotations</span><span class="p">],</span>
        <span class="n">add_input_resources</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;gnomAD v3.1.2 HGDP + 1KG dense MT&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;dense_mt&quot;</span><span class="p">:</span> <span class="n">hgdp_tgp_res</span><span class="p">[</span><span class="s2">&quot;dense_mt&quot;</span><span class="p">]}</span>
        <span class="p">},</span>
        <span class="n">output_resources</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;freq_added_ht&quot;</span><span class="p">:</span> <span class="n">hgdp_tgp_updated_callstats</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="s2">&quot;added&quot;</span><span class="p">,</span> <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">),</span>
            <span class="s2">&quot;freq_subtracted_ht&quot;</span><span class="p">:</span> <span class="n">hgdp_tgp_updated_callstats</span><span class="p">(</span>
                <span class="n">subset</span><span class="o">=</span><span class="s2">&quot;subtracted&quot;</span><span class="p">,</span> <span class="n">test</span><span class="o">=</span><span class="n">test</span>
            <span class="p">),</span>
            <span class="s2">&quot;freq_pop_diff_ht&quot;</span><span class="p">:</span> <span class="n">hgdp_tgp_updated_callstats</span><span class="p">(</span>
                <span class="n">subset</span><span class="o">=</span><span class="s2">&quot;pop_diff&quot;</span><span class="p">,</span> <span class="n">test</span><span class="o">=</span><span class="n">test</span>
            <span class="p">),</span>
        <span class="p">},</span>
    <span class="p">)</span>
    <span class="n">join_callstats_for_update</span> <span class="o">=</span> <span class="n">PipelineStepResourceCollection</span><span class="p">(</span>
        <span class="s2">&quot;--join-callstats-for-update&quot;</span><span class="p">,</span>
        <span class="n">pipeline_input_steps</span><span class="o">=</span><span class="p">[</span><span class="n">get_callstats_for_updated_samples</span><span class="p">],</span>
        <span class="n">add_input_resources</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;gnomAD v3.1.2 release sites HT&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;sites_ht&quot;</span><span class="p">:</span> <span class="n">v3_sites_ht</span><span class="p">}</span>
        <span class="p">},</span>
        <span class="n">output_resources</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;freq_join_ht&quot;</span><span class="p">:</span> <span class="n">hgdp_tgp_updated_callstats</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="s2">&quot;join&quot;</span><span class="p">,</span> <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">),</span>
        <span class="p">},</span>
    <span class="p">)</span>
    <span class="n">compute_an_for_new_variants</span> <span class="o">=</span> <span class="n">PipelineStepResourceCollection</span><span class="p">(</span>
        <span class="s2">&quot;--compute-allele-number-for-new-variants&quot;</span><span class="p">,</span>
        <span class="n">pipeline_input_steps</span><span class="o">=</span><span class="p">[</span><span class="n">update_annotations</span><span class="p">,</span> <span class="n">join_callstats_for_update</span><span class="p">],</span>
        <span class="n">output_resources</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;v3_release_an_ht&quot;</span><span class="p">:</span> <span class="n">hgdp_tgp_updated_callstats</span><span class="p">(</span>
                <span class="n">subset</span><span class="o">=</span><span class="s2">&quot;v3_release_an&quot;</span><span class="p">,</span> <span class="n">test</span><span class="o">=</span><span class="n">test</span>
            <span class="p">),</span>
        <span class="p">},</span>
    <span class="p">)</span>
    <span class="n">compute_an_for_pop_diff</span> <span class="o">=</span> <span class="n">PipelineStepResourceCollection</span><span class="p">(</span>
        <span class="s2">&quot;--compute-allele-number-for-pop-diff&quot;</span><span class="p">,</span>
        <span class="n">pipeline_input_steps</span><span class="o">=</span><span class="p">[</span><span class="n">update_annotations</span><span class="p">,</span> <span class="n">join_callstats_for_update</span><span class="p">],</span>
        <span class="n">output_resources</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;v3_pop_diff_an_ht&quot;</span><span class="p">:</span> <span class="n">hgdp_tgp_updated_callstats</span><span class="p">(</span>
                <span class="n">subset</span><span class="o">=</span><span class="s2">&quot;v3_pop_diff_an&quot;</span><span class="p">,</span> <span class="n">test</span><span class="o">=</span><span class="n">test</span>
            <span class="p">)</span>
        <span class="p">},</span>
    <span class="p">)</span>
    <span class="n">update_release_callstats</span> <span class="o">=</span> <span class="n">PipelineStepResourceCollection</span><span class="p">(</span>
        <span class="s2">&quot;--update-release-callstats&quot;</span><span class="p">,</span>
        <span class="n">pipeline_input_steps</span><span class="o">=</span><span class="p">[</span>
            <span class="n">update_annotations</span><span class="p">,</span>
            <span class="n">join_callstats_for_update</span><span class="p">,</span>
            <span class="n">compute_an_for_new_variants</span><span class="p">,</span>
            <span class="n">compute_an_for_pop_diff</span><span class="p">,</span>
        <span class="p">],</span>
        <span class="n">add_input_resources</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;v3.1 metadata&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;v3_meta_ht&quot;</span><span class="p">:</span> <span class="n">v3_meta</span><span class="p">},</span>
        <span class="p">},</span>
        <span class="n">output_resources</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;pre_v4_freq_ht&quot;</span><span class="p">:</span> <span class="n">hgdp_tgp_updated_callstats</span><span class="p">(</span>
                <span class="n">subset</span><span class="o">=</span><span class="s2">&quot;pre_validity_check&quot;</span><span class="p">,</span> <span class="n">test</span><span class="o">=</span><span class="n">test</span>
            <span class="p">)</span>
        <span class="p">},</span>
    <span class="p">)</span>
    <span class="n">apply_patch_to_freq_ht</span> <span class="o">=</span> <span class="n">PipelineStepResourceCollection</span><span class="p">(</span>
        <span class="s2">&quot;--apply-patch-to-freq-ht&quot;</span><span class="p">,</span>
        <span class="n">pipeline_input_steps</span><span class="o">=</span><span class="p">[</span><span class="n">update_annotations</span><span class="p">,</span> <span class="n">update_release_callstats</span><span class="p">],</span>
        <span class="n">add_input_resources</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;v3.1 metadata&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;v3_meta_ht&quot;</span><span class="p">:</span> <span class="n">v3_meta</span><span class="p">},</span>
        <span class="p">},</span>
        <span class="n">output_resources</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;v4_freq_ht&quot;</span><span class="p">:</span> <span class="n">v4_get_freq</span><span class="p">(</span><span class="n">data_type</span><span class="o">=</span><span class="s2">&quot;genomes&quot;</span><span class="p">,</span> <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">),</span>
        <span class="p">},</span>
    <span class="p">)</span>

    <span class="c1"># Add all steps to the gnomAD v4.0 genomes release pipeline resource collection.</span>
    <span class="n">v4_genome_release_pipeline</span><span class="o">.</span><span class="n">add_steps</span><span class="p">(</span>
        <span class="p">{</span>
            <span class="s2">&quot;get_related_to_nonsubset&quot;</span><span class="p">:</span> <span class="n">get_related_to_nonsubset</span><span class="p">,</span>
            <span class="s2">&quot;get_hgdp_tgp_v4_exome_duplicates&quot;</span><span class="p">:</span> <span class="n">hgdp_tgp_v4_exome_duplicates</span><span class="p">,</span>
            <span class="s2">&quot;update_annotations&quot;</span><span class="p">:</span> <span class="n">update_annotations</span><span class="p">,</span>
            <span class="s2">&quot;get_callstats_for_updated_samples&quot;</span><span class="p">:</span> <span class="n">get_callstats_for_updated_samples</span><span class="p">,</span>
            <span class="s2">&quot;join_callstats_for_update&quot;</span><span class="p">:</span> <span class="n">join_callstats_for_update</span><span class="p">,</span>
            <span class="s2">&quot;compute_an_for_new_variants&quot;</span><span class="p">:</span> <span class="n">compute_an_for_new_variants</span><span class="p">,</span>
            <span class="s2">&quot;compute_an_for_pop_diff&quot;</span><span class="p">:</span> <span class="n">compute_an_for_pop_diff</span><span class="p">,</span>
            <span class="s2">&quot;update_release_callstats&quot;</span><span class="p">:</span> <span class="n">update_release_callstats</span><span class="p">,</span>
            <span class="s2">&quot;apply_patch_to_freq_ht&quot;</span><span class="p">:</span> <span class="n">apply_patch_to_freq_ht</span><span class="p">,</span>
        <span class="p">}</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">v4_genome_release_pipeline</span></div>


<div class="viewcode-block" id="main"><a class="viewcode-back" href="../../../../api_reference/v4/annotations/generate_freq_genomes.html#gnomad_qc.v4.annotations.generate_freq_genomes.main">[docs]</a><span class="k">def</span> <span class="nf">main</span><span class="p">(</span><span class="n">args</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create the v4.0 genomes release Table.</span>

<span class="sd">    Update call stats to include samples from the HGDP + 1KG subset that were</span>
<span class="sd">    unintentionally excluded whole populations within the HGDP + 1KG subset that were</span>
<span class="sd">    the most genetically unique and had small sample sizes (more specifically: San,</span>
<span class="sd">    Mbuti Pygmy, Biaka Pygmy, Bougainville, and Papuan) compared to other populations</span>
<span class="sd">    within the gnomAD v3.1 callset.</span>

<span class="sd">    In order to avoid re-calculating the call stats for the whole subset / whole</span>
<span class="sd">    release, we calculate the call stats for the samples that will be added</span>
<span class="sd">    and subtracted, then merge the call stats with the old call stats in the</span>
<span class="sd">    release HT.</span>

<span class="sd">    Changes compared to the v3 release:</span>
<span class="sd">      - Some small updates to samples that are hard filtered.</span>
<span class="sd">      - Use a population PC outlier approach to filter the HGDP + 1KG samples instead</span>
<span class="sd">        of the sample QC metric outlier filtering approach used on the full dataset</span>
<span class="sd">        that caused some samples to be unintentionally excluded.</span>
<span class="sd">      - HGDP + 1KG release samples are determined using relatedness (pc_relate) run on</span>
<span class="sd">        only samples within the subset as well as relatedness to the rest of the</span>
<span class="sd">        release.</span>
<span class="sd">      - Some population labels were updated.</span>
<span class="sd">      - The `Han` and `Papuan` populations were split into more specific groupings.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">test</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">test_pcsk9</span> <span class="ow">or</span> <span class="n">args</span><span class="o">.</span><span class="n">test_x_gene</span>
    <span class="n">gene_on_chrx</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">test_x_gene</span>
    <span class="n">overwrite</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">overwrite</span>

    <span class="n">hl</span><span class="o">.</span><span class="n">init</span><span class="p">(</span>
        <span class="n">log</span><span class="o">=</span><span class="s2">&quot;/create_v4.0_genomes_freq.log&quot;</span><span class="p">,</span>
        <span class="n">default_reference</span><span class="o">=</span><span class="s2">&quot;GRCh38&quot;</span><span class="p">,</span>
        <span class="n">tmp_dir</span><span class="o">=</span><span class="s2">&quot;gs://gnomad-tmp-30day&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">v4_genome_release_resources</span> <span class="o">=</span> <span class="n">get_v4_genomes_release_resources</span><span class="p">(</span>
        <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span>
    <span class="p">)</span>
    <span class="n">v3_hgdp_tgp_meta_ht</span> <span class="o">=</span> <span class="n">v4_genome_release_resources</span><span class="o">.</span><span class="n">meta_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>
    <span class="n">v3_hgdp_tgp_dense_mt</span> <span class="o">=</span> <span class="n">v4_genome_release_resources</span><span class="o">.</span><span class="n">dense_mt</span><span class="o">.</span><span class="n">mt</span><span class="p">()</span>
    <span class="n">v3_sites_ht</span> <span class="o">=</span> <span class="n">v4_genome_release_resources</span><span class="o">.</span><span class="n">v3_sites_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>

    <span class="n">v3_vds</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="p">(</span>
        <span class="n">args</span><span class="o">.</span><span class="n">compute_allele_number_for_new_variants</span>
        <span class="ow">or</span> <span class="n">args</span><span class="o">.</span><span class="n">compute_allele_number_for_pop_diff</span>
    <span class="p">):</span>
        <span class="n">v3_vds</span> <span class="o">=</span> <span class="n">get_gnomad_v3_vds</span><span class="p">(</span><span class="n">split</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">samples_meta</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">test</span><span class="p">:</span>
        <span class="n">v3_hgdp_tgp_dense_mt</span> <span class="o">=</span> <span class="n">filter_to_test</span><span class="p">(</span>
            <span class="n">v3_hgdp_tgp_dense_mt</span><span class="p">,</span> <span class="n">gene_on_chrx</span><span class="o">=</span><span class="n">gene_on_chrx</span>
        <span class="p">)</span>
        <span class="n">v3_sites_ht</span> <span class="o">=</span> <span class="n">filter_to_test</span><span class="p">(</span><span class="n">v3_sites_ht</span><span class="p">,</span> <span class="n">gene_on_chrx</span><span class="o">=</span><span class="n">gene_on_chrx</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">v3_vds</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">v3_vds</span> <span class="o">=</span> <span class="n">filter_to_test</span><span class="p">(</span><span class="n">v3_vds</span><span class="p">,</span> <span class="n">gene_on_chrx</span><span class="o">=</span><span class="n">gene_on_chrx</span><span class="p">)</span>

    <span class="c1"># NOTE: there was a bug in the v3.1 frequency code where we used the v3.0</span>
    <span class="c1">#  freq_ht for the hotfix of depletion of homozygous frequency calculation,</span>
    <span class="c1">#  once a variant is new in v3.1 (i.e. not present in v3.0 freq_ht) and is from</span>
    <span class="c1">#  samples with heterozygous non-reference genotypes with high allele balances,</span>
    <span class="c1">#  their GT was set to missing in the v3.1 freq_ht. Here we are using a GT that was</span>
    <span class="c1">#  created in the same manner so that we can correctly merge with the existing</span>
    <span class="c1">#  v3.1.2 release sites HT.</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Annotating GT with sex ploidy adjusted genotypes and adj calculated on &quot;</span>
        <span class="s2">&quot;genotypes without sex ploidy adjustment...&quot;</span>
    <span class="p">)</span>
    <span class="n">v3_hgdp_tgp_dense_mt</span> <span class="o">=</span> <span class="n">v3_hgdp_tgp_dense_mt</span><span class="o">.</span><span class="n">annotate_entries</span><span class="p">(</span>
        <span class="n">GT</span><span class="o">=</span><span class="n">v3_hgdp_tgp_dense_mt</span><span class="o">.</span><span class="n">hom_alt_fix_investigation</span><span class="o">.</span><span class="n">sex_ploidy_adjusted_GT</span><span class="p">,</span>
        <span class="n">adj</span><span class="o">=</span><span class="n">v3_hgdp_tgp_dense_mt</span><span class="o">.</span><span class="n">hom_alt_fix_investigation</span><span class="o">.</span><span class="n">unadjusted_adj</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># Temporary hotfix for depletion of homozygous alternate genotypes</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Setting het genotypes at sites with &gt;1% AF (using v3.0 frequencies) and &gt;&quot;</span>
        <span class="s2">&quot; 0.9 AB to homalt...&quot;</span>
    <span class="p">)</span>
    <span class="c1"># Load v3.0 allele frequencies to avoid an extra frequency calculation</span>
    <span class="c1"># NOTE: Using previous callset AF works for small incremental changes to a callset, but we will need to revisit for large increments # noqa</span>
    <span class="n">freq_ht</span> <span class="o">=</span> <span class="n">release_sites</span><span class="p">(</span><span class="n">public</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">versions</span><span class="p">[</span><span class="s2">&quot;3.0&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;freq&quot;</span><span class="p">)</span>
    <span class="n">v3_hgdp_tgp_dense_mt</span> <span class="o">=</span> <span class="n">v3_hgdp_tgp_dense_mt</span><span class="o">.</span><span class="n">annotate_entries</span><span class="p">(</span>
        <span class="n">GT</span><span class="o">=</span><span class="n">hom_alt_depletion_fix</span><span class="p">(</span>
            <span class="n">v3_hgdp_tgp_dense_mt</span><span class="o">.</span><span class="n">GT</span><span class="p">,</span>
            <span class="n">het_non_ref_expr</span><span class="o">=</span><span class="n">v3_hgdp_tgp_dense_mt</span><span class="o">.</span><span class="n">_het_non_ref</span><span class="p">,</span>
            <span class="n">af_expr</span><span class="o">=</span><span class="n">freq_ht</span><span class="p">[</span><span class="n">v3_hgdp_tgp_dense_mt</span><span class="o">.</span><span class="n">row_key</span><span class="p">]</span><span class="o">.</span><span class="n">freq</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">AF</span><span class="p">,</span>
            <span class="n">ab_expr</span><span class="o">=</span><span class="n">v3_hgdp_tgp_dense_mt</span><span class="o">.</span><span class="n">AD</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">/</span> <span class="n">v3_hgdp_tgp_dense_mt</span><span class="o">.</span><span class="n">DP</span><span class="p">,</span>
            <span class="n">use_v3_1_correction</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="p">),</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">get_related_to_nonsubset</span><span class="p">:</span>
        <span class="n">res</span> <span class="o">=</span> <span class="n">v4_genome_release_resources</span><span class="o">.</span><span class="n">get_related_to_nonsubset</span>
        <span class="n">res</span><span class="o">.</span><span class="n">check_resource_existence</span><span class="p">()</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">get_hgdp_tgp_related_to_nonsubset</span><span class="p">(</span>
            <span class="n">res</span><span class="o">.</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">(),</span> <span class="n">res</span><span class="o">.</span><span class="n">v3_relatedness_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>
        <span class="p">)</span>
        <span class="n">ht</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">related_to_nonsubset_ht</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">get_hgdp_tgp_v4_exome_duplicates</span><span class="p">:</span>
        <span class="n">res</span> <span class="o">=</span> <span class="n">v4_genome_release_resources</span><span class="o">.</span><span class="n">get_hgdp_tgp_v4_exome_duplicates</span>
        <span class="n">res</span><span class="o">.</span><span class="n">check_resource_existence</span><span class="p">()</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">get_hgdp_tgp_v4_exome_duplicates</span><span class="p">(</span>
            <span class="n">res</span><span class="o">.</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">(),</span> <span class="n">res</span><span class="o">.</span><span class="n">v4_meta_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">(),</span> <span class="n">res</span><span class="o">.</span><span class="n">v4_relatedness_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>
        <span class="p">)</span>
        <span class="n">ht</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">hgdp_tgp_v4_exome_duplicate_ht</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">update_annotations</span><span class="p">:</span>
        <span class="n">res</span> <span class="o">=</span> <span class="n">v4_genome_release_resources</span><span class="o">.</span><span class="n">update_annotations</span>
        <span class="n">res</span><span class="o">.</span><span class="n">check_resource_existence</span><span class="p">()</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Adding updated sample QC annotations to meta HT...&quot;</span><span class="p">)</span>
        <span class="n">add_updated_sample_qc_annotations</span><span class="p">(</span><span class="n">v3_hgdp_tgp_meta_ht</span><span class="p">)</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
            <span class="n">res</span><span class="o">.</span><span class="n">updated_meta_ht</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">get_callstats_for_updated_samples</span><span class="p">:</span>
        <span class="n">res</span> <span class="o">=</span> <span class="n">v4_genome_release_resources</span><span class="o">.</span><span class="n">get_callstats_for_updated_samples</span>
        <span class="n">res</span><span class="o">.</span><span class="n">check_resource_existence</span><span class="p">()</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="s2">&quot;Calculating and concatenating call stats for samples to be added and &quot;</span>
            <span class="s2">&quot;samples to be subtracted from the v3.1.2 release sites HT...&quot;</span>
        <span class="p">)</span>
        <span class="n">mt</span> <span class="o">=</span> <span class="n">v3_hgdp_tgp_dense_mt</span><span class="o">.</span><span class="n">select_globals</span><span class="p">()</span>
        <span class="n">filtered_hts</span> <span class="o">=</span> <span class="n">get_updated_release_samples</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">updated_meta_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">())</span>
        <span class="k">for</span> <span class="n">ht</span><span class="p">,</span> <span class="n">name</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">filtered_hts</span><span class="p">,</span> <span class="n">JOIN_FREQS</span><span class="p">[</span><span class="mi">1</span><span class="p">:]):</span>
            <span class="n">freq_ht</span> <span class="o">=</span> <span class="n">get_hgdp_tgp_callstats_for_selected_samples</span><span class="p">(</span>
                <span class="n">mt</span><span class="p">,</span> <span class="n">ht</span><span class="p">,</span> <span class="n">compute_freq_all</span><span class="o">=</span><span class="kc">False</span> <span class="k">if</span> <span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;pop_diff&quot;</span> <span class="k">else</span> <span class="kc">True</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">name</span> <span class="o">==</span> <span class="s2">&quot;pop_diff&quot;</span><span class="p">:</span>
                <span class="n">freq_ht</span> <span class="o">=</span> <span class="n">filter_freq_arrays</span><span class="p">(</span><span class="n">freq_ht</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;pop&quot;</span><span class="p">])</span>

            <span class="n">freq_ht</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="nb">getattr</span><span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;freq_</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">_ht&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">join_callstats_for_update</span><span class="p">:</span>
        <span class="n">res</span> <span class="o">=</span> <span class="n">v4_genome_release_resources</span><span class="o">.</span><span class="n">join_callstats_for_update</span>
        <span class="n">res</span><span class="o">.</span><span class="n">check_resource_existence</span><span class="p">()</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Joining all call stats HTs with the release sites HT...&quot;</span><span class="p">)</span>
        <span class="n">join_release_ht_with_subsets</span><span class="p">(</span>
            <span class="n">v3_sites_ht</span><span class="p">,</span>
            <span class="p">{</span><span class="n">n</span><span class="p">:</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;freq_</span><span class="si">{</span><span class="n">n</span><span class="si">}</span><span class="s2">_ht&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">JOIN_FREQS</span><span class="p">[</span><span class="mi">1</span><span class="p">:]},</span>
        <span class="p">)</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">freq_join_ht</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">compute_allele_number_for_new_variants</span><span class="p">:</span>
        <span class="n">res</span> <span class="o">=</span> <span class="n">v4_genome_release_resources</span><span class="o">.</span><span class="n">compute_an_for_new_variants</span>
        <span class="n">res</span><span class="o">.</span><span class="n">check_resource_existence</span><span class="p">()</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Determine variants in the release HT that need a recomputed AN...&quot;</span><span class="p">)</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">res</span><span class="o">.</span><span class="n">freq_join_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>
        <span class="c1"># First freq in ann_array is the release sites freq and the third is the added.</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">is_missing</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">ann_array</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="o">&amp;</span> <span class="n">hl</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">ann_array</span><span class="o">.</span><span class="n">freq</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">AC</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">))</span>
        <span class="p">)</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span><span class="n">new_temp_file</span><span class="p">(</span><span class="s2">&quot;variants_for_an&quot;</span><span class="p">,</span> <span class="s2">&quot;ht&quot;</span><span class="p">))</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="s2">&quot;There are </span><span class="si">%i</span><span class="s2"> variants with an AC &gt; 0 in the added call stats HTs, but &quot;</span>
            <span class="s2">&quot;missing in the release HT...&quot;</span><span class="p">,</span>
            <span class="n">ht</span><span class="o">.</span><span class="n">count</span><span class="p">(),</span>
        <span class="p">)</span>

        <span class="n">v3_sites_meta_ht</span> <span class="o">=</span> <span class="n">v3_vds</span><span class="o">.</span><span class="n">variant_data</span><span class="o">.</span><span class="n">cols</span><span class="p">()</span>
        <span class="n">v3_sites_meta_ht</span> <span class="o">=</span> <span class="n">v3_sites_meta_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">v3_sites_meta_ht</span><span class="o">.</span><span class="n">meta</span><span class="o">.</span><span class="n">release</span><span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="s2">&quot;Annotating call stats group membership for each v3.1 release sample...&quot;</span>
        <span class="p">)</span>
        <span class="n">group_membership_ht</span> <span class="o">=</span> <span class="n">get_group_membership_ht_for_an</span><span class="p">(</span><span class="n">v3_sites_meta_ht</span><span class="p">)</span>
        <span class="n">group_membership_ht</span> <span class="o">=</span> <span class="n">group_membership_ht</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span>
            <span class="n">new_temp_file</span><span class="p">(</span><span class="s2">&quot;group_membership_all&quot;</span><span class="p">,</span> <span class="s2">&quot;ht&quot;</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="s2">&quot;Computing the AN HT of v3.1 samples for new v4.0 genomes variants...&quot;</span>
        <span class="p">)</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">compute_an_by_group_membership</span><span class="p">(</span><span class="n">v3_vds</span><span class="p">,</span> <span class="n">group_membership_ht</span><span class="p">,</span> <span class="n">ht</span><span class="p">)</span>
        <span class="n">ht</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">v3_release_an_ht</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">compute_allele_number_for_pop_diff</span><span class="p">:</span>
        <span class="n">res</span> <span class="o">=</span> <span class="n">v4_genome_release_resources</span><span class="o">.</span><span class="n">compute_an_for_pop_diff</span>
        <span class="n">res</span><span class="o">.</span><span class="n">check_resource_existence</span><span class="p">()</span>
        <span class="n">pop_diff_group_membership_ht</span> <span class="o">=</span> <span class="n">get_pop_diff_v3_vds_group_membership</span><span class="p">(</span>
            <span class="n">v3_vds</span><span class="p">,</span> <span class="n">res</span><span class="o">.</span><span class="n">updated_meta_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>
        <span class="p">)</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span><span class="n">new_temp_file</span><span class="p">(</span><span class="s2">&quot;group_membership_pop_diff&quot;</span><span class="p">,</span> <span class="s2">&quot;ht&quot;</span><span class="p">))</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="s2">&quot;Computing the AN HT of v3.1 pop diff samples for all v4.0 genomes&quot;</span>
            <span class="s2">&quot; variants. Freq meta: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">pop_diff_group_membership_ht</span><span class="o">.</span><span class="n">freq_meta</span><span class="p">),</span>
        <span class="p">)</span>
        <span class="c1"># To avoid double counting, we need to filter out the variants that are in</span>
        <span class="c1"># v4.0 genomes but not already present in pop_diff samples to get the AN.</span>
        <span class="n">join_ht</span> <span class="o">=</span> <span class="n">res</span><span class="o">.</span><span class="n">freq_join_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>
        <span class="n">join_ht</span> <span class="o">=</span> <span class="n">join_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_missing</span><span class="p">(</span><span class="n">join_ht</span><span class="o">.</span><span class="n">ann_array</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">freq</span><span class="p">))</span><span class="o">.</span><span class="n">select</span><span class="p">()</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">compute_an_by_group_membership</span><span class="p">(</span>
            <span class="n">v3_vds</span><span class="p">,</span> <span class="n">pop_diff_group_membership_ht</span><span class="p">,</span> <span class="n">join_ht</span>
        <span class="p">)</span>
        <span class="n">ht</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">v3_pop_diff_an_ht</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">update_release_callstats</span><span class="p">:</span>
        <span class="n">res</span> <span class="o">=</span> <span class="n">v4_genome_release_resources</span><span class="o">.</span><span class="n">update_release_callstats</span>
        <span class="n">res</span><span class="o">.</span><span class="n">check_resource_existence</span><span class="p">()</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Merging all call stats HTs for final v4.0 genomes call stats...&quot;</span><span class="p">)</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">generate_v4_genomes_callstats</span><span class="p">(</span>
            <span class="n">res</span><span class="o">.</span><span class="n">freq_join_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">(),</span> <span class="n">res</span><span class="o">.</span><span class="n">v3_release_an_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">(),</span> <span class="n">res</span><span class="o">.</span><span class="n">v3_pop_diff_an_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>
        <span class="p">)</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">finalize_v4_genomes_callstats</span><span class="p">(</span>
            <span class="n">ht</span><span class="p">,</span> <span class="n">v3_sites_ht</span><span class="p">,</span> <span class="n">res</span><span class="o">.</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">(),</span> <span class="n">res</span><span class="o">.</span><span class="n">updated_meta_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>
        <span class="p">)</span>
        <span class="n">ht</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">pre_v4_freq_ht</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">apply_patch_to_freq_ht</span><span class="p">:</span>
        <span class="n">res</span> <span class="o">=</span> <span class="n">v4_genome_release_resources</span><span class="o">.</span><span class="n">apply_patch_to_freq_ht</span>
        <span class="n">res</span><span class="o">.</span><span class="n">check_resource_existence</span><span class="p">()</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">patch_v4_genomes_callstats</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">pre_v4_freq_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">())</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">finalize_v4_genomes_callstats</span><span class="p">(</span>
            <span class="n">ht</span><span class="p">,</span> <span class="n">v3_sites_ht</span><span class="p">,</span> <span class="n">res</span><span class="o">.</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">(),</span> <span class="n">res</span><span class="o">.</span><span class="n">updated_meta_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>
        <span class="p">)</span>
        <span class="n">ht</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">v4_freq_ht</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">)</span></div>


<span class="k">def</span> <span class="nf">get_script_argument_parser</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get script argument parser.&quot;&quot;&quot;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span>
        <span class="n">description</span><span class="o">=</span><span class="s2">&quot;This script creates the v4.0 genomes release HT.&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--overwrite&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Overwrite data&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--slack-channel&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Slack channel to post results and notifications to.&quot;</span>
    <span class="p">)</span>

    <span class="n">test_args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_mutually_exclusive_group</span><span class="p">()</span>
    <span class="n">test_args</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--test-pcsk9&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Test on a subset of variants in PCSK9 gene.&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">test_args</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--test-x-gene&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Test on a subset of variants in TRPC5 on chrX.&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--get-related-to-nonsubset&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Get the relatedness to nonsubset samples.&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--get-hgdp-tgp-v4-exome-duplicates&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Get the duplicated samples to exomes.&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--update-annotations&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Update HGDP + 1KG sample QC annotations.&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--get-callstats-for-updated-samples&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span>
            <span class="s2">&quot;Get the call stats for the genomes with an updated release status for v4.0&quot;</span>
            <span class="s2">&quot; compared to v3.1.&quot;</span>
        <span class="p">),</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--join-callstats-for-update&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span>
            <span class="s2">&quot;Join the call stats Tables for the updated samples with the release &quot;</span>
            <span class="s2">&quot;call stats Table.&quot;</span>
        <span class="p">),</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--compute-allele-number-for-new-variants&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span>
            <span class="s2">&quot;Compute the allele number for variants that are in the updated samples &quot;</span>
            <span class="s2">&quot;but not in the release HT.&quot;</span>
        <span class="p">),</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--compute-allele-number-for-pop-diff&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span>
            <span class="s2">&quot;Compute the allele number of all v4.0 variants on the samples that have &quot;</span>
            <span class="s2">&quot;different pops in v4.0.&quot;</span>
        <span class="p">),</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--update-release-callstats&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span>
            <span class="s2">&quot;Update the release call stats by merging the call stats for the updated&quot;</span>
            <span class="s2">&quot; samples.&quot;</span>
        <span class="p">),</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--apply-patch-to-freq-ht&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Apply a patch to the final freq HT.&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">parser</span>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">get_script_argument_parser</span><span class="p">()</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">slack_channel</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">slack_notifications</span><span class="p">(</span><span class="n">slack_token</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">slack_channel</span><span class="p">):</span>
            <span class="n">main</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">main</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p></p>
  </div>

   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>