<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>gnomad_qc.v4.annotations.compute_coverage &mdash; gnomad_qc master documentation</title>
      <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/theme_overrides.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/doctools.js"></script>
        <script src="../../../_static/sphinx_highlight.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="gnomad_qc.v4.annotations.fix_freq_an" href="fix_freq_an.html" />
    <link rel="prev" title="gnomad_qc.v4.annotations" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            gnomad_qc
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../../index.html">API Reference</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../index.html">v4</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="index.html">annotations</a><ul class="current">
<li class="toctree-l4 current"><a class="current reference internal" href="#">compute_coverage</a></li>
<li class="toctree-l4"><a class="reference internal" href="fix_freq_an.html">fix_freq_an</a></li>
<li class="toctree-l4"><a class="reference internal" href="generate_freq.html">generate_freq</a></li>
<li class="toctree-l4"><a class="reference internal" href="generate_freq_genomes.html">generate_freq_genomes</a></li>
<li class="toctree-l4"><a class="reference internal" href="generate_variant_qc_annotations.html">generate_variant_qc_annotations</a></li>
<li class="toctree-l4"><a class="reference internal" href="insilico_predictors.html">insilico_predictors</a></li>
<li class="toctree-l4"><a class="reference internal" href="recover_and_complete_vep115.html">recover_and_complete_vep115</a></li>
<li class="toctree-l4"><a class="reference internal" href="vep_context_ht.html">vep_context_ht</a></li>
<li class="toctree-l4"><a class="reference internal" href="vrs_annotation_batch.html">vrs_annotation_batch</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../assessment/index.html">assessment</a></li>
<li class="toctree-l3"><a class="reference internal" href="../create_release/index.html">create_release</a></li>
<li class="toctree-l3"><a class="reference internal" href="../resources/index.html">resources</a></li>
<li class="toctree-l3"><a class="reference internal" href="../sample_qc/index.html">sample_qc</a></li>
<li class="toctree-l3"><a class="reference internal" href="../subset.html">subset</a></li>
<li class="toctree-l3"><a class="reference internal" href="../variant_qc/index.html">variant_qc</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../v5/index.html">v5</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../analyses/index.html">analyses</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/broadinstitute/gnomad_qc/tags">Change Log</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">gnomad_qc</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">gnomad_qc</a></li>
          <li class="breadcrumb-item"><a href="../index.html">gnomad_qc.v4</a></li>
          <li class="breadcrumb-item"><a href="index.html">gnomad_qc.v4.annotations</a></li>
      <li class="breadcrumb-item active">gnomad_qc.v4.annotations.compute_coverage</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../../_sources/api_reference/v4/annotations/compute_coverage.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="gnomad-qc-v4-annotations-compute-coverage">
<h1>gnomad_qc.v4.annotations.compute_coverage<a class="headerlink" href="#gnomad-qc-v4-annotations-compute-coverage" title="Permalink to this heading"></a></h1>
<p>Script to compute coverage statistics on gnomAD v4 exomes.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">gnomad_qc</span><span class="o">.</span><span class="n">v4</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">compute_coverage</span><span class="o">.</span><span class="n">py</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">overwrite</span><span class="p">]</span>
                                                    <span class="p">[</span><span class="o">--</span><span class="n">test</span><span class="o">-</span><span class="mi">2</span><span class="o">-</span><span class="n">partitions</span><span class="p">]</span>
                                                    <span class="p">[</span><span class="o">--</span><span class="n">test</span><span class="o">-</span><span class="n">chr22</span><span class="o">-</span><span class="n">chrx</span><span class="o">-</span><span class="n">chry</span><span class="p">]</span>
                                                    <span class="p">[</span><span class="o">--</span><span class="n">n</span><span class="o">-</span><span class="n">partitions</span> <span class="n">N_PARTITIONS</span><span class="p">]</span>
                                                    <span class="p">[</span><span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="nb">type</span> <span class="p">{</span><span class="n">exomes</span><span class="p">,</span><span class="n">genomes</span><span class="p">}]</span>
                                                    <span class="p">[</span><span class="o">--</span><span class="n">compute</span><span class="o">-</span><span class="n">coverage</span><span class="o">-</span><span class="n">ht</span><span class="p">]</span>
                                                    <span class="p">[</span><span class="o">--</span><span class="n">compute</span><span class="o">-</span><span class="nb">all</span><span class="o">-</span><span class="n">sites</span><span class="o">-</span><span class="n">an</span><span class="o">-</span><span class="ow">and</span><span class="o">-</span><span class="n">qual</span><span class="o">-</span><span class="n">hist</span><span class="o">-</span><span class="n">ht</span><span class="p">]</span>
                                                    <span class="p">[</span><span class="o">--</span><span class="n">stratify</span><span class="o">-</span><span class="n">by</span><span class="o">-</span><span class="n">ukb</span><span class="o">-</span><span class="ow">and</span><span class="o">-</span><span class="n">platform</span><span class="p">]</span>
                                                    <span class="p">[</span><span class="o">--</span><span class="n">calling</span><span class="o">-</span><span class="n">interval</span><span class="o">-</span><span class="n">name</span> <span class="p">{</span><span class="n">ukb</span><span class="p">,</span><span class="n">broad</span><span class="p">,</span><span class="n">intersection</span><span class="p">,</span><span class="n">union</span><span class="p">}]</span>
                                                    <span class="p">[</span><span class="o">--</span><span class="n">calling</span><span class="o">-</span><span class="n">interval</span><span class="o">-</span><span class="n">padding</span> <span class="p">{</span><span class="mi">0</span><span class="p">,</span><span class="mi">50</span><span class="p">,</span><span class="mi">150</span><span class="p">}]</span>
                                                    <span class="p">[</span><span class="o">--</span><span class="n">export</span><span class="o">-</span><span class="n">coverage</span><span class="o">-</span><span class="n">release</span><span class="o">-</span><span class="n">files</span><span class="p">]</span>
                                                    <span class="p">[</span><span class="o">--</span><span class="n">export</span><span class="o">-</span><span class="nb">all</span><span class="o">-</span><span class="n">sites</span><span class="o">-</span><span class="n">an</span><span class="o">-</span><span class="n">release</span><span class="o">-</span><span class="n">files</span><span class="p">]</span>
</pre></div>
</div>
<section id="named-arguments">
<h2>Named Arguments<a class="headerlink" href="#named-arguments" title="Permalink to this heading"></a></h2>
<dl class="option-list">
<dt><kbd>--overwrite</kbd></dt>
<dd><p>Overwrite existing hail Tables.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--test-2-partitions</kbd></dt>
<dd><p>Whether to run a test using only the first 2 partitions of the VDS test dataset.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--test-chr22-chrx-chry</kbd></dt>
<dd><p>Whether to run a test using only the chr22, chrX, and chrY chromosomes of the VDS test dataset.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--n-partitions</kbd></dt>
<dd><p>Number of partitions to use for the output Table.</p>
<p>Default: 5000</p>
</dd>
<dt><kbd>--data-type</kbd></dt>
<dd><p>Possible choices: exomes, genomes</p>
<p>Data type to compute coverage or AN and quality hist all sites Table for. One of ‘exomes’ or ‘genomes’.</p>
<p>Default: “exomes”</p>
</dd>
<dt><kbd>--compute-coverage-ht</kbd></dt>
<dd><p>Compute the coverage HT.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--compute-all-sites-an-and-qual-hist-ht</kbd></dt>
<dd><p>Compute the all sites allele number and quality histogram HT.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--stratify-by-ukb-and-platform</kbd></dt>
<dd><p>Whether to compute coverage stratified by UKB/non-UKB and platform. Only applicable if –data-type is exomes.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--calling-interval-name</kbd></dt>
<dd><p>Possible choices: ukb, broad, intersection, union</p>
<p>Name of calling intervals to use for interval coverage. One of: ‘ukb’, ‘broad’, ‘intersection’, or ‘union’. Only applicable if –data-type is exomes.</p>
<p>Default: “union”</p>
</dd>
<dt><kbd>--calling-interval-padding</kbd></dt>
<dd><p>Possible choices: 0, 50, 150</p>
<p>Number of base pair padding to use on the calling intervals. One of 0,  50, or 150 bp.  Only applicable if –data-type is exomes.</p>
<p>Default: 150</p>
</dd>
<dt><kbd>--export-coverage-release-files</kbd></dt>
<dd><p>Exports coverage release HT and TSV file.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--export-all-sites-an-release-files</kbd></dt>
<dd><p>Export all sites AN release HT and TSV file.</p>
<p>Default: False</p>
</dd>
</dl>
</section>
<section id="module-functions">
<h2>Module Functions<a class="headerlink" href="#module-functions" title="Permalink to this heading"></a></h2>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad_qc.v4.annotations.compute_coverage.get_exomes_group_membership_ht" title="gnomad_qc.v4.annotations.compute_coverage.get_exomes_group_membership_ht"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad_qc.v4.annotations.compute_coverage.get_exomes_group_membership_ht</span></code></a>(...)</p></td>
<td><p>Get exomes group membership HT for all sites allele number stratification.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad_qc.v4.annotations.compute_coverage.get_genomes_group_membership_ht" title="gnomad_qc.v4.annotations.compute_coverage.get_genomes_group_membership_ht"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad_qc.v4.annotations.compute_coverage.get_genomes_group_membership_ht</span></code></a>(meta_ht)</p></td>
<td><p>Get genomes group membership HT for all sites allele number stratification.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad_qc.v4.annotations.compute_coverage.adjust_interval_padding" title="gnomad_qc.v4.annotations.compute_coverage.adjust_interval_padding"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad_qc.v4.annotations.compute_coverage.adjust_interval_padding</span></code></a>(ht, ...)</p></td>
<td><p>Adjust interval padding in HT.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad_qc.v4.annotations.compute_coverage.compute_an_and_qual_hists_per_ref_site" title="gnomad_qc.v4.annotations.compute_coverage.compute_an_and_qual_hists_per_ref_site"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad_qc.v4.annotations.compute_coverage.compute_an_and_qual_hists_per_ref_site</span></code></a>(...)</p></td>
<td><p>Compute allele number and quality histograms per reference site.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad_qc.v4.annotations.compute_coverage.add_exomes_interval_annotations" title="gnomad_qc.v4.annotations.compute_coverage.add_exomes_interval_annotations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad_qc.v4.annotations.compute_coverage.add_exomes_interval_annotations</span></code></a>(ht)</p></td>
<td><p>Add interval annotations to exomes HT.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad_qc.v4.annotations.compute_coverage.get_coverage_resources" title="gnomad_qc.v4.annotations.compute_coverage.get_coverage_resources"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad_qc.v4.annotations.compute_coverage.get_coverage_resources</span></code></a>(...)</p></td>
<td><p>Get PipelineResourceCollection for all resources needed in the coverage pipeline.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad_qc.v4.annotations.compute_coverage.main" title="gnomad_qc.v4.annotations.compute_coverage.main"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad_qc.v4.annotations.compute_coverage.main</span></code></a>(args)</p></td>
<td><p>Compute coverage statistics, including mean, median_approx, and coverage over certain DPs.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad_qc.v4.annotations.compute_coverage.get_script_argument_parser</span></code>()</p></td>
<td><p>Get script argument parser.</p></td>
</tr>
</tbody>
</table>
<span class="target" id="module-gnomad_qc.v4.annotations.compute_coverage"></span><p>Script to compute coverage statistics on gnomAD v4 exomes.</p>
<dl class="py function">
<dt class="sig sig-object py" id="gnomad_qc.v4.annotations.compute_coverage.get_exomes_group_membership_ht">
<span class="sig-prename descclassname"><span class="pre">gnomad_qc.v4.annotations.compute_coverage.</span></span><span class="sig-name descname"><span class="pre">get_exomes_group_membership_ht</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">meta_ht</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ds_ht</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">non_ukb_ds_ht</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad_qc/v4/annotations/compute_coverage.html#get_exomes_group_membership_ht"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad_qc.v4.annotations.compute_coverage.get_exomes_group_membership_ht" title="Permalink to this definition"></a></dt>
<dd><p>Get exomes group membership HT for all sites allele number stratification.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>meta_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Metadata HT.</p></li>
<li><p><strong>ds_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Full frequency downsampling HT.</p></li>
<li><p><strong>non_ukb_ds_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Non-UKB frequency downsampling HT.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Group membership HT.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad_qc.v4.annotations.compute_coverage.get_genomes_group_membership_ht">
<span class="sig-prename descclassname"><span class="pre">gnomad_qc.v4.annotations.compute_coverage.</span></span><span class="sig-name descname"><span class="pre">get_genomes_group_membership_ht</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">meta_ht</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad_qc/v4/annotations/compute_coverage.html#get_genomes_group_membership_ht"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad_qc.v4.annotations.compute_coverage.get_genomes_group_membership_ht" title="Permalink to this definition"></a></dt>
<dd><p>Get genomes group membership HT for all sites allele number stratification.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>meta_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Metadata HT.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Group membership HT.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad_qc.v4.annotations.compute_coverage.adjust_interval_padding">
<span class="sig-prename descclassname"><span class="pre">gnomad_qc.v4.annotations.compute_coverage.</span></span><span class="sig-name descname"><span class="pre">adjust_interval_padding</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ht</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">padding</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad_qc/v4/annotations/compute_coverage.html#adjust_interval_padding"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad_qc.v4.annotations.compute_coverage.adjust_interval_padding" title="Permalink to this definition"></a></dt>
<dd><p>Adjust interval padding in HT.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This function can lead to overlapping intervals, so it is not recommended for
most applications. For example, it can be used to filter a variant list to all
variants within the returned interval list, but would not work for getting an
aggregate statistic for each interval if the desired output is independent
statistics.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – HT to adjust.</p></li>
<li><p><strong>padding</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Padding to use.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>HT with adjusted interval padding.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad_qc.v4.annotations.compute_coverage.compute_an_and_qual_hists_per_ref_site">
<span class="sig-prename descclassname"><span class="pre">gnomad_qc.v4.annotations.compute_coverage.</span></span><span class="sig-name descname"><span class="pre">compute_an_and_qual_hists_per_ref_site</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_ht</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interval_ht</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_membership_ht</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad_qc/v4/annotations/compute_coverage.html#compute_an_and_qual_hists_per_ref_site"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad_qc.v4.annotations.compute_coverage.compute_an_and_qual_hists_per_ref_site" title="Permalink to this definition"></a></dt>
<dd><p>Compute allele number and quality histograms per reference site.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vds</strong> (<a class="reference external" href="https://hail.is/docs/0.2/vds/hail.vds.VariantDataset.html#hail.vds.VariantDataset" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantDataset</span></code></a>) – Input VDS.</p></li>
<li><p><strong>ref_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Reference HT.</p></li>
<li><p><strong>interval_ht</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>]) – Interval HT.</p></li>
<li><p><strong>group_membership_ht</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>]) – Group membership HT.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>HT with allele number and quality histograms per reference site.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad_qc.v4.annotations.compute_coverage.add_exomes_interval_annotations">
<span class="sig-prename descclassname"><span class="pre">gnomad_qc.v4.annotations.compute_coverage.</span></span><span class="sig-name descname"><span class="pre">add_exomes_interval_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ht</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">padding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">150</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad_qc/v4/annotations/compute_coverage.html#add_exomes_interval_annotations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad_qc.v4.annotations.compute_coverage.add_exomes_interval_annotations" title="Permalink to this definition"></a></dt>
<dd><p>Add interval annotations to exomes HT.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Input HT.</p></li>
<li><p><strong>padding</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Number of base pair padding to use for the calling intervals.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>HT with exome interval annotations.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad_qc.v4.annotations.compute_coverage.get_coverage_resources">
<span class="sig-prename descclassname"><span class="pre">gnomad_qc.v4.annotations.compute_coverage.</span></span><span class="sig-name descname"><span class="pre">get_coverage_resources</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">test</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">calling_interval_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">calling_interval_padding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad_qc/v4/annotations/compute_coverage.html#get_coverage_resources"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad_qc.v4.annotations.compute_coverage.get_coverage_resources" title="Permalink to this definition"></a></dt>
<dd><p>Get PipelineResourceCollection for all resources needed in the coverage pipeline.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>test</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether to gather all resources for testing.</p></li>
<li><p><strong>overwrite</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether to overwrite resources if they exist.</p></li>
<li><p><strong>data_type</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – One of ‘exomes’ or ‘genomes’.</p></li>
<li><p><strong>calling_interval_name</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – Name of calling intervals to use.</p></li>
<li><p><strong>calling_interval_padding</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – Padding to use for calling intervals.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">PipelineResourceCollection</span></code></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>PipelineResourceCollection containing resources for all steps of the
coverage pipeline.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad_qc.v4.annotations.compute_coverage.main">
<span class="sig-prename descclassname"><span class="pre">gnomad_qc.v4.annotations.compute_coverage.</span></span><span class="sig-name descname"><span class="pre">main</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad_qc/v4/annotations/compute_coverage.html#main"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad_qc.v4.annotations.compute_coverage.main" title="Permalink to this definition"></a></dt>
<dd><p>Compute coverage statistics, including mean, median_approx, and coverage over certain DPs.</p>
</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="gnomad_qc.v4.annotations" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="fix_freq_an.html" class="btn btn-neutral float-right" title="gnomad_qc.v4.annotations.fix_freq_an" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p></p>
  </div>

   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>