<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>gnomad_qc.v4.annotations.generate_variant_qc_annotations &mdash; gnomad_qc master documentation</title>
      <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/theme_overrides.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/doctools.js"></script>
        <script src="../../../_static/sphinx_highlight.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="gnomad_qc.v4.annotations.insilico_predictors" href="insilico_predictors.html" />
    <link rel="prev" title="gnomad_qc.v4.annotations.generate_freq_genomes" href="generate_freq_genomes.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            gnomad_qc
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../../index.html">API Reference</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../index.html">v4</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="index.html">annotations</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="compute_coverage.html">compute_coverage</a></li>
<li class="toctree-l4"><a class="reference internal" href="fix_freq_an.html">fix_freq_an</a></li>
<li class="toctree-l4"><a class="reference internal" href="generate_freq.html">generate_freq</a></li>
<li class="toctree-l4"><a class="reference internal" href="generate_freq_genomes.html">generate_freq_genomes</a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#">generate_variant_qc_annotations</a></li>
<li class="toctree-l4"><a class="reference internal" href="insilico_predictors.html">insilico_predictors</a></li>
<li class="toctree-l4"><a class="reference internal" href="recover_and_complete_vep115.html">recover_and_complete_vep115</a></li>
<li class="toctree-l4"><a class="reference internal" href="vep_context_ht.html">vep_context_ht</a></li>
<li class="toctree-l4"><a class="reference internal" href="vrs_annotation_batch.html">vrs_annotation_batch</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../assessment/index.html">assessment</a></li>
<li class="toctree-l3"><a class="reference internal" href="../create_release/index.html">create_release</a></li>
<li class="toctree-l3"><a class="reference internal" href="../resources/index.html">resources</a></li>
<li class="toctree-l3"><a class="reference internal" href="../sample_qc/index.html">sample_qc</a></li>
<li class="toctree-l3"><a class="reference internal" href="../subset.html">subset</a></li>
<li class="toctree-l3"><a class="reference internal" href="../variant_qc/index.html">variant_qc</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../v5/index.html">v5</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../analyses/index.html">analyses</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/broadinstitute/gnomad_qc/tags">Change Log</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">gnomad_qc</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">gnomad_qc</a></li>
          <li class="breadcrumb-item"><a href="../index.html">gnomad_qc.v4</a></li>
          <li class="breadcrumb-item"><a href="index.html">gnomad_qc.v4.annotations</a></li>
      <li class="breadcrumb-item active">gnomad_qc.v4.annotations.generate_variant_qc_annotations</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../../_sources/api_reference/v4/annotations/generate_variant_qc_annotations.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="gnomad-qc-v4-annotations-generate-variant-qc-annotations">
<h1>gnomad_qc.v4.annotations.generate_variant_qc_annotations<a class="headerlink" href="#gnomad-qc-v4-annotations-generate-variant-qc-annotations" title="Permalink to this heading"></a></h1>
<p>Script to generate annotations for variant QC on gnomAD v4.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">gnomad_qc</span><span class="o">.</span><span class="n">v4</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">generate_variant_qc_annotations</span><span class="o">.</span><span class="n">py</span>
       <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">slack</span><span class="o">-</span><span class="n">channel</span> <span class="n">SLACK_CHANNEL</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">overwrite</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">test</span><span class="o">-</span><span class="n">dataset</span><span class="p">]</span>
       <span class="p">[</span><span class="o">--</span><span class="n">test</span><span class="o">-</span><span class="n">n</span><span class="o">-</span><span class="n">partitions</span> <span class="p">[</span><span class="n">TEST_N_PARTITIONS</span><span class="p">]]</span> <span class="p">[</span><span class="o">--</span><span class="n">compute</span><span class="o">-</span><span class="n">info</span><span class="p">]</span>
       <span class="p">[</span><span class="o">--</span><span class="n">compute</span><span class="o">-</span><span class="n">info</span><span class="o">-</span><span class="n">split</span><span class="o">-</span><span class="n">n</span><span class="o">-</span><span class="n">alleles</span> <span class="n">COMPUTE_INFO_SPLIT_N_ALLELES</span><span class="p">]</span>
       <span class="p">[</span><span class="o">--</span><span class="n">compute</span><span class="o">-</span><span class="n">info</span><span class="o">-</span><span class="n">over</span><span class="o">-</span><span class="n">split</span><span class="o">-</span><span class="n">n</span><span class="o">-</span><span class="n">alleles</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">combine</span><span class="o">-</span><span class="n">compute</span><span class="o">-</span><span class="n">info</span><span class="p">]</span>
       <span class="p">[</span><span class="o">--</span><span class="n">compute</span><span class="o">-</span><span class="n">info</span><span class="o">-</span><span class="n">n</span><span class="o">-</span><span class="n">partitions</span> <span class="n">COMPUTE_INFO_N_PARTITIONS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">retain</span><span class="o">-</span><span class="n">cdfs</span><span class="p">]</span>
       <span class="p">[</span><span class="o">--</span><span class="n">cdf</span><span class="o">-</span><span class="n">k</span> <span class="n">CDF_K</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">split</span><span class="o">-</span><span class="n">info</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">export</span><span class="o">-</span><span class="n">info</span><span class="o">-</span><span class="n">vcf</span><span class="p">]</span>
       <span class="p">[</span><span class="o">--</span><span class="n">export</span><span class="o">-</span><span class="n">split</span><span class="o">-</span><span class="n">info</span><span class="o">-</span><span class="n">vcf</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">run</span><span class="o">-</span><span class="n">vep</span><span class="p">]</span>
       <span class="p">[</span><span class="o">--</span><span class="n">vep</span><span class="o">-</span><span class="n">data</span><span class="o">-</span><span class="nb">type</span> <span class="p">{</span><span class="n">exomes</span><span class="p">,</span><span class="n">genomes</span><span class="p">}]</span> <span class="p">[</span><span class="o">--</span><span class="n">validate</span><span class="o">-</span><span class="n">vep</span><span class="p">]</span>
       <span class="p">[</span><span class="o">--</span><span class="n">vep</span><span class="o">-</span><span class="n">version</span> <span class="n">VEP_VERSION</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">generate</span><span class="o">-</span><span class="n">trio</span><span class="o">-</span><span class="n">stats</span><span class="p">]</span>
       <span class="p">[</span><span class="o">--</span><span class="n">releasable</span><span class="o">-</span><span class="n">trios</span><span class="o">-</span><span class="n">only</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">generate</span><span class="o">-</span><span class="n">sibling</span><span class="o">-</span><span class="n">stats</span><span class="p">]</span>
       <span class="p">[</span><span class="o">--</span><span class="n">create</span><span class="o">-</span><span class="n">variant</span><span class="o">-</span><span class="n">qc</span><span class="o">-</span><span class="n">annotation</span><span class="o">-</span><span class="n">ht</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">impute</span><span class="o">-</span><span class="n">features</span><span class="p">]</span>
       <span class="p">[</span><span class="o">--</span><span class="n">n</span><span class="o">-</span><span class="n">partitions</span> <span class="n">N_PARTITIONS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">export</span><span class="o">-</span><span class="n">true</span><span class="o">-</span><span class="n">positive</span><span class="o">-</span><span class="n">vcfs</span><span class="p">]</span>
       <span class="p">[</span><span class="o">--</span><span class="n">transmitted</span><span class="o">-</span><span class="n">singletons</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">sibling</span><span class="o">-</span><span class="n">singletons</span><span class="p">]</span>
</pre></div>
</div>
<section id="named-arguments">
<h2>Named Arguments<a class="headerlink" href="#named-arguments" title="Permalink to this heading"></a></h2>
<dl class="option-list">
<dt><kbd>--slack-channel</kbd></dt>
<dd><p>Slack channel to post results and notifications to.</p>
</dd>
<dt><kbd>--overwrite</kbd></dt>
<dd><p>Overwrite data</p>
<p>Default: False</p>
</dd>
<dt><kbd>--test-dataset</kbd></dt>
<dd><p>Use the test dataset as input.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--test-n-partitions</kbd></dt>
<dd><p>Use only 2 partitions of the VDS as input for testing purposes.</p>
</dd>
<dt><kbd>--split-info</kbd></dt>
<dd><p>Split info HT.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--export-info-vcf</kbd></dt>
<dd><p>Export info as VCF.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--export-split-info-vcf</kbd></dt>
<dd><p>Export split info as VCF.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--run-vep</kbd></dt>
<dd><p>Generates vep annotations.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--vep-data-type</kbd></dt>
<dd><p>Possible choices: exomes, genomes</p>
<p>Data type to use for VEP annotations. Default is exomes.</p>
<p>Default: “exomes”</p>
</dd>
<dt><kbd>--validate-vep</kbd></dt>
<dd><p>Validate that variants in protein-coding genes are correctly annotated by VEP.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--vep-version</kbd></dt>
<dd><p>Version of VEPed context Table to use in vep_or_lookup_vep.</p>
<p>Default: “105”</p>
</dd>
<dt><kbd>--generate-sibling-stats</kbd></dt>
<dd><p>Calculate sibling variant sharing stats.</p>
<p>Default: False</p>
</dd>
</dl>
</section>
<section id="compute-info-ht.">
<h2>Compute info HT.<a class="headerlink" href="#compute-info-ht." title="Permalink to this heading"></a></h2>
<p>Arguments relevant to computing the info HT..</p>
<dl class="option-list">
<dt><kbd>--compute-info</kbd></dt>
<dd><p>Compute info HT.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--compute-info-split-n-alleles</kbd></dt>
<dd><p>Number of alleles at a site to filter results on. If –compute-info-over-split-n-alleles is used, the results are filtered to sites with greater than or equal to the value supplied, otherwise the results are filtered to sites with less than the value supplied. By default no sites are filtered.</p>
</dd>
<dt><kbd>--compute-info-over-split-n-alleles</kbd></dt>
<dd><p>Whether to filter to sites greater than or equal to the value supplied to–compute-info-split-n-alleles. By default, sites are filtered to sites less than that value, or None if –compute-info-split-n-alleles is not supplied.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--combine-compute-info</kbd></dt>
<dd><p>Whether to combine the output from running –compute-info –compute-info-split-n-alleles with and without the –compute-info-over-split-n-alleles flag.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--compute-info-n-partitions</kbd></dt>
<dd><p>Number of desired partitions for the info HT.</p>
<p>Default: 5000</p>
</dd>
<dt><kbd>--retain-cdfs</kbd></dt>
<dd><p>If True, retains the cumulative distribution functions (CDFs) for all info annotations that are computed as a median aggregation. Keeping the CDFs is useful for annotations that require calculating the median acrosscombined datasets at a later stage. Default is False.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--cdf-k</kbd></dt>
<dd><p>Parameter controlling the accuracy vs. memory usage tradeoff when retaining CDFs. A higher value of <cite>cdf_k</cite> results in a more accurate CDF approximation but increases memory usage and computation time. Default is 200.</p>
<p>Default: 200</p>
</dd>
</dl>
</section>
<section id="arguments-used-to-generate-trio-stats.">
<h2>Arguments used to generate trio stats.<a class="headerlink" href="#arguments-used-to-generate-trio-stats." title="Permalink to this heading"></a></h2>
<dl class="option-list">
<dt><kbd>--generate-trio-stats</kbd></dt>
<dd><p>Calculates trio stats</p>
<p>Default: False</p>
</dd>
<dt><kbd>--releasable-trios-only</kbd></dt>
<dd><p>Only include releasable trios. This option is only valid when –generate-trio-stats is true.</p>
<p>Default: False</p>
</dd>
</dl>
</section>
<section id="variant-qc-annotation-ht-parameters">
<h2>Variant QC annotation HT parameters<a class="headerlink" href="#variant-qc-annotation-ht-parameters" title="Permalink to this heading"></a></h2>
<dl class="option-list">
<dt><kbd>--create-variant-qc-annotation-ht</kbd></dt>
<dd><p>Creates an annotated HT with features for variant QC.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--impute-features</kbd></dt>
<dd><p>If set, imputation is performed for variant QC features.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--n-partitions</kbd></dt>
<dd><p>Desired number of partitions for variant QC annotation HT .</p>
<p>Default: 5000</p>
</dd>
</dl>
</section>
<section id="export-true-positive-vcfs">
<h2>Export true positive VCFs<a class="headerlink" href="#export-true-positive-vcfs" title="Permalink to this heading"></a></h2>
<p>Arguments used to define true positive variant set.</p>
<dl class="option-list">
<dt><kbd>--export-true-positive-vcfs</kbd></dt>
<dd><p>Exports true positive variants (–transmitted-singletons and/or –sibling-singletons) to VCF files.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--transmitted-singletons</kbd></dt>
<dd><p>Include transmitted singletons in the exports of true positive variants to VCF files.</p>
<p>Default: False</p>
</dd>
<dt><kbd>--sibling-singletons</kbd></dt>
<dd><p>Include sibling singletons in the exports of true positive variants to VCF files.</p>
<p>Default: False</p>
</dd>
</dl>
</section>
<section id="module-functions">
<h2>Module Functions<a class="headerlink" href="#module-functions" title="Permalink to this heading"></a></h2>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.INFO_METHODS" title="gnomad_qc.v4.annotations.generate_variant_qc_annotations.INFO_METHODS"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.INFO_METHODS</span></code></a></p></td>
<td><p>List of info methods computed for variant QC.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.INFO_FEATURES" title="gnomad_qc.v4.annotations.generate_variant_qc_annotations.INFO_FEATURES"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.INFO_FEATURES</span></code></a></p></td>
<td><p>List of features info to be used for variant QC.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.NON_INFO_FEATURES" title="gnomad_qc.v4.annotations.generate_variant_qc_annotations.NON_INFO_FEATURES"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.NON_INFO_FEATURES</span></code></a></p></td>
<td><p>List of features to be used for variant QC that are not in the info field.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.TRUTH_DATA" title="gnomad_qc.v4.annotations.generate_variant_qc_annotations.TRUTH_DATA"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.TRUTH_DATA</span></code></a></p></td>
<td><p>List of truth datasets to be used for variant QC.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.extract_as_pls" title="gnomad_qc.v4.annotations.generate_variant_qc_annotations.extract_as_pls"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.extract_as_pls</span></code></a>(...)</p></td>
<td><p>Extract PLs for a specific allele from an LPL array expression.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.recompute_as_qualapprox_from_lpl" title="gnomad_qc.v4.annotations.generate_variant_qc_annotations.recompute_as_qualapprox_from_lpl"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.recompute_as_qualapprox_from_lpl</span></code></a>(mt)</p></td>
<td><p>Recompute AS_QUALapprox from LPL.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.correct_as_annotations" title="gnomad_qc.v4.annotations.generate_variant_qc_annotations.correct_as_annotations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.correct_as_annotations</span></code></a>(mt)</p></td>
<td><p>Correct allele specific annotations that are longer than the length of LA.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.run_compute_info" title="gnomad_qc.v4.annotations.generate_variant_qc_annotations.run_compute_info"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.run_compute_info</span></code></a>(mt)</p></td>
<td><p>Run compute info on a MatrixTable.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.get_reformatted_info_fields" title="gnomad_qc.v4.annotations.generate_variant_qc_annotations.get_reformatted_info_fields"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.get_reformatted_info_fields</span></code></a>(ht)</p></td>
<td><p>Reformat <cite>ht</cite> info annotations to contain all expected info fields.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.get_info_ht_for_vcf_export" title="gnomad_qc.v4.annotations.generate_variant_qc_annotations.get_info_ht_for_vcf_export"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.get_info_ht_for_vcf_export</span></code></a>(ht, ...)</p></td>
<td><p>Get info HT for VCF export.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.split_info" title="gnomad_qc.v4.annotations.generate_variant_qc_annotations.split_info"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.split_info</span></code></a>(info_ht)</p></td>
<td><p>Generate an info Table with split multi-allelic sites from the multi-allelic info Table.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.run_generate_trio_stats" title="gnomad_qc.v4.annotations.generate_variant_qc_annotations.run_generate_trio_stats"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.run_generate_trio_stats</span></code></a>(...)</p></td>
<td><p>Generate trio transmission stats from a VariantDataset and pedigree info.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.run_generate_sib_stats" title="gnomad_qc.v4.annotations.generate_variant_qc_annotations.run_generate_sib_stats"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.run_generate_sib_stats</span></code></a>(mt, ...)</p></td>
<td><p>Generate stats for the number of alternate alleles in common between sibling pairs.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.create_variant_qc_annotation_ht" title="gnomad_qc.v4.annotations.generate_variant_qc_annotations.create_variant_qc_annotation_ht"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.create_variant_qc_annotation_ht</span></code></a>(...)</p></td>
<td><p>Create a Table with all necessary annotations for variant QC.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.get_tp_ht_for_vcf_export" title="gnomad_qc.v4.annotations.generate_variant_qc_annotations.get_tp_ht_for_vcf_export"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.get_tp_ht_for_vcf_export</span></code></a>(ht)</p></td>
<td><p>Get Tables with raw and adj true positive variants to export as a VCF for use in VQSR.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.get_variant_qc_annotation_resources" title="gnomad_qc.v4.annotations.generate_variant_qc_annotations.get_variant_qc_annotation_resources"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.get_variant_qc_annotation_resources</span></code></a>(...)</p></td>
<td><p>Get PipelineResourceCollection for all resources needed in the variant QC annotation pipeline.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.main" title="gnomad_qc.v4.annotations.generate_variant_qc_annotations.main"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.main</span></code></a>(args)</p></td>
<td><p>Generate all variant annotations needed for variant QC.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.get_script_argument_parser</span></code>()</p></td>
<td><p>Get script argument parser.</p></td>
</tr>
</tbody>
</table>
<span class="target" id="module-gnomad_qc.v4.annotations.generate_variant_qc_annotations"></span><p>Script to generate annotations for variant QC on gnomAD v4.</p>
<dl class="py data">
<dt class="sig sig-object py" id="gnomad_qc.v4.annotations.generate_variant_qc_annotations.INFO_METHODS">
<span class="sig-prename descclassname"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.</span></span><span class="sig-name descname"><span class="pre">INFO_METHODS</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">['AS',</span> <span class="pre">'quasi',</span> <span class="pre">'set_long_AS_missing']</span></em><a class="headerlink" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.INFO_METHODS" title="Permalink to this definition"></a></dt>
<dd><p>List of info methods computed for variant QC.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad_qc.v4.annotations.generate_variant_qc_annotations.INFO_FEATURES">
<span class="sig-prename descclassname"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.</span></span><span class="sig-name descname"><span class="pre">INFO_FEATURES</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">['AS_MQRankSum',</span> <span class="pre">'AS_pab_max',</span> <span class="pre">'AS_MQ',</span> <span class="pre">'AS_QD',</span> <span class="pre">'AS_ReadPosRankSum',</span> <span class="pre">'AS_SOR',</span> <span class="pre">'AS_FS']</span></em><a class="headerlink" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.INFO_FEATURES" title="Permalink to this definition"></a></dt>
<dd><p>List of features info to be used for variant QC.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad_qc.v4.annotations.generate_variant_qc_annotations.NON_INFO_FEATURES">
<span class="sig-prename descclassname"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.</span></span><span class="sig-name descname"><span class="pre">NON_INFO_FEATURES</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">['variant_type',</span> <span class="pre">'allele_type',</span> <span class="pre">'n_alt_alleles',</span> <span class="pre">'was_mixed',</span> <span class="pre">'has_star']</span></em><a class="headerlink" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.NON_INFO_FEATURES" title="Permalink to this definition"></a></dt>
<dd><p>List of features to be used for variant QC that are not in the info field.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="gnomad_qc.v4.annotations.generate_variant_qc_annotations.TRUTH_DATA">
<span class="sig-prename descclassname"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.</span></span><span class="sig-name descname"><span class="pre">TRUTH_DATA</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">['hapmap',</span> <span class="pre">'omni',</span> <span class="pre">'mills',</span> <span class="pre">'kgp_phase1_hc']</span></em><a class="headerlink" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.TRUTH_DATA" title="Permalink to this definition"></a></dt>
<dd><p>List of truth datasets to be used for variant QC.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad_qc.v4.annotations.generate_variant_qc_annotations.extract_as_pls">
<span class="sig-prename descclassname"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.</span></span><span class="sig-name descname"><span class="pre">extract_as_pls</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lpl_expr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allele_idx</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad_qc/v4/annotations/generate_variant_qc_annotations.html#extract_as_pls"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.extract_as_pls" title="Permalink to this definition"></a></dt>
<dd><p>Extract PLs for a specific allele from an LPL array expression.</p>
<p>PL/LPL represents the normalized Phred-scaled likelihoods of the possible
genotypes from all considered alleles (or local alleles).</p>
<p>If three alleles are considered, LPL genotype indexes are:
[0/0, 0/1, 1/1, 0/2, 1/2, 2/2].</p>
<dl class="simple">
<dt>If we want to extract the PLs for each alternate allele, we need to extract:</dt><dd><ul class="simple">
<li><p>allele 1: [0/0, 0/1, 1/1]</p></li>
<li><p>allele 2: [0/0, 0/2, 2/2]</p></li>
</ul>
</dd>
<dt>Example:</dt><dd><ul class="simple">
<li><p>LPL: [138, 98, 154, 26, 0, 14]</p></li>
<li><p>Extract allele 1 PLs: [0/0, 0/1, 1/1] -&gt; [138, 98, 154]</p></li>
<li><p>Extract allele 2 PLs: [0/0, 0/2, 2/2] -&gt; [138, 26, 14]</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>lpl_expr</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>) – LPL ArrayExpression.</p></li>
<li><p><strong>allele_idx</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.Int32Expression.html#hail.expr.Int32Expression" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Int32Expression</span></code></a>) – The index of the alternate allele to extract PLs for.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>ArrayExpression of PLs for the specified allele.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad_qc.v4.annotations.generate_variant_qc_annotations.recompute_as_qualapprox_from_lpl">
<span class="sig-prename descclassname"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.</span></span><span class="sig-name descname"><span class="pre">recompute_as_qualapprox_from_lpl</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mt</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad_qc/v4/annotations/generate_variant_qc_annotations.html#recompute_as_qualapprox_from_lpl"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.recompute_as_qualapprox_from_lpl" title="Permalink to this definition"></a></dt>
<dd><p>Recompute AS_QUALapprox from LPL.</p>
<p>QUALapprox is the (Phred-scaled) probability that all reads at the site are hom-ref,
so QUALapprox is PL[0]. To get the QUALapprox for just one allele, pull out the
PLs for just that allele, then normalize by subtracting the smallest element from
all the entries (so the best genotype is 0) and then use the normalized PL[0]
value for that allele’s QUALapprox.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p>The first element of AS_QUALapprox is always None.</p></li>
<li><p>If the allele is a star allele, we set QUALapprox for that allele to 0.</p></li>
<li><p>If GQ == 0 and PL[0] for the allele == 1, we set QUALapprox for the allele
to 0.</p></li>
</ul>
</div>
<dl>
<dt>Example:</dt><dd><dl class="simple">
<dt>Starting Values:</dt><dd><ul class="simple">
<li><p>alleles: [‘G’, ‘*’, ‘A’, ‘C’, ‘GCTT’, ‘GT’, ‘T’]</p></li>
<li><p>LGT: 1/2</p></li>
<li><p>LA: [0, 1, 6]</p></li>
<li><p>LPL: [138, 98, 154, 26, 0, 14]</p></li>
<li><p>QUALapprox: 138</p></li>
</ul>
</dd>
<dt>Use <cite>extract_as_pls</cite> to get PLs for each allele:</dt><dd><ul class="simple">
<li><p>allele 1: [138, 98, 154]</p></li>
<li><p>allele 2: [138, 26, 14]</p></li>
</ul>
</dd>
<dt>Normalize PLs by subtracting the smallest element from all the PLs:</dt><dd><ul class="simple">
<li><p>allele 1: [138-98, 98-98, 154-98] -&gt; [40, 0, 56]</p></li>
<li><p>allele 2: [138-14, 26-14, 14-14] -&gt; [124, 12, 0]</p></li>
</ul>
</dd>
</dl>
<p>Use the first element of the allele specific PLs to generate AS_QUALapprox:
[None, 40, 124]</p>
<p>Set QUALapprox to 0 for the star allele: [None, 0, 124]</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>mt</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>) – Input MatrixTable.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>AS_QUALapprox ArrayExpression recomputed from LPL.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad_qc.v4.annotations.generate_variant_qc_annotations.correct_as_annotations">
<span class="sig-prename descclassname"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.</span></span><span class="sig-name descname"><span class="pre">correct_as_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">set_to_missing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad_qc/v4/annotations/generate_variant_qc_annotations.html#correct_as_annotations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.correct_as_annotations" title="Permalink to this definition"></a></dt>
<dd><p>Correct allele specific annotations that are longer than the length of LA.</p>
<p>For some entries in the MatrixTable, the following annotations are longer than LA,
when they should be the same length as LA:</p>
<blockquote>
<div><ul class="simple">
<li><p>AS_SB_TABLE</p></li>
<li><p>AS_RAW_MQ</p></li>
<li><p>AS_RAW_ReadPosRankSum</p></li>
<li><p>AS_RAW_MQRankSum</p></li>
</ul>
</div></blockquote>
<p>This function corrects these annotations by either dropping the alternate allele
with the index corresponding to the min value of AS_RAW_MQ, or setting them to
missing if <cite>set_to_missing</cite> is True.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mt</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>) – Input MatrixTable.</p></li>
<li><p><strong>set_to_missing</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether to set the annotations to missing instead of
correcting them.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StructExpression.html#hail.expr.StructExpression" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StructExpression</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>StructExpression with corrected allele specific annotations.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad_qc.v4.annotations.generate_variant_qc_annotations.run_compute_info">
<span class="sig-prename descclassname"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.</span></span><span class="sig-name descname"><span class="pre">run_compute_info</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_n_alleles</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_n_alleles</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">retain_cdfs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cdf_k</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad_qc/v4/annotations/generate_variant_qc_annotations.html#run_compute_info"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.run_compute_info" title="Permalink to this definition"></a></dt>
<dd><p>Run compute info on a MatrixTable.</p>
<p>..note:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Adds</span> <span class="n">a</span> <span class="n">fix</span> <span class="k">for</span> <span class="n">AS_QUALapprox</span> <span class="n">by</span> <span class="n">recomputing</span> <span class="kn">from</span> <span class="nn">LPL</span> <span class="n">because</span> <span class="n">some</span> <span class="n">were</span> <span class="n">found</span> <span class="n">to</span>
<span class="n">have</span> <span class="n">different</span> <span class="n">lengths</span> <span class="n">than</span> <span class="n">LA</span><span class="o">.</span>
</pre></div>
</div>
<dl class="simple">
<dt>Creates a Table with three different methods of computing info annotations:</dt><dd><ul class="simple">
<li><p>quasi_info: Compute info annotations using the quasi-allele specific method
defined in <cite>default_compute_info</cite>.</p></li>
<li><p>AS_info: Compute info annotations using aggregation of the allele specific
annotations in ‘gvcf_info’ after recomputing AS_QUALapprox from LPL, and
fixing the length of AS_SB_TABLE, AS_RAW_MQ, AS_RAW_ReadPosRankSum and
AS_RAW_MQRankSum.</p></li>
<li><p>set_long_AS_missing_info: Compute info annotations using aggregation of the
allele specific  annotations in ‘gvcf_info’ after setting AS_SB_TABLE,
AS_RAW_MQ, AS_RAW_ReadPosRankSum and AS_RAW_MQRankSum to missing if they have
the incorrect length.</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mt</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>) – Input MatrixTable.</p></li>
<li><p><strong>max_n_alleles</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – Maximum number of alleles for the site to be included in
computations.</p></li>
<li><p><strong>min_n_alleles</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – Minimum number of alleles for the site to be included in
computations.</p></li>
<li><p><strong>retain_cdfs</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If True, retains the cumulative distribution functions (CDFs)
for all info annotations that are computed as a median aggregation. Keeping the
CDFs is useful for annotations that require calculating the median across
combined datasets at a later stage. Default is False.</p></li>
<li><p><strong>cdf_k</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Parameter controlling the accuracy vs. memory usage tradeoff when
retaining CDFs. A higher value of <cite>cdf_k</cite> results in a more accurate CDF
approximation but increases memory usage and computation time. Default is 200.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Table with info annotations.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad_qc.v4.annotations.generate_variant_qc_annotations.get_reformatted_info_fields">
<span class="sig-prename descclassname"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.</span></span><span class="sig-name descname"><span class="pre">get_reformatted_info_fields</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ht</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">info_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad_qc/v4/annotations/generate_variant_qc_annotations.html#get_reformatted_info_fields"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.get_reformatted_info_fields" title="Permalink to this definition"></a></dt>
<dd><p>Reformat <cite>ht</cite> info annotations to contain all expected info fields.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Input Table.</p></li>
<li><p><strong>info_method</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – Shorthand name of method used to compute the info annotation
that should be reformatted. Choices are ‘AS’, ‘quasi’, or ‘set_long_AS_missing’.
If None, all info annotations will be reformatted.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Table with reformatted info annotations.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad_qc.v4.annotations.generate_variant_qc_annotations.get_info_ht_for_vcf_export">
<span class="sig-prename descclassname"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.</span></span><span class="sig-name descname"><span class="pre">get_info_ht_for_vcf_export</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ht</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">info_method</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad_qc/v4/annotations/generate_variant_qc_annotations.html#get_info_ht_for_vcf_export"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.get_info_ht_for_vcf_export" title="Permalink to this definition"></a></dt>
<dd><p>Get info HT for VCF export.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Input info HT.</p></li>
<li><p><strong>info_method</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Info method to use. One of ‘AS’, ‘quasi’, or
‘set_long_AS_missing’.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Info HT for VCF export.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad_qc.v4.annotations.generate_variant_qc_annotations.split_info">
<span class="sig-prename descclassname"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.</span></span><span class="sig-name descname"><span class="pre">split_info</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">info_ht</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad_qc/v4/annotations/generate_variant_qc_annotations.html#split_info"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.split_info" title="Permalink to this definition"></a></dt>
<dd><p>Generate an info Table with split multi-allelic sites from the multi-allelic info Table.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>gnomad_methods’ <cite>annotate_allele_info</cite> splits multi-allelic sites before the
<cite>info</cite> annotation is split to ensure that all sites in the returned Table are
annotated with allele info.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>info_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Info Table with unsplit multi-allelics.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Info Table with split multi-allelics.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad_qc.v4.annotations.generate_variant_qc_annotations.run_generate_trio_stats">
<span class="sig-prename descclassname"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.</span></span><span class="sig-name descname"><span class="pre">run_generate_trio_stats</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fam_ped</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fam_ht</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">releasable_only</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad_qc/v4/annotations/generate_variant_qc_annotations.html#run_generate_trio_stats"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.run_generate_trio_stats" title="Permalink to this definition"></a></dt>
<dd><p>Generate trio transmission stats from a VariantDataset and pedigree info.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vds</strong> (<a class="reference external" href="https://hail.is/docs/0.2/vds/hail.vds.VariantDataset.html#hail.vds.VariantDataset" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantDataset</span></code></a>) – VariantDataset to generate trio stats from.</p></li>
<li><p><strong>fam_ped</strong> (<a class="reference external" href="https://hail.is/docs/0.2/genetics/hail.genetics.Pedigree.html#hail.genetics.Pedigree" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Pedigree</span></code></a>) – Pedigree containing trio info.</p></li>
<li><p><strong>fam_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Table containing trio info.</p></li>
<li><p><strong>releasable_only</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether to only include releasable trios. Releasable trios are those where all three samples (proband, maternal, and paternal) are marked as ‘releasable’.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Table containing trio stats.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad_qc.v4.annotations.generate_variant_qc_annotations.run_generate_sib_stats">
<span class="sig-prename descclassname"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.</span></span><span class="sig-name descname"><span class="pre">run_generate_sib_stats</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rel_ht</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad_qc/v4/annotations/generate_variant_qc_annotations.html#run_generate_sib_stats"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.run_generate_sib_stats" title="Permalink to this definition"></a></dt>
<dd><p>Generate stats for the number of alternate alleles in common between sibling pairs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mt</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>) – MatrixTable to generate sibling stats from.</p></li>
<li><p><strong>rel_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Table containing relatedness info for pairs in <cite>mt</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Table containing sibling stats.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad_qc.v4.annotations.generate_variant_qc_annotations.create_variant_qc_annotation_ht">
<span class="sig-prename descclassname"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.</span></span><span class="sig-name descname"><span class="pre">create_variant_qc_annotation_ht</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">info_ht</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trio_stats_ht</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sib_stats_ht</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">impute_features</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_partitions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5000</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad_qc/v4/annotations/generate_variant_qc_annotations.html#create_variant_qc_annotation_ht"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.create_variant_qc_annotation_ht" title="Permalink to this definition"></a></dt>
<dd><p>Create a Table with all necessary annotations for variant QC.</p>
<p>Annotations that are included:</p>
<blockquote>
<div><dl class="simple">
<dt>Features for RF:</dt><dd><ul class="simple">
<li><p>variant_type</p></li>
<li><p>allele_type</p></li>
<li><p>n_alt_alleles</p></li>
<li><p>has_star</p></li>
<li><p>AS_QD</p></li>
<li><p>AS_pab_max</p></li>
<li><p>AS_MQRankSum</p></li>
<li><p>AS_SOR</p></li>
<li><p>AS_ReadPosRankSum</p></li>
</ul>
</dd>
<dt>Training sites (bool):</dt><dd><ul class="simple">
<li><p>transmitted_singleton</p></li>
<li><p>sibling_singleton</p></li>
<li><p>fail_hard_filters - (ht.QD &lt; 2) | (ht.FS &gt; 60) | (ht.MQ &lt; 30)</p></li>
</ul>
</dd>
</dl>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>info_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Info Table with split multi-allelics.</p></li>
<li><p><strong>trio_stats_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Table with trio statistics.</p></li>
<li><p><strong>sib_stats_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Table with sibling statistics.</p></li>
<li><p><strong>impute_features</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether to impute features using feature medians (this is
done by variant type).</p></li>
<li><p><strong>n_partitions</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Number of partitions to use for final annotated Table.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Hail Table with all annotations needed for variant QC.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad_qc.v4.annotations.generate_variant_qc_annotations.get_tp_ht_for_vcf_export">
<span class="sig-prename descclassname"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.</span></span><span class="sig-name descname"><span class="pre">get_tp_ht_for_vcf_export</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ht</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transmitted_singletons</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sibling_singletons</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad_qc/v4/annotations/generate_variant_qc_annotations.html#get_tp_ht_for_vcf_export"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.get_tp_ht_for_vcf_export" title="Permalink to this definition"></a></dt>
<dd><p>Get Tables with raw and adj true positive variants to export as a VCF for use in VQSR.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Input Table with transmitted singleton and sibling singleton information.</p></li>
<li><p><strong>transmitted_singletons</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether to include transmitted singletons in the
true positive variants.</p></li>
<li><p><strong>sibling_singletons</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether to include sibling singletons in the true
positive variants.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2.137)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>]</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary of ‘raw’ and ‘adj’ true positive variant sites Tables.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad_qc.v4.annotations.generate_variant_qc_annotations.get_variant_qc_annotation_resources">
<span class="sig-prename descclassname"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.</span></span><span class="sig-name descname"><span class="pre">get_variant_qc_annotation_resources</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">test</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">over_n_alleles</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">combine_compute_info</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">true_positive_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">releasable_trios_only</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vep_data_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'exomes'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vep_version</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'105'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad_qc/v4/annotations/generate_variant_qc_annotations.html#get_variant_qc_annotation_resources"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.get_variant_qc_annotation_resources" title="Permalink to this definition"></a></dt>
<dd><p>Get PipelineResourceCollection for all resources needed in the variant QC annotation pipeline.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>test</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether to gather all resources for testing.</p></li>
<li><p><strong>overwrite</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether to overwrite resources if they exist.</p></li>
<li><p><strong>over_n_alleles</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>]) – Whether to use a temporary info TableResource for results.
When True, use temporary info TableResource for only sites that have more
than the passed arg –compute-info-split-n-alleles alleles. When False, use
temporary info TableResource for only sites with fewer alleles. When None,
the finalize info ht is used instead of a temporary location. Default is None.</p></li>
<li><p><strong>combine_compute_info</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether the input for –compute-info should be the two
temporary files (with and without the –compute-info-over-split-n-alleles flag)
produced by running –compute-info with –compute-info-split-n-alleles.</p></li>
<li><p><strong>true_positive_type</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – Type of true positive variants to use for true positive
VCF path resource. Default is None.</p></li>
<li><p><strong>releasable_trios_only</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether to only include releasable trios in the trio
stats. Default is False.</p></li>
<li><p><strong>vep_data_type</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Data type to use for VEP annotations. Default is exomes.</p></li>
<li><p><strong>vep_version</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – VEP version to use (e.g., “105”, “115”). Default is “105”.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">PipelineResourceCollection</span></code></p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>PipelineResourceCollection containing resources for all steps of the
variant QC annotation pipeline.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="gnomad_qc.v4.annotations.generate_variant_qc_annotations.main">
<span class="sig-prename descclassname"><span class="pre">gnomad_qc.v4.annotations.generate_variant_qc_annotations.</span></span><span class="sig-name descname"><span class="pre">main</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/gnomad_qc/v4/annotations/generate_variant_qc_annotations.html#main"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#gnomad_qc.v4.annotations.generate_variant_qc_annotations.main" title="Permalink to this definition"></a></dt>
<dd><p>Generate all variant annotations needed for variant QC.</p>
</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="generate_freq_genomes.html" class="btn btn-neutral float-left" title="gnomad_qc.v4.annotations.generate_freq_genomes" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="insilico_predictors.html" class="btn btn-neutral float-right" title="gnomad_qc.v4.annotations.insilico_predictors" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p></p>
  </div>

   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>