
<style>
  table {
    border-collapse: collapse;
  }
  th, td {
    border: 1px solid #000;
    padding: 5px;
  }
</style>
<h1>Global Fields Specification</h1>
<p>This table provides a view of each global (non-row) field. Every leaf field (i.e., non-struct field) is listed on its own row with the full annotation. The parent portion is shown in a lighter color to indicate context. Cell colors indicate the field necessity: <span style="background-color:#fff0f0; padding:2px 6px; border-radius:4px;">"Required" in red</span>, <span style="background-color:#f0faff; padding:2px 6px; border-radius:4px;">"Optional" in blue</span>, <span style="background-color:#dedbe4; padding:2px 6px; border-radius:4px;">"Not Needed" in grey</span>.</p>
<p><strong><code>adj</code> definition:</strong>
A variant is considered to fail the "adj" criteria if any of the following conditions are met:
 <strong>GQ</strong> &lt; 20, <strong>DP</strong> &lt; 10, or <strong>AB</strong> &lt; 0.2 (heterozygous calls).</p>
<p>All histograms annotations must use the same bin edges as defined in their respective 'Description' columns.</p>
<table style="border-collapse: collapse; border: 1px solid #000; width: 100%;"><thead><tr style="background-color:#f8f8f8;"><th style="border: 1px solid #000; padding: 5px; text-align: left;">Field</th><th style="border: 1px solid #000; padding: 5px; text-align: left;">Type</th><th style="border: 1px solid #000; padding: 5px; text-align: left;">Description</th><th style="border: 1px solid #000; padding: 5px; text-align: left;">Example</th><th style="border: 1px solid #000; padding: 5px; text-align: left;">Field Necessity</th></tr></thead><tbody><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><strong>freq_meta</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;dict&lt;str, str&gt;&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Array of frequency metadata dictionaries containing the frequency aggregation group for each element of the ‘freq’ array row annotation. Each dictionary should have the following keys: 'gen_anc', 'group', 'sex'. A 'downsampling' key is optional. The 'adj' group should always be the first value of the array, and the 'raw' group should be the second value. Required 'group' values are 'adj' and 'raw'. Required 'sex' values are 'XX' and 'XY'. Specific values are not required for 'gen_anc'.</td><td style="border: 1px solid #000; padding: 5px;"><code>[{'group': 'adj'},{'group': 'raw'},{'gen_anc': 'afr', 'group': 'adj'},{'gen_anc': 'amr', 'group': 'adj'},{'group': 'adj', 'sex': 'XX'},{'group': 'adj', 'sex': 'XY'}, ...]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><strong>freq_index_dict</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>dict&lt;str, int32&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Dictionary keyed by specified label grouping combinations ('group': 'adj'/'raw', 'gen_anc': inferred genetic ancestry group, 'sex': sex karyotype), with values describing the corresponding index of each grouping entry in the ‘freq’ array row annotation. <br/><br/>If provided, keys need to be formatted in the order of 'gen_anc'_ 'sex'_'group'. If a 'downsampling' key is included, the order should be 'downsampling'_'gen_anc'_ 'sex'_'group'. The 'adj' value should always be at index 0 and the 'raw' value at index 1. A more detailed description can be found at https://gnomad.broadinstitute.org/help/v4-hts.</td><td style="border: 1px solid #000; padding: 5px;"><code>{"adj": 0, "raw": 1, "afr_adj": 2, "amr_adj": 3, "XX_adj": 4, "XX_adj": 5, ...}</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><strong>freq_meta_sample_count</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;int32&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">A sample count per sample grouping defined in the 'freq_meta' global annotation. Must be in the same order as 'freq'/'freq_meta'.</td><td style="border: 1px solid #000; padding: 5px;"><code>[730947, 730947, 16740, 15001, 50000, 680947, ...]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><strong>faf_meta</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;dict&lt;str, str&gt;&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Filtering allele frequency metadata. An ordered list containing the frequency aggregation group for each element of the ‘faf’ array row annotation.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><strong>faf_index_dict</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>dict&lt;str, int32&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Dictionary keyed by specified label grouping combinations ('group': 'adj'/'raw', 'gen_anc': inferred genetic ancestry group, 'sex': sex karyotype), with values describing the corresponding index of each grouping entry in the filtering allele frequency (‘faf’) row annotation.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><strong>age_distribution</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ... }</code></td><td style="border: 1px solid #000; padding: 5px;">Callset-wide age histogram. Cohorts for gnomAD vary in how they report age (some report the age at diagnosis, others report the age of last visit, etc), so the ages associated with the gnomAD data can be thought of as the last known age of the individual. Information on age is not available for all gnomAD samples. This field is required with the acknowledgement that age data may not be available for all samples. If no age is available for any samples, set all <code>bin_freq</code> values to <code>0</code>.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">age_distribution.</span>bin_edges</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;float64&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Bin edges for age histogram: <code>30.0 | 35.0 | 40.0 | 45.0 | 50.0 | 55.0 | 60.0 | 65.0 | 70.0 | 75.0 | 80.0</code>.</td><td style="border: 1px solid #000; padding: 5px;"><code>[30.0, 35.0, 40.0, 45.0, ...]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">age_distribution.</span>bin_freq</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;int32&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Bin frequencies for the age histogram. This is the number of records found in each bin.</td><td style="border: 1px solid #000; padding: 5px;"><code>[101, 122, 85, 4, ...]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">age_distribution.</span>n_smaller</td><td style="border: 1px solid #000; padding: 5px;"><code>int32</code></td><td style="border: 1px solid #000; padding: 5px;">Count of age values falling below lowest histogram bin edge.</td><td style="border: 1px solid #000; padding: 5px;"><code>1000</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">age_distribution.</span>n_larger</td><td style="border: 1px solid #000; padding: 5px;"><code>int32</code></td><td style="border: 1px solid #000; padding: 5px;">Count of age values falling above highest histogram bin edge.</td><td style="border: 1px solid #000; padding: 5px;"><code>30</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">downsamplings.</span>n_larger</td><td style="border: 1px solid #000; padding: 5px;"><code>dict&lt;str, array&lt;int32&gt;&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Dictionary keyed by dataset with values corresponding to available downsampled sample counts.</td><td style="border: 1px solid #000; padding: 5px;"><code>{'gnomad': [10, 100, 500...]}</code></td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><strong>filtering_model</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ... }</code></td><td style="border: 1px solid #000; padding: 5px;">The variant filtering model used and its specific cutoffs.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">filtering_model.</span>filter_name</td><td style="border: 1px solid #000; padding: 5px;"><code>str</code></td><td style="border: 1px solid #000; padding: 5px;">Variant filtering model name used in the 'filters' row annotation, indicating the variant was filtered by this model during variant QC.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">filtering_model.</span>score_name</td><td style="border: 1px solid #000; padding: 5px;"><code>str</code></td><td style="border: 1px solid #000; padding: 5px;">Name of the score used in filtering.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">filtering_model.</span>snv_cutoff</td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ... }</code></td><td style="border: 1px solid #000; padding: 5px;">SNV filtering cutoff information.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">filtering_model.snv_cutoff.</span>bin</td><td style="border: 1px solid #000; padding: 5px;"><code>int32</code></td><td style="border: 1px solid #000; padding: 5px;">Filtering percentile cutoff for SNVs.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">filtering_model.snv_cutoff.</span>min_score</td><td style="border: 1px solid #000; padding: 5px;"><code>float64</code></td><td style="border: 1px solid #000; padding: 5px;">Minimum score at SNV filtering percentile cutoff.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">filtering_model.</span>indel_cutoff</td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ... }</code></td><td style="border: 1px solid #000; padding: 5px;">Indel filtering cutoff information.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">filtering_model.indel_cutoff.</span>bin</td><td style="border: 1px solid #000; padding: 5px;"><code>int32</code></td><td style="border: 1px solid #000; padding: 5px;">Filtering percentile cutoff for indels.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">filtering_model.indel_cutoff.</span>min_score</td><td style="border: 1px solid #000; padding: 5px;"><code>float64</code></td><td style="border: 1px solid #000; padding: 5px;">Minimum score at indel filtering percentile cutoff.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">filtering_model.</span>snv_training_variables</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;str&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Variant annotations used as features in the SNV filtering model.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">filtering_model.</span>indel_training_variables</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;str&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Variant annotations used as features in the indel filtering model.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><strong>inbreeding_coeff_cutoff</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>float64</code></td><td style="border: 1px solid #000; padding: 5px;">Inbreeding Coefficient threshold used to hard filter variants</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><strong>tool_versions</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ... }</code></td><td style="border: 1px solid #000; padding: 5px;">Versions of in silico predictors used in the callset.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">tool_versions</span>.cadd_version</td><td style="border: 1px solid #000; padding: 5px;"><code>str</code></td><td style="border: 1px solid #000; padding: 5px;">Combined Annotation Dependent Depletion (CADD) version.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">tool_versions</span>.revel_version</td><td style="border: 1px solid #000; padding: 5px;"><code>str</code></td><td style="border: 1px solid #000; padding: 5px;">Rare Exome Variant Ensemble Learner (REVEL) version.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">tool_versions</span>.spliceai_version</td><td style="border: 1px solid #000; padding: 5px;"><code>str</code></td><td style="border: 1px solid #000; padding: 5px;">SpliceAI version.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">tool_versions</span>.pangolin_version</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;str&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Pangolin version.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">tool_versions</span>.phylop_version</td><td style="border: 1px solid #000; padding: 5px;"><code>str</code></td><td style="border: 1px solid #000; padding: 5px;">phyloP version.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">tool_versions</span>.dbsnp_version</td><td style="border: 1px solid #000; padding: 5px;"><code>str</code></td><td style="border: 1px solid #000; padding: 5px;">dbSNP version.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">tool_versions</span>.sift_version</td><td style="border: 1px solid #000; padding: 5px;"><code>str</code></td><td style="border: 1px solid #000; padding: 5px;">Sorting Intolerant from Tolerant (SIFT) version.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">tool_versions</span>.polyphen_version</td><td style="border: 1px solid #000; padding: 5px;"><code>str</code></td><td style="border: 1px solid #000; padding: 5px;">Polymorphism Phenotyping v2 (Polyphen-v2) version.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><strong>vrs_versions</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ... }</code></td><td style="border: 1px solid #000; padding: 5px;">The Variant Representation Specification version that was used to compute IDs on the callset. Global and row VRS annotations are optional, but the global annotaions must be filled out if the row annotations are provided.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">vrs_versions</span>.vrs_schema_version</td><td style="border: 1px solid #000; padding: 5px;"><code>str</code></td><td style="border: 1px solid #000; padding: 5px;">The version of the VRS schema that is used to represent variants and compute identifiers. Must be <code>1.3.0</code>.</td><td style="border: 1px solid #000; padding: 5px;"><code>"1.3.0"</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">vrs_versions</span>.vrs_python_version</td><td style="border: 1px solid #000; padding: 5px;"><code>str</code></td><td style="border: 1px solid #000; padding: 5px;">The version of the vrs-python library that was used to compute IDs on the callset. Must be <code>0.8.4</code>.</td><td style="border: 1px solid #000; padding: 5px;"><code>"0.8.4"</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">vrs_versions</span>.seqrepo_version</td><td style="border: 1px solid #000; padding: 5px;"><code>str</code></td><td style="border: 1px solid #000; padding: 5px;">The version of the SeqRepo database that was used in VRS computations. Must be <code>2018-11-26</code>.</td><td style="border: 1px solid #000; padding: 5px;"><code>"2018-11-26"</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><strong>vep_globals</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ... }</code></td><td style="border: 1px solid #000; padding: 5px;">Information about VEP annotations.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">vep_globals</span>.vep_version</td><td style="border: 1px solid #000; padding: 5px;"><code>str</code></td><td style="border: 1px solid #000; padding: 5px;">VEP version that was run on the callset.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">vep_globals</span>.vep_help</td><td style="border: 1px solid #000; padding: 5px;"><code>str</code></td><td style="border: 1px solid #000; padding: 5px;">Output from vep --help.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">vep_globals</span>.vep_config</td><td style="border: 1px solid #000; padding: 5px;"><code>str</code></td><td style="border: 1px solid #000; padding: 5px;">VEP configuration to run VEP version with Hail. File created using command within VEP init shell script in https://github.com/broadinstitute/gnomad_methods/tree/main.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">vep_globals</span>.gencode_version</td><td style="border: 1px solid #000; padding: 5px;"><code>str</code></td><td style="border: 1px solid #000; padding: 5px;">GENCODE version used in VEP.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">vep_globals</span>.mane_select_version</td><td style="border: 1px solid #000; padding: 5px;"><code>str</code></td><td style="border: 1px solid #000; padding: 5px;">MANE select version used in VEP.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><strong>frequency_README</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>str</code></td><td style="border: 1px solid #000; padding: 5px;">Explanation of how to use the 'freq_index_dict' global annotation to extract frequencies from the 'freq' row annotation.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><strong>date</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>str</code></td><td style="border: 1px solid #000; padding: 5px;">Date Hail Table was created.</td><td style="border: 1px solid #000; padding: 5px;"><code>"2025-04-09"</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><strong>version</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>str</code></td><td style="border: 1px solid #000; padding: 5px;">Version of the file.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr></tbody></table>
<h1>Row Fields Specification</h1>
<p>This table provides a view of each row field. Every leaf field (i.e., non-struct field) is listed on its own row with the full annotation. The parent portion is shown in a lighter color to indicate context. Cell colors indicate the field necessity: <span style="background-color:#fff0f0; padding:2px 6px; border-radius:4px;">"Required" in red</span>, <span style="background-color:#f0faff; padding:2px 6px; border-radius:4px;">"Optional" in blue</span>, <span style="background-color:#dedbe4; padding:2px 6px; border-radius:4px;">"Not Needed" in grey</span>.</p>
<p>All histograms annotations must use the same bin edges as defined in their respective 'Description' columns.</p>
<p>Data must be supplied as a "split" dataset, where multiallelic variants are split so that information for each alternate allele is in a separate row.</p>
<table style="border-collapse: collapse; border: 1px solid #000; width: 100%;"><thead><tr style="background-color:#f8f8f8;"><th style="border: 1px solid #000; padding: 5px; text-align: left;">Field</th><th style="border: 1px solid #000; padding: 5px; text-align: left;">Type</th><th style="border: 1px solid #000; padding: 5px; text-align: left;">Description</th><th style="border: 1px solid #000; padding: 5px; text-align: left;">Example</th><th style="border: 1px solid #000; padding: 5px; text-align: left;">Field Necessity</th></tr></thead><tbody><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><strong>locus</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>locus&lt;GRCh38&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Variant locus. Contains contig and position information. Must be build GRCh38.</td><td style="border: 1px solid #000; padding: 5px;"><code>chr1:12345</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><strong>alleles</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;str&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Variant alleles (reference and alternate).</td><td style="border: 1px solid #000; padding: 5px;"><code>["A", "G"]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><strong>freq</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;struct {...}&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Array of allele frequency information (AC, AN, AF, homozygote count) for each frequency aggregation group corresponding to each frequency metadata group. The 'adj' value should always be the first value of the array, and the 'raw' value should be the second value.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">freq</span>.AC</td><td style="border: 1px solid #000; padding: 5px;"><code>int32</code></td><td style="border: 1px solid #000; padding: 5px;">Alternate allele count.</td><td style="border: 1px solid #000; padding: 5px;"><code>10</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">freq</span>.AF</td><td style="border: 1px solid #000; padding: 5px;"><code>float64</code></td><td style="border: 1px solid #000; padding: 5px;">Alternate allele frequency, (AC/AN).</td><td style="border: 1px solid #000; padding: 5px;"><code>0.1</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">freq</span>.AN</td><td style="border: 1px solid #000; padding: 5px;"><code>int32</code></td><td style="border: 1px solid #000; padding: 5px;">Total number of alleles.</td><td style="border: 1px solid #000; padding: 5px;"><code>100</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">freq</span>.homozygote_count</td><td style="border: 1px solid #000; padding: 5px;"><code>int32</code></td><td style="border: 1px solid #000; padding: 5px;">Count of homozygous alternate individuals for the alternate allele.</td><td style="border: 1px solid #000; padding: 5px;"><code>2</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><strong>grpmax</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>struct {...}</code></td><td style="border: 1px solid #000; padding: 5px;">Allele frequency information (AC, AN, AF, homozygote count) for the group with maximum allele frequency.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">grpmax</span>.AC</td><td style="border: 1px solid #000; padding: 5px;"><code>int32</code></td><td style="border: 1px solid #000; padding: 5px;">Alternate allele count in the group with the maximum allele frequency.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">grpmax</span>.AF</td><td style="border: 1px solid #000; padding: 5px;"><code>float64</code></td><td style="border: 1px solid #000; padding: 5px;">Maximum alternate allele frequency, (AC/AN), across all groups.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">grpmax</span>.AN</td><td style="border: 1px solid #000; padding: 5px;"><code>int32</code></td><td style="border: 1px solid #000; padding: 5px;">Total number of alleles in the group with the maximum allele frequency.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">grpmax</span>.homozygote_count</td><td style="border: 1px solid #000; padding: 5px;"><code>int32</code></td><td style="border: 1px solid #000; padding: 5px;">Count of homozygous individuals in the group with the maximum allele frequency.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">grpmax</span>.gen_anc</td><td style="border: 1px solid #000; padding: 5px;"><code>str</code></td><td style="border: 1px solid #000; padding: 5px;">Genetic ancestry corresponding to the maximum frequency group.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><strong>faf</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;struct {...}&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Filtering allele frequency.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">faf</span>.faf95</td><td style="border: 1px solid #000; padding: 5px;"><code>float64</code></td><td style="border: 1px solid #000; padding: 5px;">FAF95: Filtering allele frequency (using Poisson 95% CI).</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">faf</span>.faf99</td><td style="border: 1px solid #000; padding: 5px;"><code>float64</code></td><td style="border: 1px solid #000; padding: 5px;">FAF99: Filtering allele frequency (using Poisson 99% CI) frequency.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><strong>fafmax</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>struct {...}</code></td><td style="border: 1px solid #000; padding: 5px;">Information about the genetic ancestry group with the maximum filtering allele frequency.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">fafmax</span>.faf95_max</td><td style="border: 1px solid #000; padding: 5px;"><code>float64</code></td><td style="border: 1px solid #000; padding: 5px;">Maximum filtering allele frequency (using Poisson 95% CI).</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">fafmax</span>.faf95_max_gen_anc</td><td style="border: 1px solid #000; padding: 5px;"><code>str</code></td><td style="border: 1px solid #000; padding: 5px;">Genetic ancestry group with the maximum filtering allele frequency (95% CI).</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">fafmax</span>.faf99_max</td><td style="border: 1px solid #000; padding: 5px;"><code>float64</code></td><td style="border: 1px solid #000; padding: 5px;">Maximum filtering allele frequency (using Poisson 99% CI).</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">fafmax</span>.faf99_max_gen_anc</td><td style="border: 1px solid #000; padding: 5px;"><code>str</code></td><td style="border: 1px solid #000; padding: 5px;">Genetic ancestry group with the maximum filtering allele frequency (99% CI).</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><strong>a_index</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>int32</code></td><td style="border: 1px solid #000; padding: 5px;">The original index of this alternate allele in the multiallelic representation (1 is the first alternate allele or the only alternate allele in a biallelic variant).</td><td style="border: 1px solid #000; padding: 5px;"><code>1</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><strong>was_split</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>bool</code></td><td style="border: 1px solid #000; padding: 5px;">True if this variant was originally multiallelic, otherwise False.</td><td style="border: 1px solid #000; padding: 5px;"><code>False</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><strong>rsid</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>set&lt;str&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">dbSNP reference SNP identification (rsID) numbers.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><strong>filters</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>set&lt;str&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Variant filters; 'AC0': Allele count is zero after filtering out low-confidence genotypes (GQ &lt; 20; DP &lt; 10; or AB &lt; 0.2 for het calls), 'AS_VQSR': Failed allele-specific VQSR filtering thresholds, 'InbreedingCoeff': GATK InbreedingCoeff &lt; -0.3. An empty set in this field indicates that the variant passed all variant filters.</td><td style="border: 1px solid #000; padding: 5px;"><code>{"AC0","AS_VQSR"}</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><strong>info</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>struct {...}</code></td><td style="border: 1px solid #000; padding: 5px;">Struct containing typical GATK allele-specific (AS) info fields and additional variant QC fields.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info</span>.FS</td><td style="border: 1px solid #000; padding: 5px;"><code>float64</code></td><td style="border: 1px solid #000; padding: 5px;">Phred-scaled p-value of Fisher's exact test for strand bias.</td><td style="border: 1px solid #000; padding: 5px;"><code>7.30e+00</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info</span>.MQ</td><td style="border: 1px solid #000; padding: 5px;"><code>float64</code></td><td style="border: 1px solid #000; padding: 5px;">Root mean square of the mapping quality of reads across all samples.</td><td style="border: 1px solid #000; padding: 5px;"><code>3.48e+01</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info</span>.MQRankSum</td><td style="border: 1px solid #000; padding: 5px;"><code>float64</code></td><td style="border: 1px solid #000; padding: 5px;">Z-score from Wilcoxon rank sum test of alternate vs. reference read mapping qualities.</td><td style="border: 1px solid #000; padding: 5px;"><code>6.70e-02</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info.</span>MQRankSum_cdf</td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ...}</code></td><td style="border: 1px solid #000; padding: 5px;">CDF summary of overall MQRankSum values. The <code>_raw</code> parameter must be set to True to return an internal representation of the CDF approximation. This annotation is generated using Hail's function: <code>hl.agg.approx_cdf(x, k=XX, _raw=True)</code>. The <code>k</code> value must be set to 200 in order to merge with the gnomAD dataset. This method is non-deterministic: computing approx_cdf multiple times will give slightly different results each time. It is currently not possible to seed the aggregator.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info.MQRankSum_cdf.</span>levels</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;int32&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">List indicating how many items are stored at each compression level.</td><td style="border: 1px solid #000; padding: 5px;"><code>[0,3]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info.MQRankSum_cdf.</span>items</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;float64&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Ordered sample of values from the MQRankSum distribution.</td><td style="border: 1px solid #000; padding: 5px;"><code>[-9.38e-01,-2.27e+00,-1.34e+00]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info.MQRankSum_cdf.</span>_compaction_counts</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;int32&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Used internally to support downstream error estimation.</td><td style="border: 1px solid #000; padding: 5px;"><code>[0]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info</span>.QUALapprox</td><td style="border: 1px solid #000; padding: 5px;"><code>int64</code></td><td style="border: 1px solid #000; padding: 5px;">Sum of PL[0] values; used to approximate the QUAL score.</td><td style="border: 1px solid #000; padding: 5px;"><code>96</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info</span>.QD</td><td style="border: 1px solid #000; padding: 5px;"><code>float32</code></td><td style="border: 1px solid #000; padding: 5px;">Variant call confidence normalized by depth of sample reads supporting a variant.</td><td style="border: 1px solid #000; padding: 5px;"><code>2.74e+00</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info</span>.ReadPosRankSum</td><td style="border: 1px solid #000; padding: 5px;"><code>float64</code></td><td style="border: 1px solid #000; padding: 5px;">Z-score from Wilcoxon rank sum test of alternate vs. reference read position bias.</td><td style="border: 1px solid #000; padding: 5px;"><code>-1.07e+00</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info.</span>ReadPosRankSum_cdf</td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ...}</code></td><td style="border: 1px solid #000; padding: 5px;">CDF summary of overall ReadPosRankSum values. The <code>_raw</code> parameter must be set to True to return an internal representation of the CDF approximation. This annotation is generated using Hail's function: <code>hl.agg.approx_cdf(x, k=XX, _raw=True)</code>. The <code>k</code> value must be set to 200 in order to merge with the gnomAD dataset. This method is non-deterministic: computing approx_cdf multiple times will give slightly different results each time. It is currently not possible to seed the aggregator.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info.ReadPosRankSum_cdf.</span>levels</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;int32&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">List indicating how many items are stored at each compression level.</td><td style="border: 1px solid #000; padding: 5px;"><code>[0,3]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info.ReadPosRankSum_cdf.</span>items</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;float64&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Ordered sample of values from the ReadPosRankSum distribution.</td><td style="border: 1px solid #000; padding: 5px;"><code>[9.67e-01,-9.67e-01,9.67e-01]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info.ReadPosRankSum_cdf.</span>_compaction_counts</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;int32&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Used internally to support downstream error estimation.</td><td style="border: 1px solid #000; padding: 5px;"><code>[0]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info</span>.SB</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;int32&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Aggregate counts of strand depth across all non-homozygous-reference calls. The values are the depth of reference allele on forward strand, depth of the reference allele on reverse strand, depth of all alternate alleles on forward strand, depth of all alternate alleles on reverse strand.</td><td style="border: 1px solid #000; padding: 5px;"><code>[21,6,4,4]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info</span>.SOR</td><td style="border: 1px solid #000; padding: 5px;"><code>float64</code></td><td style="border: 1px solid #000; padding: 5px;">Strand bias estimated by the symmetric odds ratio test.</td><td style="border: 1px solid #000; padding: 5px;"><code>9.60e-02</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info</span>.VarDP</td><td style="border: 1px solid #000; padding: 5px;"><code>int32</code></td><td style="border: 1px solid #000; padding: 5px;">Depth over variant genotypes (does not include depth of reference samples).</td><td style="border: 1px solid #000; padding: 5px;"><code>35</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info</span>.AS_FS</td><td style="border: 1px solid #000; padding: 5px;"><code>float64</code></td><td style="border: 1px solid #000; padding: 5px;">Allele-specific phred-scaled p-value of Fisher's exact test for strand bias.</td><td style="border: 1px solid #000; padding: 5px;"><code>5.10e+00</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info</span>.AS_MQ</td><td style="border: 1px solid #000; padding: 5px;"><code>float64</code></td><td style="border: 1px solid #000; padding: 5px;">Allele-specific root mean square of the mapping quality of reads across all samples.</td><td style="border: 1px solid #000; padding: 5px;"><code>3.51e+01</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info</span>.AS_MQRankSum</td><td style="border: 1px solid #000; padding: 5px;"><code>float64</code></td><td style="border: 1px solid #000; padding: 5px;">Allele-specific z-score from Wilcoxon rank sum test of alternate vs. reference read mapping qualities.</td><td style="border: 1px solid #000; padding: 5px;"><code>-5.72e-01</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info.</span>AS_MQRankSum_cdf</td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ...}</code></td><td style="border: 1px solid #000; padding: 5px;">CDF summary of allele-specific MQRankSum values. The <code>_raw</code> parameter must be set to True to return an internal representation of the CDF approximation. This annotation is generated using Hail's function: <code>hl.agg.approx_cdf(x, k=XX, _raw=True)</code>. The <code>k</code> value must be set to 200 in order to merge with the gnomAD dataset. This method is non-deterministic: computing approx_cdf multiple times will give slightly different results each time. It is currently not possible to seed the aggregator.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info.AS_MQRankSum_cdf.</span>levels</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;int32&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">List indicating how many items are stored at each compression level.</td><td style="border: 1px solid #000; padding: 5px;"><code>[0,3]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info.AS_MQRankSum_cdf.</span>items</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;float64&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Ordered sample of values from the allele-specific MQRankSum distribution.</td><td style="border: 1px solid #000; padding: 5px;"><code>[-1.38e+00,4.31e-01,-9.67e-01]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info.AS_MQRankSum_cdf.</span>_compaction_counts</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;int32&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Used internally to support downstream error estimation for allele-specific MQRankSum CDF.</td><td style="border: 1px solid #000; padding: 5px;"><code>[0]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info</span>.AS_pab_max</td><td style="border: 1px solid #000; padding: 5px;"><code>float64</code></td><td style="border: 1px solid #000; padding: 5px;">Maximum p-value over callset for binomial test of observed allele balance for a heterozygous genotype, given expectation of 0.5.</td><td style="border: 1px solid #000; padding: 5px;"><code>6.87e-01</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info</span>.AS_QUALapprox</td><td style="border: 1px solid #000; padding: 5px;"><code>int64</code></td><td style="border: 1px solid #000; padding: 5px;">Allele-specific sum of PL[0] values; used to approximate the QUAL score.</td><td style="border: 1px solid #000; padding: 5px;"><code>77</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info</span>.AS_QD</td><td style="border: 1px solid #000; padding: 5px;"><code>float32</code></td><td style="border: 1px solid #000; padding: 5px;">Allele-specific variant call confidence normalized by depth of sample reads supporting a variant.</td><td style="border: 1px solid #000; padding: 5px;"><code>2.96e+00</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info</span>.AS_ReadPosRankSum</td><td style="border: 1px solid #000; padding: 5px;"><code>float64</code></td><td style="border: 1px solid #000; padding: 5px;">Allele-specific z-score from Wilcoxon rank sum test of alternate vs. reference read position bias.</td><td style="border: 1px solid #000; padding: 5px;"><code>-1.38e+00</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info.</span>AS_ReadPosRankSum_cdf</td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ...}</code></td><td style="border: 1px solid #000; padding: 5px;">CDF summary of allele-specific ReadPosRankSum values. The <code>_raw</code> parameter must be set to True to return an internal representation of the CDF approximation. This annotation is generated using Hail's function: <code>hl.agg.approx_cdf(x, k=XX, _raw=True)</code>. The <code>k</code> value must be set to 200 in order to merge with the gnomAD dataset. This method is non-deterministic: computing approx_cdf multiple times will give slightly different results each time. It is currently not possible to seed the aggregator.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info.AS_ReadPosRankSum_cdf.</span>levels</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;int32&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">List indicating how many items are stored at each compression level.</td><td style="border: 1px solid #000; padding: 5px;"><code>[0,1]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info.AS_ReadPosRankSum_cdf.</span>items</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;float64&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Ordered sample of values from the allele-specific ReadPosRankSum distribution.</td><td style="border: 1px solid #000; padding: 5px;"><code>[-1.78e-01]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info.AS_ReadPosRankSum_cdf.</span>_compaction_counts</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;int32&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Used internally to support downstream error estimation for allele-specific ReadPosRankSum CDF.</td><td style="border: 1px solid #000; padding: 5px;"><code>[0]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info</span>.AS_SB_TABLE</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;int32&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Allele-specific forward/reverse read counts for strand bias tests.</td><td style="border: 1px solid #000; padding: 5px;"><code>[21,6,3,3]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info</span>.AS_SOR</td><td style="border: 1px solid #000; padding: 5px;"><code>float64</code></td><td style="border: 1px solid #000; padding: 5px;">Allele-specific strand bias estimated by the symmetric odds ratio test.</td><td style="border: 1px solid #000; padding: 5px;"><code>9.64e-02</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info</span>.AS_VarDP</td><td style="border: 1px solid #000; padding: 5px;"><code>int32</code></td><td style="border: 1px solid #000; padding: 5px;">Allele-specific depth over variant genotypes (does not include depth of reference samples).</td><td style="border: 1px solid #000; padding: 5px;"><code>26</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info</span>.singleton</td><td style="border: 1px solid #000; padding: 5px;"><code>bool</code></td><td style="border: 1px solid #000; padding: 5px;">Variant is seen once in the callset.</td><td style="border: 1px solid #000; padding: 5px;"><code>True</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info</span>.transmitted_singleton</td><td style="border: 1px solid #000; padding: 5px;"><code>bool</code></td><td style="border: 1px solid #000; padding: 5px;">Variant was a callset-wide doubleton that was transmitted within a family from a parent to a child (i.e., a singleton amongst unrelated samples in cohort).</td><td style="border: 1px solid #000; padding: 5px;"><code>True</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info</span>.sibling_singleton</td><td style="border: 1px solid #000; padding: 5px;"><code>bool</code></td><td style="border: 1px solid #000; padding: 5px;">Variant was a callset-wide doubleton that was present only in two siblings (i.e., a singleton amongst unrelated samples in cohort).</td><td style="border: 1px solid #000; padding: 5px;"><code>True</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info</span>.omni</td><td style="border: 1px solid #000; padding: 5px;"><code>bool</code></td><td style="border: 1px solid #000; padding: 5px;">Variant is present on the Omni 2.5 genotyping array and found in 1000 Genomes data.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info</span>.mills</td><td style="border: 1px solid #000; padding: 5px;"><code>bool</code></td><td style="border: 1px solid #000; padding: 5px;">Indel is present in the Mills and Devine data.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info</span>.monoallelic</td><td style="border: 1px solid #000; padding: 5px;"><code>bool</code></td><td style="border: 1px solid #000; padding: 5px;">All samples are homozygous alternate for the variant.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info</span>.only_het</td><td style="border: 1px solid #000; padding: 5px;"><code>bool</code></td><td style="border: 1px solid #000; padding: 5px;">All samples are heterozygous for the variant (no homozygous reference or alternate genotype calls).</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info</span>.inbreeding_coeff</td><td style="border: 1px solid #000; padding: 5px;"><code>float64</code></td><td style="border: 1px solid #000; padding: 5px;">Inbreeding coefficient, the excess heterozygosity at a variant site, computed as <code>1 - (the number of heterozygous genotypes) / (the number of heterozygous genotypes expected under Hardy-Weinberg equilibrium)</code>.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info</span>.vrs</td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ... }</code></td><td style="border: 1px solid #000; padding: 5px;">Struct containing information related to the Global Alliance for Genomic Health (GA4GH) Variant Representation Specification (VRS) standard. VRS annotations must be created using the following tool versions: (vrs_schema_version=<code>2.1.3</code>, vrs_python_version=<code>0.8.4</code>, seqrepo_version=<code>2024-12-20</code>). <br/><br/> Will potentially update the VRS versions for v4?</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info.vrs</span>.VRS_Allele_IDs</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;str&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">The computed identifiers for the GA4GH VRS Alleles corresponding to the values in the alleles column.</td><td style="border: 1px solid #000; padding: 5px;"><code>["ga4gh:VA.oTAtTrgYxm81O9fu6Mrhfo1t3eHsgg4L","ga4gh:VA.Y283OnlLjyi1T1IT_JzvW255rC6YJsW6"]</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info.vrs</span>.VRS_Starts</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;int32&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Interresidue coordinates used as the location starts for the GA4GH VRS Alleles corresponding to the values in the alleles column.</td><td style="border: 1px solid #000; padding: 5px;"><code>[10030,10030]</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info.vrs</span>.VRS_Ends</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;int32&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Interresidue coordinates used as the location ends for the GA4GH VRS Alleles corresponding to the values in the alleles column.</td><td style="border: 1px solid #000; padding: 5px;"><code>[10031,10031]</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">info.vrs</span>.VRS_States</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;str&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">The literal sequence states used for the GA4GH VRS Alleles corresponding to the values in the alleles column.</td><td style="border: 1px solid #000; padding: 5px;"><code>["T","C"]</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><strong>vep</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ... }</code></td><td style="border: 1px solid #000; padding: 5px;">VEP annotations generated by the VEP tool (to be re-annotated).</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><strong>vqsr_results</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>struct</code></td><td style="border: 1px solid #000; padding: 5px;">VQSR related variant annotations.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">vqsr_results</span>.AS_VQSLOD</td><td style="border: 1px solid #000; padding: 5px;"><code>float64</code></td><td style="border: 1px solid #000; padding: 5px;">Allele-specific log-odds ratio of being a true variant versus being a false positive under the trained VQSR Gaussian mixture model.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">vqsr_results</span>.AS_culprit</td><td style="border: 1px solid #000; padding: 5px;"><code>str</code></td><td style="border: 1px solid #000; padding: 5px;">Allele-specific worst-performing annotation in the VQSR Gaussian mixture model.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">vqsr_results</span>.positive_train_site</td><td style="border: 1px solid #000; padding: 5px;"><code>bool</code></td><td style="border: 1px solid #000; padding: 5px;">Variant was used to build the positive training set of high-quality variants for VQSR.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">vqsr_results</span>.negative_train_site</td><td style="border: 1px solid #000; padding: 5px;"><code>bool</code></td><td style="border: 1px solid #000; padding: 5px;">Variant was used to build the negative training set of low-quality variants for VQSR.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><strong>region_flags</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>struct</code></td><td style="border: 1px solid #000; padding: 5px;">Struct containing flags about regions.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">region_flags</span>.non_par</td><td style="border: 1px solid #000; padding: 5px;"><code>bool</code></td><td style="border: 1px solid #000; padding: 5px;">Variant falls within a non-pseudoautosomal region.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">region_flags</span>.lcr</td><td style="border: 1px solid #000; padding: 5px;"><code>bool</code></td><td style="border: 1px solid #000; padding: 5px;">Variant falls within a low complexity region.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">region_flags</span>.segdup</td><td style="border: 1px solid #000; padding: 5px;"><code>bool</code></td><td style="border: 1px solid #000; padding: 5px;">Variant falls within a segmental duplication region.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><strong>allele_info</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>struct</code></td><td style="border: 1px solid #000; padding: 5px;">Allele information.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">allele_info</span>.allele_type</td><td style="border: 1px solid #000; padding: 5px;"><code>str</code></td><td style="border: 1px solid #000; padding: 5px;">Allele type (one of: 'snv', 'insertion', 'deletion', or 'mixed').</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">allele_info</span>.n_alt_alleles</td><td style="border: 1px solid #000; padding: 5px;"><code>int32</code></td><td style="border: 1px solid #000; padding: 5px;">Total number of alternate alleles observed at variant locus.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">allele_info</span>.variant_type</td><td style="border: 1px solid #000; padding: 5px;"><code>str</code></td><td style="border: 1px solid #000; padding: 5px;">Variant type (one of: 'snv', 'indel', 'multi-snv', 'multi-indel', or 'mixed').</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">allele_info</span>.was_mixed</td><td style="border: 1px solid #000; padding: 5px;"><code>bool</code></td><td style="border: 1px solid #000; padding: 5px;">Variant type was mixed.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><strong>histograms</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ... }</code></td><td style="border: 1px solid #000; padding: 5px;">Variant information histograms.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.</span>qual_hists</td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ... }</code></td><td style="border: 1px solid #000; padding: 5px;">Genotype quality metric histograms for high quality genotypes.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.qual_hists.</span>gq_hist_all</td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ... }</code></td><td style="border: 1px solid #000; padding: 5px;">Histogram for GQ calculated on high quality genotypes.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.qual_hists.gq_hist_all.</span>bin_edges</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;float64&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Bin edges for the GQ histogram calculated on high quality genotypes are: <code>0.0 | 5.0 | 10.0 | 20.0 | 25.0 | 30.0 | 35.0 | 40.0 | 45.0 | 50.0 | 55.0 | 60.0 | 65.0 | 70.0 | 75.0 | 80.0 | 85.0 | 90.0 | 95.0 | 100.0</code>.</td><td style="border: 1px solid #000; padding: 5px;"><code>[0.0, 5.0, 10.0, ...]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.qual_hists.gq_hist_all.</span>bin_freq</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;int64&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Bin frequencies for the GQ histogram calculated on high quality genotypes. The number of records found in each bin.</td><td style="border: 1px solid #000; padding: 5px;"><code>[50, 56, 101, ...]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.qual_hists.gq_hist_all.</span>n_smaller</td><td style="border: 1px solid #000; padding: 5px;"><code>int64</code></td><td style="border: 1px solid #000; padding: 5px;">Count of GQ values falling below the lowest histogram bin edge, calculated on high quality genotypes.</td><td style="border: 1px solid #000; padding: 5px;"><code>0</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.qual_hists.gq_hist_all.</span>n_larger</td><td style="border: 1px solid #000; padding: 5px;"><code>int64</code></td><td style="border: 1px solid #000; padding: 5px;">Count of GQ values falling above the highest histogram bin edge, calculated on high quality genotypes.</td><td style="border: 1px solid #000; padding: 5px;"><code>0</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.qual_hists.</span>dp_hist_all</td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ... }</code></td><td style="border: 1px solid #000; padding: 5px;">Histogram for DP calculated on high quality genotypes.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.qual_hists.dp_hist_all.</span>bin_edges</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;float64&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Bin edges for the DP histogram calculated on high quality genotypes are: <code>0.0 | 5.0 | 10.0 | 20.0 | 25.0 | 30.0 | 35.0 | 40.0 | 45.0 | 50.0 | 55.0 | 60.0 | 65.0 | 70.0 | 75.0 | 80.0 | 85.0 | 90.0 | 95.0 | 100.0</code>.</td><td style="border: 1px solid #000; padding: 5px;"><code>[0.0, 5.0, 10.0, ...]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.qual_hists.dp_hist_all.</span>bin_freq</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;int64&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Bin frequencies for the DP histogram calculated on high quality genotypes. The number of records found in each bin.</td><td style="border: 1px solid #000; padding: 5px;"><code>[2, 2, 16, ...]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.qual_hists.dp_hist_all.</span>n_smaller</td><td style="border: 1px solid #000; padding: 5px;"><code>int64</code></td><td style="border: 1px solid #000; padding: 5px;">Count of DP values falling below the lowest histogram bin edge, calculated on high quality genotypes.</td><td style="border: 1px solid #000; padding: 5px;"><code>0</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.qual_hists.dp_hist_all.</span>n_larger</td><td style="border: 1px solid #000; padding: 5px;"><code>int64</code></td><td style="border: 1px solid #000; padding: 5px;">Count of DP values falling above the highest histogram bin edge, calculated on high quality genotypes.</td><td style="border: 1px solid #000; padding: 5px;"><code>500</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.qual_hists.</span>gq_hist_alt</td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ... }</code></td><td style="border: 1px solid #000; padding: 5px;">Histogram for GQ in non-reference individuals calculated on high quality genotypes.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.qual_hists.gq_hist_alt.</span>bin_edges</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;float64&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Bin edges for the histogram of GQ in non-reference individuals calculated on high quality genotypes are: <code>0.0 | 5.0 | 10.0 | 20.0 | 25.0 | 30.0 | 35.0 | 40.0 | 45.0 | 50.0 | 55.0 | 60.0 | 65.0 | 70.0 | 75.0 | 80.0 | 85.0 | 90.0 | 95.0 | 100.0</code>.</td><td style="border: 1px solid #000; padding: 5px;"><code>[0.0, 5.0, 10.0, ...]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.qual_hists.gq_hist_alt.</span>bin_freq</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;int64&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Bin frequencies for the histogram of GQ in non-reference individuals calculated on high quality genotypes. The number of records found in each bin.</td><td style="border: 1px solid #000; padding: 5px;"><code>[0, 0 , 1, ...]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.qual_hists.gq_hist_alt.</span>n_smaller</td><td style="border: 1px solid #000; padding: 5px;"><code>int64</code></td><td style="border: 1px solid #000; padding: 5px;">Count of GQ values in non-reference individuals falling below the lowest histogram bin edge, calculated on high quality genotypes.</td><td style="border: 1px solid #000; padding: 5px;"><code>0</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.qual_hists.gq_hist_alt.</span>n_larger</td><td style="border: 1px solid #000; padding: 5px;"><code>int64</code></td><td style="border: 1px solid #000; padding: 5px;">Count of GQ values in non-reference individuals falling above the highest histogram bin edge, calculated on high quality genotypes.</td><td style="border: 1px solid #000; padding: 5px;"><code>0</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.qual_hists.</span>dp_hist_alt</td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ... }</code></td><td style="border: 1px solid #000; padding: 5px;">Histogram for DP in non-reference individuals calculated on high quality genotypes.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.qual_hists.dp_hist_alt.</span>bin_edges</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;float64&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Bin edges for the histogram of DP in non-reference individuals calculated on high quality genotypes are: <code>0.0 | 5.0 | 10.0 | 20.0 | 25.0 | 30.0 | 35.0 | 40.0 | 45.0 | 50.0 | 55.0 | 60.0 | 65.0 | 70.0 | 75.0 | 80.0 | 85.0 | 90.0 | 95.0 | 100.0</code>.</td><td style="border: 1px solid #000; padding: 5px;"><code>[0.0, 5.0, 10.0, ...]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.qual_hists.dp_hist_alt.</span>bin_freq</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;int64&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Bin frequencies for the histogram of DP in non-reference individuals calculated on high quality genotypes. The number of records found in each bin.</td><td style="border: 1px solid #000; padding: 5px;"><code>[1, 1, 10, ...]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.qual_hists.dp_hist_alt.</span>n_smaller</td><td style="border: 1px solid #000; padding: 5px;"><code>int64</code></td><td style="border: 1px solid #000; padding: 5px;">Count of DP values in non-reference individuals falling below the lowest histogram bin edge, calculated on high quality genotypes.</td><td style="border: 1px solid #000; padding: 5px;"><code>0</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.qual_hists.dp_hist_alt.</span>n_larger</td><td style="border: 1px solid #000; padding: 5px;"><code>int64</code></td><td style="border: 1px solid #000; padding: 5px;">Count of DP values in non-reference individuals falling above highest histogram bin edge, calculated on high quality genotypes.</td><td style="border: 1px solid #000; padding: 5px;"><code>10</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.qual_hists.</span>ab_hist_alt</td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ... }</code></td><td style="border: 1px solid #000; padding: 5px;">Histogram for AB in heterozygous individuals calculated on high quality genotypes.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.qual_hists.ab_hist_alt.</span>bin_edges</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;float64&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Bin edges for the histogram of AB in heterozygous individuals calculated on high quality genotypes are: <code>0.00 | 0.05 | 0.10 | 0.20 | 0.25 | 0.30 | 0.35 | 0.40 | 0.45 | 0.50 | 0.55 | 0.60 | 0.65 | 0.70 | 0.75 | 0.80 | 0.85 | 0.90 | 0.95 | 1.00</code>.</td><td style="border: 1px solid #000; padding: 5px;"><code>[0.00 , 0.05 , 0.10, ...]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.qual_hists.ab_hist_alt.</span>bin_freq</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;int64&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Bin frequencies for the histogram of AB in heterozygous individuals calculated on high quality genotypes. The number of records found in each bin.</td><td style="border: 1px solid #000; padding: 5px;"><code>[0 , 0 , 5, ...]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.qual_hists.ab_hist_alt.</span>n_smaller</td><td style="border: 1px solid #000; padding: 5px;"><code>int64</code></td><td style="border: 1px solid #000; padding: 5px;">Count of AB values in heterozygous individuals falling below the lowest histogram bin edge, calculated on high quality genotypes.</td><td style="border: 1px solid #000; padding: 5px;"><code>0</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.qual_hists.ab_hist_alt.</span>n_larger</td><td style="border: 1px solid #000; padding: 5px;"><code>int64</code></td><td style="border: 1px solid #000; padding: 5px;">Count of AB values in heterozygous individuals falling above the highest histogram bin edge, calculated on high quality genotypes.</td><td style="border: 1px solid #000; padding: 5px;"><code>0</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.</span>raw_qual_hists</td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ... }</code></td><td style="border: 1px solid #000; padding: 5px;">Genotype quality metric histograms for all genotypes as opposed to high quality genotypes.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.raw_qual_hists.</span>gq_hist_all</td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ... }</code></td><td style="border: 1px solid #000; padding: 5px;">Histogram for GQ calculated on all genotypes.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.raw_qual_hists.gq_hist_all.</span>bin_edges</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;float64&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Bin edges for the GQ histogram calculated on all genotypes are: <code>0.0 | 5.0 | 10.0 | 20.0 | 25.0 | 30.0 | 35.0 | 40.0 | 45.0 | 50.0 | 55.0 | 60.0 | 65.0 | 70.0 | 75.0 | 80.0 | 85.0 | 90.0 | 95.0 | 100.0</code>.</td><td style="border: 1px solid #000; padding: 5px;"><code>[0.0, 5.0, 10.0, ...]</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.raw_qual_hists.gq_hist_all.</span>bin_freq</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;int64&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Bin frequencies for the GQ histogram calculated on all genotypes. The number of records found in each bin.</td><td style="border: 1px solid #000; padding: 5px;"><code>[60, 76, 130, ...]</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.raw_qual_hists.gq_hist_all.</span>n_smaller</td><td style="border: 1px solid #000; padding: 5px;"><code>int64</code></td><td style="border: 1px solid #000; padding: 5px;">Count of GQ values falling below lowest histogram bin edge, for GQ calculated on all genotypes.</td><td style="border: 1px solid #000; padding: 5px;"><code>0</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.raw_qual_hists.gq_hist_all.</span>n_larger</td><td style="border: 1px solid #000; padding: 5px;"><code>int64</code></td><td style="border: 1px solid #000; padding: 5px;">Count of GQ values falling above highest histogram bin edge, for GQ calculated on all genotypes.</td><td style="border: 1px solid #000; padding: 5px;"><code>0</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.raw_qual_hists.</span>dp_hist_all</td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ... }</code></td><td style="border: 1px solid #000; padding: 5px;">Histogram for DP calculated on all genotypes.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.raw_qual_hists.dp_hist_all.</span>bin_edges</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;float64&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Bin edges for the DP histogram calculated on all genotypes are: <code>0.0 | 5.0 | 10.0 | 20.0 | 25.0 | 30.0 | 35.0 | 40.0 | 45.0 | 50.0 | 55.0 | 60.0 | 65.0 | 70.0 | 75.0 | 80.0 | 85.0 | 90.0 | 95.0 | 100.0</code>.</td><td style="border: 1px solid #000; padding: 5px;"><code>[0.0, 5.0, 10.0, ...]</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.raw_qual_hists.dp_hist_all.</span>bin_freq</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;int64&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Bin frequencies for the DP histogram calculated on all genotypes. The number of records found in each bin.</td><td style="border: 1px solid #000; padding: 5px;"><code>[4, 7, 20, ...]</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.raw_qual_hists.dp_hist_all.</span>n_smaller</td><td style="border: 1px solid #000; padding: 5px;"><code>int64</code></td><td style="border: 1px solid #000; padding: 5px;">Count of DP values falling below lowest histogram bin edge, for DP calculated on all genotypes.</td><td style="border: 1px solid #000; padding: 5px;"><code>0</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.raw_qual_hists.dp_hist_all.</span>n_larger</td><td style="border: 1px solid #000; padding: 5px;"><code>int64</code></td><td style="border: 1px solid #000; padding: 5px;">Count of DP values falling above highest histogram bin edge, for DP calculated on all genotypes.</td><td style="border: 1px solid #000; padding: 5px;"><code>600</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.raw_qual_hists.</span>gq_hist_alt</td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ... }</code></td><td style="border: 1px solid #000; padding: 5px;">Histogram for GQ in non-reference individuals calculated on all genotypes.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.raw_qual_hists.gq_hist_alt.</span>bin_edges</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;float64&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Bin edges for the histogram of GQ in non-reference individuals calculated on all genotypes are: <code>0.0 | 5.0 | 10.0 | 20.0 | 25.0 | 30.0 | 35.0 | 40.0 | 45.0 | 50.0 | 55.0 | 60.0 | 65.0 | 70.0 | 75.0 | 80.0 | 85.0 | 90.0 | 95.0 | 100.0</code>.</td><td style="border: 1px solid #000; padding: 5px;"><code>[0.0, 5.0, 10.0, ...]</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.raw_qual_hists.gq_hist_alt.</span>bin_freq</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;int64&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Bin frequencies for the histogram of GQ in non-reference individuals calculated on all genotypes. The number of records found in each bin.</td><td style="border: 1px solid #000; padding: 5px;"><code>[3, 4, 10, ...]</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.raw_qual_hists.gq_hist_alt.</span>n_smaller</td><td style="border: 1px solid #000; padding: 5px;"><code>int64</code></td><td style="border: 1px solid #000; padding: 5px;">Count of GQ values in non-reference individuals falling below lowest histogram bin edge, calculated on all genotypes.</td><td style="border: 1px solid #000; padding: 5px;"><code>0</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.raw_qual_hists.gq_hist_alt.</span>n_larger</td><td style="border: 1px solid #000; padding: 5px;"><code>int64</code></td><td style="border: 1px solid #000; padding: 5px;">Count of GQ values in non-reference individuals falling above highest histogram bin edge, calculated on all genotypes.</td><td style="border: 1px solid #000; padding: 5px;"><code>0</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.raw_qual_hists.</span>dp_hist_alt</td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ... }</code></td><td style="border: 1px solid #000; padding: 5px;">Histogram for DP in non-reference individuals calculated on all genotypes.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.raw_qual_hists.dp_hist_alt.</span>bin_edges</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;float64&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Bin edges for the histogram of DP in non-reference individuals calculated on all genotypes are: <code>0.0 | 5.0 | 10.0 | 20.0 | 25.0 | 30.0 | 35.0 | 40.0 | 45.0 | 50.0 | 55.0 | 60.0 | 65.0 | 70.0 | 75.0 | 80.0 | 85.0 | 90.0 | 95.0 | 100.0</code>.</td><td style="border: 1px solid #000; padding: 5px;"><code>[0.0, 5.0, 10.0, ...]</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.raw_qual_hists.dp_hist_alt.</span>bin_freq</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;int64&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Bin frequencies for the histogram of DP in non-reference individuals calculated on all genotypes. The number of records found in each bin.</td><td style="border: 1px solid #000; padding: 5px;"><code>[2, 2, 6, ...]</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.raw_qual_hists.dp_hist_alt.</span>n_smaller</td><td style="border: 1px solid #000; padding: 5px;"><code>int64</code></td><td style="border: 1px solid #000; padding: 5px;">Count of DP values in non-reference individuals falling below lowest histogram bin edge, calculated on all genotypes.</td><td style="border: 1px solid #000; padding: 5px;"><code>0</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.raw_qual_hists.dp_hist_alt.</span>n_larger</td><td style="border: 1px solid #000; padding: 5px;"><code>int64</code></td><td style="border: 1px solid #000; padding: 5px;">Count of DP values in non-reference individuals falling above highest histogram bin edge, calculated on all genotypes.</td><td style="border: 1px solid #000; padding: 5px;"><code>11</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.raw_qual_hists.</span>ab_hist_alt</td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ... }</code></td><td style="border: 1px solid #000; padding: 5px;">Histogram for AB in heterozygous individuals calculated on all genotypes.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.raw_qual_hists.ab_hist_alt.</span>bin_edges</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;float64&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Bin edges for the histogram of AB in heterozygous individuals calculated on all genotypes are: <code>0.00 | 0.05 | 0.10 | 0.20 | 0.25 | 0.30 | 0.35 | 0.40 | 0.45 | 0.50 | 0.55 | 0.60 | 0.65 | 0.70 | 0.75 | 0.80 | 0.85 | 0.90 | 0.95 | 1.00</code>.</td><td style="border: 1px solid #000; padding: 5px;"><code>[0.00 , 0.05 , 0.10, ...]</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.raw_qual_hists.ab_hist_alt.</span>bin_freq</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;int64&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Bin frequencies for the histogram of AB in heterozygous individuals calculated on all genotypes. The number of records found in each bin.</td><td style="border: 1px solid #000; padding: 5px;"><code>[0 , 0 , 6, ...]</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.raw_qual_hists.ab_hist_alt.</span>n_smaller</td><td style="border: 1px solid #000; padding: 5px;"><code>int64</code></td><td style="border: 1px solid #000; padding: 5px;">Count of AB values in heterozygous individuals falling below lowest histogram bin edge, calculated on all genotypes.</td><td style="border: 1px solid #000; padding: 5px;"><code>0</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#f0faff;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.raw_qual_hists.ab_hist_alt.</span>n_larger</td><td style="border: 1px solid #000; padding: 5px;"><code>int64</code></td><td style="border: 1px solid #000; padding: 5px;">Count of AB values in heterozygous individuals falling above highest histogram bin edge, calculated on all genotypes.</td><td style="border: 1px solid #000; padding: 5px;"><code>0</code></td><td style="border: 1px solid #000; padding: 5px;">Optional</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.</span>age_hists</td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ... }</code></td><td style="border: 1px solid #000; padding: 5px;">Histograms containing age information for high quality genotypes.  Cohorts for gnomAD vary in how they report age (some report the age at diagnosis, others report the age of last visit, etc), so the ages associated with the gnomAD data can be thought of as the last known age of the individual. Information on age is not available for all gnomAD samples. This field is required with the acknowledgement that age data may not be available for all samples. If no age is available for any samples, set all <code>bin_freq</code> values to <code>0</code>.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.age_hists.</span>age_hist_het</td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ... }</code></td><td style="border: 1px solid #000; padding: 5px;">Histogram for age in all heterozygous samples calculated on high quality genotypes.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.age_hists.age_hist_het.</span>bin_edges</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;float64&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Bin edges for the age histogram: <code>30.0 | 35.0 | 40.0 | 45.0 | 50.0 | 55.0 | 60.0 | 65.0 | 70.0 | 75.0 | 80.0</code>.</td><td style="border: 1px solid #000; padding: 5px;"><code>[30.0, 35.0,  40.0, ...]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.age_hists.age_hist_het.</span>bin_freq</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;int64&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Bin frequencies for the age histogram. This is the number of records found in each bin.</td><td style="border: 1px solid #000; padding: 5px;"><code>[0, 3,  4, ...]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.age_hists.age_hist_het.</span>n_smaller</td><td style="border: 1px solid #000; padding: 5px;"><code>int64</code></td><td style="border: 1px solid #000; padding: 5px;">Count of age values falling below lowest histogram bin edge.</td><td style="border: 1px solid #000; padding: 5px;"><code>1</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.age_hists.age_hist_het.</span>n_larger</td><td style="border: 1px solid #000; padding: 5px;"><code>int64</code></td><td style="border: 1px solid #000; padding: 5px;">Count of age values falling above highest histogram bin edge.</td><td style="border: 1px solid #000; padding: 5px;"><code>0</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.age_hists.</span>age_hist_hom</td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ... }</code></td><td style="border: 1px solid #000; padding: 5px;">Histogram for age in all homozygous samples calculated on high quality genotypes. If variant is in the pseudoautosomal regions of chrX or chrY, this histogram also includes age counts of hemizygous samples.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.age_hists.age_hist_hom.</span>bin_edges</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;float64&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Bin edges for the age histogram: <code>30.0 | 35.0 | 40.0 | 45.0 | 50.0 | 55.0 | 60.0 | 65.0 | 70.0 | 75.0 | 80.0</code>.</td><td style="border: 1px solid #000; padding: 5px;"><code>[30.0, 35.0,  40.0, ...]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.age_hists.age_hist_hom.</span>bin_freq</td><td style="border: 1px solid #000; padding: 5px;"><code>array&lt;int64&gt;</code></td><td style="border: 1px solid #000; padding: 5px;">Bin frequencies for the age histogram. This is the number of records found in each bin.</td><td style="border: 1px solid #000; padding: 5px;"><code>[0, 2,  2, ...]</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.age_hists.age_hist_hom.</span>n_smaller</td><td style="border: 1px solid #000; padding: 5px;"><code>int64</code></td><td style="border: 1px solid #000; padding: 5px;">Count of age values falling below lowest histogram bin edge.</td><td style="border: 1px solid #000; padding: 5px;"><code>0</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#fff0f0;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">histograms.age_hists.age_hist_hom.</span>n_larger</td><td style="border: 1px solid #000; padding: 5px;"><code>int64</code></td><td style="border: 1px solid #000; padding: 5px;">Count of age values falling above highest histogram bin edge.</td><td style="border: 1px solid #000; padding: 5px;"><code>0</code></td><td style="border: 1px solid #000; padding: 5px;">Required</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><strong>in_silico_predictors</strong></td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ... }</code></td><td style="border: 1px solid #000; padding: 5px;">Variant prediction annotations.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">in_silico_predictors.</span>cadd</td><td style="border: 1px solid #000; padding: 5px;"><code>struct { ... }</code></td><td style="border: 1px solid #000; padding: 5px;">Score used to predict deleteriousness of SNVs and indels.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">in_silico_predictors.cadd.</span>phred</td><td style="border: 1px solid #000; padding: 5px;"><code>float32</code></td><td style="border: 1px solid #000; padding: 5px;">CADD Phred-like scaled C-scores ranging from 1 to 99 based on the rank of each variant relative to all possible 8.6 billion substitutions in the human reference genome. Larger values indicate increased predicted deleteriousness.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">in_silico_predictors.cadd.</span>raw_score</td><td style="border: 1px solid #000; padding: 5px;"><code>float32</code></td><td style="border: 1px solid #000; padding: 5px;">Unscaled CADD scores indicating whether a variant is likely to be "observed" (negative values) vs "simulated" (positive values). Larger values indicate increased predicted deleteriousness.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">in_silico_predictors.</span>revel_max</td><td style="border: 1px solid #000; padding: 5px;"><code>float64</code></td><td style="border: 1px solid #000; padding: 5px;">An ensemble score for predicting the pathogenicity of missense variants (based on 13 other variant predictors).</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">in_silico_predictors.</span>spliceai_ds_max</td><td style="border: 1px solid #000; padding: 5px;"><code>float32</code></td><td style="border: 1px solid #000; padding: 5px;">Maximum delta score across 4 splicing consequences, which reflects the probability of the variant being splice-altering.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">in_silico_predictors.</span>pangolin_largest_ds</td><td style="border: 1px solid #000; padding: 5px;"><code>float64</code></td><td style="border: 1px solid #000; padding: 5px;">Largest delta score across 2 splicing consequences, which reflects the probability of the variant being splice-altering.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">in_silico_predictors.</span>phylop</td><td style="border: 1px solid #000; padding: 5px;"><code>float64</code></td><td style="border: 1px solid #000; padding: 5px;">Base-wise conservation score across the 241 placental mammals in the Zoonomia project. Score ranges from -20 to 9.28, and reflects acceleration (faster evolution than expected under neutral drift, assigned negative scores) as well as conservation (slower than expected evolution, assigned positive scores).</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">in_silico_predictors.</span>sift_max</td><td style="border: 1px solid #000; padding: 5px;"><code>float64</code></td><td style="border: 1px solid #000; padding: 5px;">Score reflecting the scaled probability of the amino acid substitution being tolerated, ranging from 0 to 1.</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr><tr style="background-color:#dedbe4;"><td style="border: 1px solid #000; padding: 5px;"><span style="color:#999;">in_silico_predictors.</span>polyphen_max</td><td style="border: 1px solid #000; padding: 5px;"><code>float64</code></td><td style="border: 1px solid #000; padding: 5px;">Score that predicts the possible impact of an amino acid substitution on the structure and function of a human protein, ranging from 0.0 (tolerated) to 1.0 (deleterious).</td><td style="border: 1px solid #000; padding: 5px;">—</td><td style="border: 1px solid #000; padding: 5px;">Not Needed</td></tr></tbody></table>