<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>gnomad_qc.v4.sample_qc.generate_qc_mt &mdash; gnomad_qc master documentation</title>
      <link rel="stylesheet" href="../../../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../../../_static/theme_overrides.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../../_static/jquery.js"></script>
        <script src="../../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../../../../" id="documentation_options" src="../../../../_static/documentation_options.js"></script>
        <script src="../../../../_static/doctools.js"></script>
        <script src="../../../../_static/sphinx_highlight.js"></script>
        <script src="https://unpkg.com/mermaid@10.2.0/dist/mermaid.min.js"></script>
        <script>mermaid.initialize({startOnLoad:true});</script>
    <script src="../../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../../index.html" class="icon icon-home">
            gnomad_qc
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../api_reference/index.html">API Reference</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/broadinstitute/gnomad_qc/tags">Change Log</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../../index.html">gnomad_qc</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">gnomad_qc.v4.sample_qc.generate_qc_mt</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for gnomad_qc.v4.sample_qc.generate_qc_mt</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Script to create a dense MatrixTable filtered to a diverse set of variants for relatedness/ancestry PCA using CCDG, gnomAD v3, and UK Biobank.</span>

<span class="sd">The variant set was determined using this script:</span>
<span class="sd">https://github.com/Nealelab/ccdg_qc/blob/master/scripts/pca_variant_filter.py</span>

<span class="sd">The full gnomAD v4 VariantDataset and v4 sparse MatrixTable are filtered to these sites</span>
<span class="sd">and then densified. The resulting MatrixTables are merged and additional allele</span>
<span class="sd">frequency and callrate filters are applied, then LD-pruning is performed.</span>

<span class="sd">Additionally, the script creates a joint v3 and v4 metadata file for use in</span>
<span class="sd">relatedness/ancestry PCA.</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Union</span>

<span class="kn">import</span> <span class="nn">hail</span> <span class="k">as</span> <span class="nn">hl</span>
<span class="kn">from</span> <span class="nn">gnomad.sample_qc.pipeline</span> <span class="kn">import</span> <span class="n">get_qc_mt</span>
<span class="kn">from</span> <span class="nn">gnomad.utils.annotations</span> <span class="kn">import</span> <span class="n">annotate_adj</span>
<span class="kn">from</span> <span class="nn">gnomad.utils.slack</span> <span class="kn">import</span> <span class="n">slack_notifications</span>

<span class="kn">from</span> <span class="nn">gnomad_qc.slack_creds</span> <span class="kn">import</span> <span class="n">slack_token</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v3.resources.basics</span> <span class="kn">import</span> <span class="n">get_gnomad_v3_mt</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v3.resources.meta</span> <span class="kn">import</span> <span class="n">meta</span> <span class="k">as</span> <span class="n">v3_meta</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v4.resources.basics</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">get_checkpoint_path</span><span class="p">,</span>
    <span class="n">get_gnomad_v4_vds</span><span class="p">,</span>
    <span class="n">get_logging_path</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v4.resources.meta</span> <span class="kn">import</span> <span class="n">project_meta</span> <span class="k">as</span> <span class="n">v4_meta</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v4.resources.sample_qc</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">get_joint_qc</span><span class="p">,</span>
    <span class="n">get_predetermined_qc</span><span class="p">,</span>
    <span class="n">hard_filtered_samples</span><span class="p">,</span>
    <span class="n">joint_qc_meta</span><span class="p">,</span>
    <span class="n">predetermined_qc_sites</span><span class="p">,</span>
    <span class="n">sample_chr20_mean_dp</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="nb">format</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%(levelname)s</span><span class="s2"> (</span><span class="si">%(name)s</span><span class="s2"> </span><span class="si">%(lineno)s</span><span class="s2">): </span><span class="si">%(message)s</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s2">&quot;generate_qc_mt&quot;</span><span class="p">)</span>
<span class="n">logger</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>


<div class="viewcode-block" id="create_filtered_dense_mt"><a class="viewcode-back" href="../../../../api_reference/v4/sample_qc/generate_qc_mt.html#gnomad_qc.v4.sample_qc.generate_qc_mt.create_filtered_dense_mt">[docs]</a><span class="k">def</span> <span class="nf">create_filtered_dense_mt</span><span class="p">(</span>
    <span class="n">mtds</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">VariantDataset</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">],</span>
    <span class="n">split</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Filter a sparse MatrixTable or VariantDataset to a set of predetermined QC sites and return a dense MatrixTable.</span>

<span class="sd">    :param mtds: Input MatrixTable or VariantDataset.</span>
<span class="sd">    :param split: Whether `mtds` should have multi-allelics split before filtering</span>
<span class="sd">        variants.</span>
<span class="sd">    :return: Filtered and densified MatrixTable.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">is_vds</span> <span class="o">=</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">mtds</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">VariantDataset</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">is_vds</span><span class="p">:</span>
        <span class="n">vds</span> <span class="o">=</span> <span class="n">mtds</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Converting MatrixTable to VariantDataset...&quot;</span><span class="p">)</span>
        <span class="n">mtds</span> <span class="o">=</span> <span class="n">mtds</span><span class="o">.</span><span class="n">select_entries</span><span class="p">(</span><span class="s2">&quot;END&quot;</span><span class="p">,</span> <span class="s2">&quot;LA&quot;</span><span class="p">,</span> <span class="s2">&quot;LGT&quot;</span><span class="p">,</span> <span class="s2">&quot;GQ&quot;</span><span class="p">,</span> <span class="s2">&quot;DP&quot;</span><span class="p">,</span> <span class="s2">&quot;LAD&quot;</span><span class="p">)</span>
        <span class="n">vds</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">VariantDataset</span><span class="o">.</span><span class="n">from_merged_representation</span><span class="p">(</span><span class="n">mtds</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">split</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Performing split-multi...&quot;</span><span class="p">)</span>
        <span class="n">vds</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">split_multi</span><span class="p">(</span><span class="n">vds</span><span class="p">,</span> <span class="n">filter_changed_loci</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Filtering variants to predetermined QC variants...&quot;</span><span class="p">)</span>
    <span class="n">vds</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">filter_variants</span><span class="p">(</span><span class="n">vds</span><span class="p">,</span> <span class="n">predetermined_qc_sites</span><span class="o">.</span><span class="n">ht</span><span class="p">())</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Densifying data...&quot;</span><span class="p">)</span>
    <span class="n">mt</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">to_dense_mt</span><span class="p">(</span><span class="n">vds</span><span class="p">)</span>

    <span class="n">mt</span> <span class="o">=</span> <span class="n">mt</span><span class="o">.</span><span class="n">select_entries</span><span class="p">(</span><span class="s2">&quot;GT&quot;</span><span class="p">,</span> <span class="s2">&quot;GQ&quot;</span><span class="p">,</span> <span class="s2">&quot;DP&quot;</span><span class="p">,</span> <span class="s2">&quot;AD&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">mt</span></div>


<div class="viewcode-block" id="generate_qc_mt"><a class="viewcode-back" href="../../../../api_reference/v4/sample_qc/generate_qc_mt.html#gnomad_qc.v4.sample_qc.generate_qc_mt.generate_qc_mt">[docs]</a><span class="k">def</span> <span class="nf">generate_qc_mt</span><span class="p">(</span>
    <span class="n">v3_mt</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">,</span>
    <span class="n">v4_mt</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">,</span>
    <span class="n">bi_allelic_only</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">min_af</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.0001</span><span class="p">,</span>
    <span class="n">min_callrate</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.99</span><span class="p">,</span>
    <span class="n">min_inbreeding_coeff_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="o">-</span><span class="mf">0.8</span><span class="p">,</span>
    <span class="n">ld_r2</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span>
    <span class="n">n_partitions</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span>
    <span class="n">block_size</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2048</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Generate combined gnomAD v3 and v4 QC MatrixTable for use in relatedness and ancestry inference.</span>

<span class="sd">    :param v3_mt: Dense gnomAD v3 MatrixTable filtered to predetermined sites.</span>
<span class="sd">    :param v4_mt: Dense gnomAD v4 MatrixTable filtered to predetermined sites.</span>
<span class="sd">    :param bi_allelic_only: Whether to filter to bi-allelic variants.</span>
<span class="sd">    :param min_af: Minimum variant allele frequency to retain variant in QC MatrixTable.</span>
<span class="sd">    :param min_callrate: Minimum variant callrate to retain variant in QC MatrixTable.</span>
<span class="sd">    :param min_inbreeding_coeff_threshold: Minimum site inbreeding coefficient to</span>
<span class="sd">        retain variant in QC MatrixTable.</span>
<span class="sd">    :param ld_r2: LD-pruning cutoff.</span>
<span class="sd">    :param n_partitions: Number of partitions to repartition the MT to before LD</span>
<span class="sd">        pruning.</span>
<span class="sd">    :param block_size: Block size parameter to use for LD pruning.</span>
<span class="sd">    :return: MatrixTable of sites that pass QC filters.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">v3_count</span> <span class="o">=</span> <span class="n">v3_mt</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
    <span class="n">v4_count</span> <span class="o">=</span> <span class="n">v4_mt</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Number of (variants, samples) in the v3.1 MatrixTable: </span><span class="si">%s</span><span class="s2">...&quot;</span><span class="p">,</span> <span class="n">v3_count</span>
    <span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Number of (variants, samples) in the v4 MatrixTable: </span><span class="si">%s</span><span class="s2">...&quot;</span><span class="p">,</span> <span class="n">v4_count</span><span class="p">)</span>
    <span class="c1"># Remove v4 hard filtered samples.</span>
    <span class="n">v4_mt</span> <span class="o">=</span> <span class="n">v4_mt</span><span class="o">.</span><span class="n">anti_join_cols</span><span class="p">(</span><span class="n">hard_filtered_samples</span><span class="o">.</span><span class="n">ht</span><span class="p">())</span>

    <span class="n">samples_in_both</span> <span class="o">=</span> <span class="n">v4_mt</span><span class="o">.</span><span class="n">cols</span><span class="p">()</span><span class="o">.</span><span class="n">semi_join</span><span class="p">(</span><span class="n">v3_mt</span><span class="o">.</span><span class="n">cols</span><span class="p">())</span>
    <span class="n">n_samples_in_both</span> <span class="o">=</span> <span class="n">samples_in_both</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">n_samples_in_both</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;The following samples are found in both MatrixTables:&quot;</span><span class="p">)</span>
        <span class="n">samples_in_both</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">n_samples_in_both</span><span class="p">)</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Overlapping sample names were found in the MatrixTables!&quot;</span><span class="p">)</span>

    <span class="n">n_variants_in_both</span> <span class="o">=</span> <span class="n">v4_mt</span><span class="o">.</span><span class="n">rows</span><span class="p">()</span><span class="o">.</span><span class="n">semi_join</span><span class="p">(</span><span class="n">v3_mt</span><span class="o">.</span><span class="n">rows</span><span class="p">())</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Number of variants found in both MatrixTables: </span><span class="si">%d</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="s2">&quot;Number of variants found in only the v3.1 MatrixTable: </span><span class="si">%d</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="s2">&quot;Number of variants found in only the v4 MatrixTable: </span><span class="si">%d</span><span class="s2">&quot;</span><span class="p">,</span>
        <span class="n">n_variants_in_both</span><span class="p">,</span>
        <span class="n">v3_count</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="n">n_variants_in_both</span><span class="p">,</span>
        <span class="n">v4_count</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="n">n_variants_in_both</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Performing a union of the v3.1 and v4 predetermined QC site MatrixTable&quot;</span>
        <span class="s2">&quot; columns...&quot;</span>
    <span class="p">)</span>
    <span class="n">mt</span> <span class="o">=</span> <span class="n">v3_mt</span><span class="o">.</span><span class="n">union_cols</span><span class="p">(</span><span class="n">v4_mt</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Annotating with adj and filtering to pre LD-pruned QC sites...&quot;</span><span class="p">)</span>
    <span class="n">mt</span> <span class="o">=</span> <span class="n">annotate_adj</span><span class="p">(</span><span class="n">mt</span><span class="p">)</span>

    <span class="n">cp_path</span> <span class="o">=</span> <span class="n">get_checkpoint_path</span><span class="p">(</span><span class="s2">&quot;combined_pre_ld_prune_qc_mt&quot;</span><span class="p">,</span> <span class="n">mt</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">mt</span> <span class="o">=</span> <span class="n">get_qc_mt</span><span class="p">(</span>
        <span class="n">mt</span><span class="p">,</span>
        <span class="n">bi_allelic_only</span><span class="o">=</span><span class="n">bi_allelic_only</span><span class="p">,</span>
        <span class="n">min_af</span><span class="o">=</span><span class="n">min_af</span><span class="p">,</span>
        <span class="n">min_callrate</span><span class="o">=</span><span class="n">min_callrate</span><span class="p">,</span>
        <span class="n">min_inbreeding_coeff_threshold</span><span class="o">=</span><span class="n">min_inbreeding_coeff_threshold</span><span class="p">,</span>
        <span class="n">min_hardy_weinberg_threshold</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="c1"># Already filtered from the initial set of QC variants.</span>
        <span class="n">apply_hard_filters</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">ld_r2</span><span class="o">=</span><span class="n">ld_r2</span><span class="p">,</span>
        <span class="n">checkpoint_path</span><span class="o">=</span><span class="n">cp_path</span><span class="p">,</span>
        <span class="n">filter_lcr</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>  <span class="c1"># Already filtered from the initial set of QC variants.</span>
        <span class="n">filter_decoy</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>  <span class="c1"># Doesn&#39;t exist for hg38.</span>
        <span class="n">filter_segdup</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>  <span class="c1"># Already filtered from the initial set of QC variants.</span>
        <span class="n">n_partitions</span><span class="o">=</span><span class="n">n_partitions</span><span class="p">,</span>
        <span class="n">block_size</span><span class="o">=</span><span class="n">block_size</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Number of pre LD-pruned QC sites in gnomAD v3 + v4: </span><span class="si">%d</span><span class="s2">...&quot;</span><span class="p">,</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">read_matrix_table</span><span class="p">(</span><span class="n">cp_path</span><span class="p">)</span><span class="o">.</span><span class="n">count_rows</span><span class="p">(),</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">mt</span></div>


<div class="viewcode-block" id="generate_qc_meta_ht"><a class="viewcode-back" href="../../../../api_reference/v4/sample_qc/generate_qc_mt.html#gnomad_qc.v4.sample_qc.generate_qc_mt.generate_qc_meta_ht">[docs]</a><span class="k">def</span> <span class="nf">generate_qc_meta_ht</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Combine v3 and v4 sample metadata into a single Table for relatedness and population inference.</span>

<span class="sd">    :return: Table with v3 and v4 sample metadata</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">v3_meta_ht</span> <span class="o">=</span> <span class="n">v3_meta</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>
    <span class="n">v3_meta_ht</span> <span class="o">=</span> <span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
        <span class="n">v2_meta</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
            <span class="n">v2_known_pop</span><span class="o">=</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">v2_known_pop</span><span class="p">,</span>
            <span class="n">v2_known_subpop</span><span class="o">=</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">v2_known_subpop</span><span class="p">,</span>
            <span class="n">v2_pop</span><span class="o">=</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">v2_pop</span><span class="p">,</span>
            <span class="n">v2_subpop</span><span class="o">=</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">v2_subpop</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="n">v3_meta</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
            <span class="c1"># Annotate research project for eas samples from Osaka</span>
            <span class="n">v3_research_project</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span>
                <span class="p">(</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">project_pop</span> <span class="o">==</span> <span class="s2">&quot;eas&quot;</span><span class="p">)</span>
                <span class="o">&amp;</span> <span class="p">(</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">s</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;JPNW&quot;</span><span class="p">)),</span>
                <span class="s2">&quot;osaka&quot;</span><span class="p">,</span>
                <span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">research_project</span><span class="p">,</span>
            <span class="p">),</span>
            <span class="n">v3_project_ancestry</span><span class="o">=</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">project_ancestry</span><span class="p">,</span>
            <span class="n">v3_project_pop</span><span class="o">=</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">project_pop</span><span class="p">,</span>
            <span class="n">v3_project_subpop</span><span class="o">=</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">project_subpop</span><span class="p">,</span>
            <span class="n">v3_subpop_description</span><span class="o">=</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">subpop_description</span><span class="p">,</span>
            <span class="n">v3_subsets</span><span class="o">=</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">subsets</span><span class="p">,</span>
            <span class="n">v3_population_inference</span><span class="o">=</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">population_inference</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
                <span class="s2">&quot;training_pop&quot;</span><span class="p">,</span> <span class="s2">&quot;pca_scores&quot;</span><span class="p">,</span> <span class="s2">&quot;pop&quot;</span>
            <span class="p">),</span>
            <span class="n">v3_relatedness_relationships</span><span class="o">=</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">relatedness_inference</span><span class="o">.</span><span class="n">relationships</span><span class="p">,</span>
            <span class="n">v3_release_related</span><span class="o">=</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">sample_filters</span><span class="o">.</span><span class="n">release_related</span><span class="p">,</span>
            <span class="n">v3_qc_metrics_outlier_filters</span><span class="o">=</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">sample_filters</span><span class="o">.</span><span class="n">qc_metrics_filters</span><span class="p">,</span>
            <span class="n">v3_exclude</span><span class="o">=</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">exclude</span><span class="p">,</span>
            <span class="n">v3_exclude_reason</span><span class="o">=</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">exclude_reason</span><span class="p">,</span>
            <span class="n">v3_high_quality</span><span class="o">=</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">high_quality</span><span class="p">,</span>
            <span class="n">v3_release</span><span class="o">=</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">release</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="n">chr20_mean_dp</span><span class="o">=</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">sex_imputation</span><span class="o">.</span><span class="n">chr20_mean_dp</span><span class="p">,</span>
        <span class="n">releasable</span><span class="o">=</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">releasable</span><span class="p">,</span>
        <span class="n">broad_external</span><span class="o">=</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">broad_external</span><span class="p">,</span>
        <span class="n">hard_filters</span><span class="o">=</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">sample_filters</span><span class="o">.</span><span class="n">hard_filters</span><span class="p">,</span>
        <span class="n">hard_filtered</span><span class="o">=</span><span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">sample_filters</span><span class="o">.</span><span class="n">hard_filtered</span><span class="p">,</span>
        <span class="n">data_type</span><span class="o">=</span><span class="s2">&quot;genomes&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">v4_meta_ht</span> <span class="o">=</span> <span class="n">v4_meta</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>
    <span class="n">chr20_mean_dp_ht</span> <span class="o">=</span> <span class="n">sample_chr20_mean_dp</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>
    <span class="n">hardfilter_ht</span> <span class="o">=</span> <span class="n">hard_filtered_samples</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>

    <span class="n">v4_meta_ht</span> <span class="o">=</span> <span class="n">v4_meta_ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
        <span class="n">v2_meta</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
            <span class="n">v2_known_pop</span><span class="o">=</span><span class="n">v4_meta_ht</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">v2_meta</span><span class="o">.</span><span class="n">v2_known_pop</span><span class="p">,</span>
            <span class="n">v2_known_subpop</span><span class="o">=</span><span class="n">v4_meta_ht</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">v2_meta</span><span class="o">.</span><span class="n">v2_known_subpop</span><span class="p">,</span>
            <span class="n">v2_pop</span><span class="o">=</span><span class="n">v4_meta_ht</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">v2_meta</span><span class="o">.</span><span class="n">v2_pop</span><span class="p">,</span>
            <span class="n">v2_subpop</span><span class="o">=</span><span class="n">v4_meta_ht</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">v2_meta</span><span class="o">.</span><span class="n">v2_subpop</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="n">v4_race_ethnicity</span><span class="o">=</span><span class="n">v4_meta_ht</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">race_ethnicity</span><span class="p">,</span>
        <span class="n">chr20_mean_dp</span><span class="o">=</span><span class="n">chr20_mean_dp_ht</span><span class="p">[</span><span class="n">v4_meta_ht</span><span class="o">.</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">chr20_mean_dp</span><span class="p">,</span>
        <span class="n">releasable</span><span class="o">=</span><span class="n">v4_meta_ht</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">releasable</span><span class="p">,</span>
        <span class="n">broad_external</span><span class="o">=</span><span class="n">v4_meta_ht</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">broad_external</span><span class="p">,</span>
        <span class="n">hard_filters</span><span class="o">=</span><span class="n">hardfilter_ht</span><span class="p">[</span><span class="n">v4_meta_ht</span><span class="o">.</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">hard_filters</span><span class="p">,</span>
        <span class="n">hard_filtered</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">hardfilter_ht</span><span class="p">[</span><span class="n">v4_meta_ht</span><span class="o">.</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">hard_filters</span><span class="p">),</span>
        <span class="n">data_type</span><span class="o">=</span><span class="s2">&quot;exomes&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">samples_in_both</span> <span class="o">=</span> <span class="n">v4_meta_ht</span><span class="o">.</span><span class="n">semi_join</span><span class="p">(</span><span class="n">v3_meta_ht</span><span class="p">)</span>
    <span class="n">n_samples_in_both</span> <span class="o">=</span> <span class="n">samples_in_both</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">n_samples_in_both</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;The following samples are found in both metadata Tables:&quot;</span><span class="p">)</span>
        <span class="n">samples_in_both</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">n_samples_in_both</span><span class="p">)</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Overlapping sample names were found in the metadata Tables!&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">v3_meta_ht</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">v4_meta_ht</span><span class="p">,</span> <span class="n">unify</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div>


<div class="viewcode-block" id="main"><a class="viewcode-back" href="../../../../api_reference/v4/sample_qc/generate_qc_mt.html#gnomad_qc.v4.sample_qc.generate_qc_mt.main">[docs]</a><span class="k">def</span> <span class="nf">main</span><span class="p">(</span><span class="n">args</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create a dense MT of a diverse set of variants for relatedness/ancestry PCA.&quot;&quot;&quot;</span>
    <span class="n">hl</span><span class="o">.</span><span class="n">init</span><span class="p">(</span>
        <span class="n">log</span><span class="o">=</span><span class="s2">&quot;/generate_qc_mt.log&quot;</span><span class="p">,</span>
        <span class="n">default_reference</span><span class="o">=</span><span class="s2">&quot;GRCh38&quot;</span><span class="p">,</span>
        <span class="n">tmp_dir</span><span class="o">=</span><span class="s2">&quot;gs://gnomad-tmp-4day&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="c1"># NOTE: remove this flag when the new shuffle method is the default. This is necessary for hail 0.2.97. # noqa</span>
    <span class="n">hl</span><span class="o">.</span><span class="n">_set_flags</span><span class="p">(</span><span class="n">use_new_shuffle</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">)</span>

    <span class="n">overwrite</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">overwrite</span>
    <span class="n">ld_r2</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">ld_r2</span>
    <span class="n">test</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">test</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">create_v3_filtered_dense_mt</span><span class="p">:</span>
            <span class="c1"># NOTE: This command removes hard filtered samples.</span>
            <span class="n">mt</span> <span class="o">=</span> <span class="n">get_gnomad_v3_mt</span><span class="p">(</span><span class="n">key_by_locus_and_alleles</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">)</span>
            <span class="n">mt</span> <span class="o">=</span> <span class="n">create_filtered_dense_mt</span><span class="p">(</span><span class="n">mt</span><span class="p">,</span> <span class="n">split</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">mt</span> <span class="o">=</span> <span class="n">mt</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span>
                <span class="n">get_predetermined_qc</span><span class="p">(</span><span class="n">version</span><span class="o">=</span><span class="s2">&quot;3.1&quot;</span><span class="p">,</span> <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">)</span><span class="o">.</span><span class="n">path</span><span class="p">,</span>
                <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="s2">&quot;Number of predetermined QC variants found in the gnomAD v3&quot;</span>
                <span class="s2">&quot; MatrixTable: </span><span class="si">%d</span><span class="s2">...&quot;</span><span class="p">,</span>
                <span class="n">mt</span><span class="o">.</span><span class="n">count_rows</span><span class="p">(),</span>
            <span class="p">)</span>

        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">create_v4_filtered_dense_mt</span><span class="p">:</span>
            <span class="c1"># NOTE: This subset dense MatrixTable was created before the final hard filtering was determined. # noqa</span>
            <span class="c1"># Hard filtering is performed in `generate_qc_mt` before applying variant</span>
            <span class="c1"># filters.</span>
            <span class="n">vds</span> <span class="o">=</span> <span class="n">get_gnomad_v4_vds</span><span class="p">(</span>
                <span class="n">split</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">remove_hard_filtered_samples</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">test</span><span class="o">=</span><span class="n">test</span>
            <span class="p">)</span>
            <span class="n">mt</span> <span class="o">=</span> <span class="n">create_filtered_dense_mt</span><span class="p">(</span><span class="n">vds</span><span class="p">)</span>
            <span class="n">mt</span> <span class="o">=</span> <span class="n">mt</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span>
                <span class="n">get_predetermined_qc</span><span class="p">(</span><span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">)</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span>
            <span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="s2">&quot;Number of predetermined QC variants found in the gnomAD v4 VDS: </span><span class="si">%d</span><span class="s2">...&quot;</span><span class="p">,</span>
                <span class="n">mt</span><span class="o">.</span><span class="n">count_rows</span><span class="p">(),</span>
            <span class="p">)</span>

        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">generate_qc_mt</span><span class="p">:</span>
            <span class="n">v3_mt</span> <span class="o">=</span> <span class="n">get_predetermined_qc</span><span class="p">(</span><span class="n">version</span><span class="o">=</span><span class="s2">&quot;3.1&quot;</span><span class="p">,</span> <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">)</span><span class="o">.</span><span class="n">mt</span><span class="p">()</span>
            <span class="n">v4_mt</span> <span class="o">=</span> <span class="n">get_predetermined_qc</span><span class="p">(</span><span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">)</span><span class="o">.</span><span class="n">mt</span><span class="p">()</span>
            <span class="n">mt</span> <span class="o">=</span> <span class="n">generate_qc_mt</span><span class="p">(</span>
                <span class="n">v3_mt</span><span class="p">,</span>
                <span class="n">v4_mt</span><span class="p">,</span>
                <span class="n">bi_allelic_only</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">bi_allelic_only</span><span class="p">,</span>
                <span class="n">min_af</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">min_af</span><span class="p">,</span>
                <span class="n">min_callrate</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">min_callrate</span><span class="p">,</span>
                <span class="n">min_inbreeding_coeff_threshold</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">min_inbreeding_coeff_threshold</span><span class="p">,</span>
                <span class="n">ld_r2</span><span class="o">=</span><span class="n">ld_r2</span><span class="p">,</span>
                <span class="n">n_partitions</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">n_partitions</span><span class="p">,</span>
                <span class="n">block_size</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">block_size</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">mt</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">get_joint_qc</span><span class="p">(</span><span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">)</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">generate_qc_meta</span><span class="p">:</span>
            <span class="n">generate_qc_meta_ht</span><span class="p">()</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">joint_qc_meta</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">)</span>

    <span class="k">finally</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Copying hail log to logging bucket...&quot;</span><span class="p">)</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">copy_log</span><span class="p">(</span><span class="n">get_logging_path</span><span class="p">(</span><span class="s2">&quot;generate_qc_mt&quot;</span><span class="p">))</span></div>


<span class="k">def</span> <span class="nf">get_script_argument_parser</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get script argument parser.&quot;&quot;&quot;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">()</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--test&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Runs a test on two partitions of the VDS.&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--create-v3-filtered-dense-mt&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span>
            <span class="s2">&quot;Create a dense MatrixTable from the raw gnomAD v3.1 sparse MatrixTable&quot;</span>
            <span class="s2">&quot; filtered to predetermined QC variants.&quot;</span>
        <span class="p">),</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--create-v4-filtered-dense-mt&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span>
            <span class="s2">&quot;Create a dense MatrixTable from the raw gnomAD v4 VariantDataset filtered&quot;</span>
            <span class="s2">&quot; to predetermined QC variants.&quot;</span>
        <span class="p">),</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--generate-qc-mt&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span>
            <span class="s2">&quot;Create the final merged gnomAD v3 + v4 QC MatrixTable with all specified&quot;</span>
            <span class="s2">&quot; filters and LD-pruning.&quot;</span>
        <span class="p">),</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--bi-allelic-only&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Filter to variants that are bi-allelic.&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--min-af&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Minimum variant allele frequency to retain variant in QC MatrixTable.&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mf">0.0001</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--min-callrate&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Minimum variant callrate to retain variant in QC MatrixTable.&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mf">0.99</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--min-inbreeding-coeff-threshold&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Minimum site inbreeding coefficient to retain variant in QC MatrixTable.&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=-</span><span class="mf">0.8</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--ld-r2&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;LD-pruning cutoff.&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--n-partitions&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Desired number of partitions for output QC MatrixTable.&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mi">5000</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--block-size&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Block size parameter to use for LD pruning.&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mi">2048</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--generate-qc-meta&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Create a merged gnomAD v3 + v4 metadata Table for QC purposes.&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-o&quot;</span><span class="p">,</span>
        <span class="s2">&quot;--overwrite&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Overwrite all data from this subset.&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--slack-channel&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Slack channel to post results and notifications to.&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">parser</span>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">get_script_argument_parser</span><span class="p">()</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">slack_channel</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">slack_notifications</span><span class="p">(</span><span class="n">slack_token</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">slack_channel</span><span class="p">):</span>
            <span class="n">main</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">main</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p></p>
  </div>

   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>