<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>gnomad_qc.v5.annotations.compute_coverage &mdash; gnomad_qc master documentation</title>
      <link rel="stylesheet" href="../../../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../../../_static/theme_overrides.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../../_static/jquery.js"></script>
        <script src="../../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../../../../" id="documentation_options" src="../../../../_static/documentation_options.js"></script>
        <script src="../../../../_static/doctools.js"></script>
        <script src="../../../../_static/sphinx_highlight.js"></script>
        <script src="https://unpkg.com/mermaid@10.2.0/dist/mermaid.min.js"></script>
        <script>mermaid.initialize({startOnLoad:true});</script>
    <script src="../../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../../index.html" class="icon icon-home">
            gnomad_qc
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../api_reference/index.html">API Reference</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/broadinstitute/gnomad_qc/tags">Change Log</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../../index.html">gnomad_qc</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">gnomad_qc.v5.annotations.compute_coverage</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for gnomad_qc.v5.annotations.compute_coverage</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;Script to compute coverage, allele number, and quality histograms on all gnomAD v5 genomes (AoU v8 + updated gnomAD v4).&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">from</span> <span class="nn">functools</span> <span class="kn">import</span> <span class="n">reduce</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">List</span><span class="p">,</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Tuple</span>

<span class="kn">import</span> <span class="nn">hail</span> <span class="k">as</span> <span class="nn">hl</span>
<span class="kn">from</span> <span class="nn">gnomad.resources.grch38.gnomad</span> <span class="kn">import</span> <span class="n">CURRENT_GENOME_AN_RELEASE</span> <span class="k">as</span> <span class="n">v4_AN_RELEASE</span>
<span class="kn">from</span> <span class="nn">gnomad.resources.grch38.gnomad</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">CURRENT_GENOME_COVERAGE_RELEASE</span> <span class="k">as</span> <span class="n">v4_COVERAGE_RELEASE</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">gnomad.resources.grch38.gnomad</span> <span class="kn">import</span> <span class="n">DOWNSAMPLINGS</span><span class="p">,</span> <span class="n">public_release</span>
<span class="kn">from</span> <span class="nn">gnomad.resources.grch38.reference_data</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">telomeres_and_centromeres</span><span class="p">,</span>
    <span class="n">vep_context</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">gnomad.utils.annotations</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">annotate_downsamplings</span><span class="p">,</span>
    <span class="n">build_freq_stratification_list</span><span class="p">,</span>
    <span class="n">generate_freq_group_membership_array</span><span class="p">,</span>
    <span class="n">merge_array_expressions</span><span class="p">,</span>
    <span class="n">merge_histograms</span><span class="p">,</span>
    <span class="n">qual_hist_expr</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">gnomad.utils.sparse_mt</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">compute_stats_per_ref_site</span><span class="p">,</span>
    <span class="n">get_allele_number_agg_func</span><span class="p">,</span>
    <span class="n">get_coverage_agg_func</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">hail.utils.misc</span> <span class="kn">import</span> <span class="n">new_temp_file</span>

<span class="kn">from</span> <span class="nn">gnomad_qc.resource_utils</span> <span class="kn">import</span> <span class="n">check_resource_existence</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v4.resources.meta</span> <span class="kn">import</span> <span class="n">meta</span> <span class="k">as</span> <span class="n">v4_meta</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v5.annotations.annotation_utils</span> <span class="kn">import</span> <span class="n">annotate_adj_no_dp</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v5.resources.annotations</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">coverage_and_an_path</span><span class="p">,</span>
    <span class="n">get_aou_downsampling</span><span class="p">,</span>
    <span class="n">group_membership</span><span class="p">,</span>
    <span class="n">qual_hists</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v5.resources.basics</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">get_aou_vds</span><span class="p">,</span>
    <span class="n">get_gnomad_v5_genomes_vds</span><span class="p">,</span>
    <span class="n">get_logging_path</span><span class="p">,</span>
    <span class="n">qc_temp_prefix</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v5.resources.constants</span> <span class="kn">import</span> <span class="n">WORKSPACE_BUCKET</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v5.resources.meta</span> <span class="kn">import</span> <span class="n">meta</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v5.resources.release</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">release_all_sites_an_tsv_path</span><span class="p">,</span>
    <span class="n">release_coverage_path</span><span class="p">,</span>
    <span class="n">release_coverage_tsv_path</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="nb">format</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%(levelname)s</span><span class="s2"> (</span><span class="si">%(name)s</span><span class="s2"> </span><span class="si">%(lineno)s</span><span class="s2">): </span><span class="si">%(message)s</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s2">&quot;aou_coverage_and_an&quot;</span><span class="p">)</span>
<span class="n">logger</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>


<div class="viewcode-block" id="get_downsampling_ht"><a class="viewcode-back" href="../../../../api_reference/v5/annotations/compute_coverage.html#gnomad_qc.v5.annotations.compute_coverage.get_downsampling_ht">[docs]</a><span class="k">def</span> <span class="nf">get_downsampling_ht</span><span class="p">(</span><span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get Table with downsampling groups for all samples.</span>

<span class="sd">    v5 downsampling is only applied to the AoU dataset.</span>
<span class="sd">    Desired groups:</span>
<span class="sd">    - 10,000</span>
<span class="sd">    - 100,000</span>
<span class="sd">    - Genetic ancestry group sizes for AFR, AMR, NFE</span>
<span class="sd">    Note that the only desired genetic ancestry group sizes are AFR, AMR, and NFE,</span>
<span class="sd">    but code will also generate downsamplings for all other groups.</span>

<span class="sd">    :param ht: Input Table.</span>
<span class="sd">    :return: Table with downsampling groups.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Determining downsampling groups for AoU...&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">downsamplings</span> <span class="o">=</span> <span class="n">DOWNSAMPLINGS</span><span class="p">[</span><span class="s2">&quot;v5&quot;</span><span class="p">]</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">annotate_downsamplings</span><span class="p">(</span>
        <span class="n">ht</span><span class="p">,</span> <span class="n">downsamplings</span><span class="p">,</span> <span class="n">ht</span><span class="o">.</span><span class="n">genetic_ancestry_inference</span><span class="o">.</span><span class="n">gen_anc</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">ht</span></div>


<div class="viewcode-block" id="get_group_membership_ht"><a class="viewcode-back" href="../../../../api_reference/v5/annotations/compute_coverage.html#gnomad_qc.v5.annotations.compute_coverage.get_group_membership_ht">[docs]</a><span class="k">def</span> <span class="nf">get_group_membership_ht</span><span class="p">(</span>
    <span class="n">meta_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">project</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">ds_ht</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get genomes group membership HT for all sites allele number stratification.</span>

<span class="sd">    :param meta_ht: Meta HT.</span>
<span class="sd">    :param project: Project name. Must be &quot;aou&quot; or &quot;gnomad&quot;. If &quot;gnomad&quot;, function will filter meta HT to only consent drop samples.</span>
<span class="sd">    :param ds_ht: Optional downsampling HT. Only used for AoU.</span>
<span class="sd">    :return: Group membership HT.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">project</span> <span class="o">==</span> <span class="s2">&quot;aou&quot;</span><span class="p">:</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">generate_freq_group_membership_array</span><span class="p">(</span>
            <span class="n">meta_ht</span><span class="p">,</span>
            <span class="n">build_freq_stratification_list</span><span class="p">(</span>
                <span class="n">sex_expr</span><span class="o">=</span><span class="n">meta_ht</span><span class="o">.</span><span class="n">sex_karyotype</span><span class="p">,</span>
                <span class="n">gen_anc_expr</span><span class="o">=</span><span class="n">meta_ht</span><span class="o">.</span><span class="n">genetic_ancestry_inference</span><span class="o">.</span><span class="n">gen_anc</span><span class="p">,</span>
                <span class="n">downsampling_expr</span><span class="o">=</span><span class="n">ds_ht</span><span class="p">[</span><span class="n">meta_ht</span><span class="o">.</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">downsampling</span><span class="p">,</span>
            <span class="p">),</span>
            <span class="n">downsamplings</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">ds_ht</span><span class="o">.</span><span class="n">downsamplings</span><span class="p">),</span>
            <span class="n">ds_gen_anc_counts</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">ds_ht</span><span class="o">.</span><span class="n">ds_gen_anc_counts</span><span class="p">),</span>
        <span class="p">)</span>

        <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span>
            <span class="n">freq_meta</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">freq_meta</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
                <span class="k">lambda</span> <span class="n">d</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">dict</span><span class="p">(</span>
                    <span class="n">d</span><span class="o">.</span><span class="n">items</span><span class="p">()</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
                        <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;pop&quot;</span><span class="p">,</span> <span class="p">(</span><span class="s2">&quot;gen_anc&quot;</span><span class="p">,</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span> <span class="n">x</span><span class="p">)</span>
                    <span class="p">)</span>
                <span class="p">)</span>
            <span class="p">),</span>
        <span class="p">)</span>

    <span class="k">elif</span> <span class="n">project</span> <span class="o">==</span> <span class="s2">&quot;gnomad&quot;</span><span class="p">:</span>
        <span class="c1"># Filter to v4 consent drop samples.</span>
        <span class="c1"># NOTE: Not using v5 project meta here because this part will be run in</span>
        <span class="c1"># Dataproc.</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">meta_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
            <span class="n">meta_ht</span><span class="o">.</span><span class="n">release</span>
            <span class="o">&amp;</span> <span class="p">(</span>
                <span class="p">(</span><span class="n">meta_ht</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">research_project_key</span> <span class="o">==</span> <span class="s2">&quot;RP-1061&quot;</span><span class="p">)</span>
                <span class="o">|</span> <span class="p">(</span><span class="n">meta_ht</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">research_project_key</span> <span class="o">==</span> <span class="s2">&quot;RP-1411&quot;</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">generate_freq_group_membership_array</span><span class="p">(</span>
            <span class="n">ht</span><span class="p">,</span>
            <span class="n">build_freq_stratification_list</span><span class="p">(</span>
                <span class="n">sex_expr</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">sex_imputation</span><span class="o">.</span><span class="n">sex_karyotype</span><span class="p">,</span>
                <span class="n">gen_anc_expr</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">population_inference</span><span class="o">.</span><span class="n">pop</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span>

    <span class="k">return</span> <span class="n">ht</span></div>


<div class="viewcode-block" id="validate_vds"><a class="viewcode-back" href="../../../../api_reference/v5/annotations/compute_coverage.html#gnomad_qc.v5.annotations.compute_coverage.validate_vds">[docs]</a><span class="k">def</span> <span class="nf">validate_vds</span><span class="p">(</span><span class="n">vds</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">VariantDataset</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Validate VDS before densify.</span>

<span class="sd">    Code is taken from https://github.com/hail-is/hail/blob/858f3ab30c2bcc46d6e57fdbfe408284b4b3de53/hail/python/hail/vds/variant_dataset.py#L271</span>
<span class="sd">    at suggestion from Chris Vittal.</span>

<span class="sd">    :param vds: Input VDS.</span>
<span class="sd">    :return: None; raises ValueError if VDS is not valid.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">rd</span> <span class="o">=</span> <span class="n">vds</span><span class="o">.</span><span class="n">reference_data</span>
    <span class="n">vd</span> <span class="o">=</span> <span class="n">vds</span><span class="o">.</span><span class="n">variant_data</span>

    <span class="n">ref_cols</span> <span class="o">=</span> <span class="n">rd</span><span class="o">.</span><span class="n">col_key</span><span class="o">.</span><span class="n">collect</span><span class="p">()</span>
    <span class="n">var_cols</span> <span class="o">=</span> <span class="n">vd</span><span class="o">.</span><span class="n">col_key</span><span class="o">.</span><span class="n">collect</span><span class="p">()</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">ref_cols</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">var_cols</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;mismatch in number of columns: reference data has </span><span class="si">{</span><span class="n">ref_cols</span><span class="si">}</span><span class="s2"> columns, variant data has </span><span class="si">{</span><span class="n">var_cols</span><span class="si">}</span><span class="s2"> columns&quot;</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">ref_cols</span> <span class="o">!=</span> <span class="n">var_cols</span><span class="p">:</span>
        <span class="n">first_mismatch</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">while</span> <span class="n">ref_cols</span><span class="p">[</span><span class="n">first_mismatch</span><span class="p">]</span> <span class="o">==</span> <span class="n">var_cols</span><span class="p">[</span><span class="n">first_mismatch</span><span class="p">]:</span>
            <span class="n">first_mismatch</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;mismatch in columns keys: ref=</span><span class="si">{</span><span class="n">ref_cols</span><span class="p">[</span><span class="n">first_mismatch</span><span class="p">]</span><span class="si">}</span><span class="s2">, var=</span><span class="si">{</span><span class="n">var_cols</span><span class="p">[</span><span class="n">first_mismatch</span><span class="p">]</span><span class="si">}</span><span class="s2"> at position </span><span class="si">{</span><span class="n">first_mismatch</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="compute_all_release_stats_per_ref_site"><a class="viewcode-back" href="../../../../api_reference/v5/annotations/compute_coverage.html#gnomad_qc.v5.annotations.compute_coverage.compute_all_release_stats_per_ref_site">[docs]</a><span class="k">def</span> <span class="nf">compute_all_release_stats_per_ref_site</span><span class="p">(</span>
    <span class="n">vds</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">VariantDataset</span><span class="p">,</span>
    <span class="n">ref_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">sex_karyotype_field</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">project</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">coverage_over_x_bins</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">15</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">25</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span> <span class="mi">100</span><span class="p">],</span>
    <span class="n">interval_ht</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">group_membership_ht</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Compute coverage, allele number, and quality histograms per reference site.</span>

<span class="sd">    .. note::</span>

<span class="sd">        Running this function prior to calculating frequencies removes the need for an additional</span>
<span class="sd">        densify for frequency calculations.</span>

<span class="sd">    :param vds: Input VDS.</span>
<span class="sd">    :param ref_ht: Reference HT.</span>
<span class="sd">    :param sex_karyotype_field: Field name for sex karyotype.</span>
<span class="sd">    :param project: Project name.</span>
<span class="sd">    :param coverage_over_x_bins: List of boundaries for computing samples over X depth.</span>
<span class="sd">    :param interval_ht: Interval HT.</span>
<span class="sd">    :param group_membership_ht: Group membership HT.</span>
<span class="sd">    :return: HT with allele number and quality histograms per reference site.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">_get_hists</span><span class="p">(</span><span class="n">qual_expr</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">Expression</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">qual_hist_expr</span><span class="p">(</span>
            <span class="n">gq_expr</span><span class="o">=</span><span class="n">qual_expr</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
            <span class="n">dp_expr</span><span class="o">=</span><span class="n">qual_expr</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
            <span class="n">adj_expr</span><span class="o">=</span><span class="n">qual_expr</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">,</span>
            <span class="n">split_adj_and_raw</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="c1"># Set up coverage bins.</span>
    <span class="n">cov_bins</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">coverage_over_x_bins</span><span class="p">)</span>
    <span class="n">rev_cov_bins</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">reversed</span><span class="p">(</span><span class="n">cov_bins</span><span class="p">))</span>
    <span class="n">max_cov_bin</span> <span class="o">=</span> <span class="n">cov_bins</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">cov_bins</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">cov_bins</span><span class="p">)</span>

    <span class="n">entry_agg_funcs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;AN&quot;</span><span class="p">:</span> <span class="n">get_allele_number_agg_func</span><span class="p">(</span><span class="s2">&quot;LGT&quot;</span><span class="p">),</span>
        <span class="s2">&quot;coverage_stats&quot;</span><span class="p">:</span> <span class="n">get_coverage_agg_func</span><span class="p">(</span><span class="n">dp_field</span><span class="o">=</span><span class="s2">&quot;DP&quot;</span><span class="p">,</span> <span class="n">max_cov_bin</span><span class="o">=</span><span class="n">max_cov_bin</span><span class="p">),</span>
    <span class="p">}</span>
    <span class="n">entry_agg_group_membership</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="c1"># Only compute qual hists for AoU.</span>
    <span class="k">if</span> <span class="n">project</span> <span class="o">==</span> <span class="s2">&quot;aou&quot;</span><span class="p">:</span>
        <span class="n">entry_agg_funcs</span><span class="p">[</span><span class="s2">&quot;qual_hists&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="k">lambda</span> <span class="n">t</span><span class="p">:</span> <span class="p">[</span><span class="n">t</span><span class="o">.</span><span class="n">GQ</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">DP</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">adj</span><span class="p">],</span> <span class="n">_get_hists</span><span class="p">)</span>

        <span class="c1"># Below we use just the raw group for qual hist computations because qual hists</span>
        <span class="c1"># has its own built-in adj filtering when adj is passed as an argument and will</span>
        <span class="c1"># produce both adj and raw histograms.</span>
        <span class="n">entry_agg_group_membership</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;qual_hists&quot;</span><span class="p">:</span> <span class="p">[{</span><span class="s2">&quot;group&quot;</span><span class="p">:</span> <span class="s2">&quot;raw&quot;</span><span class="p">}]}</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Computing coverage, allele number, and optionally qual hists per reference site...&quot;</span>
    <span class="p">)</span>

    <span class="n">vmt</span> <span class="o">=</span> <span class="n">vds</span><span class="o">.</span><span class="n">variant_data</span>
    <span class="n">sex_expr</span> <span class="o">=</span> <span class="n">reduce</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">field</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="n">field</span><span class="p">],</span> <span class="n">sex_karyotype_field</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">),</span> <span class="n">vmt</span><span class="p">)</span>
    <span class="n">vmt</span> <span class="o">=</span> <span class="n">vmt</span><span class="o">.</span><span class="n">annotate_cols</span><span class="p">(</span><span class="n">sex_karyotype</span><span class="o">=</span><span class="n">sex_expr</span><span class="p">)</span>
    <span class="n">vds</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">VariantDataset</span><span class="p">(</span><span class="n">vds</span><span class="o">.</span><span class="n">reference_data</span><span class="p">,</span> <span class="n">vmt</span><span class="p">)</span>

    <span class="c1"># TODO: Save dense MT for gnomAD consent drop samples?</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">compute_stats_per_ref_site</span><span class="p">(</span>
        <span class="n">vds</span><span class="p">,</span>
        <span class="n">ref_ht</span><span class="p">,</span>
        <span class="n">entry_agg_funcs</span><span class="p">,</span>
        <span class="n">interval_ht</span><span class="o">=</span><span class="n">interval_ht</span><span class="p">,</span>
        <span class="n">group_membership_ht</span><span class="o">=</span><span class="n">group_membership_ht</span><span class="p">,</span>
        <span class="n">entry_keep_fields</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;GQ&quot;</span><span class="p">,</span> <span class="s2">&quot;DP&quot;</span><span class="p">],</span>
        <span class="n">entry_agg_group_membership</span><span class="o">=</span><span class="n">entry_agg_group_membership</span><span class="p">,</span>
        <span class="n">sex_karyotype_field</span><span class="o">=</span><span class="s2">&quot;sex_karyotype&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># This expression aggregates the DP counter in reverse order of the cov_bins and</span>
    <span class="c1"># computes the cumulative sum over them. It needs to be in reverse order because we</span>
    <span class="c1"># want the sum over samples covered by &gt; X.</span>
    <span class="k">def</span> <span class="nf">_cov_stats</span><span class="p">(</span>
        <span class="n">cov_stat</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">StructExpression</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">StructExpression</span><span class="p">:</span>
        <span class="c1"># The coverage was already floored to the max_coverage_bin, so no more</span>
        <span class="c1"># aggregation is needed for the max bin.</span>
        <span class="n">count_expr</span> <span class="o">=</span> <span class="n">cov_stat</span><span class="o">.</span><span class="n">coverage_counter</span>
        <span class="n">max_bin_expr</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">int32</span><span class="p">(</span><span class="n">count_expr</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">max_cov_bin</span><span class="p">,</span> <span class="mi">0</span><span class="p">))</span>

        <span class="c1"># For each of the other bins, coverage is summed between the boundaries.</span>
        <span class="n">bin_expr</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">range</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">len</span><span class="p">(</span><span class="n">cov_bins</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">step</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">bin_expr</span> <span class="o">=</span> <span class="n">bin_expr</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">i</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span>
                <span class="n">hl</span><span class="o">.</span><span class="n">range</span><span class="p">(</span><span class="n">cov_bins</span><span class="p">[</span><span class="n">i</span> <span class="o">-</span> <span class="mi">1</span><span class="p">],</span> <span class="n">cov_bins</span><span class="p">[</span><span class="n">i</span><span class="p">])</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
                    <span class="k">lambda</span> <span class="n">j</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">int32</span><span class="p">(</span><span class="n">count_expr</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">j</span><span class="p">,</span> <span class="mi">0</span><span class="p">))</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="n">bin_expr</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">cumulative_sum</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">max_bin_expr</span><span class="p">])</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">bin_expr</span><span class="p">))</span>
        <span class="c1"># NOTE: Keeping these as sample counts rather than fractions to join with</span>
        <span class="c1"># gnomAD v4 genomes.</span>
        <span class="n">bin_expr</span> <span class="o">=</span> <span class="p">{</span><span class="sa">f</span><span class="s2">&quot;over_</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="n">bin_expr</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">rev_cov_bins</span><span class="p">)}</span>
        <span class="k">return</span> <span class="n">cov_stat</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="o">**</span><span class="n">bin_expr</span><span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;coverage_counter&quot;</span><span class="p">)</span>

    <span class="c1"># Keep coverage stats from global adj grouping (index 0) only.</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span>
        <span class="n">coverage_stats_meta_sample_count</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">strata_sample_count</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
    <span class="p">)</span>
    <span class="n">cov_stats_expr</span> <span class="o">=</span> <span class="n">_cov_stats</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">coverage_stats</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">transmute</span><span class="p">(</span><span class="o">**</span><span class="n">cov_stats_expr</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">project</span> <span class="o">==</span> <span class="s2">&quot;aou&quot;</span><span class="p">:</span>
        <span class="c1"># `qual_hists` as returned by `compute_stats_per_ref_site` is an array of length 1 so we drop the array here.</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">qual_hists</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">qual_hists</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">ht</span></div>


<span class="k">def</span> <span class="nf">_rename_cov_annotations</span><span class="p">(</span>
    <span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">project</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">sample_count</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">coverage_over_x_bins</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">15</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">25</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span> <span class="mi">100</span><span class="p">],</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Rename coverage annotations prior to merging Tables.</span>

<span class="sd">    Function transforms mean back into sum using `sample_count` argument.</span>

<span class="sd">    :param ht: Input HT.</span>
<span class="sd">    :param project: Project name.</span>
<span class="sd">    :param sample_count: Number of samples in HT.</span>
<span class="sd">    :param coverage_over_x_bins: List of boundaries for computing samples over X.</span>
<span class="sd">        Default is [1, 5, 10, 15, 20, 25, 30, 50, 100].</span>
<span class="sd">    :return: Renamed HT.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Transform mean back into sum.</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">transmute</span><span class="p">(</span><span class="nb">sum</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">mean</span> <span class="o">*</span> <span class="n">sample_count</span><span class="p">)</span>

    <span class="c1"># Rename annotations to include project.</span>
    <span class="n">row_fields</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">row_value</span><span class="p">)</span>
    <span class="n">rename_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">f</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">f</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">project</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">row_fields</span><span class="p">}</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">rename_dict</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">project</span> <span class="o">==</span> <span class="s2">&quot;gnomad&quot;</span> <span class="ow">or</span> <span class="n">project</span> <span class="o">==</span> <span class="s2">&quot;gnomad_release&quot;</span><span class="p">:</span>
        <span class="c1"># Revert v4 genomes fraction over X bins to sample count over X bins.</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">transmute</span><span class="p">(</span>
            <span class="o">**</span><span class="p">{</span>
                <span class="sa">f</span><span class="s2">&quot;over_</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">project</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="n">hl</span><span class="o">.</span><span class="n">float64</span><span class="p">(</span><span class="n">ht</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;over_</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">project</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">])</span> <span class="o">*</span> <span class="n">sample_count</span>
                <span class="p">)</span>
                <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">coverage_over_x_bins</span>
            <span class="p">}</span>
        <span class="p">)</span>
    <span class="k">return</span> <span class="n">ht</span>


<span class="k">def</span> <span class="nf">_merge_coverage_fields</span><span class="p">(</span>
    <span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">project_1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">project_2</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">sample_count</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">operation</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">coverage_over_x_bins</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">15</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">25</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span> <span class="mi">100</span><span class="p">],</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">DictExpression</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Merge coverage fields from two Tables.</span>

<span class="sd">    .. note::</span>
<span class="sd">        - Function does not merge `median_approx` fields.</span>

<span class="sd">    :param ht: Input HT. Must have annotations from both projects.</span>
<span class="sd">    :param project_1: First project name.</span>
<span class="sd">    :param project_2: Second project name.</span>
<span class="sd">    :param sample_count: Total sample count.</span>
<span class="sd">    :param operation: Operation to perform on the coverage fields. Must be &quot;sum&quot; or &quot;diff&quot;.</span>
<span class="sd">    :param coverage_over_x_bins: List of boundaries for computing samples over X.</span>
<span class="sd">        Default is [1, 5, 10, 15, 20, 25, 30, 50, 100].</span>
<span class="sd">    :return: Merged fields.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">operation</span> <span class="o">==</span> <span class="s2">&quot;diff&quot;</span><span class="p">:</span>
        <span class="c1"># Make sure over_x fields are float64s.</span>
        <span class="n">merged_fields</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;sum_gnomad&quot;</span><span class="p">:</span> <span class="n">ht</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;sum_</span><span class="si">{</span><span class="n">project_1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">ht</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;sum_</span><span class="si">{</span><span class="n">project_2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">],</span>
            <span class="s2">&quot;total_DP_gnomad&quot;</span><span class="p">:</span> <span class="p">(</span>
                <span class="n">ht</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;total_DP_</span><span class="si">{</span><span class="n">project_1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">ht</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;total_DP_</span><span class="si">{</span><span class="n">project_2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span>
            <span class="p">),</span>
        <span class="p">}</span>
        <span class="n">merged_fields</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
            <span class="p">{</span>
                <span class="sa">f</span><span class="s2">&quot;over_</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s2">_gnomad&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="n">ht</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;over_</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">project_1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">ht</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;over_</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">project_2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span>
                <span class="p">)</span>
                <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">coverage_over_x_bins</span>
            <span class="p">}</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">merged_fields</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;mean&quot;</span><span class="p">:</span> <span class="p">(</span><span class="n">ht</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;sum_</span><span class="si">{</span><span class="n">project_1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">ht</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;sum_</span><span class="si">{</span><span class="n">project_2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">])</span> <span class="o">/</span> <span class="n">sample_count</span><span class="p">,</span>
        <span class="p">}</span>
        <span class="n">merged_fields</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
            <span class="p">{</span>
                <span class="sa">f</span><span class="s2">&quot;over_</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="p">(</span><span class="n">ht</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;over_</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">project_1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="n">ht</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;over_</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">project_2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">])</span>
                    <span class="o">/</span> <span class="n">sample_count</span>
                <span class="p">)</span>
                <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">coverage_over_x_bins</span>
            <span class="p">}</span>
        <span class="p">)</span>
    <span class="k">return</span> <span class="n">merged_fields</span>


<div class="viewcode-block" id="merge_gnomad_coverage_hts"><a class="viewcode-back" href="../../../../api_reference/v5/annotations/compute_coverage.html#gnomad_qc.v5.annotations.compute_coverage.merge_gnomad_coverage_hts">[docs]</a><span class="k">def</span> <span class="nf">merge_gnomad_coverage_hts</span><span class="p">(</span>
    <span class="n">gnomad_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">gnomad_release_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">coverage_over_x_bins</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">15</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">25</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span> <span class="mi">100</span><span class="p">],</span>
    <span class="n">v4_count</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">76215</span><span class="p">,</span>
    <span class="n">consent_drop_count</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">866</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Subtract consent drop samples from gnomAD v4 genomes release HT to create gnomAD v5 genomes coverage HT.</span>

<span class="sd">    :param gnomad_ht: gnomAD coverage HT (contains coverage for consent drop samples only).</span>
<span class="sd">    :param gnomad_release_ht: gnomAD v4 genomes coverage release HT.</span>
<span class="sd">    :param coverage_over_x_bins: List of boundaries for computing samples over X.</span>
<span class="sd">        Default is [1, 5, 10, 15, 20, 25, 30, 50, 100].</span>
<span class="sd">    :param v4_count: Number of release gnomAD v4 genome samples. Default is 76215.</span>
<span class="sd">    :param consent_drop_count: Number of consent drop gnomAD v4 genome samples. Default is 866.</span>
<span class="sd">    :return: gnomAD v5 genomes coverage HT.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Subtracting gnomAD v4 consent drop samples from gnomAD v4 genomes release HT...&quot;</span>
    <span class="p">)</span>
    <span class="n">gnomad_ht</span> <span class="o">=</span> <span class="n">_rename_cov_annotations</span><span class="p">(</span>
        <span class="n">gnomad_ht</span><span class="p">,</span> <span class="s2">&quot;gnomad&quot;</span><span class="p">,</span> <span class="n">consent_drop_count</span><span class="p">,</span> <span class="n">coverage_over_x_bins</span>
    <span class="p">)</span>
    <span class="n">gnomad_release_ht</span> <span class="o">=</span> <span class="n">_rename_cov_annotations</span><span class="p">(</span>
        <span class="n">gnomad_release_ht</span><span class="p">,</span> <span class="s2">&quot;gnomad_release&quot;</span><span class="p">,</span> <span class="n">v4_count</span><span class="p">,</span> <span class="n">coverage_over_x_bins</span>
    <span class="p">)</span>
    <span class="n">gnomad_v5_count</span> <span class="o">=</span> <span class="n">v4_count</span> <span class="o">-</span> <span class="n">consent_drop_count</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Total number of gnomAD v5 release genomes: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">gnomad_v5_count</span><span class="p">)</span>

    <span class="n">gnomad_ht</span> <span class="o">=</span> <span class="n">gnomad_ht</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gnomad_release_ht</span><span class="p">,</span> <span class="s2">&quot;right&quot;</span><span class="p">)</span>
    <span class="n">merged_fields</span> <span class="o">=</span> <span class="n">_merge_coverage_fields</span><span class="p">(</span>
        <span class="n">ht</span><span class="o">=</span><span class="n">gnomad_ht</span><span class="p">,</span>
        <span class="n">project_1</span><span class="o">=</span><span class="s2">&quot;gnomad_release&quot;</span><span class="p">,</span>
        <span class="n">project_2</span><span class="o">=</span><span class="s2">&quot;gnomad&quot;</span><span class="p">,</span>
        <span class="n">sample_count</span><span class="o">=</span><span class="n">gnomad_v5_count</span><span class="p">,</span>
        <span class="n">operation</span><span class="o">=</span><span class="s2">&quot;diff&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">gnomad_ht</span> <span class="o">=</span> <span class="n">gnomad_ht</span><span class="o">.</span><span class="n">transmute</span><span class="p">(</span><span class="o">**</span><span class="n">merged_fields</span><span class="p">)</span>

    <span class="c1"># Keep median_approx from v4 release.</span>
    <span class="n">gnomad_ht</span> <span class="o">=</span> <span class="n">gnomad_ht</span><span class="o">.</span><span class="n">transmute</span><span class="p">(</span>
        <span class="n">median_approx_gnomad</span><span class="o">=</span><span class="n">gnomad_ht</span><span class="o">.</span><span class="n">median_approx_gnomad_release</span>
    <span class="p">)</span>

    <span class="c1"># Drop unnecessary globals and add back v5 gnomAD genomes count.</span>
    <span class="n">gnomad_ht</span> <span class="o">=</span> <span class="n">gnomad_ht</span><span class="o">.</span><span class="n">select_globals</span><span class="p">()</span>
    <span class="n">gnomad_ht</span> <span class="o">=</span> <span class="n">gnomad_ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span>
        <span class="n">coverage_stats_meta_sample_count</span><span class="o">=</span><span class="n">gnomad_v5_count</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">gnomad_ht</span></div>


<div class="viewcode-block" id="join_aou_and_gnomad_coverage_ht"><a class="viewcode-back" href="../../../../api_reference/v5/annotations/compute_coverage.html#gnomad_qc.v5.annotations.compute_coverage.join_aou_and_gnomad_coverage_ht">[docs]</a><span class="k">def</span> <span class="nf">join_aou_and_gnomad_coverage_ht</span><span class="p">(</span>
    <span class="n">aou_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">gnomad_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">coverage_over_x_bins</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">15</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">25</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span> <span class="mi">100</span><span class="p">],</span>
    <span class="n">gnomad_v5_count</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">76215</span> <span class="o">-</span> <span class="mi">866</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Join AoU and gnomAD coverage HTs for release.</span>

<span class="sd">    :param aou_ht: AoU coverage HT.</span>
<span class="sd">    :param gnomad_ht: gnomAD v5 genomes coverage HT.</span>
<span class="sd">    :param coverage_over_x_bins: List of boundaries for computing samples over X.</span>
<span class="sd">        Default is [1, 5, 10, 15, 20, 25, 30, 50, 100].</span>
<span class="sd">    :param gnomad_v5_count: Number of release gnomAD v5 genome samples. Default is 76215 - 866.</span>
<span class="sd">    :return: Joined HT.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">aou_count</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">aou_ht</span><span class="o">.</span><span class="n">coverage_stats_meta_sample_count</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Total number of AoU v8 release samples: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">aou_count</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Merging AoU and gnomAD v5 coverage HTs...&quot;</span><span class="p">)</span>
    <span class="n">aou_ht</span> <span class="o">=</span> <span class="n">_rename_cov_annotations</span><span class="p">(</span><span class="n">aou_ht</span><span class="p">,</span> <span class="s2">&quot;aou&quot;</span><span class="p">,</span> <span class="n">aou_count</span><span class="p">,</span> <span class="n">coverage_over_x_bins</span><span class="p">)</span>
    <span class="n">v5_count</span> <span class="o">=</span> <span class="n">aou_count</span> <span class="o">+</span> <span class="n">gnomad_v5_count</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Total number of AoU + gnomAD v5 release genomes: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">v5_count</span><span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">aou_ht</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gnomad_ht</span><span class="p">,</span> <span class="s2">&quot;left&quot;</span><span class="p">)</span>
    <span class="n">merged_fields</span> <span class="o">=</span> <span class="n">_merge_coverage_fields</span><span class="p">(</span>
        <span class="n">ht</span><span class="o">=</span><span class="n">ht</span><span class="p">,</span>
        <span class="n">project_1</span><span class="o">=</span><span class="s2">&quot;aou&quot;</span><span class="p">,</span>
        <span class="n">project_2</span><span class="o">=</span><span class="s2">&quot;gnomad&quot;</span><span class="p">,</span>
        <span class="n">sample_count</span><span class="o">=</span><span class="n">v5_count</span><span class="p">,</span>
        <span class="n">operation</span><span class="o">=</span><span class="s2">&quot;sum&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">transmute</span><span class="p">(</span><span class="o">**</span><span class="n">merged_fields</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">ht</span><span class="o">.</span><span class="n">select_globals</span><span class="p">()</span></div>


<span class="k">def</span> <span class="nf">_rename_fields</span><span class="p">(</span>
    <span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">field_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">project</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">rename_globals</span><span class="p">:</span> <span class="nb">bool</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Rename fields by adding project name prior to merging Tables.</span>

<span class="sd">    Used for AN and qual hists merging but not coverage because</span>
<span class="sd">    coverage does not have globals and also requires extra transformations</span>
<span class="sd">    to transform v4 mean back into sum.</span>

<span class="sd">    :param ht: Input HT.</span>
<span class="sd">    :param project: Project name.</span>
<span class="sd">    :param rename_globals: Whether to rename globals.</span>
<span class="sd">    :return: Renamed HT.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">rename_globals</span><span class="p">:</span>
        <span class="n">rename_globals</span> <span class="o">=</span> <span class="p">{</span>
            <span class="sa">f</span><span class="s2">&quot;strata_meta_</span><span class="si">{</span><span class="n">project</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="n">ht</span><span class="o">.</span><span class="n">strata_meta</span><span class="p">,</span>
            <span class="sa">f</span><span class="s2">&quot;strata_sample_count_</span><span class="si">{</span><span class="n">project</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="n">ht</span><span class="o">.</span><span class="n">strata_sample_count</span><span class="p">,</span>
        <span class="p">}</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">transmute_globals</span><span class="p">(</span><span class="o">**</span><span class="n">rename_globals</span><span class="p">)</span>
    <span class="n">rename_dict</span> <span class="o">=</span> <span class="p">{</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">field_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">field_name</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">project</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">}</span>
    <span class="k">return</span> <span class="n">ht</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">rename_dict</span><span class="p">)</span>


<span class="k">def</span> <span class="nf">_merge_an_fields</span><span class="p">(</span>
    <span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">project_1</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">project_2</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">operation</span><span class="p">:</span> <span class="nb">str</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span>
    <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">ArrayExpression</span><span class="p">,</span>
    <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">ArrayExpression</span><span class="p">,</span>
    <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">DictExpression</span><span class="p">,</span>
<span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Merge AN fields from two projects.</span>

<span class="sd">    :param ht: Input HT. Must have annotations from both projects.</span>
<span class="sd">    :param project_1: First project name.</span>
<span class="sd">    :param project_2: Second project name.</span>
<span class="sd">    :param operation: Operation to perform on the AN fields. Must be &quot;sum&quot; or &quot;diff&quot;.</span>
<span class="sd">    :return: Joined AN, strata meta, and count arrays.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">joint_an</span><span class="p">,</span> <span class="n">joint_strata_meta</span><span class="p">,</span> <span class="n">count_arrays_dict</span> <span class="o">=</span> <span class="n">merge_array_expressions</span><span class="p">(</span>
        <span class="n">arrays</span><span class="o">=</span><span class="p">[</span><span class="n">ht</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;AN_</span><span class="si">{</span><span class="n">project_1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">],</span> <span class="n">ht</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;AN_</span><span class="si">{</span><span class="n">project_2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]],</span>
        <span class="n">meta</span><span class="o">=</span><span class="p">[</span>
            <span class="n">ht</span><span class="o">.</span><span class="n">index_globals</span><span class="p">()[</span><span class="sa">f</span><span class="s2">&quot;strata_meta_</span><span class="si">{</span><span class="n">project_1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">],</span>
            <span class="n">ht</span><span class="o">.</span><span class="n">index_globals</span><span class="p">()[</span><span class="sa">f</span><span class="s2">&quot;strata_meta_</span><span class="si">{</span><span class="n">project_2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">],</span>
        <span class="p">],</span>
        <span class="n">count_arrays</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;counts&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="n">ht</span><span class="o">.</span><span class="n">index_globals</span><span class="p">()[</span><span class="sa">f</span><span class="s2">&quot;strata_sample_count_</span><span class="si">{</span><span class="n">project_1</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">],</span>
                <span class="n">ht</span><span class="o">.</span><span class="n">index_globals</span><span class="p">()[</span><span class="sa">f</span><span class="s2">&quot;strata_sample_count_</span><span class="si">{</span><span class="n">project_2</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">],</span>
            <span class="p">],</span>
        <span class="p">},</span>
        <span class="n">operation</span><span class="o">=</span><span class="n">operation</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">joint_an</span><span class="p">,</span> <span class="n">joint_strata_meta</span><span class="p">,</span> <span class="n">count_arrays_dict</span>


<div class="viewcode-block" id="merge_gnomad_an_hts"><a class="viewcode-back" href="../../../../api_reference/v5/annotations/compute_coverage.html#gnomad_qc.v5.annotations.compute_coverage.merge_gnomad_an_hts">[docs]</a><span class="k">def</span> <span class="nf">merge_gnomad_an_hts</span><span class="p">(</span>
    <span class="n">gnomad_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">gnomad_release_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Subtract consent drop samples from gnomAD v4 genomes release HT to create gnomAD v5 genomes AN HT.</span>

<span class="sd">    :param gnomad_ht: gnomAD AN HT (contains AN for consent drop samples only).</span>
<span class="sd">    :param gnomad_release_ht: gnomAD v4 genomes release AN HT.</span>
<span class="sd">    :return: gnomAD v5 genomes AN HT.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Subtracting gnomAD v4 consent drop samples from gnomAD v4 genomes release HT...&quot;</span>
    <span class="p">)</span>
    <span class="n">gnomad_ht</span> <span class="o">=</span> <span class="n">_rename_fields</span><span class="p">(</span><span class="n">gnomad_ht</span><span class="p">,</span> <span class="s2">&quot;AN&quot;</span><span class="p">,</span> <span class="s2">&quot;gnomad&quot;</span><span class="p">,</span> <span class="n">rename_globals</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">gnomad_release_ht</span> <span class="o">=</span> <span class="n">_rename_fields</span><span class="p">(</span>
        <span class="n">gnomad_release_ht</span><span class="p">,</span> <span class="s2">&quot;AN&quot;</span><span class="p">,</span> <span class="s2">&quot;gnomad_release&quot;</span><span class="p">,</span> <span class="n">rename_globals</span><span class="o">=</span><span class="kc">True</span>
    <span class="p">)</span>

    <span class="n">gnomad_ht</span> <span class="o">=</span> <span class="n">gnomad_ht</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gnomad_release_ht</span><span class="p">,</span> <span class="s2">&quot;right&quot;</span><span class="p">)</span>
    <span class="n">joint_an</span><span class="p">,</span> <span class="n">joint_strata_meta</span><span class="p">,</span> <span class="n">count_arrays_dict</span> <span class="o">=</span> <span class="n">_merge_an_fields</span><span class="p">(</span>
        <span class="n">ht</span><span class="o">=</span><span class="n">gnomad_ht</span><span class="p">,</span>
        <span class="n">project_1</span><span class="o">=</span><span class="s2">&quot;gnomad_release&quot;</span><span class="p">,</span>
        <span class="n">project_2</span><span class="o">=</span><span class="s2">&quot;gnomad&quot;</span><span class="p">,</span>
        <span class="n">operation</span><span class="o">=</span><span class="s2">&quot;diff&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">gnomad_ht</span> <span class="o">=</span> <span class="n">gnomad_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">AN_gnomad</span><span class="o">=</span><span class="n">joint_an</span><span class="p">)</span>
    <span class="n">gnomad_ht</span> <span class="o">=</span> <span class="n">gnomad_ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span>
        <span class="n">strata_meta_gnomad</span><span class="o">=</span><span class="n">joint_strata_meta</span><span class="p">,</span>
        <span class="n">strata_sample_count_gnomad</span><span class="o">=</span><span class="n">count_arrays_dict</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">],</span>
    <span class="p">)</span>
    <span class="n">gnomad_ht</span> <span class="o">=</span> <span class="n">gnomad_ht</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span>
        <span class="s2">&quot;strata_meta_gnomad_release&quot;</span><span class="p">,</span>
        <span class="s2">&quot;strata_sample_count_gnomad_release&quot;</span><span class="p">,</span>
        <span class="s2">&quot;coverage_stats_meta_sample_count&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">gnomad_ht</span> <span class="o">=</span> <span class="n">gnomad_ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;AN_gnomad&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">gnomad_ht</span></div>


<div class="viewcode-block" id="join_aou_and_gnomad_an_ht"><a class="viewcode-back" href="../../../../api_reference/v5/annotations/compute_coverage.html#gnomad_qc.v5.annotations.compute_coverage.join_aou_and_gnomad_an_ht">[docs]</a><span class="k">def</span> <span class="nf">join_aou_and_gnomad_an_ht</span><span class="p">(</span>
    <span class="n">aou_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">gnomad_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Join AoU and gnomAD AN HTs for release.</span>

<span class="sd">    :param aou_ht: AoU AN HT.</span>
<span class="sd">    :param gnomad_ht: gnomAD v5 genomes AN HT.</span>
<span class="sd">    :return: Joined HT.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">aou_ht</span> <span class="o">=</span> <span class="n">_rename_fields</span><span class="p">(</span><span class="n">aou_ht</span><span class="p">,</span> <span class="s2">&quot;AN&quot;</span><span class="p">,</span> <span class="s2">&quot;aou&quot;</span><span class="p">,</span> <span class="n">rename_globals</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># TODO: should we only merge the overall adj AN (like coverage)?</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Merging AoU and gnomAD v5 AN HTs...&quot;</span><span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">aou_ht</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gnomad_ht</span><span class="p">,</span> <span class="s2">&quot;left&quot;</span><span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span><span class="n">new_temp_file</span><span class="p">(</span><span class="s2">&quot;aou_and_gnomad_join&quot;</span><span class="p">,</span> <span class="s2">&quot;ht&quot;</span><span class="p">))</span>

    <span class="n">joint_an</span><span class="p">,</span> <span class="n">joint_strata_meta</span><span class="p">,</span> <span class="n">count_arrays_dict</span> <span class="o">=</span> <span class="n">_merge_an_fields</span><span class="p">(</span>
        <span class="n">ht</span><span class="o">=</span><span class="n">ht</span><span class="p">,</span>
        <span class="n">project_1</span><span class="o">=</span><span class="s2">&quot;aou&quot;</span><span class="p">,</span>
        <span class="n">project_2</span><span class="o">=</span><span class="s2">&quot;gnomad&quot;</span><span class="p">,</span>
        <span class="n">operation</span><span class="o">=</span><span class="s2">&quot;sum&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">AN</span><span class="o">=</span><span class="n">joint_an</span><span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span>
        <span class="n">strata_meta</span><span class="o">=</span><span class="n">joint_strata_meta</span><span class="p">,</span>
        <span class="n">strata_sample_count</span><span class="o">=</span><span class="n">count_arrays_dict</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">],</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">ht</span></div>


<div class="viewcode-block" id="join_aou_and_gnomad_qual_hists_ht"><a class="viewcode-back" href="../../../../api_reference/v5/annotations/compute_coverage.html#gnomad_qc.v5.annotations.compute_coverage.join_aou_and_gnomad_qual_hists_ht">[docs]</a><span class="k">def</span> <span class="nf">join_aou_and_gnomad_qual_hists_ht</span><span class="p">(</span>
    <span class="n">aou_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">gnomad_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Join AoU and gnomAD qual hists HTs for release.</span>

<span class="sd">    .. note::</span>
<span class="sd">        We did not compute qual hists for the gnomAD v4 genomes release</span>
<span class="sd">        (https://github.com/broadinstitute/gnomad_qc/blob/e65bdbb5768113c0129199a875d845da245690e2/gnomad_qc/v4/annotations/generate_freq_genomes.py#L1139).</span>
<span class="sd">        This means we will not also not recompute hists on the gnomAD v4 genomes for v5,</span>
<span class="sd">        which also means we will not subtract values from the samples to drop for consent reasons.</span>

<span class="sd">    :param aou_ht: AoU qual hists HT.</span>
<span class="sd">    :param gnomad_ht: gnomAD qual hists HT.</span>
<span class="sd">    :return: Joined HT.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">aou_ht</span> <span class="o">=</span> <span class="n">_rename_fields</span><span class="p">(</span><span class="n">aou_ht</span><span class="p">,</span> <span class="s2">&quot;qual_hists&quot;</span><span class="p">,</span> <span class="s2">&quot;aou&quot;</span><span class="p">,</span> <span class="n">rename_globals</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">gnomad_ht</span> <span class="o">=</span> <span class="n">_rename_fields</span><span class="p">(</span><span class="n">gnomad_ht</span><span class="p">,</span> <span class="s2">&quot;qual_hists&quot;</span><span class="p">,</span> <span class="s2">&quot;gnomad&quot;</span><span class="p">,</span> <span class="n">rename_globals</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">aou_ht</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gnomad_ht</span><span class="p">,</span> <span class="s2">&quot;left&quot;</span><span class="p">)</span>
    <span class="n">qual_hists</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;gq_hist_all&quot;</span><span class="p">,</span>
        <span class="s2">&quot;dp_hist_all&quot;</span><span class="p">,</span>
    <span class="p">]</span>
    <span class="n">hist_structs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;qual_hists&quot;</span><span class="p">:</span> <span class="n">qual_hists</span><span class="p">,</span>
        <span class="s2">&quot;raw_qual_hists&quot;</span><span class="p">:</span> <span class="n">qual_hists</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">hists_expr</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">hist_struct</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
            <span class="o">**</span><span class="p">{</span>
                <span class="n">h</span><span class="p">:</span> <span class="n">merge_histograms</span><span class="p">(</span>
                    <span class="p">[</span>
                        <span class="n">ht</span><span class="o">.</span><span class="n">qual_hists_aou</span><span class="p">[</span><span class="n">hist_struct</span><span class="p">][</span><span class="n">h</span><span class="p">],</span>
                        <span class="n">ht</span><span class="o">.</span><span class="n">qual_hists_gnomad</span><span class="p">[</span><span class="n">hist_struct</span><span class="p">][</span><span class="n">h</span><span class="p">],</span>
                    <span class="p">],</span>
                    <span class="n">operation</span><span class="o">=</span><span class="s2">&quot;sum&quot;</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="k">for</span> <span class="n">h</span> <span class="ow">in</span> <span class="n">hists</span>
            <span class="p">}</span>
        <span class="p">)</span>
        <span class="k">for</span> <span class="n">hist_struct</span><span class="p">,</span> <span class="n">hists</span> <span class="ow">in</span> <span class="n">hist_structs</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
    <span class="p">}</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="o">**</span><span class="n">hists_expr</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">ht</span></div>


<div class="viewcode-block" id="main"><a class="viewcode-back" href="../../../../api_reference/v5/annotations/compute_coverage.html#gnomad_qc.v5.annotations.compute_coverage.main">[docs]</a><span class="k">def</span> <span class="nf">main</span><span class="p">(</span><span class="n">args</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compute all sites coverage, allele number, and quality histograms for v5 genomes (AoU v8 + gnomAD v4).&quot;&quot;&quot;</span>
    <span class="n">project</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">project_name</span>
    <span class="n">environment</span> <span class="o">=</span> <span class="s2">&quot;rwb&quot;</span> <span class="k">if</span> <span class="n">project</span> <span class="o">==</span> <span class="s2">&quot;aou&quot;</span> <span class="k">else</span> <span class="s2">&quot;dataproc&quot;</span>
    <span class="k">if</span> <span class="n">environment</span> <span class="o">==</span> <span class="s2">&quot;rwb&quot;</span><span class="p">:</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">init</span><span class="p">(</span>
            <span class="n">log</span><span class="o">=</span><span class="s2">&quot;/home/jupyter/workspaces/gnomadproduction/compute_coverage.log&quot;</span><span class="p">,</span>
            <span class="n">tmp_dir</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;gs://</span><span class="si">{</span><span class="n">WORKSPACE_BUCKET</span><span class="si">}</span><span class="s2">/tmp/4_day&quot;</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">init</span><span class="p">(</span>
            <span class="n">log</span><span class="o">=</span><span class="s2">&quot;compute_coverage.log&quot;</span><span class="p">,</span>
            <span class="n">tmp_dir</span><span class="o">=</span><span class="s2">&quot;gs://gnomad-tmp-4day&quot;</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="n">hl</span><span class="o">.</span><span class="n">default_reference</span><span class="p">(</span><span class="s2">&quot;GRCh38&quot;</span><span class="p">)</span>

    <span class="n">test_2_partitions</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">test_2_partitions</span>
    <span class="n">test_chr22_chrx_chry</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">test_chr22_chrx_chry</span>
    <span class="n">test</span> <span class="o">=</span> <span class="n">test_2_partitions</span> <span class="ow">or</span> <span class="n">test_chr22_chrx_chry</span> <span class="ow">or</span> <span class="n">args</span><span class="o">.</span><span class="n">test</span>
    <span class="n">overwrite</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">overwrite</span>
    <span class="n">n_partitions</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">n_partitions</span>

    <span class="n">chrom</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="n">test_chr22_chrx_chry</span><span class="p">:</span>
        <span class="n">chrom</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;chr22&quot;</span><span class="p">,</span> <span class="s2">&quot;chrX&quot;</span><span class="p">,</span> <span class="s2">&quot;chrY&quot;</span><span class="p">]</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">cov_and_an_ht_path</span> <span class="o">=</span> <span class="n">coverage_and_an_path</span><span class="p">(</span>
            <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">,</span>
            <span class="n">data_set</span><span class="o">=</span><span class="n">project</span><span class="p">,</span>
            <span class="n">environment</span><span class="o">=</span><span class="n">environment</span><span class="p">,</span>
        <span class="p">)</span><span class="o">.</span><span class="n">path</span>
        <span class="n">downsampling_ht_path</span> <span class="o">=</span> <span class="n">get_aou_downsampling</span><span class="p">(</span>
            <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">,</span> <span class="n">environment</span><span class="o">=</span><span class="n">environment</span>
        <span class="p">)</span><span class="o">.</span><span class="n">path</span>
        <span class="n">meta_ht_path</span> <span class="o">=</span> <span class="n">meta</span><span class="p">(</span><span class="n">data_type</span><span class="o">=</span><span class="s2">&quot;genomes&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">path</span>
        <span class="n">group_membership_ht_path</span> <span class="o">=</span> <span class="n">group_membership</span><span class="p">(</span>
            <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">,</span> <span class="n">data_set</span><span class="o">=</span><span class="n">project</span><span class="p">,</span> <span class="n">environment</span><span class="o">=</span><span class="n">environment</span>
        <span class="p">)</span><span class="o">.</span><span class="n">path</span>
        <span class="n">meta_ht</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="n">project</span> <span class="o">==</span> <span class="s2">&quot;aou&quot;</span><span class="p">:</span>
            <span class="n">meta_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="n">meta_ht_path</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">write_aou_downsampling_ht</span><span class="p">:</span>
            <span class="n">check_resource_existence</span><span class="p">(</span>
                <span class="n">output_step_resources</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;downsampling_ht&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">downsampling_ht_path</span><span class="p">]},</span>
                <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="n">ht</span> <span class="o">=</span> <span class="n">meta_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                <span class="p">(</span><span class="n">meta_ht</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">project</span> <span class="o">==</span> <span class="n">project</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">meta_ht</span><span class="o">.</span><span class="n">release</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="n">ds_ht</span> <span class="o">=</span> <span class="n">get_downsampling_ht</span><span class="p">(</span><span class="n">ht</span><span class="p">)</span>
            <span class="n">ds_ht</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">downsampling_ht_path</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">write_group_membership_ht</span><span class="p">:</span>
            <span class="n">check_resource_existence</span><span class="p">(</span>
                <span class="n">output_step_resources</span><span class="o">=</span><span class="p">{</span>
                    <span class="s2">&quot;group_membership_ht&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">group_membership_ht_path</span><span class="p">]</span>
                <span class="p">},</span>
                <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="k">if</span> <span class="n">project</span> <span class="o">==</span> <span class="s2">&quot;gnomad&quot;</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Writing gnomAD group membership HT...&quot;</span><span class="p">)</span>
                <span class="n">v4_meta_ht_path</span> <span class="o">=</span> <span class="n">v4_meta</span><span class="p">(</span><span class="n">data_type</span><span class="o">=</span><span class="s2">&quot;genomes&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">path</span>

                <span class="n">group_membership_ht</span> <span class="o">=</span> <span class="n">get_group_membership_ht</span><span class="p">(</span>
                    <span class="n">hl</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="n">v4_meta_ht_path</span><span class="p">),</span> <span class="n">project</span><span class="o">=</span><span class="n">project</span>
                <span class="p">)</span>
                <span class="n">group_membership_ht</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">group_membership_ht_path</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Writing AoU group membership HT...&quot;</span><span class="p">)</span>
                <span class="n">meta_ht</span> <span class="o">=</span> <span class="n">meta_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                    <span class="p">(</span><span class="n">meta_ht</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">project</span> <span class="o">==</span> <span class="n">project</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">meta_ht</span><span class="o">.</span><span class="n">release</span><span class="p">)</span>
                <span class="p">)</span>
                <span class="n">group_membership_ht</span> <span class="o">=</span> <span class="n">get_group_membership_ht</span><span class="p">(</span>
                    <span class="n">meta_ht</span><span class="o">=</span><span class="n">meta_ht</span><span class="p">,</span>
                    <span class="n">project</span><span class="o">=</span><span class="n">project</span><span class="p">,</span>
                    <span class="n">ds_ht</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="n">downsampling_ht_path</span><span class="p">),</span>
                <span class="p">)</span>
                <span class="n">group_membership_ht</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                    <span class="n">group_membership_ht_path</span><span class="p">,</span>
                    <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">,</span>
                <span class="p">)</span>

        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">compute_all_cov_release_stats_ht</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="s2">&quot;Computing coverage, all sites allele number, and optionally quality histograms HT for </span><span class="si">%s</span><span class="s2">...&quot;</span><span class="p">,</span>
                <span class="n">project</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">check_resource_existence</span><span class="p">(</span>
                <span class="n">output_step_resources</span><span class="o">=</span><span class="p">{</span>
                    <span class="s2">&quot;coverage_and_an_ht&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">cov_and_an_ht_path</span><span class="p">],</span>
                <span class="p">},</span>
                <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="c1"># Context Table is used because it contains every locus in the GRCh38</span>
            <span class="c1"># reference as opposed to a ref-blocked VDS reference dataset.</span>
            <span class="n">ref_ht</span> <span class="o">=</span> <span class="n">vep_context</span><span class="o">.</span><span class="n">versions</span><span class="p">[</span><span class="s2">&quot;105&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">test_chr22_chrx_chry</span><span class="p">:</span>
                <span class="n">ref_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">filter_intervals</span><span class="p">(</span>
                    <span class="n">ref_ht</span><span class="p">,</span> <span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">parse_locus_interval</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">chrom</span><span class="p">]</span>
                <span class="p">)</span>
            <span class="k">elif</span> <span class="n">test_2_partitions</span><span class="p">:</span>
                <span class="n">ref_ht</span> <span class="o">=</span> <span class="n">ref_ht</span><span class="o">.</span><span class="n">_filter_partitions</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">))</span>

            <span class="c1"># Retain only &#39;locus&#39; annotation from context Table.</span>
            <span class="n">ref_ht</span> <span class="o">=</span> <span class="n">ref_ht</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">select</span><span class="p">()</span><span class="o">.</span><span class="n">distinct</span><span class="p">()</span>

            <span class="c1"># The AoU dataset should have removed centromeres and telomeres, but</span>
            <span class="c1"># this serves as an additional safeguard.</span>
            <span class="n">ref_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">filter_intervals</span><span class="p">(</span>
                <span class="n">ref_ht</span><span class="p">,</span>
                <span class="n">telomeres_and_centromeres</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span><span class="o">.</span><span class="n">interval</span><span class="o">.</span><span class="n">collect</span><span class="p">(),</span>
                <span class="n">keep</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">ref_ht</span> <span class="o">=</span> <span class="n">ref_ht</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">new_temp_file</span><span class="p">(</span><span class="s2">&quot;ref&quot;</span><span class="p">,</span> <span class="s2">&quot;ht&quot;</span><span class="p">))</span>

            <span class="k">if</span> <span class="n">project</span> <span class="o">==</span> <span class="s2">&quot;aou&quot;</span><span class="p">:</span>
                <span class="n">sex_karyotype_field</span> <span class="o">=</span> <span class="s2">&quot;meta.sex_karyotype&quot;</span>
                <span class="n">vds</span> <span class="o">=</span> <span class="n">get_aou_vds</span><span class="p">(</span>
                    <span class="n">release_only</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                    <span class="n">filter_partitions</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span> <span class="k">if</span> <span class="n">test_2_partitions</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
                    <span class="n">annotate_meta</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                    <span class="n">chrom</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chr22&quot;</span><span class="p">,</span> <span class="s2">&quot;chrX&quot;</span><span class="p">,</span> <span class="s2">&quot;chrY&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">test_chr22_chrx_chry</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="c1"># NOTE: AoU v8 VDS does not have DP annotation, so using custom function</span>
                <span class="c1"># to annotate adj.</span>
                <span class="c1"># Also adding DP here to make sure `compute_stats_per_ref_site`</span>
                <span class="c1"># doesn&#39;t throw an error.</span>
                <span class="n">vmt</span> <span class="o">=</span> <span class="n">vds</span><span class="o">.</span><span class="n">variant_data</span>
                <span class="n">vmt</span> <span class="o">=</span> <span class="n">annotate_adj_no_dp</span><span class="p">(</span><span class="n">vmt</span><span class="p">)</span>
                <span class="n">vmt</span> <span class="o">=</span> <span class="n">vmt</span><span class="o">.</span><span class="n">annotate_entries</span><span class="p">(</span><span class="n">DP</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">vmt</span><span class="o">.</span><span class="n">LAD</span><span class="p">))</span>
                <span class="n">vds</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">VariantDataset</span><span class="p">(</span><span class="n">vds</span><span class="o">.</span><span class="n">reference_data</span><span class="p">,</span> <span class="n">vmt</span><span class="p">)</span>

                <span class="k">if</span> <span class="n">test</span><span class="p">:</span>
                    <span class="n">meta_ht</span> <span class="o">=</span> <span class="n">meta_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                        <span class="p">(</span><span class="n">meta_ht</span><span class="o">.</span><span class="n">project_meta</span><span class="o">.</span><span class="n">project</span> <span class="o">==</span> <span class="n">project</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">meta_ht</span><span class="o">.</span><span class="n">release</span><span class="p">)</span>
                    <span class="p">)</span><span class="o">.</span><span class="n">select</span><span class="p">()</span>
                    <span class="n">meta_ht</span> <span class="o">=</span> <span class="n">meta_ht</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="mf">0.001</span><span class="p">)</span>
                    <span class="n">vds</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">vds</span><span class="o">.</span><span class="n">filter_samples</span><span class="p">(</span><span class="n">vds</span><span class="p">,</span> <span class="n">meta_ht</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">sex_karyotype_field</span> <span class="o">=</span> <span class="s2">&quot;meta.sex_imputation.sex_karyotype&quot;</span>
                <span class="n">vds</span> <span class="o">=</span> <span class="n">get_gnomad_v5_genomes_vds</span><span class="p">(</span>
                    <span class="n">release_only</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                    <span class="n">consent_drop_only</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                    <span class="n">filter_partitions</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span> <span class="k">if</span> <span class="n">test_2_partitions</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
                    <span class="n">annotate_meta</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                    <span class="n">chrom</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chr22&quot;</span><span class="p">,</span> <span class="s2">&quot;chrX&quot;</span><span class="p">,</span> <span class="s2">&quot;chrY&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">test_chr22_chrx_chry</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
                <span class="p">)</span>

            <span class="n">validate_vds</span><span class="p">(</span><span class="n">vds</span><span class="p">)</span>

            <span class="n">cov_and_an_ht</span> <span class="o">=</span> <span class="n">compute_all_release_stats_per_ref_site</span><span class="p">(</span>
                <span class="n">vds</span><span class="p">,</span>
                <span class="n">ref_ht</span><span class="p">,</span>
                <span class="n">sex_karyotype_field</span><span class="o">=</span><span class="n">sex_karyotype_field</span><span class="p">,</span>
                <span class="n">project</span><span class="o">=</span><span class="n">project</span><span class="p">,</span>
                <span class="n">group_membership_ht</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="n">group_membership_ht_path</span><span class="p">),</span>
            <span class="p">)</span>
            <span class="n">cov_and_an_ht</span> <span class="o">=</span> <span class="n">cov_and_an_ht</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span>
                <span class="n">new_temp_file</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">project</span><span class="si">}</span><span class="s2">_cov_and_an&quot;</span><span class="p">,</span> <span class="s2">&quot;ht&quot;</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="c1"># Write out the intermediate HT.</span>
            <span class="n">cov_and_an_ht</span> <span class="o">=</span> <span class="n">cov_and_an_ht</span><span class="o">.</span><span class="n">naive_coalesce</span><span class="p">(</span><span class="n">n_partitions</span><span class="p">)</span>
            <span class="n">cov_and_an_ht</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">cov_and_an_ht_path</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">merge_gnomad_coverage</span><span class="p">:</span>
            <span class="n">merged_gnomad_coverage_ht_path</span> <span class="o">=</span> <span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="p">(</span><span class="n">qc_temp_prefix</span><span class="p">())</span><span class="si">}</span><span class="s2">gnomad_v5_genomes_coverage.ht&quot;</span>
            <span class="p">)</span>
            <span class="n">check_resource_existence</span><span class="p">(</span>
                <span class="n">output_step_resources</span><span class="o">=</span><span class="p">{</span>
                    <span class="s2">&quot;merged_gnomad_coverage_ht&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">merged_gnomad_coverage_ht_path</span><span class="p">],</span>
                <span class="p">},</span>
                <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="n">gnomad_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="n">cov_and_an_ht_path</span><span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;AN&quot;</span><span class="p">)</span>
            <span class="n">gnomad_release_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span>
                <span class="n">release_coverage_path</span><span class="p">(</span>
                    <span class="n">release_version</span><span class="o">=</span><span class="n">v4_COVERAGE_RELEASE</span><span class="p">,</span>
                    <span class="n">public</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">test_chr22_chrx_chry</span><span class="p">:</span>
                <span class="n">gnomad_release_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">filter_intervals</span><span class="p">(</span>
                    <span class="n">gnomad_release_ht</span><span class="p">,</span> <span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">parse_locus_interval</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">chrom</span><span class="p">]</span>
                <span class="p">)</span>
            <span class="k">elif</span> <span class="n">test_2_partitions</span><span class="p">:</span>
                <span class="n">gnomad_release_ht</span> <span class="o">=</span> <span class="n">gnomad_release_ht</span><span class="o">.</span><span class="n">_filter_partitions</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">))</span>

            <span class="n">ht</span> <span class="o">=</span> <span class="n">merge_gnomad_coverage_hts</span><span class="p">(</span><span class="n">gnomad_ht</span><span class="p">,</span> <span class="n">gnomad_release_ht</span><span class="p">)</span>
            <span class="n">ht</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">merged_gnomad_coverage_ht_path</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">merge_gnomad_an</span><span class="p">:</span>
            <span class="n">merged_gnomad_an_ht_path</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">qc_temp_prefix</span><span class="p">()</span><span class="si">}</span><span class="s2">gnomad_v5_genomes_an.ht&quot;</span>
            <span class="n">check_resource_existence</span><span class="p">(</span>
                <span class="n">output_step_resources</span><span class="o">=</span><span class="p">{</span>
                    <span class="s2">&quot;merged_gnomad_an_ht&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">merged_gnomad_an_ht_path</span><span class="p">],</span>
                <span class="p">},</span>
                <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="n">gnomad_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="n">cov_and_an_ht_path</span><span class="p">)</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;AN&quot;</span><span class="p">)</span>
            <span class="n">gnomad_release_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span>
                <span class="n">release_coverage_path</span><span class="p">(</span>
                    <span class="n">release_version</span><span class="o">=</span><span class="n">v4_AN_RELEASE</span><span class="p">,</span>
                    <span class="n">public</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                    <span class="n">coverage_type</span><span class="o">=</span><span class="s2">&quot;allele_number&quot;</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">)</span>

            <span class="k">if</span> <span class="n">test_chr22_chrx_chry</span><span class="p">:</span>
                <span class="n">gnomad_release_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">filter_intervals</span><span class="p">(</span>
                    <span class="n">gnomad_release_ht</span><span class="p">,</span> <span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">parse_locus_interval</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">chrom</span><span class="p">]</span>
                <span class="p">)</span>
            <span class="k">elif</span> <span class="n">test_2_partitions</span><span class="p">:</span>
                <span class="n">gnomad_release_ht</span> <span class="o">=</span> <span class="n">gnomad_release_ht</span><span class="o">.</span><span class="n">_filter_partitions</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">))</span>

            <span class="n">ht</span> <span class="o">=</span> <span class="n">merge_gnomad_an_hts</span><span class="p">(</span><span class="n">gnomad_ht</span><span class="p">,</span> <span class="n">gnomad_release_ht</span><span class="p">)</span>
            <span class="n">ht</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">merged_gnomad_an_ht_path</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">export_coverage_release_files</span><span class="p">:</span>
            <span class="n">cov_ht_path</span> <span class="o">=</span> <span class="n">release_coverage_path</span><span class="p">(</span>
                <span class="n">public</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">,</span>
                <span class="n">coverage_type</span><span class="o">=</span><span class="s2">&quot;coverage&quot;</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">cov_tsv_path</span> <span class="o">=</span> <span class="n">release_coverage_tsv_path</span><span class="p">(</span><span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">)</span>
            <span class="n">gnomad_coverage_ht_path</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">qc_temp_prefix</span><span class="p">()</span><span class="si">}</span><span class="s2">gnomad_v5_genomes_coverage.ht&quot;</span>
            <span class="n">check_resource_existence</span><span class="p">(</span>
                <span class="n">input_step_resources</span><span class="o">=</span><span class="p">{</span>
                    <span class="s2">&quot;gnomad_coverage_ht&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">gnomad_coverage_ht_path</span><span class="p">],</span>
                <span class="p">},</span>
                <span class="n">output_step_resources</span><span class="o">=</span><span class="p">{</span>
                    <span class="s2">&quot;cov_release_ht&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">cov_ht_path</span><span class="p">],</span>
                    <span class="s2">&quot;cov_tsv&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">cov_tsv_path</span><span class="p">],</span>
                <span class="p">},</span>
                <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Exporting coverage HT and TSV...&quot;</span><span class="p">)</span>
            <span class="n">aou_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="n">cov_and_an_ht_path</span><span class="p">)</span>
            <span class="n">aou_ht</span> <span class="o">=</span> <span class="n">aou_ht</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;AN&quot;</span><span class="p">,</span> <span class="s2">&quot;qual_hists&quot;</span><span class="p">)</span>
            <span class="n">gnomad_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="n">gnomad_coverage_ht_path</span><span class="p">)</span>

            <span class="n">ht</span> <span class="o">=</span> <span class="n">join_aou_and_gnomad_coverage_ht</span><span class="p">(</span><span class="n">aou_ht</span><span class="p">,</span> <span class="n">gnomad_ht</span><span class="p">)</span>
            <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span><span class="n">new_temp_file</span><span class="p">(</span><span class="s2">&quot;aou_and_gnomad_cov_join&quot;</span><span class="p">,</span> <span class="s2">&quot;ht&quot;</span><span class="p">))</span>
            <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">naive_coalesce</span><span class="p">(</span><span class="n">n_partitions</span><span class="p">)</span>
            <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span><span class="n">cov_ht_path</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">)</span>
            <span class="n">ht</span><span class="o">.</span><span class="n">export</span><span class="p">(</span><span class="n">cov_tsv_path</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">export_an_release_files</span><span class="p">:</span>
            <span class="n">an_ht_path</span> <span class="o">=</span> <span class="n">release_coverage_path</span><span class="p">(</span>
                <span class="n">public</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">,</span>
                <span class="n">coverage_type</span><span class="o">=</span><span class="s2">&quot;allele_number&quot;</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">an_tsv_path</span> <span class="o">=</span> <span class="n">release_all_sites_an_tsv_path</span><span class="p">(</span><span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">)</span>
            <span class="n">gnomad_an_ht_path</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">qc_temp_prefix</span><span class="p">()</span><span class="si">}</span><span class="s2">gnomad_v5_genomes_an.ht&quot;</span>
            <span class="n">check_resource_existence</span><span class="p">(</span>
                <span class="n">input_step_resources</span><span class="o">=</span><span class="p">{</span>
                    <span class="s2">&quot;gnomad_an_ht&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">gnomad_an_ht_path</span><span class="p">],</span>
                <span class="p">},</span>
                <span class="n">output_step_resources</span><span class="o">=</span><span class="p">{</span>
                    <span class="s2">&quot;an_release_ht&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">an_ht_path</span><span class="p">],</span>
                    <span class="s2">&quot;an_release_tsv&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">an_tsv_path</span><span class="p">],</span>
                <span class="p">},</span>
                <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Exporting AN HT and TSV...&quot;</span><span class="p">)</span>
            <span class="n">aou_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="n">cov_and_an_ht_path</span><span class="p">)</span>
            <span class="n">aou_ht</span> <span class="o">=</span> <span class="n">aou_ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;AN&quot;</span><span class="p">)</span>
            <span class="n">gnomad_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="n">gnomad_an_ht_path</span><span class="p">)</span>

            <span class="n">ht</span> <span class="o">=</span> <span class="n">join_aou_and_gnomad_an_ht</span><span class="p">(</span><span class="n">aou_ht</span><span class="p">,</span> <span class="n">gnomad_ht</span><span class="p">)</span>
            <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span><span class="n">new_temp_file</span><span class="p">(</span><span class="s2">&quot;aou_and_gnomad_an_join&quot;</span><span class="p">,</span> <span class="s2">&quot;ht&quot;</span><span class="p">))</span>
            <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;AN&quot;</span><span class="p">)</span>
            <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">select_globals</span><span class="p">(</span>
                <span class="n">strata_meta</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">strata_meta</span><span class="p">,</span>
                <span class="n">strata_sample_count</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">strata_sample_count</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">naive_coalesce</span><span class="p">(</span><span class="n">n_partitions</span><span class="p">)</span>
            <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span><span class="n">an_ht_path</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">)</span>

            <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">transmute</span><span class="p">(</span><span class="n">AN</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">AN</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">ht</span><span class="o">.</span><span class="n">export</span><span class="p">(</span><span class="n">an_tsv_path</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">merge_qual_hists</span><span class="p">:</span>
            <span class="n">qual_hists_path</span> <span class="o">=</span> <span class="n">qual_hists</span><span class="p">(</span><span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">,</span> <span class="n">environment</span><span class="o">=</span><span class="n">environment</span><span class="p">)</span><span class="o">.</span><span class="n">path</span>
            <span class="n">check_resource_existence</span><span class="p">(</span>
                <span class="n">output_step_resources</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;qual_hists_ht&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">qual_hists_path</span><span class="p">]},</span>
                <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Merging qual hists HTs...&quot;</span><span class="p">)</span>
            <span class="n">aou_ht</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">hl</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="n">cov_and_an_ht_path</span><span class="p">)</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;qual_hists&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">select_globals</span><span class="p">()</span>
            <span class="p">)</span>
            <span class="n">gnomad_ht</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">public_release</span><span class="p">(</span><span class="n">data_type</span><span class="o">=</span><span class="s2">&quot;genomes&quot;</span><span class="p">)</span>
                <span class="o">.</span><span class="n">ht</span><span class="p">()</span>
                <span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;histograms&quot;</span><span class="p">)</span>
                <span class="o">.</span><span class="n">select_globals</span><span class="p">()</span>
            <span class="p">)</span>

            <span class="c1"># Drop age hists because they are handled in the frequency script</span>
            <span class="c1"># and re-key by locus.</span>
            <span class="n">gnomad_ht</span> <span class="o">=</span> <span class="n">gnomad_ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
                <span class="n">qual_hists</span><span class="o">=</span><span class="n">gnomad_ht</span><span class="o">.</span><span class="n">histograms</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;age_hists&quot;</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="n">gnomad_ht</span> <span class="o">=</span> <span class="n">gnomad_ht</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">)</span>

            <span class="n">ht</span> <span class="o">=</span> <span class="n">join_aou_and_gnomad_qual_hists_ht</span><span class="p">(</span><span class="n">aou_ht</span><span class="p">,</span> <span class="n">gnomad_ht</span><span class="p">)</span>
            <span class="n">ht</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">qual_hists_path</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">)</span>

    <span class="k">finally</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Copying hail log to logging bucket...&quot;</span><span class="p">)</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">copy_log</span><span class="p">(</span><span class="n">get_logging_path</span><span class="p">(</span><span class="s2">&quot;compute_coverage&quot;</span><span class="p">,</span> <span class="n">environment</span><span class="o">=</span><span class="n">environment</span><span class="p">))</span></div>


<span class="k">def</span> <span class="nf">get_script_argument_parser</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get script argument parser.&quot;&quot;&quot;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">()</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--project-name&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Project name. Determines environment where script will run.&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;aou&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;aou&quot;</span><span class="p">,</span> <span class="s2">&quot;gnomad&quot;</span><span class="p">],</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--overwrite&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Overwrite existing hail Tables.&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--n-partitions&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Number of partitions to use for the output Table.&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mi">5000</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">test_group</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_mutually_exclusive_group</span><span class="p">()</span>
    <span class="n">test_group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--test-2-partitions&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span>
            <span class="s2">&quot;Whether to run a test using only the first 2 partitions of the VDS test&quot;</span>
            <span class="s2">&quot; dataset.&quot;</span>
        <span class="p">),</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">test_group</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--test-chr22-chrx-chry&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span>
            <span class="s2">&quot;Whether to run a test using only the chr22, chrX, and chrY chromosomes of&quot;</span>
            <span class="s2">&quot; the VDS test dataset.&quot;</span>
        <span class="p">),</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">group_membership_args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span>
        <span class="s2">&quot;Get gnomAD genomes group membership HT.&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group_membership_args</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--write-group-membership-ht&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Write group membership HT.&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">group_membership_args</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--test&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Write test group membership HT to test path.&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--write-aou-downsampling-ht&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Write v5 downsampling HT.&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--compute-all-cov-release-stats-ht&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Compute the all sites coverage, allele number, and quality histogram HT.&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">coverage_args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span>
        <span class="s2">&quot;Compute coverage release stats HT.&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">coverage_args</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--merge-gnomad-coverage&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Subtract consent drop samples from v4 release HT to create gnomAD v5 genomes coverage HT.&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">coverage_args</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--export-coverage-release-files&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Join and export AoU + gnomAD v4 coverage release HT and TSV file.&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">an_args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span>
        <span class="s2">&quot;Compute AN release stats HT.&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">an_args</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--merge-gnomad-an&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Subtract consent drop samples from v4 release HT to create gnomAD v5 genomes AN HT.&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">an_args</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--export-an-release-files&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Exports joint AoU + gnomAD v4 AN release HT and TSV file.&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--merge-qual-hists&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Merge variant quality histograms from AoU v8 and gnomAD v4 genomes.&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">parser</span>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">get_script_argument_parser</span><span class="p">()</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">write_aou_downsampling_ht</span> <span class="ow">and</span> <span class="n">args</span><span class="o">.</span><span class="n">project_name</span> <span class="o">!=</span> <span class="s2">&quot;aou&quot;</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;--write-aou-downsampling-ht requires --project-name to be &#39;aou&#39;.&quot;</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">merge_gnomad_coverage</span> <span class="ow">and</span> <span class="n">args</span><span class="o">.</span><span class="n">project_name</span> <span class="o">!=</span> <span class="s2">&quot;gnomad&quot;</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;--merge-gnomad-coverage requires --project-name to be &#39;gnomad&#39;.&quot;</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">merge_gnomad_an</span> <span class="ow">and</span> <span class="n">args</span><span class="o">.</span><span class="n">project_name</span> <span class="o">!=</span> <span class="s2">&quot;gnomad&quot;</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;--merge-gnomad-an requires --project-name to be &#39;gnomad&#39;.&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">export_coverage_release_files</span> <span class="ow">and</span> <span class="n">args</span><span class="o">.</span><span class="n">project_name</span> <span class="o">!=</span> <span class="s2">&quot;aou&quot;</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;--export-coverage-release-files requires --project-name to be &#39;aou&#39;.&quot;</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">export_an_release_files</span> <span class="ow">and</span> <span class="n">args</span><span class="o">.</span><span class="n">project_name</span> <span class="o">!=</span> <span class="s2">&quot;aou&quot;</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;--export-an-release-files requires --project-name to be &#39;aou&#39;.&quot;</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">merge_qual_hists</span> <span class="ow">and</span> <span class="n">args</span><span class="o">.</span><span class="n">project_name</span> <span class="o">!=</span> <span class="s2">&quot;aou&quot;</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;--merge-qual-hists requires --project-name to be &#39;aou&#39;.&quot;</span><span class="p">)</span>

    <span class="n">main</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p></p>
  </div>

   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>