<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>gnomad_qc.v4.create_release.validate_and_export_vcf &mdash; gnomad_qc master documentation</title>
      <link rel="stylesheet" href="../../../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../../../_static/theme_overrides.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../../_static/jquery.js"></script>
        <script src="../../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script data-url_root="../../../../" id="documentation_options" src="../../../../_static/documentation_options.js"></script>
        <script src="../../../../_static/doctools.js"></script>
        <script src="../../../../_static/sphinx_highlight.js"></script>
        <script src="https://unpkg.com/mermaid@10.2.0/dist/mermaid.min.js"></script>
        <script>mermaid.initialize({startOnLoad:true});</script>
    <script src="../../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../../index.html" class="icon icon-home">
            gnomad_qc
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../api_reference/index.html">API Reference</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/broadinstitute/gnomad_qc/tags">Change Log</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../../index.html">gnomad_qc</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">gnomad_qc.v4.create_release.validate_and_export_vcf</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for gnomad_qc.v4.create_release.validate_and_export_vcf</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;Script to validate and export gnomAD VCFs.&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">pickle</span>
<span class="kn">from</span> <span class="nn">copy</span> <span class="kn">import</span> <span class="n">deepcopy</span>
<span class="kn">from</span> <span class="nn">pprint</span> <span class="kn">import</span> <span class="n">pprint</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">List</span><span class="p">,</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Set</span><span class="p">,</span> <span class="n">Tuple</span>

<span class="kn">import</span> <span class="nn">hail</span> <span class="k">as</span> <span class="nn">hl</span>
<span class="kn">from</span> <span class="nn">gnomad.assessment.validity_checks</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">check_global_and_row_annot_lengths</span><span class="p">,</span>
    <span class="n">pprint_global_anns</span><span class="p">,</span>
    <span class="n">validate_release_t</span><span class="p">,</span>
    <span class="n">vcf_field_check</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">gnomad.resources.grch38.gnomad</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">GEN_ANC_GROUPS</span><span class="p">,</span>
    <span class="n">HGDP_GEN_ANC_GROUPS</span><span class="p">,</span>
    <span class="n">SUBSETS</span><span class="p">,</span>
    <span class="n">TGP_GEN_ANC_GROUPS</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">gnomad.sample_qc.ancestry</span> <span class="kn">import</span> <span class="n">GEN_ANC_NAMES</span>
<span class="kn">from</span> <span class="nn">gnomad.utils.filtering</span> <span class="kn">import</span> <span class="n">remove_fields_from_constant</span>
<span class="kn">from</span> <span class="nn">gnomad.utils.vcf</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">ALLELE_TYPE_FIELDS</span><span class="p">,</span>
    <span class="n">AS_FIELDS</span><span class="p">,</span>
    <span class="n">AS_VQSR_FIELDS</span><span class="p">,</span>
    <span class="n">FAF_GEN_ANC_GROUPS</span><span class="p">,</span>
    <span class="n">HISTS</span><span class="p">,</span>
    <span class="n">IN_SILICO_ANNOTATIONS_INFO_DICT</span><span class="p">,</span>
    <span class="n">INFO_DICT</span><span class="p">,</span>
    <span class="n">JOINT_FILTERS_INFO_DICT</span><span class="p">,</span>
    <span class="n">JOINT_REGION_FLAG_FIELDS</span><span class="p">,</span>
    <span class="n">JOINT_REGION_FLAGS_INFO_DICT</span><span class="p">,</span>
    <span class="n">REGION_FLAG_FIELDS</span><span class="p">,</span>
    <span class="n">SEXES</span><span class="p">,</span>
    <span class="n">SITE_FIELDS</span><span class="p">,</span>
    <span class="n">VRS_FIELDS_DICT</span><span class="p">,</span>
    <span class="n">add_as_info_dict</span><span class="p">,</span>
    <span class="n">adjust_vcf_incompatible_types</span><span class="p">,</span>
    <span class="n">build_vcf_export_reference</span><span class="p">,</span>
    <span class="n">create_label_groups</span><span class="p">,</span>
    <span class="n">make_hist_bin_edges_expr</span><span class="p">,</span>
    <span class="n">make_hist_dict</span><span class="p">,</span>
    <span class="n">make_info_dict</span><span class="p">,</span>
    <span class="n">make_vcf_filter_dict</span><span class="p">,</span>
    <span class="n">rekey_new_reference</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">gnomad.utils.vep</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">CURRENT_VEP_VERSION</span><span class="p">,</span>
    <span class="n">VEP_CSQ_FIELDS</span><span class="p">,</span>
    <span class="n">VEP_CSQ_HEADER</span><span class="p">,</span>
    <span class="n">vep_struct_to_csq</span><span class="p">,</span>
<span class="p">)</span>

<span class="kn">from</span> <span class="nn">gnomad_qc.resource_utils</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">PipelineResourceCollection</span><span class="p">,</span>
    <span class="n">PipelineStepResourceCollection</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v4.resources.basics</span> <span class="kn">import</span> <span class="n">get_logging_path</span><span class="p">,</span> <span class="n">qc_temp_prefix</span>
<span class="kn">from</span> <span class="nn">gnomad_qc.v4.resources.release</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">append_to_vcf_header_path</span><span class="p">,</span>
    <span class="n">release_header_path</span><span class="p">,</span>
    <span class="n">release_sites</span><span class="p">,</span>
    <span class="n">release_vcf_path</span><span class="p">,</span>
    <span class="n">validated_release_ht</span><span class="p">,</span>
<span class="p">)</span>

<span class="c1"># Add new site fields</span>
<span class="n">NEW_SITE_FIELDS</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;monoallelic&quot;</span><span class="p">,</span>
    <span class="s2">&quot;only_het&quot;</span><span class="p">,</span>
    <span class="s2">&quot;transmitted_singleton&quot;</span><span class="p">,</span>
<span class="p">]</span>
<span class="n">SITE_FIELDS</span> <span class="o">=</span> <span class="n">deepcopy</span><span class="p">(</span><span class="n">SITE_FIELDS</span><span class="p">)</span>
<span class="n">SITE_FIELDS</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">NEW_SITE_FIELDS</span><span class="p">)</span>
<span class="n">SITE_FIELDS</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;exomes&quot;</span><span class="p">:</span> <span class="n">SITE_FIELDS</span> <span class="o">+</span> <span class="p">[</span><span class="s2">&quot;sibling_singleton&quot;</span><span class="p">],</span>
    <span class="s2">&quot;genomes&quot;</span><span class="p">:</span> <span class="n">SITE_FIELDS</span><span class="p">,</span>
<span class="p">}</span>

<span class="c1"># Add updatedVQSR fields</span>
<span class="n">NEW_AS_VQSR_FIELDS</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;negative_train_site&quot;</span><span class="p">,</span> <span class="s2">&quot;positive_train_site&quot;</span><span class="p">]</span>
<span class="n">AS_VQSR_FIELDSS</span> <span class="o">=</span> <span class="n">deepcopy</span><span class="p">(</span><span class="n">AS_VQSR_FIELDS</span><span class="p">)</span>
<span class="n">AS_VQSR_FIELDS</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">NEW_AS_VQSR_FIELDS</span><span class="p">)</span>

<span class="c1"># Update inbreeding coeff</span>
<span class="n">MISSING_AS_FIELDS</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;InbreedingCoeff&quot;</span><span class="p">]</span>
<span class="n">AS_FIELDS</span> <span class="o">=</span> <span class="n">remove_fields_from_constant</span><span class="p">(</span><span class="n">AS_FIELDS</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;InbreedingCoeff&quot;</span><span class="p">])</span>
<span class="n">AS_FIELDS</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;inbreeding_coeff&quot;</span><span class="p">)</span>

<span class="c1"># Drop decoy, still doesn&#39;t exist on 38</span>
<span class="n">REGION_FLAG_FIELDS</span> <span class="o">=</span> <span class="n">deepcopy</span><span class="p">(</span><span class="n">REGION_FLAG_FIELDS</span><span class="p">)</span>
<span class="n">REGION_FLAG_FIELDS</span> <span class="o">=</span> <span class="n">remove_fields_from_constant</span><span class="p">(</span>
    <span class="n">REGION_FLAG_FIELDS</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;decoy&quot;</span><span class="p">,</span> <span class="s2">&quot;nonpar&quot;</span><span class="p">]</span>
<span class="p">)</span>
<span class="n">REGION_FLAG_FIELDS</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;exomes&quot;</span><span class="p">:</span> <span class="n">REGION_FLAG_FIELDS</span>
    <span class="o">+</span> <span class="p">[</span>
        <span class="s2">&quot;fail_interval_qc&quot;</span><span class="p">,</span>
        <span class="s2">&quot;outside_ukb_capture_region&quot;</span><span class="p">,</span>
        <span class="s2">&quot;outside_broad_capture_region&quot;</span><span class="p">,</span>
    <span class="p">],</span>
    <span class="s2">&quot;genomes&quot;</span><span class="p">:</span> <span class="n">REGION_FLAG_FIELDS</span><span class="p">,</span>
    <span class="s2">&quot;joint&quot;</span><span class="p">:</span> <span class="n">JOINT_REGION_FLAG_FIELDS</span><span class="p">,</span>
<span class="p">}</span>

<span class="c1"># Remove original alleles for containing non-releasable alleles</span>
<span class="n">ALLELE_TYPE_FIELDS</span> <span class="o">=</span> <span class="n">deepcopy</span><span class="p">(</span><span class="n">ALLELE_TYPE_FIELDS</span><span class="p">)</span>
<span class="n">ALLELE_TYPE_FIELDS</span> <span class="o">=</span> <span class="n">remove_fields_from_constant</span><span class="p">(</span>
    <span class="n">ALLELE_TYPE_FIELDS</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;original_alleles&quot;</span><span class="p">]</span>
<span class="p">)</span>
<span class="c1"># Remove original alleles for containing non-releasable alleles</span>
<span class="n">ALLELE_TYPE_FIELDS</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;exomes&quot;</span><span class="p">:</span> <span class="n">ALLELE_TYPE_FIELDS</span><span class="p">,</span>
    <span class="s2">&quot;genomes&quot;</span><span class="p">:</span> <span class="n">remove_fields_from_constant</span><span class="p">(</span><span class="n">ALLELE_TYPE_FIELDS</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;has_star&quot;</span><span class="p">]),</span>
<span class="p">}</span>

<span class="n">INSILICO_FIELDS</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;cadd&quot;</span><span class="p">,</span>
    <span class="s2">&quot;revel_max&quot;</span><span class="p">,</span>
    <span class="s2">&quot;spliceai_ds_max&quot;</span><span class="p">,</span>
    <span class="s2">&quot;pangolin_largest_ds&quot;</span><span class="p">,</span>
    <span class="s2">&quot;phylop&quot;</span><span class="p">,</span>
    <span class="s2">&quot;sift_max&quot;</span><span class="p">,</span>
    <span class="s2">&quot;polyphen_max&quot;</span><span class="p">,</span>
<span class="p">]</span>

<span class="n">GENOME_SUBSETS_TO_DROP</span> <span class="o">=</span> <span class="n">remove_fields_from_constant</span><span class="p">(</span>
    <span class="n">deepcopy</span><span class="p">(</span><span class="n">SUBSETS</span><span class="p">[</span><span class="s2">&quot;v3&quot;</span><span class="p">]),</span> <span class="p">[</span><span class="s2">&quot;hgdp&quot;</span><span class="p">,</span> <span class="s2">&quot;tgp&quot;</span><span class="p">]</span>
<span class="p">)</span>
<span class="n">SUBSETS</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;exomes&quot;</span><span class="p">:</span> <span class="n">deepcopy</span><span class="p">(</span><span class="n">SUBSETS</span><span class="p">[</span><span class="s2">&quot;v4&quot;</span><span class="p">]),</span>
    <span class="s2">&quot;genomes&quot;</span><span class="p">:</span> <span class="n">remove_fields_from_constant</span><span class="p">(</span>
        <span class="n">deepcopy</span><span class="p">(</span><span class="n">SUBSETS</span><span class="p">[</span><span class="s2">&quot;v3&quot;</span><span class="p">]),</span> <span class="n">GENOME_SUBSETS_TO_DROP</span>
    <span class="p">),</span>
    <span class="s2">&quot;joint&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;&quot;</span><span class="p">],</span>
<span class="p">}</span>

<span class="n">GEN_ANC_GROUPS</span> <span class="o">=</span> <span class="n">deepcopy</span><span class="p">(</span><span class="n">GEN_ANC_GROUPS</span><span class="p">[</span><span class="s2">&quot;v4&quot;</span><span class="p">])</span>
<span class="n">GEN_ANC_GROUPS</span><span class="p">[</span><span class="s2">&quot;joint&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">GEN_ANC_GROUPS</span><span class="p">[</span><span class="s2">&quot;exomes&quot;</span><span class="p">])</span> <span class="o">|</span> <span class="nb">set</span><span class="p">(</span><span class="n">GEN_ANC_GROUPS</span><span class="p">[</span><span class="s2">&quot;genomes&quot;</span><span class="p">])</span>

<span class="c1"># Remove unnecessary pop names from GEN_ANC_NAMES dict</span>
<span class="n">GEN_ANC_GROUPS</span> <span class="o">=</span> <span class="p">{</span>
    <span class="n">d</span><span class="p">:</span> <span class="p">{</span><span class="n">pop</span><span class="p">:</span> <span class="n">GEN_ANC_NAMES</span><span class="p">[</span><span class="n">pop</span><span class="p">]</span> <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">pops</span><span class="p">}</span> <span class="k">for</span> <span class="n">d</span><span class="p">,</span> <span class="n">pops</span> <span class="ow">in</span> <span class="n">GEN_ANC_GROUPS</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
<span class="p">}</span>

<span class="n">SAMPLE_SUM_SETS_AND_POPS</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;exomes&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;non_ukb&quot;</span><span class="p">:</span> <span class="n">GEN_ANC_GROUPS</span><span class="p">[</span><span class="s2">&quot;exomes&quot;</span><span class="p">]},</span>
    <span class="s2">&quot;genomes&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;hgdp&quot;</span><span class="p">:</span> <span class="n">HGDP_GEN_ANC_GROUPS</span><span class="p">,</span> <span class="s2">&quot;tgp&quot;</span><span class="p">:</span> <span class="n">TGP_GEN_ANC_GROUPS</span><span class="p">},</span>
    <span class="s2">&quot;joint&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">}</span>

<span class="c1"># Row annotaions and their associated global annotations for length comparison</span>
<span class="n">LEN_COMP_GLOBAL_ROWS</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;freq&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;freq_meta&quot;</span><span class="p">,</span> <span class="s2">&quot;freq_index_dict&quot;</span><span class="p">,</span> <span class="s2">&quot;freq_meta_sample_count&quot;</span><span class="p">],</span>
    <span class="s2">&quot;faf&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;faf_meta&quot;</span><span class="p">,</span> <span class="s2">&quot;faf_index_dict&quot;</span><span class="p">],</span>
<span class="p">}</span>

<span class="c1"># Joint row annotations and their associated global annotations for length comparison</span>
<span class="n">LEN_COMP_JOINT_GLOBAL_ROWS</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;joint_freq&quot;</span><span class="p">:</span> <span class="p">[</span>
        <span class="s2">&quot;joint_freq_meta&quot;</span><span class="p">,</span>
        <span class="s2">&quot;joint_freq_index_dict&quot;</span><span class="p">,</span>
        <span class="s2">&quot;joint_freq_meta_sample_count&quot;</span><span class="p">,</span>
    <span class="p">],</span>
    <span class="s2">&quot;joint_faf&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;joint_faf_meta&quot;</span><span class="p">,</span> <span class="s2">&quot;joint_faf_index_dict&quot;</span><span class="p">],</span>
<span class="p">}</span>


<span class="c1"># VCF INFO fields to reorder.</span>
<span class="n">VCF_INFO_REORDER</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;AC&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AN&quot;</span><span class="p">,</span>
    <span class="s2">&quot;AF&quot;</span><span class="p">,</span>
    <span class="s2">&quot;grpmax&quot;</span><span class="p">,</span>
    <span class="s2">&quot;fafmax_faf95_max&quot;</span><span class="p">,</span>
    <span class="s2">&quot;fafmax_faf95_max_gen_anc&quot;</span><span class="p">,</span>
<span class="p">]</span>


<span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span>
    <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">%(asctime)s</span><span class="s2"> (</span><span class="si">%(name)s</span><span class="s2"> </span><span class="si">%(lineno)s</span><span class="s2">): </span><span class="si">%(message)s</span><span class="s2">&quot;</span><span class="p">,</span>
    <span class="n">datefmt</span><span class="o">=</span><span class="s2">&quot;%m/</span><span class="si">%d</span><span class="s2">/%Y %I:%M:%S %p&quot;</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>
<span class="n">logger</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>


<div class="viewcode-block" id="get_export_resources"><a class="viewcode-back" href="../../../../api_reference/v4/create_release/validate_and_export_vcf.html#gnomad_qc.v4.create_release.validate_and_export_vcf.get_export_resources">[docs]</a><span class="k">def</span> <span class="nf">get_export_resources</span><span class="p">(</span>
    <span class="n">overwrite</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">data_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;exomes&quot;</span><span class="p">,</span>
    <span class="n">test</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">bool</span><span class="p">]</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">contig</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">PipelineResourceCollection</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get export resources.</span>

<span class="sd">    :param overwrite: Whether to overwrite existing files.</span>
<span class="sd">    :param data_type: Data type to get resources for. One of &quot;exomes&quot; or &quot;genomes&quot;.</span>
<span class="sd">        Default is &quot;exomes&quot;.</span>
<span class="sd">    :param test: Whether to use test resources.</span>
<span class="sd">    :param contig: Contig to get resources for. Default is None.</span>
<span class="sd">    :return: Export resources.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">export_pipeline</span> <span class="o">=</span> <span class="n">PipelineResourceCollection</span><span class="p">(</span>
        <span class="n">pipeline_name</span><span class="o">=</span><span class="s2">&quot;validate_and_export_vcf&quot;</span><span class="p">,</span>
        <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">validate_release_ht</span> <span class="o">=</span> <span class="n">PipelineStepResourceCollection</span><span class="p">(</span>
        <span class="s2">&quot;--validate-release-ht&quot;</span><span class="p">,</span>
        <span class="n">input_resources</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;create_release_sites_ht.py&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;release_ht&quot;</span><span class="p">:</span> <span class="n">release_sites</span><span class="p">(</span><span class="n">data_type</span><span class="o">=</span><span class="n">data_type</span><span class="p">,</span> <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">)</span>
            <span class="p">}</span>
        <span class="p">},</span>
        <span class="n">output_resources</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;validated_ht&quot;</span><span class="p">:</span> <span class="n">validated_release_ht</span><span class="p">(</span><span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">,</span> <span class="n">data_type</span><span class="o">=</span><span class="n">data_type</span><span class="p">)</span>
        <span class="p">},</span>
    <span class="p">)</span>
    <span class="n">prepare_vcf_header</span> <span class="o">=</span> <span class="n">PipelineStepResourceCollection</span><span class="p">(</span>
        <span class="s2">&quot;--prepare-vcf-header&quot;</span><span class="p">,</span>
        <span class="n">pipeline_input_steps</span><span class="o">=</span><span class="p">[</span><span class="n">validate_release_ht</span><span class="p">],</span>
        <span class="n">add_input_resources</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;create_release_sites_ht.py&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;release_ht&quot;</span><span class="p">:</span> <span class="n">release_sites</span><span class="p">(</span><span class="n">data_type</span><span class="o">=</span><span class="n">data_type</span><span class="p">)</span>
            <span class="p">}</span>
        <span class="p">},</span>
        <span class="n">output_resources</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;vcf_header_path&quot;</span><span class="p">:</span> <span class="n">release_header_path</span><span class="p">(</span><span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">,</span> <span class="n">data_type</span><span class="o">=</span><span class="n">data_type</span><span class="p">)</span>
        <span class="p">},</span>
    <span class="p">)</span>
    <span class="n">export_vcf</span> <span class="o">=</span> <span class="n">PipelineStepResourceCollection</span><span class="p">(</span>
        <span class="s2">&quot;--export-vcf&quot;</span><span class="p">,</span>
        <span class="n">pipeline_input_steps</span><span class="o">=</span><span class="p">[</span><span class="n">validate_release_ht</span><span class="p">,</span> <span class="n">prepare_vcf_header</span><span class="p">],</span>
        <span class="n">output_resources</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;release_vcf&quot;</span><span class="p">:</span> <span class="n">release_vcf_path</span><span class="p">(</span>
                <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">,</span>
                <span class="n">data_type</span><span class="o">=</span><span class="n">data_type</span><span class="p">,</span>
                <span class="n">contig</span><span class="o">=</span><span class="n">contig</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="p">},</span>
    <span class="p">)</span>

    <span class="n">export_pipeline</span><span class="o">.</span><span class="n">add_steps</span><span class="p">(</span>
        <span class="p">{</span>
            <span class="s2">&quot;validate_release_ht&quot;</span><span class="p">:</span> <span class="n">validate_release_ht</span><span class="p">,</span>
            <span class="s2">&quot;prepare_vcf_header&quot;</span><span class="p">:</span> <span class="n">prepare_vcf_header</span><span class="p">,</span>
            <span class="s2">&quot;export_vcf&quot;</span><span class="p">:</span> <span class="n">export_vcf</span><span class="p">,</span>
        <span class="p">}</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">export_pipeline</span></div>


<div class="viewcode-block" id="filter_to_test"><a class="viewcode-back" href="../../../../api_reference/v4/create_release/validate_and_export_vcf.html#gnomad_qc.v4.create_release.validate_and_export_vcf.filter_to_test">[docs]</a><span class="k">def</span> <span class="nf">filter_to_test</span><span class="p">(</span><span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">num_partitions</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">2</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Filter Table to `num_partitions` partitions on chr20, chrX, and chrY for testing.</span>

<span class="sd">    :param ht: Input Table to filter.</span>
<span class="sd">    :param num_partitions: Number of partitions to grab from each chromosome.</span>
<span class="sd">    :return: Input Table filtered to `num_partitions` on chr20, chrX, and chrY.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Filtering to </span><span class="si">%d</span><span class="s2"> partitions on chr20, chrX, and chrY (for tests only)...&quot;</span><span class="p">,</span>
        <span class="n">num_partitions</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">ht_chr20</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">filter_intervals</span><span class="p">(</span><span class="n">ht</span><span class="p">,</span> <span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">parse_locus_interval</span><span class="p">(</span><span class="s2">&quot;chr20&quot;</span><span class="p">)])</span>
    <span class="n">ht_chr20</span> <span class="o">=</span> <span class="n">ht_chr20</span><span class="o">.</span><span class="n">_filter_partitions</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">num_partitions</span><span class="p">))</span>

    <span class="n">ht_chrx</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">filter_intervals</span><span class="p">(</span><span class="n">ht</span><span class="p">,</span> <span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">parse_locus_interval</span><span class="p">(</span><span class="s2">&quot;chrX&quot;</span><span class="p">)])</span>
    <span class="n">ht_chrx</span> <span class="o">=</span> <span class="n">ht_chrx</span><span class="o">.</span><span class="n">_filter_partitions</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">num_partitions</span><span class="p">))</span>

    <span class="n">ht_chry</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">filter_intervals</span><span class="p">(</span><span class="n">ht</span><span class="p">,</span> <span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">parse_locus_interval</span><span class="p">(</span><span class="s2">&quot;chrY&quot;</span><span class="p">)])</span>
    <span class="n">ht_chry</span> <span class="o">=</span> <span class="n">ht_chry</span><span class="o">.</span><span class="n">_filter_partitions</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">num_partitions</span><span class="p">))</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht_chr20</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">ht_chrx</span><span class="p">,</span> <span class="n">ht_chry</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">ht</span></div>


<div class="viewcode-block" id="select_type_from_joint_ht"><a class="viewcode-back" href="../../../../api_reference/v4/create_release/validate_and_export_vcf.html#gnomad_qc.v4.create_release.validate_and_export_vcf.select_type_from_joint_ht">[docs]</a><span class="k">def</span> <span class="nf">select_type_from_joint_ht</span><span class="p">(</span><span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">data_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Select all fields from the joint HT that are relevant to `data_type`.</span>

<span class="sd">    :param ht: Joint release HT.</span>
<span class="sd">    :param data_type: Data type to select in joint HT. One of &quot;exomes&quot;, &quot;genomes&quot;, or</span>
<span class="sd">        &quot;joint&quot;.</span>
<span class="sd">    :return: Joint HT with fields relevant to `data_type`.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">global_fields</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">data_type</span><span class="si">}</span><span class="s2">_globals&quot;</span><span class="p">]</span>
    <span class="n">row_fields</span> <span class="o">=</span> <span class="p">[</span><span class="n">data_type</span><span class="p">,</span> <span class="s2">&quot;region_flags&quot;</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;joint&quot;</span><span class="p">:</span>
        <span class="n">row_fields</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;freq_comparison_stats&quot;</span><span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">select_globals</span><span class="p">(</span><span class="o">*</span><span class="n">global_fields</span><span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="o">*</span><span class="n">row_fields</span><span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">transmute_globals</span><span class="p">(</span><span class="o">**</span><span class="n">ht</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">data_type</span><span class="si">}</span><span class="s2">_globals&quot;</span><span class="p">])</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">transmute</span><span class="p">(</span><span class="o">**</span><span class="n">ht</span><span class="p">[</span><span class="n">data_type</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">ht</span></div>


<div class="viewcode-block" id="unfurl_nested_annotations"><a class="viewcode-back" href="../../../../api_reference/v4/create_release/validate_and_export_vcf.html#gnomad_qc.v4.create_release.validate_and_export_vcf.unfurl_nested_annotations">[docs]</a><span class="k">def</span> <span class="nf">unfurl_nested_annotations</span><span class="p">(</span>
    <span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">entries_to_remove</span><span class="p">:</span> <span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">data_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;exomes&quot;</span><span class="p">,</span>
    <span class="n">joint_included</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">hist_prefix</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="n">freq_comparison_included</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">for_joint_validation</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">StructExpression</span><span class="p">,</span> <span class="n">Set</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create dictionary keyed by the variant annotation labels to be extracted from variant annotation arrays.</span>

<span class="sd">    The values of the returned dictionary are Hail Expressions describing how to access</span>
<span class="sd">    the corresponding values.</span>

<span class="sd">    :param ht: Table containing the nested variant annotation arrays to be unfurled.</span>
<span class="sd">    :param entries_to_remove: Optional Set of frequency entries to remove for vcf_export.</span>
<span class="sd">    :param data_type: Data type to unfurl nested annotations for. One of &quot;exomes&quot;,</span>
<span class="sd">        &quot;genomes&quot;, or &quot;joint&quot;. Default is &quot;exomes&quot;.</span>
<span class="sd">    :param joint_included: Whether joint frequency data is included in the exome or</span>
<span class="sd">        genome HT. Default is False.</span>
<span class="sd">    :param hist_prefix: Prefix to use for histograms. Default is &quot;&quot;.</span>
<span class="sd">    :param freq_comparison_included: Whether frequency comparison data is included in</span>
<span class="sd">        the HT. Default is False.</span>
<span class="sd">    :param for_joint_validation: Whether to prepare HT for joint validation. Default is</span>
<span class="sd">        False.</span>
<span class="sd">    :return: StructExpression containing variant annotations and their corresponding</span>
<span class="sd">        expressions and updated entries, set of frequency entries to remove from the</span>
<span class="sd">        VCF, and a dict of fields to rename when `for_joint_validation` is True.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">expr_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">rename_dict</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="c1"># Unfurl freq index dict</span>
    <span class="c1"># Cycles through each key and index (e.g., k=afr_XX, i=31)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Unfurling freq data...&quot;</span><span class="p">)</span>
    <span class="n">freq_idx</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">freq_index_dict</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">freq_idx</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">ht</span><span class="o">.</span><span class="n">freq</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">field_name</span> <span class="o">=</span> <span class="n">f</span> <span class="k">if</span> <span class="n">f</span> <span class="o">!=</span> <span class="s2">&quot;homozygote_count&quot;</span> <span class="k">else</span> <span class="s2">&quot;nhomalt&quot;</span>
            <span class="n">expr_dict</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">field_name</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">freq</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">f</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">for_joint_validation</span><span class="p">:</span>
                <span class="n">rename_dict</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">field_name</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">field_name</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">data_type</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2">&quot;</span>

    <span class="k">if</span> <span class="n">joint_included</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Unfurling joint freq data...&quot;</span><span class="p">)</span>
        <span class="n">joint_freq_idx</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">joint_freq_index_dict</span><span class="p">)</span>
        <span class="n">expr_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
            <span class="p">{</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">f</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">f</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="s1">&#39;homozygote_count&#39;</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;nhomalt&#39;</span><span class="si">}</span><span class="s2">_joint_</span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="n">ht</span><span class="o">.</span><span class="n">joint_freq</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">f</span><span class="p">]</span>
                <span class="p">)</span>
                <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">joint_freq_idx</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
                <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">ht</span><span class="o">.</span><span class="n">joint_freq</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
            <span class="p">}</span>
        <span class="p">)</span>

    <span class="c1"># This creates fields like grpmax, AC_grpmax_non_ukb...</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Adding grpmax data...&quot;</span><span class="p">)</span>
    <span class="n">grpmax_idx</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">grpmax</span>
    <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;exomes&quot;</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">for_joint_validation</span><span class="p">:</span>
        <span class="n">grpmax_dict</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">grpmax_idx</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">grpmax_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                <span class="p">{</span><span class="sa">f</span><span class="s2">&quot;grpmax</span><span class="si">{</span><span class="s1">&#39;_&#39;</span><span class="o">+</span><span class="n">s</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">s</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="s1">&#39;gnomad&#39;</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;&#39;</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="n">grpmax_idx</span><span class="p">[</span><span class="n">s</span><span class="p">]</span><span class="o">.</span><span class="n">gen_anc</span><span class="p">}</span>
            <span class="p">)</span>
            <span class="n">grpmax_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                <span class="p">{</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">f</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">f</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="s1">&#39;homozygote_count&#39;</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;nhomalt&#39;</span><span class="si">}</span><span class="s2">_grpmax</span><span class="si">{</span><span class="s1">&#39;_&#39;</span><span class="o">+</span><span class="n">s</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">s</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="s1">&#39;gnomad&#39;</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;&#39;</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="n">grpmax_idx</span><span class="p">[</span>
                        <span class="n">s</span>
                    <span class="p">][</span>
                        <span class="n">f</span>
                    <span class="p">]</span>
                    <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">grpmax_idx</span><span class="p">[</span><span class="n">s</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
                    <span class="k">if</span> <span class="n">f</span> <span class="o">!=</span> <span class="s2">&quot;gen_anc&quot;</span>
                <span class="p">}</span>
            <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">grpmax_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;grpmax&quot;</span><span class="p">:</span> <span class="n">grpmax_idx</span><span class="o">.</span><span class="n">gen_anc</span><span class="p">}</span>
        <span class="n">grpmax_rename</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">f</span><span class="p">:</span> <span class="n">f</span> <span class="k">if</span> <span class="n">f</span> <span class="o">!=</span> <span class="s2">&quot;homozygote_count&quot;</span> <span class="k">else</span> <span class="s2">&quot;nhomalt&quot;</span>
            <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">grpmax_idx</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">f</span> <span class="o">!=</span> <span class="s2">&quot;gen_anc&quot;</span>
        <span class="p">}</span>
        <span class="n">grpmax_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
            <span class="p">{</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">v</span><span class="si">}</span><span class="s2">_grpmax&quot;</span><span class="p">:</span> <span class="n">grpmax_idx</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">grpmax_rename</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">for_joint_validation</span><span class="p">:</span>
            <span class="n">rename_dict</span><span class="p">[</span><span class="s2">&quot;grpmax&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;grpmax_</span><span class="si">{</span><span class="n">data_type</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="n">rename_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                <span class="p">{</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">v</span><span class="si">}</span><span class="s2">_grpmax&quot;</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">v</span><span class="si">}</span><span class="s2">_grpmax_</span><span class="si">{</span><span class="n">data_type</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">grpmax_rename</span><span class="o">.</span><span class="n">values</span><span class="p">()</span>
                <span class="p">}</span>
            <span class="p">)</span>
    <span class="n">expr_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">grpmax_dict</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">joint_included</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Adding joint grpmax data...&quot;</span><span class="p">)</span>
        <span class="n">joint_grpmax_idx</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">joint_grpmax</span>
        <span class="n">joint_grpmax_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;grpmax_joint&quot;</span><span class="p">:</span> <span class="n">joint_grpmax_idx</span><span class="o">.</span><span class="n">gen_anc</span><span class="p">}</span>
        <span class="n">joint_grpmax_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
            <span class="p">{</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">f</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">f</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="s1">&#39;homozygote_count&#39;</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;nhomalt&#39;</span><span class="si">}</span><span class="s2">_grpmax_joint&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="n">joint_grpmax_idx</span><span class="p">[</span><span class="n">f</span><span class="p">]</span>
                <span class="p">)</span>
                <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="p">[</span><span class="n">f</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">joint_grpmax_idx</span><span class="o">.</span><span class="n">_fields</span> <span class="k">if</span> <span class="n">f</span> <span class="o">!=</span> <span class="s2">&quot;gen_anc&quot;</span><span class="p">]</span>
            <span class="p">}</span>
        <span class="p">)</span>
        <span class="n">expr_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">joint_grpmax_dict</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Unfurling faf data...&quot;</span><span class="p">)</span>
    <span class="n">faf_idx</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">faf_index_dict</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">faf_idx</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">ht</span><span class="o">.</span><span class="n">faf</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">expr_dict</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">f</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">faf</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">f</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">for_joint_validation</span><span class="p">:</span>
                <span class="n">rename_dict</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">f</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">f</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">data_type</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2">&quot;</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Unfurling fafmax data...&quot;</span><span class="p">)</span>
    <span class="n">fafmax_idx</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">fafmax</span>
    <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;exomes&quot;</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">for_joint_validation</span><span class="p">:</span>
        <span class="n">fafmax_dict</span> <span class="o">=</span> <span class="p">{</span>
            <span class="sa">f</span><span class="s2">&quot;fafmax_</span><span class="si">{</span><span class="n">f</span><span class="si">}{</span><span class="s1">&#39;_&#39;</span><span class="o">+</span><span class="n">s</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">s</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="s1">&#39;gnomad&#39;</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;&#39;</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="n">fafmax_idx</span><span class="p">[</span><span class="n">s</span><span class="p">][</span><span class="n">f</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">fafmax_idx</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">fafmax_idx</span><span class="p">[</span><span class="n">s</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
        <span class="p">}</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">fafmax_dict</span> <span class="o">=</span> <span class="p">{</span><span class="sa">f</span><span class="s2">&quot;fafmax_</span><span class="si">{</span><span class="n">f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="n">fafmax_idx</span><span class="p">[</span><span class="n">f</span><span class="p">]</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">fafmax_idx</span><span class="o">.</span><span class="n">keys</span><span class="p">()}</span>
        <span class="k">if</span> <span class="n">for_joint_validation</span><span class="p">:</span>
            <span class="n">rename_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                <span class="p">{</span><span class="sa">f</span><span class="s2">&quot;fafmax_</span><span class="si">{</span><span class="n">f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;fafmax_</span><span class="si">{</span><span class="n">f</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">data_type</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">fafmax_idx</span><span class="o">.</span><span class="n">keys</span><span class="p">()}</span>
            <span class="p">)</span>
    <span class="n">expr_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">fafmax_dict</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">joint_included</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Unfurling joint faf data...&quot;</span><span class="p">)</span>
        <span class="n">joint_faf_idx</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">joint_faf_index_dict</span><span class="p">)</span>
        <span class="n">expr_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
            <span class="p">{</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">f</span><span class="si">}</span><span class="s2">_joint_</span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="n">ht</span><span class="o">.</span><span class="n">joint_faf</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">f</span><span class="p">]</span>
                <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">ht</span><span class="o">.</span><span class="n">joint_faf</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
                <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">joint_faf_idx</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
            <span class="p">}</span>
        <span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Unfurling joint fafmax data...&quot;</span><span class="p">)</span>
        <span class="n">joint_fafmax_idx</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">joint_fafmax</span>
        <span class="n">joint_fafmax_dict</span> <span class="o">=</span> <span class="p">{</span>
            <span class="sa">f</span><span class="s2">&quot;fafmax_</span><span class="si">{</span><span class="n">f</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">f</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="s1">&#39;joint_fafmax_data_type&#39;</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;data_type&#39;</span><span class="si">}</span><span class="s2">_joint&quot;</span><span class="p">:</span> <span class="p">(</span>
                <span class="n">joint_fafmax_idx</span><span class="p">[</span><span class="n">f</span><span class="p">]</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">joint_fafmax_idx</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
        <span class="p">}</span>
        <span class="n">expr_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">joint_fafmax_dict</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Unfurling age hists...&quot;</span><span class="p">)</span>
    <span class="n">hist_idx</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">histograms</span><span class="o">.</span><span class="n">age_hists</span>
    <span class="n">age_hists</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;age_hist_het&quot;</span><span class="p">,</span> <span class="s2">&quot;age_hist_hom&quot;</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">hist</span> <span class="ow">in</span> <span class="n">age_hists</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">hist_idx</span><span class="p">[</span><span class="n">hist</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">expr_dict</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">hist</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">hl</span><span class="o">.</span><span class="n">delimit</span><span class="p">(</span><span class="n">hist_idx</span><span class="p">[</span><span class="n">hist</span><span class="p">][</span><span class="n">f</span><span class="p">],</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;|&quot;</span><span class="p">)</span>
                <span class="k">if</span> <span class="s2">&quot;bin&quot;</span> <span class="ow">in</span> <span class="n">f</span>
                <span class="k">else</span> <span class="n">hist_idx</span><span class="p">[</span><span class="n">hist</span><span class="p">][</span><span class="n">f</span><span class="p">]</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">for_joint_validation</span><span class="p">:</span>
                <span class="n">rename_dict</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">hist</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">hist</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">f</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">data_type</span><span class="si">}</span><span class="s2">&quot;</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Unfurling variant quality histograms...&quot;</span><span class="p">)</span>
    <span class="c1"># Add underscore to hist_prefix if it isn&#39;t empty</span>
    <span class="k">if</span> <span class="n">hist_prefix</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
        <span class="n">hist_prefix</span> <span class="o">+=</span> <span class="s2">&quot;_&quot;</span>

    <span class="c1"># Histograms to export are:</span>
    <span class="c1"># gq_hist_alt, gq_hist_all, dp_hist_alt, dp_hist_all, ab_hist_alt</span>
    <span class="c1"># We previously dropped:</span>
    <span class="c1"># _n_smaller for all hists</span>
    <span class="c1"># _bin_edges for all hists</span>
    <span class="c1"># _n_larger for all hists EXCEPT DP hists</span>
    <span class="k">for</span> <span class="n">hist</span> <span class="ow">in</span> <span class="n">HISTS</span><span class="p">:</span>
        <span class="n">hist_dict</span> <span class="o">=</span> <span class="p">{</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">hist</span><span class="si">}</span><span class="s2">_bin_freq&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">delimit</span><span class="p">(</span>
                <span class="n">ht</span><span class="o">.</span><span class="n">histograms</span><span class="o">.</span><span class="n">qual_hists</span><span class="p">[</span><span class="n">hist</span><span class="p">]</span><span class="o">.</span><span class="n">bin_freq</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;|&quot;</span>
            <span class="p">),</span>
        <span class="p">}</span>
        <span class="n">expr_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">hist_dict</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">for_joint_validation</span><span class="p">:</span>
            <span class="n">rename_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                <span class="p">{</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">hist</span><span class="si">}</span><span class="s2">_bin_freq&quot;</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">hist_prefix</span><span class="si">}{</span><span class="n">hist</span><span class="si">}</span><span class="s2">_bin_freq_</span><span class="si">{</span><span class="n">data_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">}</span>
            <span class="p">)</span>

        <span class="k">if</span> <span class="s2">&quot;dp&quot;</span> <span class="ow">in</span> <span class="n">hist</span><span class="p">:</span>
            <span class="n">expr_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                <span class="p">{</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">hist</span><span class="si">}</span><span class="s2">_n_larger&quot;</span><span class="p">:</span> <span class="n">ht</span><span class="o">.</span><span class="n">histograms</span><span class="o">.</span><span class="n">qual_hists</span><span class="p">[</span><span class="n">hist</span><span class="p">]</span><span class="o">.</span><span class="n">n_larger</span><span class="p">},</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">for_joint_validation</span><span class="p">:</span>
                <span class="n">rename_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                    <span class="p">{</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">hist</span><span class="si">}</span><span class="s2">_n_larger&quot;</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">hist_prefix</span><span class="si">}{</span><span class="n">hist</span><span class="si">}</span><span class="s2">_n_larger_</span><span class="si">{</span><span class="n">data_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">}</span>
                <span class="p">)</span>

    <span class="k">if</span> <span class="n">freq_comparison_included</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Unfurling contingency table test results...&quot;</span><span class="p">)</span>
        <span class="n">contingency_idx</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">freq_index_dict</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">contingency_idx</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">ht</span><span class="o">.</span><span class="n">freq_comparison_stats</span><span class="o">.</span><span class="n">contingency_table_test</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                <span class="n">key</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;CTT_</span><span class="si">{</span><span class="n">f</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="n">expr</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">freq_comparison_stats</span><span class="o">.</span><span class="n">contingency_table_test</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="n">f</span><span class="p">]</span>
                <span class="n">expr_dict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">expr</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Unfurling Cochran-Mantel-Haenszel test results...&quot;</span><span class="p">)</span>
        <span class="n">expr_dict</span><span class="p">[</span><span class="s2">&quot;CMH_chisq&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">ht</span><span class="o">.</span><span class="n">freq_comparison_stats</span><span class="o">.</span><span class="n">cochran_mantel_haenszel_test</span><span class="o">.</span><span class="n">chisq</span>
        <span class="p">)</span>
        <span class="n">expr_dict</span><span class="p">[</span><span class="s2">&quot;CMH_p_value&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">ht</span><span class="o">.</span><span class="n">freq_comparison_stats</span><span class="o">.</span><span class="n">cochran_mantel_haenszel_test</span><span class="o">.</span><span class="n">p_value</span>
        <span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Unfurling unionized stats...&quot;</span><span class="p">)</span>
        <span class="n">expr_dict</span><span class="p">[</span><span class="s2">&quot;stat_union_p_value&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">freq_comparison_stats</span><span class="o">.</span><span class="n">stat_union</span><span class="o">.</span><span class="n">p_value</span>
        <span class="n">expr_dict</span><span class="p">[</span><span class="s2">&quot;stat_union_test_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">ht</span><span class="o">.</span><span class="n">freq_comparison_stats</span><span class="o">.</span><span class="n">stat_union</span><span class="o">.</span><span class="n">stat_test_name</span>
        <span class="p">)</span>
        <span class="n">expr_dict</span><span class="p">[</span><span class="s2">&quot;stat_union_gen_ancs&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">freq_comparison_stats</span><span class="o">.</span><span class="n">stat_union</span><span class="o">.</span><span class="n">gen_ancs</span>

    <span class="k">return</span> <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span><span class="o">**</span><span class="n">expr_dict</span><span class="p">),</span> <span class="n">entries_to_remove</span><span class="p">,</span> <span class="n">rename_dict</span></div>


<div class="viewcode-block" id="make_info_expr"><a class="viewcode-back" href="../../../../api_reference/v4/create_release/validate_and_export_vcf.html#gnomad_qc.v4.create_release.validate_and_export_vcf.make_info_expr">[docs]</a><span class="k">def</span> <span class="nf">make_info_expr</span><span class="p">(</span>
    <span class="n">t</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">data_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;exomes&quot;</span><span class="p">,</span>
    <span class="n">for_joint_validation</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">Expression</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Make Hail expression for variant annotations to be included in VCF INFO field.</span>

<span class="sd">    :param t: Table containing variant annotations to be reformatted for VCF export.</span>
<span class="sd">    :param data_type: Data type to make info expression for. One of &quot;exomes&quot;, &quot;genomes&quot;,</span>
<span class="sd">        or &quot;joint&quot;. Default is &quot;exomes&quot;.</span>
<span class="sd">    :param for_joint_validation: Whether to prepare HT for joint validation. Default is False.</span>
<span class="sd">    :return: Dictionary containing Hail expressions for relevant INFO annotations.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">vcf_info_dict</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="c1"># Set data type to joint if for_joint_validation is True so the correct region flag</span>
    <span class="c1"># fields are used.</span>
    <span class="k">if</span> <span class="n">for_joint_validation</span><span class="p">:</span>
        <span class="n">data_type</span> <span class="o">=</span> <span class="s2">&quot;joint&quot;</span>

    <span class="k">if</span> <span class="s2">&quot;region_flags&quot;</span> <span class="ow">in</span> <span class="n">t</span><span class="o">.</span><span class="n">row</span><span class="p">:</span>
        <span class="c1"># Add region_flag to info dict</span>
        <span class="k">for</span> <span class="n">field</span> <span class="ow">in</span> <span class="n">REGION_FLAG_FIELDS</span><span class="p">[</span><span class="n">data_type</span><span class="p">]:</span>
            <span class="n">vcf_info_dict</span><span class="p">[</span><span class="n">field</span><span class="p">]</span> <span class="o">=</span> <span class="n">t</span><span class="p">[</span><span class="s2">&quot;region_flags&quot;</span><span class="p">][</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">field</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">for_joint_validation</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">vcf_info_dict</span>

    <span class="c1"># Add site-level annotations and AS annotations to vcf_info_dict</span>
    <span class="k">for</span> <span class="n">field</span> <span class="ow">in</span> <span class="n">SITE_FIELDS</span><span class="p">[</span><span class="n">data_type</span><span class="p">]</span> <span class="o">+</span> <span class="n">AS_FIELDS</span><span class="p">:</span>
        <span class="n">vcf_info_dict</span><span class="p">[</span><span class="n">field</span><span class="p">]</span> <span class="o">=</span> <span class="n">t</span><span class="p">[</span><span class="s2">&quot;release_ht_info&quot;</span><span class="p">][</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">field</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span>

    <span class="k">for</span> <span class="n">field</span> <span class="ow">in</span> <span class="n">AS_VQSR_FIELDS</span><span class="p">:</span>
        <span class="n">vcf_info_dict</span><span class="p">[</span><span class="n">field</span><span class="p">]</span> <span class="o">=</span> <span class="n">t</span><span class="p">[</span><span class="s2">&quot;vqsr_results&quot;</span><span class="p">][</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">field</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span>

    <span class="c1"># Add allele_info fields to info dict</span>
    <span class="k">for</span> <span class="n">field</span> <span class="ow">in</span> <span class="n">ALLELE_TYPE_FIELDS</span><span class="p">[</span><span class="n">data_type</span><span class="p">]:</span>
        <span class="n">vcf_info_dict</span><span class="p">[</span><span class="n">field</span><span class="p">]</span> <span class="o">=</span> <span class="n">t</span><span class="p">[</span><span class="s2">&quot;allele_info&quot;</span><span class="p">][</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">field</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span>

    <span class="c1"># Add in silico annotations to info dict</span>
    <span class="n">insilico_idx</span> <span class="o">=</span> <span class="n">t</span><span class="o">.</span><span class="n">in_silico_predictors</span>
    <span class="k">for</span> <span class="n">field</span> <span class="ow">in</span> <span class="n">INSILICO_FIELDS</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">field</span> <span class="o">==</span> <span class="s2">&quot;cadd&quot;</span><span class="p">:</span>
            <span class="n">vcf_info_dict</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">field</span><span class="si">}</span><span class="s2">_raw_score&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">insilico_idx</span><span class="p">[</span><span class="n">field</span><span class="p">][</span><span class="s2">&quot;raw_score&quot;</span><span class="p">]</span>
            <span class="n">vcf_info_dict</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">field</span><span class="si">}</span><span class="s2">_phred&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">insilico_idx</span><span class="p">[</span><span class="n">field</span><span class="p">][</span><span class="s2">&quot;phred&quot;</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">vcf_info_dict</span><span class="p">[</span><span class="n">field</span><span class="p">]</span> <span class="o">=</span> <span class="n">insilico_idx</span><span class="p">[</span><span class="n">field</span><span class="p">]</span>

    <span class="c1"># Add VRS annotations to info dict</span>
    <span class="k">for</span> <span class="n">field</span> <span class="ow">in</span> <span class="n">VRS_FIELDS_DICT</span><span class="p">:</span>
        <span class="n">vcf_info_dict</span><span class="p">[</span><span class="n">field</span><span class="p">]</span> <span class="o">=</span> <span class="n">t</span><span class="p">[</span><span class="s2">&quot;release_ht_info&quot;</span><span class="p">][</span><span class="s2">&quot;vrs&quot;</span><span class="p">][</span><span class="n">field</span><span class="p">]</span>

    <span class="c1"># Add vep annotations to info dict</span>
    <span class="n">vcf_info_dict</span><span class="p">[</span><span class="s2">&quot;vep&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">t</span><span class="p">[</span><span class="s2">&quot;vep&quot;</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">vcf_info_dict</span></div>


<div class="viewcode-block" id="prepare_ht_for_validation"><a class="viewcode-back" href="../../../../api_reference/v4/create_release/validate_and_export_vcf.html#gnomad_qc.v4.create_release.validate_and_export_vcf.prepare_ht_for_validation">[docs]</a><span class="k">def</span> <span class="nf">prepare_ht_for_validation</span><span class="p">(</span>
    <span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">data_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;exomes&quot;</span><span class="p">,</span>
    <span class="n">freq_entries_to_remove</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">vcf_info_reorder</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">joint_included</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">for_joint_validation</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">freq_comparison_included</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Prepare HT for validity checks and export.</span>

<span class="sd">    :param ht: Release Hail Table.</span>
<span class="sd">    :param data_type: Data type to prepare HT for. One of &quot;exomes&quot;, &quot;genomes&quot;, or</span>
<span class="sd">        &quot;joint&quot;. Default is &quot;exomes&quot;.</span>
<span class="sd">    :param freq_entries_to_remove: List of entries to remove from freq.</span>
<span class="sd">    :param vcf_info_reorder: Order of VCF INFO fields.</span>
<span class="sd">    :param joint_included: Whether joint frequency data is included in the HT. Default</span>
<span class="sd">        is False.</span>
<span class="sd">    :param for_joint_validation: Whether to prepare HT for joint validation. Default is</span>
<span class="sd">        True.</span>
<span class="sd">    :param freq_comparison_included: Whether frequency comparison data is included in</span>
<span class="sd">        the HT. Default is False.</span>
<span class="sd">    :return: Hail Table prepared for validity checks and export and a dictionary of</span>
<span class="sd">        fields to rename when `for_joint_validation` is True.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Unfurling nested gnomAD frequency annotations and add to INFO field...&quot;</span>
    <span class="p">)</span>
    <span class="n">info_struct</span><span class="p">,</span> <span class="n">freq_entries_to_remove</span><span class="p">,</span> <span class="n">rename_dict</span> <span class="o">=</span> <span class="n">unfurl_nested_annotations</span><span class="p">(</span>
        <span class="n">ht</span><span class="p">,</span>
        <span class="n">entries_to_remove</span><span class="o">=</span><span class="n">freq_entries_to_remove</span><span class="p">,</span>
        <span class="n">data_type</span><span class="o">=</span><span class="n">data_type</span><span class="p">,</span>
        <span class="n">joint_included</span><span class="o">=</span><span class="n">joint_included</span><span class="p">,</span>
        <span class="n">freq_comparison_included</span><span class="o">=</span><span class="n">freq_comparison_included</span><span class="p">,</span>
        <span class="n">for_joint_validation</span><span class="o">=</span><span class="n">for_joint_validation</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Constructing INFO field&quot;</span><span class="p">)</span>
    <span class="c1"># Remove SIFT and Polyphen from CSQ fields or they will be inserted with</span>
    <span class="c1"># missing values by vep_struct_to_csq. These fields are processed separately</span>
    <span class="c1"># as in silico annotations.</span>
    <span class="n">csq_fields</span> <span class="o">=</span> <span class="s2">&quot;|&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="p">[</span>
            <span class="n">c</span>
            <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">VEP_CSQ_FIELDS</span><span class="p">[</span><span class="n">CURRENT_VEP_VERSION</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;|&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">c</span> <span class="o">!=</span> <span class="s2">&quot;PolyPhen&quot;</span> <span class="ow">and</span> <span class="n">c</span> <span class="o">!=</span> <span class="s2">&quot;SIFT&quot;</span>
        <span class="p">]</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">for_joint_validation</span><span class="p">:</span>
        <span class="n">ann_expr</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;info&quot;</span><span class="p">:</span> <span class="n">info_struct</span><span class="p">}</span>
        <span class="k">if</span> <span class="s2">&quot;region_flags&quot;</span> <span class="ow">in</span> <span class="n">ht</span><span class="o">.</span><span class="n">row</span><span class="p">:</span>
            <span class="n">ann_expr</span><span class="p">[</span><span class="s2">&quot;region_flags&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">region_flags</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">ann_expr</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;region_flag&quot;</span><span class="p">:</span> <span class="n">ht</span><span class="o">.</span><span class="n">region_flags</span><span class="p">,</span>
            <span class="s2">&quot;release_ht_info&quot;</span><span class="p">:</span> <span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="p">,</span>
            <span class="s2">&quot;info&quot;</span><span class="p">:</span> <span class="n">info_struct</span><span class="p">,</span>
            <span class="s2">&quot;rsid&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">str</span><span class="p">(</span><span class="s2">&quot;;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">rsid</span><span class="p">),</span>
            <span class="s2">&quot;vep&quot;</span><span class="p">:</span> <span class="n">vep_struct_to_csq</span><span class="p">(</span>
                <span class="n">ht</span><span class="o">.</span><span class="n">vep</span><span class="p">,</span> <span class="n">csq_fields</span><span class="o">=</span><span class="n">csq_fields</span><span class="p">,</span> <span class="n">has_polyphen_sift</span><span class="o">=</span><span class="kc">False</span>
            <span class="p">),</span>
        <span class="p">}</span>

    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="o">**</span><span class="n">ann_expr</span><span class="p">)</span>

    <span class="c1"># Add variant annotations to INFO field</span>
    <span class="c1"># This adds the following:</span>
    <span class="c1">#   region flag for problematic regions</span>
    <span class="c1">#   annotations in ht.release_ht_info (site and allele-specific annotations),</span>
    <span class="c1">#   info struct (unfurled data obtained above),</span>
    <span class="c1">#   dbSNP rsIDs</span>
    <span class="c1">#   all VEP annotations</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="n">info</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
            <span class="o">**</span><span class="n">make_info_expr</span><span class="p">(</span>
                <span class="n">ht</span><span class="p">,</span>
                <span class="n">data_type</span><span class="o">=</span><span class="n">data_type</span><span class="p">,</span>
                <span class="n">for_joint_validation</span><span class="o">=</span><span class="n">for_joint_validation</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">for_joint_validation</span><span class="p">:</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span>
            <span class="n">freq_entries_to_remove</span><span class="o">=</span><span class="p">(</span>
                <span class="n">freq_entries_to_remove</span>
                <span class="k">if</span> <span class="n">freq_entries_to_remove</span>
                <span class="k">else</span> <span class="n">hl</span><span class="o">.</span><span class="n">empty_set</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">tstr</span><span class="p">)</span>
            <span class="p">),</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span>
            <span class="n">vep_csq_header</span><span class="o">=</span><span class="n">process_vep_csq_header</span><span class="p">(</span><span class="n">VEP_CSQ_HEADER</span><span class="p">),</span>
            <span class="n">freq_entries_to_remove</span><span class="o">=</span><span class="p">(</span>
                <span class="n">freq_entries_to_remove</span>
                <span class="k">if</span> <span class="n">freq_entries_to_remove</span>
                <span class="k">else</span> <span class="n">hl</span><span class="o">.</span><span class="n">empty_set</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">tstr</span><span class="p">)</span>
            <span class="p">),</span>
        <span class="p">)</span>

    <span class="c1"># Select relevant fields for VCF export</span>
    <span class="k">if</span> <span class="n">for_joint_validation</span><span class="p">:</span>
        <span class="k">if</span> <span class="s2">&quot;filters&quot;</span> <span class="ow">in</span> <span class="n">ht</span><span class="o">.</span><span class="n">row</span><span class="p">:</span>
            <span class="n">filters_expr</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">filters</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">filters_expr</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">empty_set</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">tstr</span><span class="p">)</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;info&quot;</span><span class="p">,</span> <span class="n">filters</span><span class="o">=</span><span class="n">filters_expr</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;info&quot;</span><span class="p">,</span> <span class="s2">&quot;filters&quot;</span><span class="p">,</span> <span class="s2">&quot;rsid&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">vcf_info_reorder</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Rearranging fields to desired order...&quot;</span><span class="p">)</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
            <span class="n">info</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="o">*</span><span class="n">vcf_info_reorder</span><span class="p">,</span> <span class="o">*</span><span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="o">*</span><span class="n">vcf_info_reorder</span><span class="p">))</span>
        <span class="p">)</span>

    <span class="k">return</span> <span class="n">ht</span><span class="p">,</span> <span class="n">rename_dict</span></div>


<div class="viewcode-block" id="populate_subset_info_dict"><a class="viewcode-back" href="../../../../api_reference/v4/create_release/validate_and_export_vcf.html#gnomad_qc.v4.create_release.validate_and_export_vcf.populate_subset_info_dict">[docs]</a><span class="k">def</span> <span class="nf">populate_subset_info_dict</span><span class="p">(</span>
    <span class="n">subset</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">description_text</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">data_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;exomes&quot;</span><span class="p">,</span>
    <span class="n">pops</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">GEN_ANC_GROUPS</span><span class="p">[</span><span class="s2">&quot;exomes&quot;</span><span class="p">],</span>
    <span class="n">faf_pops</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="n">FAF_GEN_ANC_GROUPS</span><span class="p">,</span>
    <span class="n">sexes</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">SEXES</span><span class="p">,</span>
    <span class="n">label_delimiter</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="p">,</span>
    <span class="n">freq_comparison_included</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Call `make_info_dict` to populate INFO dictionary for the requested `subset`.</span>

<span class="sd">    Creates:</span>
<span class="sd">        - INFO fields for AC, AN, AF, nhomalt for each combination of sample genetic</span>
<span class="sd">            ancestry group, sex both for adj and raw data.</span>
<span class="sd">        - INFO fields for filtering allele frequency (faf) annotations.</span>

<span class="sd">    :param subset: Sample subset in dataset. &quot;&quot; is used as a placeholder for the full</span>
<span class="sd">        dataset.</span>
<span class="sd">    :param description_text: Text describing the sample subset that should be added to</span>
<span class="sd">        the INFO description.</span>
<span class="sd">    :param data_type: One of &quot;exomes&quot;, &quot;genomes&quot;, or &quot;joint&quot;. Default is &quot;exomes&quot;.</span>
<span class="sd">    :param pops: Dict of sample global genetic ancestry names for the gnomAD data type.</span>
<span class="sd">    :param faf_pops: Dict with gnomAD version (keys) and faf genentic ancestry group</span>
<span class="sd">        names (values). Default is FAF_GEN_ANC_GROUPS.</span>
<span class="sd">    :param sexes: gnomAD sample sexes used in VCF export. Default is SEXES.</span>
<span class="sd">    :param label_delimiter: String to use as delimiter when making group label</span>
<span class="sd">        combinations. Default is &#39;_&#39;.</span>
<span class="sd">    :param freq_comparison_included: Whether frequency comparison data is included in the HT.</span>
<span class="sd">    :return: Dictionary containing Subset specific INFO header fields.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">vcf_info_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="c1"># Remove unnecessary pop names from FAF_GEN_ANC_GROUPS dict depending on data type</span>
    <span class="c1"># and version of FAF_GEN_ANC_GROUPS.</span>
    <span class="n">faf_pops_version</span> <span class="o">=</span> <span class="s2">&quot;v3&quot;</span> <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;genomes&quot;</span> <span class="ow">or</span> <span class="n">subset</span> <span class="o">==</span> <span class="s2">&quot;genomes&quot;</span> <span class="k">else</span> <span class="s2">&quot;v4&quot;</span>
    <span class="n">faf_pops</span> <span class="o">=</span> <span class="p">{</span><span class="n">pop</span><span class="p">:</span> <span class="n">GEN_ANC_NAMES</span><span class="p">[</span><span class="n">pop</span><span class="p">]</span> <span class="k">for</span> <span class="n">pop</span> <span class="ow">in</span> <span class="n">faf_pops</span><span class="p">[</span><span class="n">faf_pops_version</span><span class="p">]}</span>

    <span class="c1"># Add FAF fields to dict.</span>
    <span class="n">faf_label_groups</span> <span class="o">=</span> <span class="n">create_label_groups</span><span class="p">(</span>
        <span class="n">gen_ancs</span><span class="o">=</span><span class="n">faf_pops</span><span class="p">,</span> <span class="n">sexes</span><span class="o">=</span><span class="n">sexes</span><span class="p">,</span> <span class="n">all_groups</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;adj&quot;</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="k">for</span> <span class="n">label_group</span> <span class="ow">in</span> <span class="n">faf_label_groups</span><span class="p">:</span>
        <span class="n">vcf_info_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
            <span class="n">make_info_dict</span><span class="p">(</span>
                <span class="n">prefix</span><span class="o">=</span><span class="n">subset</span><span class="p">,</span>
                <span class="n">prefix_before_metric</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                <span class="n">gen_anc_names</span><span class="o">=</span><span class="n">faf_pops</span><span class="p">,</span>
                <span class="n">label_groups</span><span class="o">=</span><span class="n">label_group</span><span class="p">,</span>
                <span class="n">label_delimiter</span><span class="o">=</span><span class="n">label_delimiter</span><span class="p">,</span>
                <span class="n">faf</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">description_text</span><span class="o">=</span><span class="n">description_text</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="p">)</span>
    <span class="c1"># Add AC, AN, AF, nhomalt fields to dict.</span>
    <span class="n">label_groups</span> <span class="o">=</span> <span class="n">create_label_groups</span><span class="p">(</span><span class="n">gen_ancs</span><span class="o">=</span><span class="n">pops</span><span class="p">,</span> <span class="n">sexes</span><span class="o">=</span><span class="n">sexes</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">label_group</span> <span class="ow">in</span> <span class="n">label_groups</span><span class="p">:</span>
        <span class="n">vcf_info_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
            <span class="n">make_info_dict</span><span class="p">(</span>
                <span class="n">prefix</span><span class="o">=</span><span class="n">subset</span><span class="p">,</span>
                <span class="n">prefix_before_metric</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                <span class="n">gen_anc_names</span><span class="o">=</span><span class="n">pops</span><span class="p">,</span>
                <span class="n">label_groups</span><span class="o">=</span><span class="n">label_group</span><span class="p">,</span>
                <span class="n">label_delimiter</span><span class="o">=</span><span class="n">label_delimiter</span><span class="p">,</span>
                <span class="n">description_text</span><span class="o">=</span><span class="n">description_text</span><span class="p">,</span>
                <span class="n">callstats</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="p">)</span>

    <span class="c1"># Add grpmax.</span>
    <span class="n">vcf_info_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
        <span class="n">make_info_dict</span><span class="p">(</span>
            <span class="n">suffix</span><span class="o">=</span><span class="n">subset</span><span class="p">,</span>
            <span class="n">label_delimiter</span><span class="o">=</span><span class="n">label_delimiter</span><span class="p">,</span>
            <span class="n">gen_anc_names</span><span class="o">=</span><span class="n">pops</span><span class="p">,</span>
            <span class="n">grpmax</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">description_text</span><span class="o">=</span><span class="n">description_text</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="c1"># Add fafmax.</span>
    <span class="n">vcf_info_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
        <span class="n">make_info_dict</span><span class="p">(</span>
            <span class="n">suffix</span><span class="o">=</span><span class="n">subset</span><span class="p">,</span>
            <span class="n">label_delimiter</span><span class="o">=</span><span class="n">label_delimiter</span><span class="p">,</span>
            <span class="n">gen_anc_names</span><span class="o">=</span><span class="n">pops</span><span class="p">,</span>
            <span class="n">fafmax</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">description_text</span><span class="o">=</span><span class="n">description_text</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="k">if</span> <span class="n">freq_comparison_included</span><span class="p">:</span>
        <span class="n">ctt_label_groups</span> <span class="o">=</span> <span class="n">create_label_groups</span><span class="p">(</span><span class="n">gen_ancs</span><span class="o">=</span><span class="n">pops</span><span class="p">,</span> <span class="n">sexes</span><span class="o">=</span><span class="n">sexes</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">label_group</span> <span class="ow">in</span> <span class="n">ctt_label_groups</span><span class="p">:</span>
            <span class="n">vcf_info_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                <span class="n">make_info_dict</span><span class="p">(</span>
                    <span class="n">freq_ctt</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                    <span class="n">label_groups</span><span class="o">=</span><span class="n">label_group</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="n">vcf_info_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
            <span class="n">make_info_dict</span><span class="p">(</span>
                <span class="n">freq_cmh</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="n">vcf_info_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
            <span class="n">make_info_dict</span><span class="p">(</span>
                <span class="n">freq_stat_union</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="p">)</span>

    <span class="k">return</span> <span class="n">vcf_info_dict</span></div>


<div class="viewcode-block" id="populate_info_dict"><a class="viewcode-back" href="../../../../api_reference/v4/create_release/validate_and_export_vcf.html#gnomad_qc.v4.create_release.validate_and_export_vcf.populate_info_dict">[docs]</a><span class="k">def</span> <span class="nf">populate_info_dict</span><span class="p">(</span>
    <span class="n">info_fields</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">bin_edges</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span>
    <span class="n">age_hist_distribution</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">info_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="n">INFO_DICT</span><span class="p">,</span>
    <span class="n">subset_list</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">SUBSETS</span><span class="p">[</span><span class="s2">&quot;exomes&quot;</span><span class="p">],</span>
    <span class="n">pops</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">GEN_ANC_GROUPS</span><span class="p">[</span><span class="s2">&quot;exomes&quot;</span><span class="p">],</span>
    <span class="n">faf_pops</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="n">FAF_GEN_ANC_GROUPS</span><span class="p">,</span>
    <span class="n">sexes</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="n">SEXES</span><span class="p">,</span>
    <span class="n">in_silico_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="n">IN_SILICO_ANNOTATIONS_INFO_DICT</span><span class="p">,</span>
    <span class="n">vrs_fields_dict</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="n">VRS_FIELDS_DICT</span><span class="p">,</span>
    <span class="n">label_delimiter</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="p">,</span>
    <span class="n">data_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;exomes&quot;</span><span class="p">,</span>
    <span class="n">freq_comparison_included</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">extra_suffix</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">extra_description_text</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Call `make_info_dict` and `make_hist_dict` to populate INFO dictionary.</span>

<span class="sd">    Used during VCF export.</span>

<span class="sd">    Creates:</span>
<span class="sd">        - INFO fields for age histograms (bin freq, n_smaller, and n_larger for</span>
<span class="sd">          heterozygous and homozygous variant carriers).</span>
<span class="sd">        - INFO fields for grpmax AC, AN, AF, nhomalt, and grpmax genetic ancestry group.</span>
<span class="sd">        - INFO fields for AC, AN, AF, nhomalt for each combination of sample genetic</span>
<span class="sd">          ancestry group, sex both for adj and raw data.</span>
<span class="sd">        - INFO fields for filtering allele frequency (faf) annotations.</span>
<span class="sd">        - INFO fields for variant histograms (hist_bin_freq for each histogram and</span>
<span class="sd">          hist_n_larger for DP histograms).</span>

<span class="sd">    :param info_fields: List of info fields to add to the info dict. Default is None.</span>
<span class="sd">    :param bin_edges: Dictionary of variant annotation histograms and their associated</span>
<span class="sd">        bin edges.</span>
<span class="sd">    :param age_hist_distribution: Pipe-delimited string of overall age histogram bin</span>
<span class="sd">        frequency.</span>
<span class="sd">    :param info_dict: INFO dict to be populated.</span>
<span class="sd">    :param subset_list: List of sample subsets in dataset. Default is SUBSETS[&quot;exomes&quot;].</span>
<span class="sd">    :param pops: Dict of sample global genetic ancestry names for the gnomAD data type.</span>
<span class="sd">    :param faf_pops: Dict with gnomAD version (keys) and faf genentic ancestry group</span>
<span class="sd">        names (values). Default is FAF_GEN_ANC_GROUPS.</span>
<span class="sd">    :param sexes: gnomAD sample sexes used in VCF export. Default is SEXES.</span>
<span class="sd">    :param in_silico_dict: Dictionary of in silico predictor score descriptions.</span>
<span class="sd">    :param vrs_fields_dict: Dictionary with VRS annotations.</span>
<span class="sd">    :param label_delimiter: String to use as delimiter when making group label</span>
<span class="sd">        combinations.</span>
<span class="sd">    :param data_type: Data type to populate info dict for. One of &quot;exomes&quot; or</span>
<span class="sd">        &quot;genomes&quot;. Default is &quot;exomes&quot;.</span>
<span class="sd">    :param freq_comparison_included: Whether frequency comparison data is included in the HT.</span>
<span class="sd">    :param extra_suffix: Suffix to add to INFO field.</span>
<span class="sd">    :param extra_description_text: Extra description text to add to INFO field.</span>
<span class="sd">    :return: Updated INFO dictionary for VCF export.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">vcf_info_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;joint&quot;</span><span class="p">:</span>
        <span class="c1"># vcf_info_dict stays empty if data_type is &quot;joint&quot; and subset is not &quot;joint&quot;</span>
        <span class="k">if</span> <span class="s2">&quot;joint&quot;</span> <span class="ow">in</span> <span class="n">subset_list</span><span class="p">:</span>
            <span class="n">vcf_info_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">JOINT_REGION_FLAGS_INFO_DICT</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Get existing info fields from predefined info_dict, e.g. `FS`,</span>
        <span class="c1"># `non_par`, `negative_train_site`...</span>
        <span class="n">vcf_info_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">info_dict</span><span class="p">)</span>
        <span class="n">vcf_info_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">f</span><span class="p">:</span> <span class="n">vcf_info_dict</span><span class="p">[</span><span class="n">f</span><span class="p">]</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">info_fields</span> <span class="k">if</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">vcf_info_dict</span><span class="p">}</span>
        <span class="c1"># Add allele-specific fields to info dict, including AS_VQSR_FIELDS</span>
        <span class="n">vcf_info_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
            <span class="n">add_as_info_dict</span><span class="p">(</span><span class="n">info_dict</span><span class="o">=</span><span class="n">info_dict</span><span class="p">,</span> <span class="n">as_fields</span><span class="o">=</span><span class="n">AS_FIELDS</span> <span class="o">+</span> <span class="n">AS_VQSR_FIELDS</span><span class="p">)</span>
        <span class="p">)</span>

    <span class="k">for</span> <span class="n">subset</span> <span class="ow">in</span> <span class="n">subset_list</span><span class="p">:</span>
        <span class="n">subset_pops</span> <span class="o">=</span> <span class="n">deepcopy</span><span class="p">(</span><span class="n">pops</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;joint&quot;</span><span class="p">:</span>
            <span class="n">description_text</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot; in </span><span class="si">{</span><span class="n">subset</span><span class="si">}</span><span class="s2"> dataset&quot;</span> <span class="k">if</span> <span class="n">subset</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">description_text</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span> <span class="k">if</span> <span class="n">subset</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span> <span class="k">else</span> <span class="sa">f</span><span class="s2">&quot; in </span><span class="si">{</span><span class="n">subset</span><span class="si">}</span><span class="s2"> subset&quot;</span>

        <span class="n">vcf_info_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
            <span class="n">populate_subset_info_dict</span><span class="p">(</span>
                <span class="n">subset</span><span class="o">=</span><span class="n">subset</span><span class="p">,</span>
                <span class="n">description_text</span><span class="o">=</span><span class="n">description_text</span><span class="p">,</span>
                <span class="n">data_type</span><span class="o">=</span><span class="n">data_type</span><span class="p">,</span>
                <span class="n">pops</span><span class="o">=</span><span class="n">subset_pops</span><span class="p">,</span>
                <span class="n">faf_pops</span><span class="o">=</span><span class="n">faf_pops</span><span class="p">,</span>
                <span class="n">sexes</span><span class="o">=</span><span class="n">sexes</span><span class="p">,</span>
                <span class="n">label_delimiter</span><span class="o">=</span><span class="n">label_delimiter</span><span class="p">,</span>
                <span class="n">freq_comparison_included</span><span class="o">=</span><span class="n">freq_comparison_included</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">age_hist_distribution</span><span class="p">:</span>
        <span class="n">age_hist_distribution</span> <span class="o">=</span> <span class="s2">&quot;|&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">age_hist_distribution</span><span class="p">)</span>

    <span class="c1"># Add age histogram data to info dict.</span>
    <span class="n">vcf_info_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
        <span class="n">make_info_dict</span><span class="p">(</span>
            <span class="n">suffix</span><span class="o">=</span><span class="n">extra_suffix</span><span class="p">,</span>
            <span class="n">label_delimiter</span><span class="o">=</span><span class="n">label_delimiter</span><span class="p">,</span>
            <span class="n">bin_edges</span><span class="o">=</span><span class="n">bin_edges</span><span class="p">,</span>
            <span class="n">age_hist_distribution</span><span class="o">=</span><span class="n">age_hist_distribution</span><span class="p">,</span>
            <span class="n">description_text</span><span class="o">=</span><span class="n">extra_description_text</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="c1"># Add variant quality histograms to info dict.</span>
    <span class="n">vcf_info_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
        <span class="n">make_hist_dict</span><span class="p">(</span>
            <span class="n">bin_edges</span><span class="p">,</span>
            <span class="n">adj</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">drop_n_smaller_larger</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">suffix</span><span class="o">=</span><span class="n">extra_suffix</span><span class="p">,</span>
            <span class="n">description_text</span><span class="o">=</span><span class="n">extra_description_text</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;joint&quot;</span><span class="p">:</span>
        <span class="n">vcf_info_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">JOINT_FILTERS_INFO_DICT</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">vcf_info_dict</span>

    <span class="c1"># Add in silico prediction annotations to info_dict.</span>
    <span class="n">vcf_info_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">in_silico_dict</span><span class="p">)</span>

    <span class="c1"># Add VRS annotations to info_dict.</span>
    <span class="n">vcf_info_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">vrs_fields_dict</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">vcf_info_dict</span></div>


<div class="viewcode-block" id="prepare_vcf_header_dict"><a class="viewcode-back" href="../../../../api_reference/v4/create_release/validate_and_export_vcf.html#gnomad_qc.v4.create_release.validate_and_export_vcf.prepare_vcf_header_dict">[docs]</a><span class="k">def</span> <span class="nf">prepare_vcf_header_dict</span><span class="p">(</span>
    <span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">validated_ht</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">],</span>
    <span class="n">info_fields</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">bin_edges</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span>
    <span class="n">age_hist_distribution</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">subset_list</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">pops</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span>
    <span class="n">data_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;exomes&quot;</span><span class="p">,</span>
    <span class="n">joint_included</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">freq_comparison_included</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">extra_suffix</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">extra_description_text</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Prepare VCF header dictionary.</span>

<span class="sd">    :param ht: Input Table</span>
<span class="sd">    :param validated_ht: Validated HT with unfurled info fields.</span>
<span class="sd">    :param info_fields: List of info fields to add to the info dict.</span>
<span class="sd">    :param bin_edges: Dictionary of variant annotation histograms and their associated</span>
<span class="sd">        bin edges.</span>
<span class="sd">    :param age_hist_distribution: Pipe-delimited string of overall age histogram bin</span>
<span class="sd">        frequency.</span>
<span class="sd">    :param subset_list: List of sample subsets in dataset.</span>
<span class="sd">    :param pops: List of sample global genetic ancestry group names for gnomAD data type.</span>
<span class="sd">    :param data_type: Data type to prepare VCF header for. One of &quot;exomes&quot; or &quot;genomes&quot;.</span>
<span class="sd">        Default is &quot;exomes&quot;.</span>
<span class="sd">    :param joint_included: Whether joint frequency data is included in the HT. Default is False.</span>
<span class="sd">    :param freq_comparison_included: Whether frequency comparison data is included in the HT.</span>
<span class="sd">    :param extra_suffix: Suffix to add to INFO field.</span>
<span class="sd">    :param extra_description_text: Extra description text to add to INFO field.</span>
<span class="sd">    :return: Prepared VCF header dictionary.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">data_type</span> <span class="o">!=</span> <span class="s2">&quot;joint&quot;</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Making FILTER dict for VCF...&quot;</span><span class="p">)</span>
        <span class="n">filter_dict</span> <span class="o">=</span> <span class="n">make_vcf_filter_dict</span><span class="p">(</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">filtering_model</span><span class="o">.</span><span class="n">snv_cutoff</span><span class="o">.</span><span class="n">min_score</span><span class="p">),</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">filtering_model</span><span class="o">.</span><span class="n">indel_cutoff</span><span class="o">.</span><span class="n">min_score</span><span class="p">),</span>
            <span class="n">inbreeding_cutoff</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">inbreeding_coeff_cutoff</span><span class="p">),</span>
            <span class="n">variant_qc_filter</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">filtering_model</span><span class="o">.</span><span class="n">filter_name</span><span class="p">),</span>
        <span class="p">)</span>
        <span class="c1"># subset = &quot;&quot; represents full dataset in VCF header construction, the</span>
        <span class="c1"># logic in gnomad_methods is built around this.</span>
        <span class="n">subset_list</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="s2">&quot;joint&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">joint_included</span> <span class="k">else</span> <span class="p">[</span><span class="s2">&quot;&quot;</span><span class="p">])</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Making INFO dict for VCF...&quot;</span><span class="p">)</span>
    <span class="n">vcf_info_dict</span> <span class="o">=</span> <span class="n">populate_info_dict</span><span class="p">(</span>
        <span class="n">info_fields</span><span class="o">=</span><span class="n">info_fields</span><span class="p">,</span>
        <span class="n">bin_edges</span><span class="o">=</span><span class="n">bin_edges</span><span class="p">,</span>
        <span class="n">age_hist_distribution</span><span class="o">=</span><span class="n">age_hist_distribution</span><span class="p">,</span>
        <span class="n">subset_list</span><span class="o">=</span><span class="n">subset_list</span><span class="p">,</span>
        <span class="n">pops</span><span class="o">=</span><span class="n">pops</span><span class="p">,</span>
        <span class="n">data_type</span><span class="o">=</span><span class="n">data_type</span><span class="p">,</span>
        <span class="n">freq_comparison_included</span><span class="o">=</span><span class="n">freq_comparison_included</span><span class="p">,</span>
        <span class="n">extra_suffix</span><span class="o">=</span><span class="n">extra_suffix</span><span class="p">,</span>
        <span class="n">extra_description_text</span><span class="o">=</span><span class="n">extra_description_text</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">data_type</span> <span class="o">!=</span> <span class="s2">&quot;joint&quot;</span><span class="p">:</span>
        <span class="n">vcf_info_dict</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
            <span class="p">{</span><span class="s2">&quot;vep&quot;</span><span class="p">:</span> <span class="p">{</span><span class="s2">&quot;Description&quot;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">validated_ht</span><span class="o">.</span><span class="n">vep_csq_header</span><span class="p">)}}</span>
        <span class="p">)</span>

    <span class="c1"># Adjust keys to remove adj tags before exporting to VCF.</span>
    <span class="n">new_vcf_info_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">i</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_adj&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">):</span> <span class="n">j</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">vcf_info_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>

    <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;joint&quot;</span><span class="p">:</span>
        <span class="n">header_dict</span> <span class="o">=</span> <span class="n">new_vcf_info_dict</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">header_dict</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;info&quot;</span><span class="p">:</span> <span class="n">new_vcf_info_dict</span><span class="p">,</span>
            <span class="s2">&quot;filter&quot;</span><span class="p">:</span> <span class="n">filter_dict</span><span class="p">,</span>
        <span class="p">}</span>

    <span class="k">return</span> <span class="n">header_dict</span></div>


<div class="viewcode-block" id="get_downsamplings_fields"><a class="viewcode-back" href="../../../../api_reference/v4/create_release/validate_and_export_vcf.html#gnomad_qc.v4.create_release.validate_and_export_vcf.get_downsamplings_fields">[docs]</a><span class="k">def</span> <span class="nf">get_downsamplings_fields</span><span class="p">(</span><span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get downsampling specific annotations from info struct.</span>

<span class="sd">    .. note::</span>

<span class="sd">        This retrieves any annotation that contains any downsampling value in its name.</span>

<span class="sd">    :param ht: Input Table.</span>
<span class="sd">    :return: List of downsampling specific annotations to drop from info struct.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">ds</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">flatten</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">downsamplings</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>
    <span class="n">ds</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">ds</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">d</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">str</span><span class="p">(</span><span class="n">d</span><span class="p">)))</span>
    <span class="n">ds_fields</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
        <span class="nb">set</span><span class="p">([</span><span class="n">field</span> <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">ds</span> <span class="k">for</span> <span class="n">field</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="p">)</span> <span class="k">if</span> <span class="nb">str</span><span class="p">(</span><span class="n">d</span><span class="p">)</span> <span class="ow">in</span> <span class="n">field</span><span class="p">])</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">ds_fields</span></div>


<div class="viewcode-block" id="format_validated_ht_for_export"><a class="viewcode-back" href="../../../../api_reference/v4/create_release/validate_and_export_vcf.html#gnomad_qc.v4.create_release.validate_and_export_vcf.format_validated_ht_for_export">[docs]</a><span class="k">def</span> <span class="nf">format_validated_ht_for_export</span><span class="p">(</span>
    <span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
    <span class="n">data_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;exomes&quot;</span><span class="p">,</span>
    <span class="n">vcf_info_reorder</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="n">VCF_INFO_REORDER</span><span class="p">,</span>
    <span class="n">info_fields_to_drop</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Format validated HT for export.</span>

<span class="sd">    Drop downsamplings frequency stats from info, rearrange info, and make sure fields</span>
<span class="sd">    are VCF compatible.</span>

<span class="sd">    :param ht: Validated HT.</span>
<span class="sd">    :param data_type: Data type to format validated HT for. One of &quot;exomes&quot; or &quot;genomes&quot;.</span>
<span class="sd">        Default is &quot;exomes&quot;.</span>
<span class="sd">    :param vcf_info_reorder: Order of VCF INFO fields. These will be placed in front of</span>
<span class="sd">        all other fields in the order specified.</span>
<span class="sd">    :param info_fields_to_drop: List of info fields to drop from the info struct.</span>
<span class="sd">    :return: Formatted HT and list rename row annotations.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">info_fields_to_drop</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">info_fields_to_drop</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Dropping fields from info struct...&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;exomes&quot;</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Getting downsampling annotations to drop from info struct...&quot;</span><span class="p">)</span>
        <span class="n">ds_fields</span> <span class="o">=</span> <span class="n">get_downsamplings_fields</span><span class="p">(</span><span class="n">ht</span><span class="p">)</span>
        <span class="n">info_fields_to_drop</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">ds_fields</span><span class="p">)</span>
    <span class="c1"># Drop any info annotation with &quot;hgdp&quot; or &quot;tgp&quot; in the name for genomes.</span>
    <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;genomes&quot;</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Dropping hgdp and tgp annotations from info struct...&quot;</span><span class="p">)</span>
        <span class="n">info_fields_to_drop</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span>
            <span class="p">[</span><span class="n">f</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="p">)</span> <span class="k">if</span> <span class="p">((</span><span class="s2">&quot;hgdp&quot;</span> <span class="ow">in</span> <span class="n">f</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="s2">&quot;tgp&quot;</span> <span class="ow">in</span> <span class="n">f</span><span class="p">))]</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;joint&quot;</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">dt</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;exomes&quot;</span><span class="p">,</span> <span class="s2">&quot;genomes&quot;</span><span class="p">,</span> <span class="s2">&quot;joint&quot;</span><span class="p">]:</span>
            <span class="n">info_fields_to_drop</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span>
                <span class="p">[</span>
                    <span class="sa">f</span><span class="s2">&quot;age_hist_het_bin_edges_</span><span class="si">{</span><span class="n">dt</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
                    <span class="sa">f</span><span class="s2">&quot;age_hist_hom_bin_edges_</span><span class="si">{</span><span class="n">dt</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="p">]</span>
            <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Add age_histogram bin edges to info fields to drop...&quot;</span><span class="p">)</span>
        <span class="n">info_fields_to_drop</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="s2">&quot;age_hist_het_bin_edges&quot;</span><span class="p">,</span> <span class="s2">&quot;age_hist_hom_bin_edges&quot;</span><span class="p">])</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Adding &#39;SB&#39; to info fields to drop...&quot;</span><span class="p">)</span>
        <span class="n">info_fields_to_drop</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;SB&quot;</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Dropping the following fields from info struct: </span><span class="si">%s</span><span class="s2">...&quot;</span><span class="p">,</span>
        <span class="n">pprint</span><span class="p">(</span><span class="n">info_fields_to_drop</span><span class="p">),</span>
    <span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">info</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="o">*</span><span class="n">info_fields_to_drop</span><span class="p">))</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Dropping _&#39;adj&#39; from info fields...&quot;</span><span class="p">)</span>
    <span class="n">row_annots</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="p">)</span>
    <span class="n">new_row_annots</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_adj&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">row_annots</span><span class="p">]</span>
    <span class="n">info_annot_mapping</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span>
        <span class="nb">zip</span><span class="p">(</span><span class="n">new_row_annots</span><span class="p">,</span> <span class="p">[</span><span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">row_annots</span><span class="p">])</span>
    <span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">transmute</span><span class="p">(</span><span class="n">info</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span><span class="o">**</span><span class="n">info_annot_mapping</span><span class="p">))</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Adjusting VCF incompatible types...&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">data_type</span> <span class="o">!=</span> <span class="s2">&quot;joint&quot;</span><span class="p">:</span>
        <span class="c1"># Reformat AS_SB_TABLE for use in adjust_vcf_incompatible_types</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
            <span class="n">info</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
                <span class="n">AS_SB_TABLE</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">AS_SB_TABLE</span><span class="p">[:</span><span class="mi">2</span><span class="p">],</span> <span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">AS_SB_TABLE</span><span class="p">[</span><span class="mi">2</span><span class="p">:]])</span>
            <span class="p">)</span>
        <span class="p">)</span>
    <span class="c1"># The Table is already split so there are no annotations that need to be</span>
    <span class="c1"># pipe delimited.</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">adjust_vcf_incompatible_types</span><span class="p">(</span><span class="n">ht</span><span class="p">,</span> <span class="n">pipe_delimited_annotations</span><span class="o">=</span><span class="p">[])</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Rearranging fields to desired order...&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;joint&quot;</span><span class="p">:</span>
        <span class="n">special_items</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;exomes&quot;</span><span class="p">:</span> <span class="s2">&quot;exomes_filters&quot;</span><span class="p">,</span> <span class="s2">&quot;genomes&quot;</span><span class="p">:</span> <span class="s2">&quot;genomes_filters&quot;</span><span class="p">}</span>
        <span class="n">new_vcf_info_reorder</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">dt</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;joint&quot;</span><span class="p">,</span> <span class="s2">&quot;exomes&quot;</span><span class="p">,</span> <span class="s2">&quot;genomes&quot;</span><span class="p">]:</span>
            <span class="n">new_vcf_info_reorder</span> <span class="o">+=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">f</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">dt</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">vcf_info_reorder</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">dt</span> <span class="ow">in</span> <span class="n">special_items</span><span class="p">:</span>
                <span class="n">new_vcf_info_reorder</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">special_items</span><span class="p">[</span><span class="n">dt</span><span class="p">])</span>
        <span class="n">vcf_info_reorder</span> <span class="o">=</span> <span class="n">new_vcf_info_reorder</span>

    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="n">info</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="o">*</span><span class="n">vcf_info_reorder</span><span class="p">,</span> <span class="o">*</span><span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="o">*</span><span class="n">vcf_info_reorder</span><span class="p">))</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">ht</span><span class="p">,</span> <span class="n">new_row_annots</span></div>


<div class="viewcode-block" id="process_vep_csq_header"><a class="viewcode-back" href="../../../../api_reference/v4/create_release/validate_and_export_vcf.html#gnomad_qc.v4.create_release.validate_and_export_vcf.process_vep_csq_header">[docs]</a><span class="k">def</span> <span class="nf">process_vep_csq_header</span><span class="p">(</span><span class="n">vep_csq_header</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">VEP_CSQ_HEADER</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Process VEP CSQ header string, delimited by &#39;|&#39;, to remove polyphen and sift annotations.</span>

<span class="sd">    :param vep_csq_header: VEP CSQ header.</span>
<span class="sd">    :return: Processed VEP CSQ header.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Processing VEP CSQ header...&quot;</span><span class="p">)</span>
    <span class="n">vep_csq_header</span> <span class="o">=</span> <span class="n">vep_csq_header</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;|&quot;</span><span class="p">)</span>
    <span class="n">vep_csq_header</span> <span class="o">=</span> <span class="p">[</span><span class="n">f</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">vep_csq_header</span> <span class="k">if</span> <span class="n">f</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;PolyPhen&quot;</span><span class="p">,</span> <span class="s2">&quot;SIFT&quot;</span><span class="p">]]</span>
    <span class="n">vep_csq_header</span> <span class="o">=</span> <span class="s2">&quot;|&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">vep_csq_header</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">vep_csq_header</span></div>


<div class="viewcode-block" id="check_globals_for_retired_terms"><a class="viewcode-back" href="../../../../api_reference/v4/create_release/validate_and_export_vcf.html#gnomad_qc.v4.create_release.validate_and_export_vcf.check_globals_for_retired_terms">[docs]</a><span class="k">def</span> <span class="nf">check_globals_for_retired_terms</span><span class="p">(</span><span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Check list of dictionaries to see if the keys in the dictionaries contain either &#39;pop and &#39;oth&#39;.</span>

<span class="sd">    :param ht: Input Table</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Checking globals for retired terms...&quot;</span><span class="p">)</span>
    <span class="n">errors</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">field</span> <span class="ow">in</span> <span class="n">ht</span><span class="o">.</span><span class="n">globals</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">field</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;meta&quot;</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">ht</span><span class="p">[</span><span class="n">field</span><span class="p">]):</span>
                <span class="k">if</span> <span class="s2">&quot;pop&quot;</span> <span class="ow">in</span> <span class="n">d</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                    <span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Found retired term &#39;pop&#39; in global </span><span class="si">{</span><span class="n">field</span><span class="si">}</span><span class="s2"> annotation: </span><span class="si">{</span><span class="n">d</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="p">)</span>
                <span class="k">if</span> <span class="s2">&quot;oth&quot;</span> <span class="ow">in</span> <span class="n">d</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
                    <span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Found retired term &#39;oth&#39; in global </span><span class="si">{</span><span class="n">field</span><span class="si">}</span><span class="s2"> annotation: </span><span class="si">{</span><span class="n">d</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="p">)</span>
        <span class="k">if</span> <span class="s2">&quot;index_dict&quot;</span> <span class="ow">in</span> <span class="n">field</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">ht</span><span class="p">[</span><span class="n">field</span><span class="p">])</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                <span class="k">if</span> <span class="s2">&quot;oth&quot;</span> <span class="ow">in</span> <span class="n">k</span><span class="p">:</span>
                    <span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="sa">f</span><span class="s2">&quot;Found retired term &#39;oth&#39; in global </span><span class="si">{</span><span class="n">field</span><span class="si">}</span><span class="s2"> annotation: </span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2">&quot;</span>
                    <span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">errors</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Failed retired term check&quot;</span><span class="p">)</span>
        <span class="n">pprint</span><span class="p">(</span><span class="n">errors</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Passed retired term check: No retired terms found in globals.&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="get_joint_filters"><a class="viewcode-back" href="../../../../api_reference/v4/create_release/validate_and_export_vcf.html#gnomad_qc.v4.create_release.validate_and_export_vcf.get_joint_filters">[docs]</a><span class="k">def</span> <span class="nf">get_joint_filters</span><span class="p">(</span><span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Transform exomes and genomes filters to joint filters.</span>

<span class="sd">    :param ht: Input Table.</span>
<span class="sd">    :return: Table with joint filters transformed from exomes and genomes filters.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">exomes_filters</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">exomes_filters</span>
    <span class="n">genomes_filters</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">genomes_filters</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="n">filters</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">set</span><span class="p">(</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">case</span><span class="p">()</span>
            <span class="o">.</span><span class="n">when</span><span class="p">(</span>
                <span class="p">((</span><span class="n">hl</span><span class="o">.</span><span class="n">len</span><span class="p">(</span><span class="n">exomes_filters</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="o">|</span> <span class="n">hl</span><span class="o">.</span><span class="n">is_missing</span><span class="p">(</span><span class="n">exomes_filters</span><span class="p">))</span>
                <span class="o">&amp;</span> <span class="p">((</span><span class="n">hl</span><span class="o">.</span><span class="n">len</span><span class="p">(</span><span class="n">genomes_filters</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="o">|</span> <span class="n">hl</span><span class="o">.</span><span class="n">is_missing</span><span class="p">(</span><span class="n">genomes_filters</span><span class="p">)),</span>
                <span class="p">[</span><span class="s2">&quot;PASS&quot;</span><span class="p">],</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">when</span><span class="p">(</span>
                <span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">len</span><span class="p">(</span><span class="n">exomes_filters</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span>
                <span class="o">&amp;</span> <span class="p">((</span><span class="n">hl</span><span class="o">.</span><span class="n">len</span><span class="p">(</span><span class="n">genomes_filters</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="o">|</span> <span class="n">hl</span><span class="o">.</span><span class="n">is_missing</span><span class="p">(</span><span class="n">genomes_filters</span><span class="p">)),</span>
                <span class="p">[</span><span class="s2">&quot;EXOMES_FILTERED&quot;</span><span class="p">],</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">when</span><span class="p">(</span>
                <span class="p">((</span><span class="n">hl</span><span class="o">.</span><span class="n">len</span><span class="p">(</span><span class="n">exomes_filters</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="o">|</span> <span class="n">hl</span><span class="o">.</span><span class="n">is_missing</span><span class="p">(</span><span class="n">exomes_filters</span><span class="p">))</span>
                <span class="o">&amp;</span> <span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">len</span><span class="p">(</span><span class="n">genomes_filters</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">),</span>
                <span class="p">[</span><span class="s2">&quot;GENOMES_FILTERED&quot;</span><span class="p">],</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">when</span><span class="p">(</span>
                <span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">len</span><span class="p">(</span><span class="n">exomes_filters</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">len</span><span class="p">(</span><span class="n">genomes_filters</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">),</span>
                <span class="p">[</span><span class="s2">&quot;BOTH_FILTERED&quot;</span><span class="p">],</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">default</span><span class="p">([</span><span class="s2">&quot;MISSING_FILTERS&quot;</span><span class="p">])</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">ht</span></div>


<div class="viewcode-block" id="main"><a class="viewcode-back" href="../../../../api_reference/v4/create_release/validate_and_export_vcf.html#gnomad_qc.v4.create_release.validate_and_export_vcf.main">[docs]</a><span class="k">def</span> <span class="nf">main</span><span class="p">(</span><span class="n">args</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Validate release Table and export VCFs.&quot;&quot;&quot;</span>
    <span class="n">hl</span><span class="o">.</span><span class="n">init</span><span class="p">(</span>
        <span class="n">log</span><span class="o">=</span><span class="s2">&quot;/validate_and_export_vcf.log&quot;</span><span class="p">,</span>
        <span class="n">default_reference</span><span class="o">=</span><span class="s2">&quot;GRCh38&quot;</span><span class="p">,</span>
        <span class="n">tmp_dir</span><span class="o">=</span><span class="s2">&quot;gs://gnomad-tmp-30day&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="c1"># SSA Logs are easier to troubleshoot with and this hits a</span>
    <span class="c1"># class too large error without this no whole stage codegen flag</span>
    <span class="c1"># after I added joint_fafmax -- just the straw that broke hails back</span>
    <span class="n">hl</span><span class="o">.</span><span class="n">_set_flags</span><span class="p">(</span><span class="n">use_ssa_logs</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="n">no_whole_stage_codegen</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">)</span>

    <span class="n">overwrite</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">overwrite</span>
    <span class="n">test</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">test</span>
    <span class="n">data_type</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">data_type</span>
    <span class="n">contig</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">contig</span>
    <span class="n">joint_included</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">joint_included</span>
    <span class="n">for_joint</span> <span class="o">=</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;joint&quot;</span>

    <span class="n">resources</span> <span class="o">=</span> <span class="n">get_export_resources</span><span class="p">(</span>
        <span class="n">overwrite</span><span class="p">,</span>
        <span class="n">data_type</span><span class="p">,</span>
        <span class="n">test</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">contig</span> <span class="ow">and</span> <span class="n">test</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;Test argument cannot be used with contig argument as test uses the dataset&quot;</span>
            <span class="s2">&quot; filtered to test gene(s). &quot;</span>
        <span class="p">)</span>
    <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;joint&quot;</span> <span class="ow">and</span> <span class="n">joint_included</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;Joint data type and joint included cannot both be true. joint_included&quot;</span>
            <span class="s2">&quot;is only used for exomes and genomes data types.&quot;</span>
        <span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">validate_release_ht</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running </span><span class="si">%s</span><span class="s2"> release HT validation...&quot;</span><span class="p">,</span> <span class="n">data_type</span><span class="p">)</span>
            <span class="n">res</span> <span class="o">=</span> <span class="n">resources</span><span class="o">.</span><span class="n">validate_release_ht</span>
            <span class="n">res</span><span class="o">.</span><span class="n">check_resource_existence</span><span class="p">()</span>
            <span class="n">ht</span> <span class="o">=</span> <span class="n">res</span><span class="o">.</span><span class="n">release_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="s2">&quot;Checking globals for retired terms and checking their associated row&quot;</span>
                <span class="s2">&quot; annotation lengths...&quot;</span>
            <span class="p">)</span>
            <span class="n">check_globals_for_retired_terms</span><span class="p">(</span><span class="n">ht</span><span class="p">)</span>
            <span class="n">pprint_global_anns</span><span class="p">(</span><span class="n">ht</span><span class="p">)</span>
            <span class="n">validate_hts</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;joint&quot;</span><span class="p">:</span>
                <span class="n">iter_data_types</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;exomes&quot;</span><span class="p">,</span> <span class="s2">&quot;genomes&quot;</span><span class="p">,</span> <span class="s2">&quot;joint&quot;</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">iter_data_types</span> <span class="o">=</span> <span class="p">[</span><span class="n">data_type</span><span class="p">]</span>

            <span class="k">for</span> <span class="n">dt</span> <span class="ow">in</span> <span class="n">iter_data_types</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">for_joint</span><span class="p">:</span>
                    <span class="n">dt_ht</span> <span class="o">=</span> <span class="n">select_type_from_joint_ht</span><span class="p">(</span><span class="n">ht</span><span class="p">,</span> <span class="n">dt</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">dt_ht</span> <span class="o">=</span> <span class="n">ht</span>

                <span class="n">len_comp_global_rows</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">LEN_COMP_GLOBAL_ROWS</span> <span class="o">+</span> <span class="n">LEN_COMP_JOINT_GLOBAL_ROWS</span>
                    <span class="k">if</span> <span class="n">joint_included</span>
                    <span class="k">else</span> <span class="n">LEN_COMP_GLOBAL_ROWS</span>
                <span class="p">)</span>
                <span class="n">check_global_and_row_annot_lengths</span><span class="p">(</span><span class="n">dt_ht</span><span class="p">,</span> <span class="n">len_comp_global_rows</span><span class="p">)</span>

                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Preparing </span><span class="si">%s</span><span class="s2"> HT for validity checks and export...&quot;</span><span class="p">,</span> <span class="n">dt</span><span class="p">)</span>
                <span class="n">dt_ht</span><span class="p">,</span> <span class="n">rename_dict</span> <span class="o">=</span> <span class="n">prepare_ht_for_validation</span><span class="p">(</span>
                    <span class="n">dt_ht</span><span class="p">,</span>
                    <span class="n">data_type</span><span class="o">=</span><span class="n">dt</span><span class="p">,</span>
                    <span class="n">joint_included</span><span class="o">=</span><span class="n">joint_included</span><span class="p">,</span>
                    <span class="n">freq_comparison_included</span><span class="o">=</span><span class="p">(</span><span class="n">dt</span> <span class="o">==</span> <span class="s2">&quot;joint&quot;</span><span class="p">),</span>
                    <span class="n">for_joint_validation</span><span class="o">=</span><span class="n">for_joint</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="n">dt_ht</span> <span class="o">=</span> <span class="n">dt_ht</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">new_temp_file</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;validate_</span><span class="si">{</span><span class="n">dt</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;ht&quot;</span><span class="p">))</span>
                <span class="n">site_gt_check_expr</span> <span class="o">=</span> <span class="kc">None</span>

                <span class="k">if</span> <span class="n">data_type</span> <span class="o">!=</span> <span class="s2">&quot;joint&quot;</span><span class="p">:</span>
                    <span class="n">site_gt_check_expr</span> <span class="o">=</span> <span class="p">{</span>
                        <span class="s2">&quot;monoallelic&quot;</span><span class="p">:</span> <span class="n">dt_ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">monoallelic</span><span class="p">,</span>
                        <span class="s2">&quot;only_het&quot;</span><span class="p">:</span> <span class="n">dt_ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">only_het</span><span class="p">,</span>
                    <span class="p">}</span>

                <span class="n">validate_release_t</span><span class="p">(</span>
                    <span class="n">dt_ht</span><span class="p">,</span>
                    <span class="n">subsets</span><span class="o">=</span><span class="n">SUBSETS</span><span class="p">[</span><span class="n">data_type</span><span class="p">],</span>
                    <span class="n">gen_anc_groups</span><span class="o">=</span><span class="n">GEN_ANC_GROUPS</span><span class="p">[</span><span class="n">dt</span><span class="p">],</span>
                    <span class="n">site_gt_check_expr</span><span class="o">=</span><span class="n">site_gt_check_expr</span><span class="p">,</span>
                    <span class="n">verbose</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">verbose</span><span class="p">,</span>
                    <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;_&quot;</span><span class="p">,</span>
                    <span class="n">sample_sum_sets_and_gen_anc_groups</span><span class="o">=</span><span class="n">SAMPLE_SUM_SETS_AND_POPS</span><span class="p">[</span>
                        <span class="n">data_type</span>
                    <span class="p">],</span>
                    <span class="n">variant_filter_field</span><span class="o">=</span><span class="s2">&quot;AS_VQSR&quot;</span><span class="p">,</span>
                    <span class="n">problematic_regions</span><span class="o">=</span><span class="n">REGION_FLAG_FIELDS</span><span class="p">[</span><span class="n">data_type</span><span class="p">],</span>
                    <span class="n">single_filter_count</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                    <span class="n">filters_check</span><span class="o">=</span><span class="kc">False</span> <span class="k">if</span> <span class="n">dt</span> <span class="o">==</span> <span class="s2">&quot;joint&quot;</span> <span class="k">else</span> <span class="kc">True</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="k">if</span> <span class="n">for_joint</span><span class="p">:</span>
                    <span class="n">ordered_rename_dict</span> <span class="o">=</span> <span class="p">{</span>
                        <span class="n">key</span><span class="p">:</span> <span class="n">rename_dict</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="n">key</span><span class="p">)</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">dt_ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
                    <span class="p">}</span>
                    <span class="n">dt_ht</span> <span class="o">=</span> <span class="n">dt_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">info</span><span class="o">=</span><span class="n">dt_ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">ordered_rename_dict</span><span class="p">))</span>
                    <span class="k">if</span> <span class="n">dt</span> <span class="o">!=</span> <span class="s2">&quot;joint&quot;</span><span class="p">:</span>
                        <span class="n">dt_ht</span> <span class="o">=</span> <span class="n">dt_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
                            <span class="n">info</span><span class="o">=</span><span class="n">dt_ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="o">**</span><span class="p">{</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">dt</span><span class="si">}</span><span class="s2">_filters&quot;</span><span class="p">:</span> <span class="n">dt_ht</span><span class="o">.</span><span class="n">filters</span><span class="p">})</span>
                        <span class="p">)</span>
                        <span class="n">dt_ht</span> <span class="o">=</span> <span class="n">dt_ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="s2">&quot;info&quot;</span><span class="p">)</span>

                    <span class="n">dt_ht</span> <span class="o">=</span> <span class="n">dt_ht</span><span class="o">.</span><span class="n">select_globals</span><span class="p">(</span>
                        <span class="o">**</span><span class="p">{</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">dt</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">:</span> <span class="n">dt_ht</span><span class="p">[</span><span class="n">f</span><span class="p">]</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">dt_ht</span><span class="o">.</span><span class="n">globals</span><span class="p">}</span>
                    <span class="p">)</span>

                <span class="n">validate_hts</span><span class="p">[</span><span class="n">dt</span><span class="p">]</span> <span class="o">=</span> <span class="n">dt_ht</span>

            <span class="n">ht</span> <span class="o">=</span> <span class="n">validate_hts</span><span class="p">[</span><span class="n">data_type</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">for_joint</span><span class="p">:</span>
                <span class="n">in_joint_ht</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
                <span class="k">for</span> <span class="n">dt</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;exomes&quot;</span><span class="p">,</span> <span class="s2">&quot;genomes&quot;</span><span class="p">]:</span>
                    <span class="n">info_expr</span> <span class="o">=</span> <span class="n">validate_hts</span><span class="p">[</span><span class="n">dt</span><span class="p">][</span><span class="n">ht</span><span class="o">.</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">info</span>
                    <span class="n">info_expr</span> <span class="o">=</span> <span class="n">info_expr</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
                        <span class="o">*</span><span class="p">[</span><span class="n">f</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">info_expr</span> <span class="k">if</span> <span class="n">f</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">in_joint_ht</span><span class="p">]</span>
                    <span class="p">)</span>
                    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">info</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="o">**</span><span class="n">info_expr</span><span class="p">))</span>
                    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span><span class="o">**</span><span class="n">validate_hts</span><span class="p">[</span><span class="n">dt</span><span class="p">]</span><span class="o">.</span><span class="n">index_globals</span><span class="p">())</span>
                <span class="n">ht</span> <span class="o">=</span> <span class="n">get_joint_filters</span><span class="p">(</span><span class="n">ht</span><span class="p">)</span>
            <span class="n">ht</span><span class="o">.</span><span class="n">describe</span><span class="p">()</span>

            <span class="c1"># Note: Checkpoint saves time in validity checks and final export by not</span>
            <span class="c1"># needing to run the VCF HT prep on each chromosome -- more needs to happen</span>
            <span class="c1"># before ready for export, but this is an intermediate write.</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Writing prepared VCF HT for validity checks and export...&quot;</span><span class="p">)</span>
            <span class="n">ht</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">validated_ht</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="n">overwrite</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">prepare_vcf_header</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Preparing VCF header dict...&quot;</span><span class="p">)</span>
            <span class="n">res</span> <span class="o">=</span> <span class="n">resources</span><span class="o">.</span><span class="n">prepare_vcf_header</span>
            <span class="n">res</span><span class="o">.</span><span class="n">check_resource_existence</span><span class="p">()</span>
            <span class="n">ht</span> <span class="o">=</span> <span class="n">res</span><span class="o">.</span><span class="n">release_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>
            <span class="n">validated_ht</span> <span class="o">=</span> <span class="n">res</span><span class="o">.</span><span class="n">validated_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">for_joint</span><span class="p">:</span>
                <span class="c1"># v4 Genomes drops subsets from VCF</span>
                <span class="n">subsets</span> <span class="o">=</span> <span class="n">SUBSETS</span><span class="p">[</span><span class="s2">&quot;exomes&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">data_type</span> <span class="o">==</span> <span class="s2">&quot;exomes&quot;</span> <span class="k">else</span> <span class="p">[]</span>
                <span class="n">header_dict</span> <span class="o">=</span> <span class="n">prepare_vcf_header_dict</span><span class="p">(</span>
                    <span class="n">ht</span><span class="p">,</span>
                    <span class="n">validated_ht</span><span class="o">=</span><span class="n">validated_ht</span><span class="p">,</span>
                    <span class="n">info_fields</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">validated_ht</span><span class="o">.</span><span class="n">info</span><span class="p">),</span>
                    <span class="n">bin_edges</span><span class="o">=</span><span class="n">make_hist_bin_edges_expr</span><span class="p">(</span>
                        <span class="n">ht</span><span class="p">,</span>
                        <span class="n">include_age_hists</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                    <span class="p">),</span>
                    <span class="n">age_hist_distribution</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">age_distribution</span><span class="o">.</span><span class="n">bin_freq</span><span class="p">),</span>
                    <span class="n">subset_list</span><span class="o">=</span><span class="n">subsets</span><span class="p">,</span>
                    <span class="n">pops</span><span class="o">=</span><span class="n">GEN_ANC_GROUPS</span><span class="p">[</span><span class="n">data_type</span><span class="p">],</span>
                    <span class="n">data_type</span><span class="o">=</span><span class="n">data_type</span><span class="p">,</span>
                    <span class="n">joint_included</span><span class="o">=</span><span class="n">joint_included</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">header_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;filter&quot;</span><span class="p">:</span> <span class="n">make_vcf_filter_dict</span><span class="p">(</span><span class="n">joint</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span> <span class="s2">&quot;info&quot;</span><span class="p">:</span> <span class="p">{}}</span>
                <span class="k">for</span> <span class="n">dt</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;exomes&quot;</span><span class="p">,</span> <span class="s2">&quot;genomes&quot;</span><span class="p">,</span> <span class="s2">&quot;joint&quot;</span><span class="p">]:</span>
                    <span class="n">dt_ht</span> <span class="o">=</span> <span class="n">select_type_from_joint_ht</span><span class="p">(</span><span class="n">ht</span><span class="p">,</span> <span class="n">dt</span><span class="p">)</span>
                    <span class="n">temp_header_dict</span> <span class="o">=</span> <span class="n">prepare_vcf_header_dict</span><span class="p">(</span>
                        <span class="n">dt_ht</span><span class="p">,</span>
                        <span class="n">validated_ht</span><span class="o">=</span><span class="n">validated_ht</span><span class="p">,</span>
                        <span class="n">info_fields</span><span class="o">=</span><span class="p">[</span><span class="n">f</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">validated_ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">if</span> <span class="n">dt</span> <span class="ow">in</span> <span class="n">f</span><span class="p">],</span>
                        <span class="n">bin_edges</span><span class="o">=</span><span class="n">make_hist_bin_edges_expr</span><span class="p">(</span>
                            <span class="n">dt_ht</span><span class="p">,</span> <span class="n">include_age_hists</span><span class="o">=</span><span class="kc">True</span>
                        <span class="p">),</span>
                        <span class="n">age_hist_distribution</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">eval</span><span class="p">(</span><span class="n">dt_ht</span><span class="o">.</span><span class="n">age_distribution</span><span class="o">.</span><span class="n">bin_freq</span><span class="p">),</span>
                        <span class="n">subset_list</span><span class="o">=</span><span class="p">[</span><span class="n">dt</span><span class="p">],</span>
                        <span class="n">pops</span><span class="o">=</span><span class="n">GEN_ANC_GROUPS</span><span class="p">[</span><span class="n">dt</span><span class="p">],</span>
                        <span class="n">data_type</span><span class="o">=</span><span class="s2">&quot;joint&quot;</span><span class="p">,</span>
                        <span class="n">joint_included</span><span class="o">=</span><span class="n">joint_included</span><span class="p">,</span>
                        <span class="n">freq_comparison_included</span><span class="o">=</span><span class="p">(</span><span class="n">dt</span> <span class="o">==</span> <span class="s2">&quot;joint&quot;</span><span class="p">),</span>
                        <span class="n">extra_suffix</span><span class="o">=</span><span class="n">dt</span><span class="p">,</span>
                        <span class="n">extra_description_text</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot; in </span><span class="si">{</span><span class="n">dt</span><span class="si">}</span><span class="s2"> dataset&quot;</span><span class="p">,</span>
                    <span class="p">)</span>
                    <span class="n">header_dict</span><span class="p">[</span><span class="s2">&quot;info&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">temp_header_dict</span><span class="p">)</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Writing VCF header dict...&quot;</span><span class="p">)</span>
            <span class="k">with</span> <span class="n">hl</span><span class="o">.</span><span class="n">hadoop_open</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">vcf_header_path</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">p</span><span class="p">:</span>
                <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">header_dict</span><span class="p">,</span> <span class="n">p</span><span class="p">,</span> <span class="n">protocol</span><span class="o">=</span><span class="n">pickle</span><span class="o">.</span><span class="n">HIGHEST_PROTOCOL</span><span class="p">)</span>

        <span class="c1"># NOTE: The following step is not yet implemented for joint and should not</span>
        <span class="c1"># be reviewed for it</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">export_vcf</span><span class="p">:</span>
            <span class="n">contig</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;chr</span><span class="si">{</span><span class="n">contig</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">if</span> <span class="n">contig</span> <span class="k">else</span> <span class="kc">None</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Exporting VCF</span><span class="si">{</span><span class="sa">f</span><span class="s1">&#39; for </span><span class="si">{</span><span class="n">contig</span><span class="si">}</span><span class="s1">&#39;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">contig</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;&#39;</span><span class="si">}</span><span class="s2">...&quot;</span><span class="p">)</span>
            <span class="n">res</span> <span class="o">=</span> <span class="n">resources</span><span class="o">.</span><span class="n">export_vcf</span>
            <span class="n">res</span><span class="o">.</span><span class="n">check_resource_existence</span><span class="p">()</span>
            <span class="n">ht</span> <span class="o">=</span> <span class="n">res</span><span class="o">.</span><span class="n">validated_ht</span><span class="o">.</span><span class="n">ht</span><span class="p">()</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Reading release HT from </span><span class="si">%s</span><span class="s2">...&quot;</span><span class="p">,</span> <span class="n">res</span><span class="o">.</span><span class="n">validated_ht</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>
            <span class="k">with</span> <span class="n">hl</span><span class="o">.</span><span class="n">hadoop_open</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">vcf_header_path</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                <span class="n">header_dict</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">contig</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Filtering to </span><span class="si">{</span><span class="n">contig</span><span class="si">}</span><span class="s2">...&quot;</span><span class="p">)</span>
                <span class="n">ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">filter_intervals</span><span class="p">(</span>
                    <span class="n">ht</span><span class="p">,</span> <span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">parse_locus_interval</span><span class="p">(</span><span class="n">contig</span><span class="p">,</span> <span class="n">reference_genome</span><span class="o">=</span><span class="s2">&quot;GRCh38&quot;</span><span class="p">)]</span>
                <span class="p">)</span>

            <span class="n">ht</span><span class="p">,</span> <span class="n">new_row_annots</span> <span class="o">=</span> <span class="n">format_validated_ht_for_export</span><span class="p">(</span><span class="n">ht</span><span class="p">,</span> <span class="n">data_type</span><span class="o">=</span><span class="n">data_type</span><span class="p">)</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running check on VCF fields and info dict...&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">vcf_field_check</span><span class="p">(</span><span class="n">ht</span><span class="p">,</span> <span class="n">header_dict</span><span class="p">,</span> <span class="n">new_row_annots</span><span class="p">):</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Did not pass VCF field check&quot;</span><span class="p">)</span>

            <span class="n">output_path</span> <span class="o">=</span> <span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">qc_temp_prefix</span><span class="p">(</span><span class="n">data_type</span><span class="o">=</span><span class="n">data_type</span><span class="p">)</span><span class="si">}</span><span class="s2">gnomad.</span><span class="si">{</span><span class="n">data_type</span><span class="si">}</span><span class="s2">.test</span><span class="si">{</span><span class="s1">&#39;.chr&#39;</span><span class="o">+</span><span class="n">contig</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">contig</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s1">&#39;&#39;</span><span class="si">}</span><span class="s2">.vcf.bgz&quot;</span>
                <span class="k">if</span> <span class="n">test</span>
                <span class="k">else</span> <span class="n">release_vcf_path</span><span class="p">(</span><span class="n">contig</span><span class="o">=</span><span class="n">contig</span><span class="p">,</span> <span class="n">data_type</span><span class="o">=</span><span class="n">data_type</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="c1"># Match header order to info field order</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Matching header order to info field order...&quot;</span><span class="p">)</span>
            <span class="n">ordered_vcf_info_dict</span> <span class="o">=</span> <span class="p">{</span>
                <span class="n">f</span><span class="p">:</span> <span class="n">header_dict</span><span class="p">[</span><span class="s2">&quot;info&quot;</span><span class="p">][</span><span class="n">f</span><span class="p">]</span>
                <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">header_dict</span><span class="p">[</span><span class="s2">&quot;info&quot;</span><span class="p">]</span>
            <span class="p">}</span>
            <span class="n">header_dict</span><span class="o">.</span><span class="n">update</span><span class="p">({</span><span class="s2">&quot;info&quot;</span><span class="p">:</span> <span class="n">ordered_vcf_info_dict</span><span class="p">})</span>

            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Exporting VCF...&quot;</span><span class="p">)</span>
            <span class="n">export_reference</span> <span class="o">=</span> <span class="n">build_vcf_export_reference</span><span class="p">(</span>
                <span class="s2">&quot;gnomAD_GRCh38&quot;</span><span class="p">,</span> <span class="n">keep_chrM</span><span class="o">=</span><span class="kc">False</span>
            <span class="p">)</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">export_vcf</span><span class="p">(</span>
                <span class="n">rekey_new_reference</span><span class="p">(</span><span class="n">ht</span><span class="p">,</span> <span class="n">export_reference</span><span class="p">),</span>
                <span class="n">output_path</span><span class="p">,</span>
                <span class="n">metadata</span><span class="o">=</span><span class="n">header_dict</span><span class="p">,</span>
                <span class="n">append_to_header</span><span class="o">=</span><span class="n">append_to_vcf_header_path</span><span class="p">(</span><span class="n">data_type</span><span class="o">=</span><span class="n">data_type</span><span class="p">),</span>
                <span class="n">tabix</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="p">)</span>

    <span class="k">finally</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Copying log to logging bucket...&quot;</span><span class="p">)</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">copy_log</span><span class="p">(</span><span class="n">get_logging_path</span><span class="p">(</span><span class="s2">&quot;validity_checks_and_export&quot;</span><span class="p">))</span></div>


<span class="k">def</span> <span class="nf">get_script_argument_parser</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get script argument parser.&quot;&quot;&quot;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">()</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--validate-release-ht&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Run release HT validation&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--verbose&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Log successes in addition to failures during validation&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--overwrite&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Overwrite all data and start from raw inputs&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--test&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="p">(</span>
            <span class="s2">&quot;For validation, test will run on chr20, chrX, and chrY. For VCF export,&quot;</span>
            <span class="s2">&quot; test runs on PCSK9 region. Outputs to test bucket.&quot;</span>
        <span class="p">),</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-d&quot;</span><span class="p">,</span>
        <span class="s2">&quot;--data-type&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Data type to run validity checks on.&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;exomes&quot;</span><span class="p">,</span>
        <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;exomes&quot;</span><span class="p">,</span> <span class="s2">&quot;genomes&quot;</span><span class="p">,</span> <span class="s2">&quot;joint&quot;</span><span class="p">],</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--prepare-vcf-header&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Prepare VCF header dict.&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--export-vcf&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Export VCF.&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--contig&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Contig to export VCF for.&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;--joint-included&quot;</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Whether joint frequency data is included in the release HT.&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">parser</span>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">get_script_argument_parser</span><span class="p">()</span>
    <span class="n">main</span><span class="p">(</span><span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">())</span>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p></p>
  </div>

   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>